{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "e318f7c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "params_export_flag = False\n",
    "\n",
    "input_directory = \"../results/gene_sets_overlaps/\"\n",
    "\n",
    "input_config_filepaths = None"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28aed973",
   "metadata": {},
   "source": [
    "# Comparative genomics - mouse / human binding sites orthologies"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c78ab77",
   "metadata": {},
   "source": [
    "## Overview"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9646b823",
   "metadata": {},
   "source": [
    "NOTE: previous nomenclature used \"peaks\" instead of binding sites. In this context they both designate any short (< 100 nt) genomic interval associated with experimentally measured signal of RBP binding as\n",
    "identified in previous steps."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff7136f8",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "9593f672",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandera\n",
    "import functools\n",
    "import pandas as pd\n",
    "from dotmap import DotMap\n",
    "import numpy as np\n",
    "import os\n",
    "import sys\n",
    "from pathlib import Path\n",
    "import yaml\n",
    "\n",
    "from numbers import Number\n",
    "from typing import Sequence, List, Set, Union, Tuple, Optional\n",
    "\n",
    "import pybedtools as pbt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "9b875aa1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib as mpl\n",
    "import matplotlib.font_manager\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib_venn as venn\n",
    "import seaborn as sns\n",
    "\n",
    "from matplotlib import rc\n",
    "import matplotlib_venn as venn\n",
    "import matplotlib_inline\n",
    "\n",
    "try:\n",
    "    matplotlib.font_manager.findfont(\"Arial\")\n",
    "    mpl.rcParams[\"font.sans-serif\"] = [\"Arial\"]\n",
    "except Exception as e:\n",
    "    print(f\"Arial font not found: {e}\")\n",
    "    mpl.rcParams[\"font.sans-serif\"] = [\"DejaVu Sans\"]\n",
    "\n",
    "rc(\"text\", usetex=False)\n",
    "\n",
    "\n",
    "FONT_SMALL_SIZE = 16\n",
    "FONT_MEDIUM_SIZE = 20\n",
    "FONT_BIGGER_SIZE = 22\n",
    "\n",
    "mpl.rc(\"font\", size=FONT_SMALL_SIZE)  # controls default text sizes\n",
    "mpl.rc(\"axes\", titlesize=FONT_SMALL_SIZE)  # fontsize of the axes title\n",
    "mpl.rc(\"axes\", labelsize=FONT_MEDIUM_SIZE)  # fontsize of the x and y labels\n",
    "mpl.rc(\"xtick\", labelsize=FONT_SMALL_SIZE)  # fontsize of the tick labels\n",
    "mpl.rc(\"ytick\", labelsize=FONT_SMALL_SIZE)  # fontsize of the tick labels\n",
    "mpl.rc(\"legend\", fontsize=FONT_SMALL_SIZE)  # legend fontsize\n",
    "mpl.rc(\"figure\", titlesize=FONT_BIGGER_SIZE)  # fontsize of the figure title\n",
    "\n",
    "matplotlib_inline.backend_inline.set_matplotlib_formats(\"svg\")\n",
    "\n",
    "DEFAULT_FIGSIZE = (10, 6)\n",
    "plt.rcParams[\"figure.figsize\"] = DEFAULT_FIGSIZE\n",
    "plt.rcParams[\"axes.spines.top\"] = True\n",
    "plt.rcParams[\"axes.spines.right\"] = True\n",
    "plt.rcParams[\"font.size\"] = FONT_SMALL_SIZE\n",
    "plt.rcParams[\"figure.titlesize\"] = FONT_BIGGER_SIZE\n",
    "plt.rcParams[\"xtick.labelsize\"] = FONT_MEDIUM_SIZE\n",
    "plt.rcParams[\"ytick.labelsize\"] = FONT_MEDIUM_SIZE\n",
    "plt.rcParams[\"axes.axisbelow\"] = True"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42c6d143",
   "metadata": {},
   "source": [
    "## Definitions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "a08ad03d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def has_array_one_value_greater_than_others(arr: Sequence[Number]) -> bool:\n",
    "    if len(arr) == 0:\n",
    "        raise ValueError(\"Empty array\")\n",
    "\n",
    "    if len(arr) == 1:\n",
    "        return True\n",
    "\n",
    "    else:\n",
    "        arr_s = sorted(arr, reverse=True)\n",
    "        if arr_s[0] > arr_s[1]:\n",
    "            return True\n",
    "        else:\n",
    "            return False\n",
    "\n",
    "\n",
    "def build_identifier_bed6(row_bed6: pd.Series) -> str:\n",
    "    \"\"\"Build a sequence identifier as {chrom}:{start}-{end}:{strand} from BED6 row.\"\"\"\n",
    "    return row_bed6[\"chrom\"] + \":\" + str(row_bed6[\"start\"]) + \"-\" + str(row_bed6[\"end\"]) + \":\" + row_bed6[\"strand\"]\n",
    "\n",
    "\n",
    "def series_extract_set_annotation(column: pd.Series) -> list[set[str]]:\n",
    "    expected_format_delimiters = \"NULL;NULL.NULL;;NULL\"\n",
    "    list_extracted_annotations = (\n",
    "        column.replace(np.nan, expected_format_delimiters)\n",
    "        .str.split(\";;\")\n",
    "        .map(lambda v: list(map(lambda elem: elem.split(\";\")[0], v)))\n",
    "        .map(set)\n",
    "    ).tolist()\n",
    "    return list_extracted_annotations\n",
    "\n",
    "\n",
    "def series_extract_first_annotation(column: pd.Series) -> list[str]:\n",
    "    expected_format_delimiters = \"NULL;NULL.NULL;;NULL\"\n",
    "    list_extracted_annotations = (\n",
    "        column.replace(np.nan, expected_format_delimiters).str.split(\";;\").str[0].str.split(\";\").str[0]\n",
    "    ).tolist()\n",
    "    return list_extracted_annotations\n",
    "\n",
    "\n",
    "def series_extract_first_gene_id(column: pd.Series) -> list[str]:\n",
    "    expected_format_delimiters = \";NULL.NULL;;NULL\"\n",
    "    list_extracted_ids = (\n",
    "        column.replace(np.nan, expected_format_delimiters)\n",
    "        .str.split(\";;\")\n",
    "        .str[0]\n",
    "        .str.split(\";\")\n",
    "        .str[-1]\n",
    "        .str.split(\".\")\n",
    "        .str[0]\n",
    "    ).tolist()\n",
    "    return list_extracted_ids\n",
    "\n",
    "\n",
    "def series_extract_set_geneid(column: pd.Series) -> list[set[str]]:\n",
    "    expected_format_delimiters = \"NULL;NULL.NULL;;NULL\"\n",
    "    list_extracted_annotations = (\n",
    "        column.replace(np.nan, expected_format_delimiters)\n",
    "        .str.split(\";;\")\n",
    "        .map(lambda v: list(map(lambda elem: elem.split(\";\")[-1].split(\".\")[0], v)))\n",
    "        .map(set)\n",
    "    ).tolist()\n",
    "    return list_extracted_annotations\n",
    "\n",
    "\n",
    "def complete_bed6_table_with_additional_columns(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    df[\"identifier\"] = df.apply(lambda row: build_identifier_bed6(row), axis=1)\n",
    "    df[\"gene_id\"] = series_extract_first_gene_id(df[\"name\"])\n",
    "    df[\"main_annot\"] = series_extract_first_annotation(df[\"name\"])\n",
    "    df[\"annot_set\"] = series_extract_set_annotation(df[\"name\"])\n",
    "    df[\"gene_id_set\"] = series_extract_set_geneid(df[\"name\"])\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "def load_bed6_table_and_complete_with_additional_columns(path: os.PathLike, columns: list[str]) -> pd.DataFrame:\n",
    "    df = pd.read_csv(\n",
    "        path,\n",
    "        header=None,\n",
    "        index_col=None,\n",
    "        sep=\"\\t\",\n",
    "        names=columns,\n",
    "        comment=\"#\",\n",
    "    )\n",
    "\n",
    "    df = complete_bed6_table_with_additional_columns(df)\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "b0d0541a",
   "metadata": {},
   "outputs": [],
   "source": [
    "class MousePeakIdentifierAndGeneOrthologiesSchema(pandera.DataFrameModel):\n",
    "    identifier_mouse_source: pandera.typing.Series[str] = pandera.Field(alias=\"identifier.mouse_source\", coerce=True)\n",
    "    identifier_human_target: pandera.typing.Series[str] = pandera.Field(alias=\"identifier.human_target\", coerce=True)\n",
    "    orth_first_order: pandera.typing.Series[str] = pandera.Field(alias=\"orth_first_order\", coerce=True, nullable=True)\n",
    "    orth_second_order: pandera.typing.Series[str] = pandera.Field(alias=\"orth_second_order\", coerce=True, nullable=True)\n",
    "    orth_third_order_details: pandera.typing.Series[str] = pandera.Field(\n",
    "        alias=\"orth_third_order.details\", coerce=True, nullable=True\n",
    "    )\n",
    "    orth_third_order: pandera.typing.Series[str] = pandera.Field(alias=\"orth_third_order\", coerce=True, nullable=True)\n",
    "    gene_orthology_status: pandera.typing.Series[str] = pandera.Field(alias=\"gene_orthology_status\", coerce=True)\n",
    "    orthologous_gene_set: pandera.typing.Series[str] = pandera.Field(alias=\"orthologous_gene_set\", coerce=True)\n",
    "    annotated_lifted_region_bound_gene_set: pandera.typing.Series[str] = pandera.Field(\n",
    "        alias=\"annotated_lifted_region_bound_gene_set\", coerce=True\n",
    "    )\n",
    "\n",
    "    class Config:\n",
    "        strict = False\n",
    "\n",
    "\n",
    "def read_mouse_peaks_identifiers_with_gene_sets_orthologies(\n",
    "    filepath: os.PathLike,\n",
    ") -> pandera.typing.DataFrame[MousePeakIdentifierAndGeneOrthologiesSchema]:\n",
    "    df = pd.read_csv(filepath, sep=\"\\t\")\n",
    "    return MousePeakIdentifierAndGeneOrthologiesSchema.validate(df)\n",
    "\n",
    "\n",
    "##\n",
    "\n",
    "\n",
    "class HumanPeakIdentifierAndGeneOrthologiesSchema(pandera.DataFrameModel):\n",
    "    identifier_human_source: pandera.typing.Series[str] = pandera.Field(alias=\"identifier.human_source\", coerce=True)\n",
    "    identifier_mouse_target: pandera.typing.Series[str] = pandera.Field(alias=\"identifier.mouse_target\", coerce=True)\n",
    "    orth_first_order: pandera.typing.Series[str] = pandera.Field(alias=\"orth_first_order\", coerce=True, nullable=True)\n",
    "    orth_second_order: pandera.typing.Series[str] = pandera.Field(alias=\"orth_second_order\", coerce=True, nullable=True)\n",
    "    orth_third_order_details: pandera.typing.Series[str] = pandera.Field(\n",
    "        alias=\"orth_third_order.details\", coerce=True, nullable=True\n",
    "    )\n",
    "    orth_third_order: pandera.typing.Series[str] = pandera.Field(alias=\"orth_third_order\", coerce=True, nullable=True)\n",
    "    gene_orthology_status: pandera.typing.Series[str] = pandera.Field(alias=\"gene_orthology_status\", coerce=True)\n",
    "    orthologous_gene_set: pandera.typing.Series[str] = pandera.Field(alias=\"orthologous_gene_set\", coerce=True)\n",
    "    annotated_lifted_region_bound_gene_set: pandera.typing.Series[str] = pandera.Field(\n",
    "        alias=\"annotated_lifted_region_bound_gene_set\", coerce=True\n",
    "    )\n",
    "\n",
    "    class Config:\n",
    "        strict = False\n",
    "\n",
    "\n",
    "def read_human_peaks_identifiers_with_gene_sets_orthologies(\n",
    "    filepath: os.PathLike,\n",
    ") -> pandera.typing.DataFrame[HumanPeakIdentifierAndGeneOrthologiesSchema]:\n",
    "    df = pd.read_csv(filepath, sep=\"\\t\")\n",
    "    return HumanPeakIdentifierAndGeneOrthologiesSchema.validate(df)\n",
    "\n",
    "\n",
    "##\n",
    "\n",
    "\n",
    "def eval_peaks_identifiers_with_gene_set_orthologies(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    \"\"\"Ugly hack to cast the strings in the dataframe to python objects.\"\"\"\n",
    "\n",
    "    columns_to_cast = [\n",
    "        \"orth_third_order.details\",\n",
    "        \"annotated_lifted_region_bound_gene_set\",\n",
    "        \"orthologous_gene_set\",\n",
    "    ]\n",
    "\n",
    "    for c in columns_to_cast:\n",
    "        df[c] = df[c].map(eval)\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "###\n",
    "\n",
    "\n",
    "class UnmappedPeakSchema(pandera.DataFrameModel):\n",
    "    identifier: pandera.typing.Series[str] = pandera.Field(\n",
    "        coerce=True,\n",
    "    )\n",
    "    name: pandera.typing.Series[str] = pandera.Field(coerce=True)\n",
    "    gene_id: pandera.typing.Series[str] = pandera.Field(coerce=True)\n",
    "\n",
    "\n",
    "##\n",
    "\n",
    "\n",
    "class MappingPeakMetaSchema(pandera.DataFrameModel):\n",
    "    \"\"\"Schema for the mouse peaks metadata.\"\"\"\n",
    "\n",
    "    name: pandera.typing.Series[str] = pandera.Field(coerce=True)\n",
    "    gene_id: pandera.typing.Series[str] = pandera.Field(coerce=True, nullable=True)\n",
    "    main_annot: pandera.typing.Series[str] = pandera.Field(coerce=True, nullable=True)\n",
    "    annot_set: pandera.typing.Series[str] = pandera.Field(coerce=True, nullable=True)\n",
    "    gene_id_set: pandera.typing.Series[str] = pandera.Field(coerce=True, nullable=True)\n",
    "\n",
    "\n",
    "def read_mapping_peak_meta(fp: Path) -> pandera.typing.DataFrame[MappingPeakMetaSchema]:\n",
    "    \"\"\"Read the mapping peak metadata from a file.\"\"\"\n",
    "    df = pd.read_csv(fp, sep=\"\\t\", dtype=str, index_col=0)\n",
    "    return MappingPeakMetaSchema.validate(df)\n",
    "\n",
    "\n",
    "##\n",
    "\n",
    "\n",
    "class MappingPeakBed6Schema(pandera.DataFrameModel):\n",
    "    chrom: pandera.typing.Series[str] = pandera.Field(coerce=True)\n",
    "    start: pandera.typing.Series[int] = pandera.Field(coerce=True)\n",
    "    end: pandera.typing.Series[int] = pandera.Field(coerce=True)\n",
    "    name: pandera.typing.Series[str] = pandera.Field(coerce=True)\n",
    "    score: pandera.typing.Series[float] = pandera.Field(coerce=True)\n",
    "    strand: pandera.typing.Series[str] = pandera.Field(coerce=True)\n",
    "\n",
    "\n",
    "def read_mapping_peak_bed6(fp: Path) -> pandera.typing.DataFrame[MappingPeakBed6Schema]:\n",
    "    \"\"\"Read the mapping peak bed6 file.\"\"\"\n",
    "    df = pd.read_csv(fp, sep=\"\\t\", index_col=0)\n",
    "    return MappingPeakBed6Schema.validate(df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "b23165e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_identifier_bed6(row_bed6: pd.Series) -> str:\n",
    "    \"\"\"Build a sequence identifier as {chrom}:{start}-{end}:{strand} from BED6 row.\"\"\"\n",
    "    return row_bed6[\"chrom\"] + \":\" + str(row_bed6[\"start\"]) + \"-\" + str(row_bed6[\"end\"]) + \":\" + row_bed6[\"strand\"]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66c61e97",
   "metadata": {},
   "source": [
    "## Initialization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "558d79cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_directory = Path(input_directory)\n",
    "if not input_directory.exists():\n",
    "    raise FileNotFoundError(\n",
    "        f\"Input directory {input_directory} does not exist. \"\n",
    "        f\"Please run the `orthologies.gene_sets_overlaps.py.ipynb` notebook first.\"\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "a75534ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed6_cols = [\"chrom\", \"start\", \"end\", \"name\", \"score\", \"strand\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "be0b3a2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "if input_config_filepaths is not None:\n",
    "    if Path(input_config_filepaths).exists():\n",
    "        with open(input_config_filepaths, \"r\") as f:\n",
    "            filepaths = yaml.safe_load(f)\n",
    "    else:\n",
    "        filepaths = {}\n",
    "\n",
    "# NOTE: hard-coded filepaths override config filepaths if defined\n",
    "local_filepaths = {\n",
    "    \"mouse_chromsizes\": \"../../../roquin_lmoyon/general_storage/ucsc/chromsizes/mm10.chrom.sizes\",\n",
    "    \"human_chromsizes\": \"../../../roquin_lmoyon/general_storage/ucsc/chromsizes/hg38.chrom.sizes\",\n",
    "    \"orthologs\": \"../../../roquin_lmoyon/roquin_storage/orthologs/merged.human_mouse.tsv.gz\",\n",
    "    #\n",
    "    \"mouse_peaks_full\": \"../../../roquin_lmoyon/roquin_storage/20240715-liftover_results/roquin/mouse/test/bs_mouse.bed\",\n",
    "    \"mouse_peaks_mappable\": \"../../../roquin_lmoyon/roquin_storage/20240715-liftover_results/roquin/mouse/test/bs_mouse_mappable.bed\",\n",
    "    \"mouse_peaks_unmappable\": \"../../../roquin_lmoyon/roquin_storage/20240715-liftover_results/roquin/mouse/test/unmapped_bs.txt\",\n",
    "    \"mouse_peaks_lifted\": \"../../../roquin_lmoyon/roquin_storage/20240715-liftover_results/roquin/mouse/test/lifted_mm_to_hg.bed\",\n",
    "    #\n",
    "    \"human_peaks_full\": \"../../../roquin_lmoyon/roquin_storage/20240715-liftover_results/roquin/human/test/bs_human.bed\",\n",
    "    \"human_peaks_mappable\": \"../../../roquin_lmoyon/roquin_storage/20240715-liftover_results/roquin/human/test/bs_human_mappable.bed\",\n",
    "    \"human_peaks_unmappable\": \"../../../roquin_lmoyon/roquin_storage/20240715-liftover_results/roquin/human/test/unmapped_bs.txt\",\n",
    "    \"human_peaks_lifted\": \"../../../roquin_lmoyon/roquin_storage/20240715-liftover_results/roquin/human/test/lifted_hg_to_mm.bed\",\n",
    "    #\n",
    "    #\n",
    "}\n",
    "\n",
    "filepaths = {\n",
    "    **{k: v for k, v in local_filepaths.items() if v is not None},\n",
    "    **local_filepaths,\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "45a8654e",
   "metadata": {},
   "outputs": [],
   "source": [
    "if params_export_flag:\n",
    "    outputdir = Path(\"../results\") / \"peaks_orthologies\"\n",
    "\n",
    "    if not outputdir.exists():\n",
    "        outputdir.mkdir()\n",
    "\n",
    "    print(outputdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9ff647a",
   "metadata": {},
   "source": [
    "## Load"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "42030aff",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA = DotMap()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70b281cb",
   "metadata": {},
   "source": [
    "### Misc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "73f55b83",
   "metadata": {},
   "outputs": [],
   "source": [
    "fp = filepaths[\"mouse_chromsizes\"]\n",
    "DATA[\"mouse\"][\"chromsizes_fp\"] = fp\n",
    "DATA[\"mouse\"][\"chromsizes\"] = pd.read_csv(fp, sep=\"\\t\", header=None).set_index(0)[1].to_dict()\n",
    "\n",
    "fp = filepaths[\"human_chromsizes\"]\n",
    "DATA[\"human\"][\"chromsizes_fp\"] = fp\n",
    "DATA[\"human\"][\"chromsizes\"] = pd.read_csv(fp, sep=\"\\t\", header=None).set_index(0)[1].to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "18a4a445",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>human_ensid</th>\n",
       "      <th>human_name</th>\n",
       "      <th>mouse_ensid</th>\n",
       "      <th>mouse_name</th>\n",
       "      <th>relationtype_id</th>\n",
       "      <th>supported_by_resource</th>\n",
       "      <th>support_level</th>\n",
       "      <th>support_details</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>ENSMUSG00000067377</td>\n",
       "      <td>Tspan6</td>\n",
       "      <td>1:1</td>\n",
       "      <td>biomart;;omabrowser;;mgi</td>\n",
       "      <td>5</td>\n",
       "      <td>biomart:human_to_mouse;mouse_to_human;;omabrow...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>ENSMUSG00000031250</td>\n",
       "      <td>Tnmd</td>\n",
       "      <td>1:1</td>\n",
       "      <td>biomart;;omabrowser;;mgi</td>\n",
       "      <td>5</td>\n",
       "      <td>biomart:human_to_mouse;mouse_to_human;;omabrow...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>ENSMUSG00000078919</td>\n",
       "      <td>Dpm1</td>\n",
       "      <td>n:1</td>\n",
       "      <td>biomart;;omabrowser;;mgi</td>\n",
       "      <td>5</td>\n",
       "      <td>biomart:human_to_mouse;mouse_to_human;;omabrow...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       human_ensid human_name         mouse_ensid mouse_name relationtype_id  \\\n",
       "0  ENSG00000000003     TSPAN6  ENSMUSG00000067377     Tspan6             1:1   \n",
       "1  ENSG00000000005       TNMD  ENSMUSG00000031250       Tnmd             1:1   \n",
       "2  ENSG00000000419       DPM1  ENSMUSG00000078919       Dpm1             n:1   \n",
       "\n",
       "      supported_by_resource  support_level  \\\n",
       "0  biomart;;omabrowser;;mgi              5   \n",
       "1  biomart;;omabrowser;;mgi              5   \n",
       "2  biomart;;omabrowser;;mgi              5   \n",
       "\n",
       "                                     support_details  \n",
       "0  biomart:human_to_mouse;mouse_to_human;;omabrow...  \n",
       "1  biomart:human_to_mouse;mouse_to_human;;omabrow...  \n",
       "2  biomart:human_to_mouse;mouse_to_human;;omabrow...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(32490, 8)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Table of orthologous genes\n",
    "fp = filepaths[\"orthologs\"]\n",
    "df = pd.read_csv(\n",
    "    fp,\n",
    "    header=0,\n",
    "    index_col=None,\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "\n",
    "\n",
    "display(df.head(3))\n",
    "display(df.shape)\n",
    "\n",
    "DATA[\"orthologs\"] = df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d26cb55",
   "metadata": {},
   "source": [
    "### Peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "63334acc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MOUSE PEAKS\n",
      "Number of peaks in the 'full set': 19,759\n",
      "Number of peaks in the 'mappable' set: 17,426\n",
      "Number of peaks in the 'unmapped' set: 2,333\n",
      "Number of peaks in the 'lifted' set: 17,426\n"
     ]
    }
   ],
   "source": [
    "print(\"MOUSE PEAKS\")\n",
    "\n",
    "####\n",
    "\n",
    "df = load_bed6_table_and_complete_with_additional_columns(path=filepaths[\"mouse_peaks_full\"], columns=bed6_cols)\n",
    "\n",
    "print(f\"Number of peaks in the 'full set': {df.shape[0]:,}\")\n",
    "DATA[\"mouse\"][\"peaks\"][\"full\"] = df\n",
    "\n",
    "\n",
    "####\n",
    "\n",
    "df = load_bed6_table_and_complete_with_additional_columns(\n",
    "    path=filepaths[\"mouse_peaks_mappable\"],  # NOTE: content is the same as mouse_bs_mapped.bed\n",
    "    columns=bed6_cols,\n",
    ")\n",
    "print(f\"Number of peaks in the 'mappable' set: {df.shape[0]:,}\")\n",
    "DATA[\"mouse\"][\"peaks\"][\"mappable\"] = df\n",
    "\n",
    "####\n",
    "\n",
    "df = load_bed6_table_and_complete_with_additional_columns(path=filepaths[\"mouse_peaks_unmappable\"], columns=bed6_cols)\n",
    "\n",
    "print(f\"Number of peaks in the 'unmapped' set: {df.shape[0]:,}\")\n",
    "DATA[\"mouse\"][\"peaks\"][\"unmapped\"] = df\n",
    "\n",
    "####\n",
    "\n",
    "\n",
    "df = load_bed6_table_and_complete_with_additional_columns(\n",
    "    path=filepaths[\"mouse_peaks_lifted\"],\n",
    "    columns=bed6_cols + [\"source_identifier\"],\n",
    ")\n",
    "\n",
    "DATA[\"mouse\"][\"peaks\"][\"lifted\"] = df\n",
    "print(f\"Number of peaks in the 'lifted' set: {df.shape[0]:,}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "490d6561",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HUMAN PEAKS\n",
      "Number of peaks in the 'full' set: 33,492\n",
      "Number of peaks in the 'mappable' set: 29,383\n",
      "Number of peaks in the 'unmapped' set: 4,109\n",
      "Number of peaks in the 'lifted' set: 29,383\n"
     ]
    }
   ],
   "source": [
    "print(\"HUMAN PEAKS\")\n",
    "\n",
    "####\n",
    "\n",
    "df = load_bed6_table_and_complete_with_additional_columns(path=filepaths[\"human_peaks_full\"], columns=bed6_cols)\n",
    "\n",
    "print(f\"Number of peaks in the 'full' set: {df.shape[0]:,}\")\n",
    "DATA[\"human\"][\"peaks\"][\"full\"] = df\n",
    "\n",
    "####\n",
    "\n",
    "df = load_bed6_table_and_complete_with_additional_columns(\n",
    "    path=filepaths[\"human_peaks_mappable\"],  # Same as human_bs_mapped.bed\n",
    "    columns=bed6_cols,\n",
    ")\n",
    "\n",
    "print(f\"Number of peaks in the 'mappable' set: {df.shape[0]:,}\")\n",
    "DATA[\"human\"][\"peaks\"][\"mappable\"] = df\n",
    "\n",
    "####\n",
    "\n",
    "df = load_bed6_table_and_complete_with_additional_columns(path=filepaths[\"human_peaks_unmappable\"], columns=bed6_cols)\n",
    "\n",
    "print(f\"Number of peaks in the 'unmapped' set: {df.shape[0]:,}\")\n",
    "DATA[\"human\"][\"peaks\"][\"unmapped\"] = df\n",
    "\n",
    "####\n",
    "\n",
    "df = load_bed6_table_and_complete_with_additional_columns(\n",
    "    path=filepaths[\"human_peaks_lifted\"],\n",
    "    columns=bed6_cols + [\"source_identifier\"],\n",
    ")\n",
    "\n",
    "print(f\"Number of peaks in the 'lifted' set: {df.shape[0]:,}\")\n",
    "DATA[\"human\"][\"peaks\"][\"lifted\"] = df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90925f7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Verifications :\n",
    "# - regions should be unique\n",
    "# - mappable + unmappable = full\n",
    "\n",
    "m_full = DATA.mouse.peaks.full.loc[:, [\"identifier\", \"name\", \"gene_id\"]].copy()\n",
    "m_map = DATA.mouse.peaks.mappable.loc[:, [\"identifier\", \"name\", \"gene_id\"]].copy()\n",
    "m_umap = DATA.mouse.peaks.unmapped.loc[:, [\"identifier\", \"name\", \"gene_id\"]].copy()\n",
    "\n",
    "assert m_full[\"identifier\"].duplicated().sum() == 0\n",
    "assert m_map[\"identifier\"].duplicated().sum() == 0\n",
    "assert m_umap[\"identifier\"].duplicated().sum() == 0\n",
    "assert set(m_map[\"identifier\"].unique()) | set(m_umap[\"identifier\"].unique()) == set(m_full[\"identifier\"].unique())\n",
    "\n",
    "\n",
    "h_full = DATA.human.peaks.full.loc[:, [\"identifier\", \"name\", \"gene_id\"]].copy()\n",
    "h_map = DATA.human.peaks.mappable.loc[:, [\"identifier\", \"name\", \"gene_id\"]].copy()\n",
    "h_umap = DATA.human.peaks.unmapped.loc[:, [\"identifier\", \"name\", \"gene_id\"]].copy()\n",
    "\n",
    "assert h_full[\"identifier\"].duplicated().sum() == 0\n",
    "assert h_map[\"identifier\"].duplicated().sum() == 0\n",
    "assert h_umap[\"identifier\"].duplicated().sum() == 0\n",
    "assert set(h_map[\"identifier\"].unique()) | set(h_umap[\"identifier\"].unique()) == set(h_full[\"identifier\"].unique())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "77c98538",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mouse peaks full shape: (19759, 3)\n",
      "Mouse peaks 3' UTR shape: (12565, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>name</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:6274997-6275006:+</td>\n",
       "      <td>three_prime_UTR;Rb1cc1;protein_coding;ENSMUSG0...</td>\n",
       "      <td>ENSMUSG00000025907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:7170406-7170415:+</td>\n",
       "      <td>three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...</td>\n",
       "      <td>ENSMUSG00000051285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1:7170431-7170440:+</td>\n",
       "      <td>three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...</td>\n",
       "      <td>ENSMUSG00000051285</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               identifier                                               name  \\\n",
       "0  chr1:6274997-6275006:+  three_prime_UTR;Rb1cc1;protein_coding;ENSMUSG0...   \n",
       "1  chr1:7170406-7170415:+  three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...   \n",
       "2  chr1:7170431-7170440:+  three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...   \n",
       "\n",
       "              gene_id  \n",
       "0  ENSMUSG00000025907  \n",
       "1  ENSMUSG00000051285  \n",
       "2  ENSMUSG00000051285  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Human peaks full shape: (33492, 3)\n",
      "Human peaks 3' UTR shape: (22338, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>name</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:1211398-1211407:-</td>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>ENSG00000186827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:1211415-1211424:-</td>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>ENSG00000186827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1:1211431-1211440:-</td>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>ENSG00000186827</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               identifier                                               name  \\\n",
       "0  chr1:1211398-1211407:-  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "1  chr1:1211415-1211424:-  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "2  chr1:1211431-1211440:-  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "\n",
       "           gene_id  \n",
       "0  ENSG00000186827  \n",
       "1  ENSG00000186827  \n",
       "2  ENSG00000186827  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "m_full_3utr = m_full.loc[lambda df: df[\"name\"].str.startswith(\"three_prime_UTR\"), :].copy()\n",
    "\n",
    "print(\"Mouse peaks full shape:\", m_full.shape)\n",
    "print(\"Mouse peaks 3' UTR shape:\", m_full_3utr.shape)\n",
    "display(m_full_3utr.head(3))\n",
    "\n",
    "\n",
    "h_full_3utr = h_full.loc[lambda df: df[\"name\"].str.startswith(\"three_prime_UTR\"), :].copy()\n",
    "print(\"Human peaks full shape:\", h_full.shape)\n",
    "print(\"Human peaks 3' UTR shape:\", h_full_3utr.shape)\n",
    "display(h_full_3utr.head(3))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9086bf3",
   "metadata": {},
   "source": [
    "### Results from gene sets overlaps"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4787d867",
   "metadata": {},
   "source": [
    "#### Mouse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e4763f7d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "m_lift_3utr.shape=(11880, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier.mouse_source</th>\n",
       "      <th>identifier.human_target</th>\n",
       "      <th>orth_first_order</th>\n",
       "      <th>orth_second_order</th>\n",
       "      <th>orth_third_order.details</th>\n",
       "      <th>orth_third_order</th>\n",
       "      <th>gene_orthology_status</th>\n",
       "      <th>orthologous_gene_set</th>\n",
       "      <th>annotated_lifted_region_bound_gene_set</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:6274997-6275006:+</td>\n",
       "      <td>chr8:52623049-52623059:-</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSMUSG00000025907, [ENSG00000023287])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSG00000023287}</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:7170406-7170415:+</td>\n",
       "      <td>chr8:51819766-51819775:-</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSMUSG00000051285, [ENSG00000168300])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSG00000168300}</td>\n",
       "      <td>{ENSG00000168300}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1:7170431-7170440:+</td>\n",
       "      <td>chr8:51819741-51819750:-</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSMUSG00000051285, [ENSG00000168300])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSG00000168300}</td>\n",
       "      <td>{ENSG00000168300}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  identifier.mouse_source   identifier.human_target orth_first_order  \\\n",
       "0  chr1:6274997-6275006:+  chr8:52623049-52623059:-             True   \n",
       "1  chr1:7170406-7170415:+  chr8:51819766-51819775:-             True   \n",
       "2  chr1:7170431-7170440:+  chr8:51819741-51819750:-             True   \n",
       "\n",
       "  orth_second_order                   orth_third_order.details  \\\n",
       "0              True  [(ENSMUSG00000025907, [ENSG00000023287])]   \n",
       "1              True  [(ENSMUSG00000051285, [ENSG00000168300])]   \n",
       "2              True  [(ENSMUSG00000051285, [ENSG00000168300])]   \n",
       "\n",
       "  orth_third_order gene_orthology_status orthologous_gene_set  \\\n",
       "0             True                 first    {ENSG00000023287}   \n",
       "1             True                 first    {ENSG00000168300}   \n",
       "2             True                 first    {ENSG00000168300}   \n",
       "\n",
       "  annotated_lifted_region_bound_gene_set  \n",
       "0                                     {}  \n",
       "1                      {ENSG00000168300}  \n",
       "2                      {ENSG00000168300}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fp = input_directory / \"m_lift_3utr.tsv.gz\"\n",
    "m_lift_3utr = read_mouse_peaks_identifiers_with_gene_sets_orthologies(fp)\n",
    "m_lift_3utr = eval_peaks_identifiers_with_gene_set_orthologies(m_lift_3utr)\n",
    "\n",
    "print(f\"{m_lift_3utr.shape=}\")\n",
    "display(m_lift_3utr.head(3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6ed76af2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "m_umap_3utr.shape=(685, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>name</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:36526658-36526667:+</td>\n",
       "      <td>three_prime_UTR;Cnnm3;protein_coding;ENSMUSG00...</td>\n",
       "      <td>ENSMUSG00000001138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:36526988-36526997:+</td>\n",
       "      <td>three_prime_UTR;Cnnm3;protein_coding;ENSMUSG00...</td>\n",
       "      <td>ENSMUSG00000001138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1:36527018-36527027:+</td>\n",
       "      <td>three_prime_UTR;Cnnm3;protein_coding;ENSMUSG00...</td>\n",
       "      <td>ENSMUSG00000001138</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 identifier  \\\n",
       "0  chr1:36526658-36526667:+   \n",
       "1  chr1:36526988-36526997:+   \n",
       "2  chr1:36527018-36527027:+   \n",
       "\n",
       "                                                name             gene_id  \n",
       "0  three_prime_UTR;Cnnm3;protein_coding;ENSMUSG00...  ENSMUSG00000001138  \n",
       "1  three_prime_UTR;Cnnm3;protein_coding;ENSMUSG00...  ENSMUSG00000001138  \n",
       "2  three_prime_UTR;Cnnm3;protein_coding;ENSMUSG00...  ENSMUSG00000001138  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fp = input_directory / \"m_umap_3utr.tsv.gz\"\n",
    "m_umap_3utr = UnmappedPeakSchema.validate(pd.read_csv(fp, sep=\"\\t\"))\n",
    "\n",
    "print(f\"{m_umap_3utr.shape=}\")\n",
    "display(m_umap_3utr.head(3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "43e3dcb7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1312\n",
      "824\n",
      "1015\n",
      "703\n"
     ]
    }
   ],
   "source": [
    "fp = input_directory / \"m_genes_bound_orth_complete.txt\"\n",
    "m_genes_bound_orth_complete = pd.read_csv(fp, sep=\"\\t\", header=None).iloc[:, 0].tolist()\n",
    "\n",
    "print(len(m_genes_bound_orth_complete))\n",
    "\n",
    "\n",
    "fp = input_directory / \"m_genes_no_bound_orth_complete.txt\"\n",
    "m_genes_no_bound_orth_complete = pd.read_csv(fp, sep=\"\\t\", header=None).iloc[:, 0].tolist()\n",
    "\n",
    "print(len(m_genes_no_bound_orth_complete))\n",
    "\n",
    "\n",
    "##\n",
    "\n",
    "\n",
    "fp = input_directory / \"m_genes_bound_orth_complete_3utr.txt\"\n",
    "m_genes_bound_orth_complete_3utr = pd.read_csv(fp, sep=\"\\t\", header=None).iloc[:, 0].tolist()\n",
    "\n",
    "print(len(m_genes_bound_orth_complete_3utr))\n",
    "\n",
    "\n",
    "fp = input_directory / \"m_genes_no_bound_orth_complete_3utr.txt\"\n",
    "m_genes_no_bound_orth_complete_3utr = pd.read_csv(fp, sep=\"\\t\", header=None).iloc[:, 0].tolist()\n",
    "\n",
    "print(len(m_genes_no_bound_orth_complete_3utr))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "7ab8f57e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "m_id_to_meta.shape=(48126, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>main_annot</th>\n",
       "      <th>annot_set</th>\n",
       "      <th>gene_id_set</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:6274997-6275006:+</th>\n",
       "      <td>three_prime_UTR;Rb1cc1;protein_coding;ENSMUSG0...</td>\n",
       "      <td>ENSMUSG00000025907</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{'three_prime_UTR'}</td>\n",
       "      <td>{'ENSMUSG00000025907'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:7170406-7170415:+</th>\n",
       "      <td>three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...</td>\n",
       "      <td>ENSMUSG00000051285</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{'three_prime_UTR'}</td>\n",
       "      <td>{'ENSMUSG00000051285'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:7170431-7170440:+</th>\n",
       "      <td>three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...</td>\n",
       "      <td>ENSMUSG00000051285</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{'three_prime_UTR'}</td>\n",
       "      <td>{'ENSMUSG00000051285'}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                     name  \\\n",
       "identifier                                                                  \n",
       "chr1:6274997-6275006:+  three_prime_UTR;Rb1cc1;protein_coding;ENSMUSG0...   \n",
       "chr1:7170406-7170415:+  three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...   \n",
       "chr1:7170431-7170440:+  three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...   \n",
       "\n",
       "                                   gene_id       main_annot  \\\n",
       "identifier                                                    \n",
       "chr1:6274997-6275006:+  ENSMUSG00000025907  three_prime_UTR   \n",
       "chr1:7170406-7170415:+  ENSMUSG00000051285  three_prime_UTR   \n",
       "chr1:7170431-7170440:+  ENSMUSG00000051285  three_prime_UTR   \n",
       "\n",
       "                                  annot_set             gene_id_set  \n",
       "identifier                                                           \n",
       "chr1:6274997-6275006:+  {'three_prime_UTR'}  {'ENSMUSG00000025907'}  \n",
       "chr1:7170406-7170415:+  {'three_prime_UTR'}  {'ENSMUSG00000051285'}  \n",
       "chr1:7170431-7170440:+  {'three_prime_UTR'}  {'ENSMUSG00000051285'}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "m_id_to_meta_3utr.shape=(30917, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>main_annot</th>\n",
       "      <th>annot_set</th>\n",
       "      <th>gene_id_set</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:6274997-6275006:+</th>\n",
       "      <td>three_prime_UTR;Rb1cc1;protein_coding;ENSMUSG0...</td>\n",
       "      <td>ENSMUSG00000025907</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{'three_prime_UTR'}</td>\n",
       "      <td>{'ENSMUSG00000025907'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:7170406-7170415:+</th>\n",
       "      <td>three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...</td>\n",
       "      <td>ENSMUSG00000051285</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{'three_prime_UTR'}</td>\n",
       "      <td>{'ENSMUSG00000051285'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:7170431-7170440:+</th>\n",
       "      <td>three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...</td>\n",
       "      <td>ENSMUSG00000051285</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{'three_prime_UTR'}</td>\n",
       "      <td>{'ENSMUSG00000051285'}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                     name  \\\n",
       "identifier                                                                  \n",
       "chr1:6274997-6275006:+  three_prime_UTR;Rb1cc1;protein_coding;ENSMUSG0...   \n",
       "chr1:7170406-7170415:+  three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...   \n",
       "chr1:7170431-7170440:+  three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...   \n",
       "\n",
       "                                   gene_id       main_annot  \\\n",
       "identifier                                                    \n",
       "chr1:6274997-6275006:+  ENSMUSG00000025907  three_prime_UTR   \n",
       "chr1:7170406-7170415:+  ENSMUSG00000051285  three_prime_UTR   \n",
       "chr1:7170431-7170440:+  ENSMUSG00000051285  three_prime_UTR   \n",
       "\n",
       "                                  annot_set             gene_id_set  \n",
       "identifier                                                           \n",
       "chr1:6274997-6275006:+  {'three_prime_UTR'}  {'ENSMUSG00000025907'}  \n",
       "chr1:7170406-7170415:+  {'three_prime_UTR'}  {'ENSMUSG00000051285'}  \n",
       "chr1:7170431-7170440:+  {'three_prime_UTR'}  {'ENSMUSG00000051285'}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fp = input_directory / \"m_id_to_meta.tsv.gz\"\n",
    "m_id_to_meta = read_mapping_peak_meta(fp)\n",
    "\n",
    "print(f\"{m_id_to_meta.shape=}\")\n",
    "display(m_id_to_meta.head(3))\n",
    "\n",
    "\n",
    "fp = input_directory / \"m_id_to_meta_3utr.tsv.gz\"\n",
    "m_id_to_meta_3utr = read_mapping_peak_meta(fp)\n",
    "\n",
    "print(f\"{m_id_to_meta_3utr.shape=}\")\n",
    "display(m_id_to_meta_3utr.head(3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "ed4cc687",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "m_id_to_bed6_3utr.shape=(31695, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:6274997-6275006:+</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6274997</td>\n",
       "      <td>6275006</td>\n",
       "      <td>three_prime_UTR;Rb1cc1;protein_coding;ENSMUSG0...</td>\n",
       "      <td>9.344450</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:7170406-7170415:+</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7170406</td>\n",
       "      <td>7170415</td>\n",
       "      <td>three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...</td>\n",
       "      <td>5.094060</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:7170431-7170440:+</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7170431</td>\n",
       "      <td>7170440</td>\n",
       "      <td>three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...</td>\n",
       "      <td>26.054101</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       chrom    start      end  \\\n",
       "identifier                                       \n",
       "chr1:6274997-6275006:+  chr1  6274997  6275006   \n",
       "chr1:7170406-7170415:+  chr1  7170406  7170415   \n",
       "chr1:7170431-7170440:+  chr1  7170431  7170440   \n",
       "\n",
       "                                                                     name  \\\n",
       "identifier                                                                  \n",
       "chr1:6274997-6275006:+  three_prime_UTR;Rb1cc1;protein_coding;ENSMUSG0...   \n",
       "chr1:7170406-7170415:+  three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...   \n",
       "chr1:7170431-7170440:+  three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...   \n",
       "\n",
       "                            score strand  \n",
       "identifier                                \n",
       "chr1:6274997-6275006:+   9.344450      +  \n",
       "chr1:7170406-7170415:+   5.094060      +  \n",
       "chr1:7170431-7170440:+  26.054101      +  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "m_id_to_bed6.shape=(49141, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:6274997-6275006:+</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6274997</td>\n",
       "      <td>6275006</td>\n",
       "      <td>three_prime_UTR;Rb1cc1;protein_coding;ENSMUSG0...</td>\n",
       "      <td>9.344450</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:7170406-7170415:+</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7170406</td>\n",
       "      <td>7170415</td>\n",
       "      <td>three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...</td>\n",
       "      <td>5.094060</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:7170431-7170440:+</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7170431</td>\n",
       "      <td>7170440</td>\n",
       "      <td>three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...</td>\n",
       "      <td>26.054101</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       chrom    start      end  \\\n",
       "identifier                                       \n",
       "chr1:6274997-6275006:+  chr1  6274997  6275006   \n",
       "chr1:7170406-7170415:+  chr1  7170406  7170415   \n",
       "chr1:7170431-7170440:+  chr1  7170431  7170440   \n",
       "\n",
       "                                                                     name  \\\n",
       "identifier                                                                  \n",
       "chr1:6274997-6275006:+  three_prime_UTR;Rb1cc1;protein_coding;ENSMUSG0...   \n",
       "chr1:7170406-7170415:+  three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...   \n",
       "chr1:7170431-7170440:+  three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...   \n",
       "\n",
       "                            score strand  \n",
       "identifier                                \n",
       "chr1:6274997-6275006:+   9.344450      +  \n",
       "chr1:7170406-7170415:+   5.094060      +  \n",
       "chr1:7170431-7170440:+  26.054101      +  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fp = input_directory / \"m_id_to_bed6_3utr.tsv.gz\"\n",
    "m_id_to_bed6_3utr = read_mapping_peak_bed6(fp)\n",
    "\n",
    "print(f\"{m_id_to_bed6_3utr.shape=}\")\n",
    "display(m_id_to_bed6_3utr.head(3))\n",
    "\n",
    "##\n",
    "\n",
    "fp = input_directory / \"m_id_to_bed6.tsv.gz\"\n",
    "m_id_to_bed6 = read_mapping_peak_bed6(fp)\n",
    "\n",
    "print(f\"{m_id_to_bed6.shape=}\")\n",
    "display(m_id_to_bed6.head(3))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "863c0b93",
   "metadata": {},
   "source": [
    "### Human"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0df1905f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Human peaks full shape: (33492, 7)\n",
      "Human peaks 3' UTR shape: (22338, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>identifier</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1211398</td>\n",
       "      <td>1211407</td>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>281.428788</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:1211398-1211407:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1211415</td>\n",
       "      <td>1211424</td>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>231.774900</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:1211415-1211424:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1211431</td>\n",
       "      <td>1211440</td>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>206.707600</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:1211431-1211440:-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom    start      end                                               name  \\\n",
       "0  chr1  1211398  1211407  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "1  chr1  1211415  1211424  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "2  chr1  1211431  1211440  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "\n",
       "        score strand              identifier  \n",
       "0  281.428788      -  chr1:1211398-1211407:-  \n",
       "1  231.774900      -  chr1:1211415-1211424:-  \n",
       "2  206.707600      -  chr1:1211431-1211440:-  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "h_full = pd.read_csv(filepaths[\"human_peaks_full\"], sep=\"\\t\", header=None, names=bed6_cols)\n",
    "h_full[\"identifier\"] = h_full.apply(build_identifier_bed6, axis=1)\n",
    "\n",
    "h_full_3utr = h_full.loc[lambda df: df[\"name\"].str.startswith(\"three_prime_UTR\"), :].copy()\n",
    "\n",
    "print(\"Human peaks full shape:\", h_full.shape)\n",
    "print(\"Human peaks 3' UTR shape:\", h_full_3utr.shape)\n",
    "display(h_full_3utr.head(3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "6e1926f2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "h_lift_3utr.shape=(20706, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier.human_source</th>\n",
       "      <th>identifier.mouse_target</th>\n",
       "      <th>orth_first_order</th>\n",
       "      <th>orth_second_order</th>\n",
       "      <th>orth_third_order.details</th>\n",
       "      <th>orth_third_order</th>\n",
       "      <th>gene_orthology_status</th>\n",
       "      <th>orthologous_gene_set</th>\n",
       "      <th>annotated_lifted_region_bound_gene_set</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:1211398-1211407:-</td>\n",
       "      <td>chr4:156016532-156016539:+</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSG00000186827, [ENSMUSG00000029075])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSMUSG00000029075}</td>\n",
       "      <td>{ENSMUSG00000029075}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:1211415-1211424:-</td>\n",
       "      <td>chr4:156016517-156016524:+</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSG00000186827, [ENSMUSG00000029075])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSMUSG00000029075}</td>\n",
       "      <td>{ENSMUSG00000029075}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1:1211431-1211440:-</td>\n",
       "      <td>chr4:156016501-156016510:+</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSG00000186827, [ENSMUSG00000029075])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSMUSG00000029075}</td>\n",
       "      <td>{ENSMUSG00000029075}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  identifier.human_source     identifier.mouse_target orth_first_order  \\\n",
       "0  chr1:1211398-1211407:-  chr4:156016532-156016539:+             True   \n",
       "1  chr1:1211415-1211424:-  chr4:156016517-156016524:+             True   \n",
       "2  chr1:1211431-1211440:-  chr4:156016501-156016510:+             True   \n",
       "\n",
       "  orth_second_order                   orth_third_order.details  \\\n",
       "0              True  [(ENSG00000186827, [ENSMUSG00000029075])]   \n",
       "1              True  [(ENSG00000186827, [ENSMUSG00000029075])]   \n",
       "2              True  [(ENSG00000186827, [ENSMUSG00000029075])]   \n",
       "\n",
       "  orth_third_order gene_orthology_status  orthologous_gene_set  \\\n",
       "0             True                 first  {ENSMUSG00000029075}   \n",
       "1             True                 first  {ENSMUSG00000029075}   \n",
       "2             True                 first  {ENSMUSG00000029075}   \n",
       "\n",
       "  annotated_lifted_region_bound_gene_set  \n",
       "0                   {ENSMUSG00000029075}  \n",
       "1                   {ENSMUSG00000029075}  \n",
       "2                   {ENSMUSG00000029075}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fp = input_directory / \"h_lift_3utr.tsv.gz\"\n",
    "h_lift_3utr = read_human_peaks_identifiers_with_gene_sets_orthologies(fp)\n",
    "h_lift_3utr = eval_peaks_identifiers_with_gene_set_orthologies(h_lift_3utr)\n",
    "\n",
    "print(f\"{h_lift_3utr.shape=}\")\n",
    "display(h_lift_3utr.head(3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "0e8ee5c9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "h_umap_3utr.shape=(1632, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>name</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:12207382-12207391:+</td>\n",
       "      <td>three_prime_UTR;TNFRSF1B;protein_coding;ENSG00...</td>\n",
       "      <td>ENSG00000028137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:23961916-23961925:+</td>\n",
       "      <td>three_prime_UTR;PNRC2;protein_coding;ENSG00000...</td>\n",
       "      <td>ENSG00000189266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1:23961935-23961944:+</td>\n",
       "      <td>three_prime_UTR;PNRC2;protein_coding;ENSG00000...</td>\n",
       "      <td>ENSG00000189266</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 identifier  \\\n",
       "0  chr1:12207382-12207391:+   \n",
       "1  chr1:23961916-23961925:+   \n",
       "2  chr1:23961935-23961944:+   \n",
       "\n",
       "                                                name          gene_id  \n",
       "0  three_prime_UTR;TNFRSF1B;protein_coding;ENSG00...  ENSG00000028137  \n",
       "1  three_prime_UTR;PNRC2;protein_coding;ENSG00000...  ENSG00000189266  \n",
       "2  three_prime_UTR;PNRC2;protein_coding;ENSG00000...  ENSG00000189266  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fp = input_directory / \"h_umap_3utr.tsv.gz\"\n",
    "h_umap_3utr = UnmappedPeakSchema.validate(pd.read_csv(fp, sep=\"\\t\"))\n",
    "\n",
    "print(f\"{h_umap_3utr.shape=}\")\n",
    "display(h_umap_3utr.head(3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "c98b03df",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1302\n",
      "1667\n",
      "1000\n",
      "1480\n"
     ]
    }
   ],
   "source": [
    "fp = input_directory / \"h_genes_bound_orth_complete.txt\"\n",
    "h_genes_bound_orth_complete = pd.read_csv(fp, sep=\"\\t\", header=None).iloc[:, 0].tolist()\n",
    "\n",
    "print(len(h_genes_bound_orth_complete))\n",
    "\n",
    "\n",
    "fp = input_directory / \"h_genes_no_bound_orth_complete.txt\"\n",
    "h_genes_no_bound_orth_complete = pd.read_csv(fp, sep=\"\\t\", header=None).iloc[:, 0].tolist()\n",
    "\n",
    "print(len(h_genes_no_bound_orth_complete))\n",
    "\n",
    "\n",
    "##\n",
    "\n",
    "\n",
    "fp = input_directory / \"h_genes_bound_orth_complete_3utr.txt\"\n",
    "h_genes_bound_orth_complete_3utr = pd.read_csv(fp, sep=\"\\t\", header=None).iloc[:, 0].tolist()\n",
    "\n",
    "print(len(h_genes_bound_orth_complete_3utr))\n",
    "\n",
    "\n",
    "fp = input_directory / \"h_genes_no_bound_orth_complete_3utr.txt\"\n",
    "h_genes_no_bound_orth_complete_3utr = pd.read_csv(fp, sep=\"\\t\", header=None).iloc[:, 0].tolist()\n",
    "\n",
    "print(len(h_genes_no_bound_orth_complete_3utr))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "42328dbd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "h_id_to_meta.shape=(49904, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>main_annot</th>\n",
       "      <th>annot_set</th>\n",
       "      <th>gene_id_set</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:1211398-1211407:-</th>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>ENSG00000186827</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{'three_prime_UTR'}</td>\n",
       "      <td>{'ENSG00000186827'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:1211415-1211424:-</th>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>ENSG00000186827</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{'three_prime_UTR'}</td>\n",
       "      <td>{'ENSG00000186827'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:1211431-1211440:-</th>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>ENSG00000186827</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{'three_prime_UTR'}</td>\n",
       "      <td>{'ENSG00000186827'}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                     name  \\\n",
       "identifier                                                                  \n",
       "chr1:1211398-1211407:-  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "chr1:1211415-1211424:-  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "chr1:1211431-1211440:-  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "\n",
       "                                gene_id       main_annot            annot_set  \\\n",
       "identifier                                                                      \n",
       "chr1:1211398-1211407:-  ENSG00000186827  three_prime_UTR  {'three_prime_UTR'}   \n",
       "chr1:1211415-1211424:-  ENSG00000186827  three_prime_UTR  {'three_prime_UTR'}   \n",
       "chr1:1211431-1211440:-  ENSG00000186827  three_prime_UTR  {'three_prime_UTR'}   \n",
       "\n",
       "                                gene_id_set  \n",
       "identifier                                   \n",
       "chr1:1211398-1211407:-  {'ENSG00000186827'}  \n",
       "chr1:1211415-1211424:-  {'ENSG00000186827'}  \n",
       "chr1:1211431-1211440:-  {'ENSG00000186827'}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "h_id_to_meta_3utr.shape=(33847, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>main_annot</th>\n",
       "      <th>annot_set</th>\n",
       "      <th>gene_id_set</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:1211398-1211407:-</th>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>ENSG00000186827</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{'three_prime_UTR'}</td>\n",
       "      <td>{'ENSG00000186827'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:1211415-1211424:-</th>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>ENSG00000186827</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{'three_prime_UTR'}</td>\n",
       "      <td>{'ENSG00000186827'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:1211431-1211440:-</th>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>ENSG00000186827</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{'three_prime_UTR'}</td>\n",
       "      <td>{'ENSG00000186827'}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                     name  \\\n",
       "identifier                                                                  \n",
       "chr1:1211398-1211407:-  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "chr1:1211415-1211424:-  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "chr1:1211431-1211440:-  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "\n",
       "                                gene_id       main_annot            annot_set  \\\n",
       "identifier                                                                      \n",
       "chr1:1211398-1211407:-  ENSG00000186827  three_prime_UTR  {'three_prime_UTR'}   \n",
       "chr1:1211415-1211424:-  ENSG00000186827  three_prime_UTR  {'three_prime_UTR'}   \n",
       "chr1:1211431-1211440:-  ENSG00000186827  three_prime_UTR  {'three_prime_UTR'}   \n",
       "\n",
       "                                gene_id_set  \n",
       "identifier                                   \n",
       "chr1:1211398-1211407:-  {'ENSG00000186827'}  \n",
       "chr1:1211415-1211424:-  {'ENSG00000186827'}  \n",
       "chr1:1211431-1211440:-  {'ENSG00000186827'}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fp = input_directory / \"h_id_to_meta.tsv.gz\"\n",
    "h_id_to_meta = read_mapping_peak_meta(fp)\n",
    "\n",
    "print(f\"{h_id_to_meta.shape=}\")\n",
    "display(h_id_to_meta.head(3))\n",
    "\n",
    "##\n",
    "\n",
    "fp = input_directory / \"h_id_to_meta_3utr.tsv.gz\"\n",
    "h_id_to_meta_3utr = read_mapping_peak_meta(fp)\n",
    "\n",
    "print(f\"{h_id_to_meta_3utr.shape=}\")\n",
    "display(h_id_to_meta_3utr.head(3))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a4c6c42",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "h_id_to_bed6_3utr.shape=(34652, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:1211398-1211407:-</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1211398</td>\n",
       "      <td>1211407</td>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>281.428788</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:1211415-1211424:-</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1211415</td>\n",
       "      <td>1211424</td>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>231.774900</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:1211431-1211440:-</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1211431</td>\n",
       "      <td>1211440</td>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>206.707600</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       chrom    start      end  \\\n",
       "identifier                                       \n",
       "chr1:1211398-1211407:-  chr1  1211398  1211407   \n",
       "chr1:1211415-1211424:-  chr1  1211415  1211424   \n",
       "chr1:1211431-1211440:-  chr1  1211431  1211440   \n",
       "\n",
       "                                                                     name  \\\n",
       "identifier                                                                  \n",
       "chr1:1211398-1211407:-  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "chr1:1211415-1211424:-  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "chr1:1211431-1211440:-  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "\n",
       "                             score strand  \n",
       "identifier                                 \n",
       "chr1:1211398-1211407:-  281.428788      -  \n",
       "chr1:1211415-1211424:-  231.774900      -  \n",
       "chr1:1211431-1211440:-  206.707600      -  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "h_id_to_bed6.shape=(50917, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:1211398-1211407:-</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1211398</td>\n",
       "      <td>1211407</td>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>281.428788</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:1211415-1211424:-</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1211415</td>\n",
       "      <td>1211424</td>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>231.774900</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:1211431-1211440:-</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1211431</td>\n",
       "      <td>1211440</td>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>206.707600</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       chrom    start      end  \\\n",
       "identifier                                       \n",
       "chr1:1211398-1211407:-  chr1  1211398  1211407   \n",
       "chr1:1211415-1211424:-  chr1  1211415  1211424   \n",
       "chr1:1211431-1211440:-  chr1  1211431  1211440   \n",
       "\n",
       "                                                                     name  \\\n",
       "identifier                                                                  \n",
       "chr1:1211398-1211407:-  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "chr1:1211415-1211424:-  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "chr1:1211431-1211440:-  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "\n",
       "                             score strand  \n",
       "identifier                                 \n",
       "chr1:1211398-1211407:-  281.428788      -  \n",
       "chr1:1211415-1211424:-  231.774900      -  \n",
       "chr1:1211431-1211440:-  206.707600      -  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fp = input_directory / \"h_id_to_bed6_3utr.tsv.gz\"\n",
    "h_id_to_bed6_3utr = read_mapping_peak_bed6(fp)\n",
    "\n",
    "print(f\"{h_id_to_bed6_3utr.shape=}\")\n",
    "display(h_id_to_bed6_3utr.head(3))\n",
    "\n",
    "\n",
    "##\n",
    "\n",
    "\n",
    "fp = input_directory / \"h_id_to_bed6.tsv.gz\"\n",
    "h_id_to_bed6 = read_mapping_peak_bed6(fp)\n",
    "\n",
    "print(f\"{h_id_to_bed6.shape=}\")\n",
    "display(h_id_to_bed6.head(3))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8af43cf9",
   "metadata": {},
   "source": [
    "## Loaded data processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "b1a63164",
   "metadata": {},
   "outputs": [],
   "source": [
    "m_lift = (\n",
    "    DATA.mouse.peaks.lifted.loc[:, [\"source_identifier\", \"identifier\"]]\n",
    "    .copy()\n",
    "    .rename(columns={\"source_identifier\": \"identifier.mouse_source\", \"identifier\": \"identifier.human_target\"})\n",
    ")\n",
    "\n",
    "h_lift = (\n",
    "    DATA.human.peaks.lifted.loc[:, [\"source_identifier\", \"identifier\"]]\n",
    "    .copy()\n",
    "    .rename(columns={\"source_identifier\": \"identifier.human_source\", \"identifier\": \"identifier.mouse_target\"})\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e0ff1f9",
   "metadata": {},
   "source": [
    "## Peaks orthologies"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "183ca2d3",
   "metadata": {},
   "source": [
    "Here we want to count the number of **peaks** after assigning them to the following sets:\n",
    "\n",
    "- **mouse_specific_unmapped** : the gene is bound only in mouse, with peak(s) that cannot be mapped in human.\n",
    "- **mouse_specific_specifically_bound** : the gene is bound in mouse, with peak(s) that can be mapped in human, but there: no peaks in the 3'UTR.\n",
    "- **mouse_common_unmapped** : the gene is bound in both species, with peak(s) that cannot be mapped in human.\n",
    "- **mouse_common_wo_ortholog** : the gene is bound in both species, with peak(s) that can be mapped in human, but the peak(s) are not in genomic-distance proximity.\n",
    "- **mouse_common_with_ortholog**  : the gene is bound in both species, with peak(s) that can be mapped in human, and the peak(s) are in genomic-distance proximity.\n",
    "\n",
    "With the same categories for human."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4d5ad0d",
   "metadata": {},
   "source": [
    "### Gather tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "2d9e5664",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ENSMUSG00000018143', 'ENSMUSG00000031924']"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "['ENSMUSG00000037958', 'ENSMUSG00000027177']"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "['ENSG00000137642', 'ENSG00000188404']"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "['ENSG00000169813', 'ENSG00000213741']"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "PEAKS_m_lift_3utr = m_lift_3utr.copy()\n",
    "PEAKS_m_umap_3utr = m_umap_3utr.copy()\n",
    "\n",
    "PEAKS_h_lift_3utr = h_lift_3utr.copy()\n",
    "PEAKS_h_umap_3utr = h_umap_3utr.copy()\n",
    "\n",
    "\n",
    "display(list(m_genes_bound_orth_complete_3utr)[:2])\n",
    "display(list(m_genes_no_bound_orth_complete_3utr)[:2])\n",
    "display(list(h_genes_bound_orth_complete_3utr)[:2])\n",
    "display(list(h_genes_no_bound_orth_complete_3utr)[:2])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef2eb00c",
   "metadata": {},
   "source": [
    "### Annotate peaks on the gene orthologous binding status"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "cd85a76b",
   "metadata": {},
   "outputs": [],
   "source": [
    "PEAKS_m_lift_3utr = pd.merge(\n",
    "    PEAKS_m_lift_3utr,\n",
    "    m_id_to_meta_3utr,\n",
    "    left_on=\"identifier.mouse_source\",\n",
    "    right_index=True,\n",
    "    how=\"left\",\n",
    ")\n",
    "\n",
    "PEAKS_m_lift_3utr[\"is_gene_bound_in_common\"] = PEAKS_m_lift_3utr[\"gene_id\"].isin(m_genes_bound_orth_complete_3utr)\n",
    "PEAKS_m_umap_3utr[\"is_gene_bound_in_common\"] = PEAKS_m_umap_3utr[\"gene_id\"].isin(m_genes_bound_orth_complete_3utr)\n",
    "\n",
    "##\n",
    "\n",
    "PEAKS_h_lift_3utr = pd.merge(\n",
    "    PEAKS_h_lift_3utr,\n",
    "    h_id_to_meta_3utr,\n",
    "    left_on=\"identifier.human_source\",\n",
    "    right_index=True,\n",
    "    how=\"left\",\n",
    ")\n",
    "\n",
    "PEAKS_h_lift_3utr[\"is_gene_bound_in_common\"] = PEAKS_h_lift_3utr[\"gene_id\"].isin(h_genes_bound_orth_complete_3utr)\n",
    "PEAKS_h_umap_3utr[\"is_gene_bound_in_common\"] = PEAKS_h_umap_3utr[\"gene_id\"].isin(h_genes_bound_orth_complete_3utr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1547383",
   "metadata": {},
   "source": [
    "### Annotate peaks on the peak-level orthologous binding status"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbb25223",
   "metadata": {},
   "source": [
    "Here we use the lift-over peaks to check whether there is a peak in the target species in proximity."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "4223ddc8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9.0\n",
      "9.0\n"
     ]
    }
   ],
   "source": [
    "print(\n",
    "    m_id_to_bed6_3utr.loc[PEAKS_m_lift_3utr[\"identifier.mouse_source\"], :]\n",
    "    .apply(lambda row: row[\"end\"] - row[\"start\"], axis=1)\n",
    "    .mean()\n",
    ")\n",
    "print(\n",
    "    h_id_to_bed6_3utr.loc[PEAKS_h_lift_3utr[\"identifier.human_source\"], :]\n",
    "    .apply(lambda row: row[\"end\"] - row[\"start\"], axis=1)\n",
    "    .mean()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c9b08f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "EXTEND = (\n",
    "    10  # Peaks are 9-nt ; after lift-over, we check whether there is a peak upstream or downstream up to 10-nt away.\n",
    ")\n",
    "FRACTION = 0.9  # We still want 90% of the target-species intervals to overlap.\n",
    "intersect_columns = [f\"A.{c}\" for c in bed6_cols] + [f\"B.{c}\" for c in bed6_cols] + [\"ol\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e25673e",
   "metadata": {},
   "source": [
    "#### Mouse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "8b1b70b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    8720\n",
       "True     3160\n",
       "Name: is_peak_closeby, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# MOUSE PEAKS: USE HUMAN-LIFTED COORDINATES AND COMPARE TO 3'UTR HUMAN PEAKS\n",
    "\n",
    "bed6_lift_mouse_peaks_3utr = pd.merge(\n",
    "    PEAKS_m_lift_3utr.loc[:, [\"identifier.mouse_source\", \"identifier.human_target\"]],\n",
    "    h_id_to_bed6.reset_index(),\n",
    "    left_on=\"identifier.human_target\",\n",
    "    right_on=\"identifier\",\n",
    "    how=\"left\",\n",
    ").loc[:, [\"chrom\", \"start\", \"end\", \"identifier.mouse_source\", \"score\", \"strand\"]]\n",
    "\n",
    "bt_mouse_lift_peaks_3utr = pbt.BedTool.from_dataframe(bed6_lift_mouse_peaks_3utr)\n",
    "\n",
    "# Caution: we have to index on the human-peaks only, as the `hh_id_to_bed6_3utr` structure\n",
    "# contains info for both human peaks and mouse-lifted-to-human peaks.\n",
    "bt_human_peaks_3utr = pbt.BedTool.from_dataframe(\n",
    "    h_id_to_bed6_3utr.loc[h_full_3utr[\"identifier\"].values, :]\n",
    "    .reset_index()\n",
    "    .loc[:, [\"chrom\", \"start\", \"end\", \"identifier\", \"score\", \"strand\"]]\n",
    ")\n",
    "\n",
    "# Extend and intersect\n",
    "m_intersect_full = (\n",
    "    bt_mouse_lift_peaks_3utr.slop(b=EXTEND, g=DATA.human.chromsizes_fp)\n",
    "    .intersect(\n",
    "        bt_human_peaks_3utr,\n",
    "        wao=True,\n",
    "        s=True,\n",
    "        F=FRACTION,\n",
    "    )\n",
    "    .to_dataframe(names=intersect_columns)\n",
    ")\n",
    "\n",
    "PEAKS_m_lift_3utr[\"is_peak_closeby\"] = PEAKS_m_lift_3utr[\"identifier.mouse_source\"].isin(\n",
    "    m_intersect_full.loc[lambda df: df[\"ol\"] > 0, \"A.name\"].unique()\n",
    ")\n",
    "\n",
    "display(PEAKS_m_lift_3utr[\"is_peak_closeby\"].value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52d7afd8",
   "metadata": {},
   "source": [
    "#### Human"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "a1920a4c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    17464\n",
       "True      3242\n",
       "Name: is_peak_closeby, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# HUMAN PEAKS: USE MOUSE-LIFTED COORDINATES AND COMPARE TO 3'UTR MOUSE PEAKS\n",
    "\n",
    "bed6_lift_human_peaks_3utr = pd.merge(\n",
    "    PEAKS_h_lift_3utr.loc[:, [\"identifier.human_source\", \"identifier.mouse_target\"]],\n",
    "    m_id_to_bed6.reset_index(),\n",
    "    left_on=\"identifier.mouse_target\",\n",
    "    right_on=\"identifier\",\n",
    "    how=\"left\",\n",
    ").loc[:, [\"chrom\", \"start\", \"end\", \"identifier.human_source\", \"score\", \"strand\"]]\n",
    "\n",
    "bt_human_lift_peaks_3utr = pbt.BedTool.from_dataframe(bed6_lift_human_peaks_3utr)\n",
    "\n",
    "# Caution: we have to index on the human-peaks only, as the `hh_id_to_bed6_3utr` structure\n",
    "# contains info for both human peaks and mouse-lifted-to-human peaks.\n",
    "bt_mouse_peaks_3utr = pbt.BedTool.from_dataframe(\n",
    "    m_id_to_bed6_3utr.loc[m_full_3utr[\"identifier\"].values, :]\n",
    "    .reset_index()\n",
    "    .loc[:, [\"chrom\", \"start\", \"end\", \"identifier\", \"score\", \"strand\"]]\n",
    ")\n",
    "\n",
    "# Extend and intersect\n",
    "h_intersect_full = (\n",
    "    bt_human_lift_peaks_3utr.slop(b=EXTEND, g=DATA.mouse.chromsizes_fp)\n",
    "    .intersect(\n",
    "        bt_mouse_peaks_3utr,\n",
    "        wao=True,\n",
    "        s=True,\n",
    "        F=FRACTION,\n",
    "    )\n",
    "    .to_dataframe(names=intersect_columns)\n",
    ")\n",
    "\n",
    "PEAKS_h_lift_3utr[\"is_peak_closeby\"] = PEAKS_h_lift_3utr[\"identifier.human_source\"].isin(\n",
    "    h_intersect_full.loc[lambda df: df[\"ol\"] > 0, \"A.name\"].unique()\n",
    ")\n",
    "\n",
    "display(PEAKS_h_lift_3utr[\"is_peak_closeby\"].value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c20fce3b",
   "metadata": {},
   "source": [
    "### Assign final status of peaks - mouse"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c84061c7",
   "metadata": {},
   "source": [
    "#### Expected cases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55372082",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(False, False)\n",
      "(True, False)\n",
      "(True, True)\n"
     ]
    }
   ],
   "source": [
    "m_peaks_orthology_label_3utr = []\n",
    "\n",
    "m_peaks_orthology_label_3utr.append(\n",
    "    PEAKS_m_umap_3utr.loc[lambda df: (df[\"is_gene_bound_in_common\"]), [\"identifier\"]].assign(\n",
    "        status=\"not-aligned_bound-ortholog\"\n",
    "    )\n",
    ")\n",
    "\n",
    "m_peaks_orthology_label_3utr.append(\n",
    "    PEAKS_m_umap_3utr.loc[lambda df: (~df[\"is_gene_bound_in_common\"]), [\"identifier\"]].assign(\n",
    "        status=\"not-aligned_species-specific\"\n",
    "    )\n",
    ")\n",
    "\n",
    "# Unexpected case: (False, True), i.e. the peak location is orthologously bound, but the annotated genes were not identified\n",
    "# as orthologous to one another. We check these in the next section.\n",
    "bool_to_label = {\n",
    "    (False, False): \"aligned_species-specific\",\n",
    "    (True, False): \"aligned_bound-ortholog_distant\",\n",
    "    (True, True): \"aligned_bound-ortholog_matched\",\n",
    "}\n",
    "\n",
    "grouped_PEAKS_m_lift_3utr = PEAKS_m_lift_3utr.groupby([\"is_gene_bound_in_common\", \"is_peak_closeby\"])\n",
    "\n",
    "for group_id, group_label in bool_to_label.items():\n",
    "    print(group_id)\n",
    "    m_peaks_orthology_label_3utr.append(\n",
    "        grouped_PEAKS_m_lift_3utr.get_group(group_id)\n",
    "        .loc[:, [\"identifier.mouse_source\"]]\n",
    "        .assign(status=group_label)\n",
    "        .rename(columns={\"identifier.mouse_source\": \"identifier\"})\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bef3a248",
   "metadata": {},
   "source": [
    "#### Unexpected group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a5d8b5c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier.mouse_source</th>\n",
       "      <th>identifier.human_target</th>\n",
       "      <th>orth_first_order</th>\n",
       "      <th>orth_second_order</th>\n",
       "      <th>orth_third_order.details</th>\n",
       "      <th>orth_third_order</th>\n",
       "      <th>gene_orthology_status</th>\n",
       "      <th>orthologous_gene_set</th>\n",
       "      <th>annotated_lifted_region_bound_gene_set</th>\n",
       "      <th>name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>main_annot</th>\n",
       "      <th>annot_set</th>\n",
       "      <th>gene_id_set</th>\n",
       "      <th>is_gene_bound_in_common</th>\n",
       "      <th>is_peak_closeby</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7642</th>\n",
       "      <td>chr6:48678394-48678403:+</td>\n",
       "      <td>chr7:150520911-150520916:+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(ENSMUSG00000051124, None)]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not_orthologous</td>\n",
       "      <td>{}</td>\n",
       "      <td>{ENSG00000179144}</td>\n",
       "      <td>three_prime_UTR;Gimap9;protein_coding;ENSMUSG0...</td>\n",
       "      <td>ENSMUSG00000051124</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{'three_prime_UTR'}</td>\n",
       "      <td>{'ENSMUSG00000051124'}</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       identifier.mouse_source     identifier.human_target orth_first_order  \\\n",
       "7642  chr6:48678394-48678403:+  chr7:150520911-150520916:+              NaN   \n",
       "\n",
       "     orth_second_order      orth_third_order.details orth_third_order  \\\n",
       "7642               NaN  [(ENSMUSG00000051124, None)]              NaN   \n",
       "\n",
       "     gene_orthology_status orthologous_gene_set  \\\n",
       "7642       not_orthologous                   {}   \n",
       "\n",
       "     annotated_lifted_region_bound_gene_set  \\\n",
       "7642                      {ENSG00000179144}   \n",
       "\n",
       "                                                   name             gene_id  \\\n",
       "7642  three_prime_UTR;Gimap9;protein_coding;ENSMUSG0...  ENSMUSG00000051124   \n",
       "\n",
       "           main_annot            annot_set             gene_id_set  \\\n",
       "7642  three_prime_UTR  {'three_prime_UTR'}  {'ENSMUSG00000051124'}   \n",
       "\n",
       "      is_gene_bound_in_common  is_peak_closeby  \n",
       "7642                    False             True  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A.name</th>\n",
       "      <th>B.name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9550</th>\n",
       "      <td>chr6:48678394-48678403:+</td>\n",
       "      <td>chr7:150520917-150520926:+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        A.name                      B.name\n",
       "9550  chr6:48678394-48678403:+  chr7:150520917-150520926:+"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>main_annot</th>\n",
       "      <th>annot_set</th>\n",
       "      <th>gene_id_set</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr6:48678394-48678403:+</th>\n",
       "      <td>three_prime_UTR;Gimap9;protein_coding;ENSMUSG0...</td>\n",
       "      <td>ENSMUSG00000051124</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{'three_prime_UTR'}</td>\n",
       "      <td>{'ENSMUSG00000051124'}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                       name  \\\n",
       "identifier                                                                    \n",
       "chr6:48678394-48678403:+  three_prime_UTR;Gimap9;protein_coding;ENSMUSG0...   \n",
       "\n",
       "                                     gene_id       main_annot  \\\n",
       "identifier                                                      \n",
       "chr6:48678394-48678403:+  ENSMUSG00000051124  three_prime_UTR   \n",
       "\n",
       "                                    annot_set             gene_id_set  \n",
       "identifier                                                             \n",
       "chr6:48678394-48678403:+  {'three_prime_UTR'}  {'ENSMUSG00000051124'}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>main_annot</th>\n",
       "      <th>annot_set</th>\n",
       "      <th>gene_id_set</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr7:150520917-150520926:+</th>\n",
       "      <td>three_prime_UTR;GIMAP7;protein_coding;ENSG0000...</td>\n",
       "      <td>ENSG00000179144</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{'three_prime_UTR'}</td>\n",
       "      <td>{'ENSG00000179144'}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                         name  \\\n",
       "identifier                                                                      \n",
       "chr7:150520917-150520926:+  three_prime_UTR;GIMAP7;protein_coding;ENSG0000...   \n",
       "\n",
       "                                    gene_id       main_annot  \\\n",
       "identifier                                                     \n",
       "chr7:150520917-150520926:+  ENSG00000179144  three_prime_UTR   \n",
       "\n",
       "                                      annot_set          gene_id_set  \n",
       "identifier                                                            \n",
       "chr7:150520917-150520926:+  {'three_prime_UTR'}  {'ENSG00000179144'}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print(grouped_PEAKS_m_lift_3utr.get_group((False, True)).shape[0])\n",
    "\n",
    "assert grouped_PEAKS_m_lift_3utr.get_group((False, True)).shape[0] == 1\n",
    "\n",
    "# Expected 1 ; detailed below.\n",
    "\n",
    "display(grouped_PEAKS_m_lift_3utr.get_group((False, True)))\n",
    "\n",
    "# Mouse: chr6:48678394-48678403:+ ; lifted to chr7:150520911-150520916:+\n",
    "#\n",
    "\n",
    "display(m_intersect_full.loc[lambda df: df[\"A.name\"] == \"chr6:48678394-48678403:+\", [\"A.name\", \"B.name\"]])\n",
    "\n",
    "# Human peak closeby: chr7:150520917-150520926:+\n",
    "\n",
    "display(m_id_to_meta_3utr.loc[[\"chr6:48678394-48678403:+\"], :])\n",
    "display(h_id_to_meta_3utr.loc[[\"chr7:150520917-150520926:+\"], :])\n",
    "\n",
    "# This peak is annotated in the 3'UTR of 'ENSMUSG00000051124' = GIMAP7 in human\n",
    "# While the mouse peak is annotated in 'ENSMUSG00000051124' = Gimap9.\n",
    "\n",
    "# => Exploring synteny and orthology of these genes and their neighbours in Genomicus :\n",
    "# this seems to be a cluster of duplications of the Gimap family. Orthologies might be difficult to establish.\n",
    "# As per our orthology resources, we only have the Gimap7 / GIMAP7 estalished orthology.\n",
    "# So the two peaks, despite mapping at the same location, are likely associated to different duplications of\n",
    "# the ancestral gene.\n",
    "\n",
    "# DECISION => Add the locus to the \"aligned_species-specific\" category.\n",
    "\n",
    "m_peaks_orthology_label_3utr.append(\n",
    "    pd.DataFrame([pd.Series({\"identifier\": \"chr6:48678394-48678403:+\", \"status\": \"aligned_species-specific\"})])\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a3fe509",
   "metadata": {},
   "source": [
    "#### Final table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82d700fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "m_peaks_orthology_label_3utr = pd.concat(m_peaks_orthology_label_3utr).reset_index(drop=True)\n",
    "assert m_peaks_orthology_label_3utr[\"identifier\"].duplicated().sum() == 0\n",
    "assert set(m_peaks_orthology_label_3utr[\"identifier\"].unique()) == set(m_full_3utr[\"identifier\"].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "741a9150",
   "metadata": {},
   "source": [
    "### Assign final status of peaks - human"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32dee4fd",
   "metadata": {},
   "source": [
    "#### Expected cases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "faa347bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(False, False)\n",
      "(True, False)\n",
      "(True, True)\n"
     ]
    }
   ],
   "source": [
    "h_peaks_orthology_label_3utr = []\n",
    "\n",
    "h_peaks_orthology_label_3utr.append(\n",
    "    PEAKS_h_umap_3utr.loc[lambda df: (df[\"is_gene_bound_in_common\"]), [\"identifier\"]].assign(\n",
    "        status=\"not-aligned_bound-ortholog\"\n",
    "    )\n",
    ")\n",
    "\n",
    "h_peaks_orthology_label_3utr.append(\n",
    "    PEAKS_h_umap_3utr.loc[lambda df: (~df[\"is_gene_bound_in_common\"]), [\"identifier\"]].assign(\n",
    "        status=\"not-aligned_species-specific\"\n",
    "    )\n",
    ")\n",
    "\n",
    "grouped_PEAKS_h_lift_3utr = PEAKS_h_lift_3utr.groupby([\"is_gene_bound_in_common\", \"is_peak_closeby\"])\n",
    "\n",
    "for group_id, group_label in bool_to_label.items():\n",
    "    print(group_id)\n",
    "    h_peaks_orthology_label_3utr.append(\n",
    "        grouped_PEAKS_h_lift_3utr.get_group(group_id)\n",
    "        .loc[:, [\"identifier.human_source\"]]\n",
    "        .assign(status=group_label)\n",
    "        .rename(columns={\"identifier.human_source\": \"identifier\"})\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e1132fe",
   "metadata": {},
   "source": [
    "#### Unexpected group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "0d165d92",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier.human_source</th>\n",
       "      <th>identifier.mouse_target</th>\n",
       "      <th>orth_first_order</th>\n",
       "      <th>orth_second_order</th>\n",
       "      <th>orth_third_order.details</th>\n",
       "      <th>orth_third_order</th>\n",
       "      <th>gene_orthology_status</th>\n",
       "      <th>orthologous_gene_set</th>\n",
       "      <th>annotated_lifted_region_bound_gene_set</th>\n",
       "      <th>name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>main_annot</th>\n",
       "      <th>annot_set</th>\n",
       "      <th>gene_id_set</th>\n",
       "      <th>is_gene_bound_in_common</th>\n",
       "      <th>is_peak_closeby</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14847</th>\n",
       "      <td>chr6:26270966-26270975:-</td>\n",
       "      <td>chr13:23535854-23535863:+</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSG00000273983, [ENSMUSG00000099517])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSMUSG00000099517}</td>\n",
       "      <td>{}</td>\n",
       "      <td>three_prime_UTR;H3C8;protein_coding;ENSG000002...</td>\n",
       "      <td>ENSG00000273983</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{'three_prime_UTR'}</td>\n",
       "      <td>{'ENSG00000273983'}</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        identifier.human_source    identifier.mouse_target orth_first_order  \\\n",
       "14847  chr6:26270966-26270975:-  chr13:23535854-23535863:+             True   \n",
       "\n",
       "      orth_second_order                   orth_third_order.details  \\\n",
       "14847              True  [(ENSG00000273983, [ENSMUSG00000099517])]   \n",
       "\n",
       "      orth_third_order gene_orthology_status  orthologous_gene_set  \\\n",
       "14847             True                 first  {ENSMUSG00000099517}   \n",
       "\n",
       "      annotated_lifted_region_bound_gene_set  \\\n",
       "14847                                     {}   \n",
       "\n",
       "                                                    name          gene_id  \\\n",
       "14847  three_prime_UTR;H3C8;protein_coding;ENSG000002...  ENSG00000273983   \n",
       "\n",
       "            main_annot            annot_set          gene_id_set  \\\n",
       "14847  three_prime_UTR  {'three_prime_UTR'}  {'ENSG00000273983'}   \n",
       "\n",
       "       is_gene_bound_in_common  is_peak_closeby  \n",
       "14847                    False             True  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A.name</th>\n",
       "      <th>B.name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16846</th>\n",
       "      <td>chr6:26270966-26270975:-</td>\n",
       "      <td>chr13:23535860-23535869:+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         A.name                     B.name\n",
       "16846  chr6:26270966-26270975:-  chr13:23535860-23535869:+"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>main_annot</th>\n",
       "      <th>annot_set</th>\n",
       "      <th>gene_id_set</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr13:23535860-23535869:+</th>\n",
       "      <td>three_prime_UTR;H3c8;protein_coding;ENSMUSG000...</td>\n",
       "      <td>ENSMUSG00000099517</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{'three_prime_UTR'}</td>\n",
       "      <td>{'ENSMUSG00000099517'}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                        name  \\\n",
       "identifier                                                                     \n",
       "chr13:23535860-23535869:+  three_prime_UTR;H3c8;protein_coding;ENSMUSG000...   \n",
       "\n",
       "                                      gene_id       main_annot  \\\n",
       "identifier                                                       \n",
       "chr13:23535860-23535869:+  ENSMUSG00000099517  three_prime_UTR   \n",
       "\n",
       "                                     annot_set             gene_id_set  \n",
       "identifier                                                              \n",
       "chr13:23535860-23535869:+  {'three_prime_UTR'}  {'ENSMUSG00000099517'}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>main_annot</th>\n",
       "      <th>annot_set</th>\n",
       "      <th>gene_id_set</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr6:26270966-26270975:-</th>\n",
       "      <td>three_prime_UTR;H3C8;protein_coding;ENSG000002...</td>\n",
       "      <td>ENSG00000273983</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{'three_prime_UTR'}</td>\n",
       "      <td>{'ENSG00000273983'}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                       name  \\\n",
       "identifier                                                                    \n",
       "chr6:26270966-26270975:-  three_prime_UTR;H3C8;protein_coding;ENSG000002...   \n",
       "\n",
       "                                  gene_id       main_annot  \\\n",
       "identifier                                                   \n",
       "chr6:26270966-26270975:-  ENSG00000273983  three_prime_UTR   \n",
       "\n",
       "                                    annot_set          gene_id_set  \n",
       "identifier                                                          \n",
       "chr6:26270966-26270975:-  {'three_prime_UTR'}  {'ENSG00000273983'}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>human_ensid</th>\n",
       "      <th>human_name</th>\n",
       "      <th>mouse_ensid</th>\n",
       "      <th>mouse_name</th>\n",
       "      <th>relationtype_id</th>\n",
       "      <th>supported_by_resource</th>\n",
       "      <th>support_level</th>\n",
       "      <th>support_details</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>30096</th>\n",
       "      <td>ENSG00000273983</td>\n",
       "      <td>H3C8</td>\n",
       "      <td>ENSMUSG00000099517</td>\n",
       "      <td>H3c8</td>\n",
       "      <td>n:n</td>\n",
       "      <td>omabrowser;;mgi</td>\n",
       "      <td>3</td>\n",
       "      <td>omabrowser:human_to_mouse;mouse_to_human;;mgi:...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           human_ensid human_name         mouse_ensid mouse_name  \\\n",
       "30096  ENSG00000273983       H3C8  ENSMUSG00000099517       H3c8   \n",
       "\n",
       "      relationtype_id supported_by_resource  support_level  \\\n",
       "30096             n:n       omabrowser;;mgi              3   \n",
       "\n",
       "                                         support_details  \n",
       "30096  omabrowser:human_to_mouse;mouse_to_human;;mgi:...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "False\n",
      "True\n",
      "True\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array(['stop_codon;H3c8;protein_coding;ENSMUSG00000099517.2'],\n",
       "      dtype=object)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>identifier</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>main_annot</th>\n",
       "      <th>annot_set</th>\n",
       "      <th>gene_id_set</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23628</th>\n",
       "      <td>chr6</td>\n",
       "      <td>26270966</td>\n",
       "      <td>26270975</td>\n",
       "      <td>three_prime_UTR;H3C8;protein_coding;ENSG000002...</td>\n",
       "      <td>149.55341</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6:26270966-26270975:-</td>\n",
       "      <td>ENSG00000273983</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{three_prime_UTR}</td>\n",
       "      <td>{ENSG00000273983}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom     start       end  \\\n",
       "23628  chr6  26270966  26270975   \n",
       "\n",
       "                                                    name      score strand  \\\n",
       "23628  three_prime_UTR;H3C8;protein_coding;ENSG000002...  149.55341      -   \n",
       "\n",
       "                     identifier          gene_id       main_annot  \\\n",
       "23628  chr6:26270966-26270975:-  ENSG00000273983  three_prime_UTR   \n",
       "\n",
       "               annot_set        gene_id_set  \n",
       "23628  {three_prime_UTR}  {ENSG00000273983}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>source_identifier</th>\n",
       "      <th>identifier</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>main_annot</th>\n",
       "      <th>annot_set</th>\n",
       "      <th>gene_id_set</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20497</th>\n",
       "      <td>chr13</td>\n",
       "      <td>23535854</td>\n",
       "      <td>23535863</td>\n",
       "      <td>stop_codon;H3c8;protein_coding;ENSMUSG00000099...</td>\n",
       "      <td>149.55341</td>\n",
       "      <td>+</td>\n",
       "      <td>chr6:26270966-26270975:-</td>\n",
       "      <td>chr13:23535854-23535863:+</td>\n",
       "      <td>ENSMUSG00000099517</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>{stop_codon}</td>\n",
       "      <td>{ENSMUSG00000099517}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom     start       end  \\\n",
       "20497  chr13  23535854  23535863   \n",
       "\n",
       "                                                    name      score strand  \\\n",
       "20497  stop_codon;H3c8;protein_coding;ENSMUSG00000099...  149.55341      +   \n",
       "\n",
       "              source_identifier                 identifier  \\\n",
       "20497  chr6:26270966-26270975:-  chr13:23535854-23535863:+   \n",
       "\n",
       "                  gene_id  main_annot     annot_set           gene_id_set  \n",
       "20497  ENSMUSG00000099517  stop_codon  {stop_codon}  {ENSMUSG00000099517}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>identifier</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>main_annot</th>\n",
       "      <th>annot_set</th>\n",
       "      <th>gene_id_set</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4440</th>\n",
       "      <td>chr13</td>\n",
       "      <td>23535860</td>\n",
       "      <td>23535869</td>\n",
       "      <td>three_prime_UTR;H3c8;protein_coding;ENSMUSG000...</td>\n",
       "      <td>32.99104</td>\n",
       "      <td>+</td>\n",
       "      <td>chr13:23535860-23535869:+</td>\n",
       "      <td>ENSMUSG00000099517</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{three_prime_UTR}</td>\n",
       "      <td>{ENSMUSG00000099517}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom     start       end  \\\n",
       "4440  chr13  23535860  23535869   \n",
       "\n",
       "                                                   name     score strand  \\\n",
       "4440  three_prime_UTR;H3c8;protein_coding;ENSMUSG000...  32.99104      +   \n",
       "\n",
       "                     identifier             gene_id       main_annot  \\\n",
       "4440  chr13:23535860-23535869:+  ENSMUSG00000099517  three_prime_UTR   \n",
       "\n",
       "              annot_set           gene_id_set  \n",
       "4440  {three_prime_UTR}  {ENSMUSG00000099517}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>source_identifier</th>\n",
       "      <th>identifier</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>main_annot</th>\n",
       "      <th>annot_set</th>\n",
       "      <th>gene_id_set</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4025</th>\n",
       "      <td>chr6</td>\n",
       "      <td>26270955</td>\n",
       "      <td>26270969</td>\n",
       "      <td>three_prime_UTR;H3C8;protein_coding;ENSG000002...</td>\n",
       "      <td>32.99104</td>\n",
       "      <td>-</td>\n",
       "      <td>chr13:23535860-23535869:+</td>\n",
       "      <td>chr6:26270955-26270969:-</td>\n",
       "      <td>ENSG00000273983</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{three_prime_UTR}</td>\n",
       "      <td>{ENSG00000273983}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom     start       end  \\\n",
       "4025  chr6  26270955  26270969   \n",
       "\n",
       "                                                   name     score strand  \\\n",
       "4025  three_prime_UTR;H3C8;protein_coding;ENSG000002...  32.99104      -   \n",
       "\n",
       "              source_identifier                identifier          gene_id  \\\n",
       "4025  chr13:23535860-23535869:+  chr6:26270955-26270969:-  ENSG00000273983   \n",
       "\n",
       "           main_annot          annot_set        gene_id_set  \n",
       "4025  three_prime_UTR  {three_prime_UTR}  {ENSG00000273983}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print(grouped_PEAKS_h_lift_3utr.get_group((False, True)).shape[0])\n",
    "\n",
    "assert grouped_PEAKS_h_lift_3utr.get_group((False, True)).shape[0] == 1\n",
    "\n",
    "# Expected 1 ; detailed below.\n",
    "\n",
    "display(grouped_PEAKS_h_lift_3utr.get_group((False, True)))\n",
    "\n",
    "\n",
    "# Human : chr6:26270966-26270975:- ; lifted to chr13:23535854-23535863:+\n",
    "\n",
    "\n",
    "display(h_intersect_full.loc[lambda df: df[\"A.name\"] == \"chr6:26270966-26270975:-\", [\"A.name\", \"B.name\"]])\n",
    "\n",
    "# Mouse peak closeby: chr13:23535860-23535869:+\n",
    "\n",
    "display(m_id_to_meta_3utr.loc[[\"chr13:23535860-23535869:+\"], :])\n",
    "display(h_id_to_meta_3utr.loc[[\"chr6:26270966-26270975:-\"], :])\n",
    "\n",
    "tmp_h_s = \"chr6:26270966-26270975:-\"\n",
    "tmp_h_l = \"chr13:23535854-23535863:+\"\n",
    "tmp_m_s = \"chr13:23535860-23535869:+\"\n",
    "tmp_m_l = \"chr6:26270955-26270969:-\"\n",
    "\n",
    "# 'ENSMUSG00000099517' in Mouse = H3C8\n",
    "# 'ENSG00000273983' in Human = H3C8\n",
    "\n",
    "display(\n",
    "    DATA[\"orthologs\"].loc[\n",
    "        lambda df: (df[\"mouse_ensid\"] == \"ENSMUSG00000099517\") & (df[\"human_ensid\"] == \"ENSG00000273983\"), :\n",
    "    ]\n",
    ")\n",
    "\n",
    "# There is an n:n orthology in the table.\n",
    "\n",
    "print(\"ENSG00000273983\" in h_genes_bound_orth_complete)\n",
    "print(\"ENSG00000273983\" in h_genes_bound_orth_complete_3utr)\n",
    "\n",
    "print(\"ENSMUSG00000099517\" in m_genes_bound_orth_complete)\n",
    "print(\"ENSMUSG00000099517\" in m_genes_bound_orth_complete_3utr)\n",
    "\n",
    "# The gene sets are unintuitive : the genes\n",
    "\n",
    "display(DATA.human.peaks.lifted.loc[lambda df: df[\"identifier\"] == \"chr13:23535854-23535863:+\", \"name\"].values)\n",
    "\n",
    "\n",
    "tmp_h_s = \"chr6:26270966-26270975:-\"\n",
    "tmp_h_l = \"chr13:23535854-23535863:+\"\n",
    "tmp_m_s = \"chr13:23535860-23535869:+\"\n",
    "tmp_m_l = \"chr6:26270955-26270969:-\"\n",
    "\n",
    "display(DATA.human.peaks.full.loc[lambda df: df[\"identifier\"] == tmp_h_s, :])\n",
    "display(DATA.human.peaks.lifted.loc[lambda df: df[\"identifier\"] == tmp_h_l, :])\n",
    "display(DATA.mouse.peaks.full.loc[lambda df: df[\"identifier\"] == tmp_m_s, :])\n",
    "display(DATA.mouse.peaks.lifted.loc[lambda df: df[\"identifier\"] == tmp_m_l, :])\n",
    "\n",
    "\n",
    "# THIS IS ACTUALLY FROM THE HUMAN PEAK THAT WE SEE THE ISSUE:\n",
    "# chr6:26270966-26270975:- is lifed to chr13:23535854-23535863:+\n",
    "# This chr13:23535854-23535863:+ region is actually annotated as STOP_CODON.\n",
    "#\n",
    "# => We can consider this peak as binding at the same location."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "5e680f84",
   "metadata": {},
   "outputs": [],
   "source": [
    "# DECISION => Add the locus to the \"aligned_species-specific\" category.\n",
    "\n",
    "h_peaks_orthology_label_3utr.append(\n",
    "    pd.DataFrame([pd.Series({\"identifier\": \"chr6:26270966-26270975:-\", \"status\": \"aligned_bound-ortholog_matched\"})])\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19cc3f96",
   "metadata": {},
   "source": [
    "#### Final table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "11bef3aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "h_peaks_orthology_label_3utr = pd.concat(h_peaks_orthology_label_3utr).reset_index(drop=True)\n",
    "assert h_peaks_orthology_label_3utr[\"identifier\"].duplicated().sum() == 0\n",
    "assert set(h_peaks_orthology_label_3utr[\"identifier\"].unique()) == set(h_full_3utr[\"identifier\"].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1b25b33",
   "metadata": {},
   "source": [
    "## Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f7edf0a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "mouse.aligned_bound-ortholog_distant    5956\n",
       "mouse.aligned_bound-ortholog_matched    3159\n",
       "mouse.aligned_species-specific          2765\n",
       "mouse.not-aligned_species-specific       404\n",
       "mouse.not-aligned_bound-ortholog         281\n",
       "human.aligned_bound-ortholog_distant    8839\n",
       "human.aligned_species-specific          8625\n",
       "human.aligned_bound-ortholog_matched    3242\n",
       "human.not-aligned_species-specific      1049\n",
       "human.not-aligned_bound-ortholog         583\n",
       "Name: count, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "counts_peaks_orthology_label_3utr = (\n",
    "    pd.concat(\n",
    "        [\n",
    "            m_peaks_orthology_label_3utr.assign(status=lambda df: \"mouse.\" + df[\"status\"])[\"status\"].value_counts(),\n",
    "            h_peaks_orthology_label_3utr.assign(status=lambda df: \"human.\" + df[\"status\"])[\"status\"].value_counts(),\n",
    "        ],\n",
    "        axis=0,\n",
    "    )\n",
    "    .fillna(0)\n",
    "    .astype(int)\n",
    "    .rename(\"count\")\n",
    ")\n",
    "\n",
    "display(counts_peaks_orthology_label_3utr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ead3711f",
   "metadata": {},
   "outputs": [],
   "source": [
    "PEAKS_COUNTS_ORDERED_LABELS = [\n",
    "    \"mouse.not-aligned_species-specific\",\n",
    "    \"mouse.aligned_species-specific\",\n",
    "    \"mouse.not-aligned_bound-ortholog\",\n",
    "    \"mouse.aligned_bound-ortholog_distant\",\n",
    "    \"mouse.aligned_bound-ortholog_matched\",\n",
    "    \"human.aligned_bound-ortholog_matched\",\n",
    "    \"human.aligned_bound-ortholog_distant\",\n",
    "    \"human.aligned_species-specific\",\n",
    "    \"human.not-aligned_bound-ortholog\",\n",
    "    \"human.not-aligned_species-specific\",\n",
    "]\n",
    "\n",
    "for c in counts_peaks_orthology_label_3utr.index.values:\n",
    "    if c not in PEAKS_COUNTS_ORDERED_LABELS:\n",
    "        print(c)\n",
    "        raise ValueError()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be01f70c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "UserWarning: set_ticklabels() should only be used with a fixed number of ticks, i.e. after set_ticks() or using a FixedLocator.\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"utf-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       "  \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<svg xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"627.846563pt\" height=\"699.664375pt\" viewBox=\"0 0 627.846563 699.664375\" xmlns=\"http://www.w3.org/2000/svg\" version=\"1.1\">\n",
       " <metadata>\n",
       "  <rdf:RDF xmlns:dc=\"http://purl.org/dc/elements/1.1/\" xmlns:cc=\"http://creativecommons.org/ns#\" xmlns:rdf=\"http://www.w3.org/1999/02/22-rdf-syntax-ns#\">\n",
       "   <cc:Work>\n",
       "    <dc:type rdf:resource=\"http://purl.org/dc/dcmitype/StillImage\"/>\n",
       "    <dc:date>2025-07-14T10:43:33.374140</dc:date>\n",
       "    <dc:format>image/svg+xml</dc:format>\n",
       "    <dc:creator>\n",
       "     <cc:Agent>\n",
       "      <dc:title>Matplotlib v3.8.2, https://matplotlib.org/</dc:title>\n",
       "     </cc:Agent>\n",
       "    </dc:creator>\n",
       "   </cc:Work>\n",
       "  </rdf:RDF>\n",
       " </metadata>\n",
       " <defs>\n",
       "  <style type=\"text/css\">*{stroke-linejoin: round; stroke-linecap: butt}</style>\n",
       " </defs>\n",
       " <g id=\"figure_1\">\n",
       "  <g id=\"patch_1\">\n",
       "   <path d=\"M 0 699.664375 \n",
       "L 627.846563 699.664375 \n",
       "L 627.846563 0 \n",
       "L 0 0 \n",
       "z\n",
       "\" style=\"fill: #ffffff\"/>\n",
       "  </g>\n",
       "  <g id=\"axes_1\">\n",
       "   <g id=\"patch_2\">\n",
       "    <path d=\"M 68.246563 320.78 \n",
       "L 620.646562 320.78 \n",
       "L 620.646562 7.2 \n",
       "L 68.246563 7.2 \n",
       "z\n",
       "\" style=\"fill: #ffffff\"/>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_1\">\n",
       "    <g id=\"xtick_1\">\n",
       "     <g id=\"line2d_1\">\n",
       "      <defs>\n",
       "       <path id=\"m33278d9a19\" d=\"M 0 0 \n",
       "L 0 3.5 \n",
       "\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </defs>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m33278d9a19\" x=\"95.866563\" y=\"320.78\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_1\">\n",
       "      <!-- mouse.not-aligned_species-specific -->\n",
       "      <g transform=\"translate(91.657187 644.595625) rotate(-90) scale(0.2 -0.2)\">\n",
       "       <defs>\n",
       "        <path id=\"ArialMT-6d\" d=\"M 422 0 \n",
       "L 422 3319 \n",
       "L 925 3319 \n",
       "L 925 2853 \n",
       "Q 1081 3097 1340 3245 \n",
       "Q 1600 3394 1931 3394 \n",
       "Q 2300 3394 2536 3241 \n",
       "Q 2772 3088 2869 2813 \n",
       "Q 3263 3394 3894 3394 \n",
       "Q 4388 3394 4653 3120 \n",
       "Q 4919 2847 4919 2278 \n",
       "L 4919 0 \n",
       "L 4359 0 \n",
       "L 4359 2091 \n",
       "Q 4359 2428 4304 2576 \n",
       "Q 4250 2725 4106 2815 \n",
       "Q 3963 2906 3769 2906 \n",
       "Q 3419 2906 3187 2673 \n",
       "Q 2956 2441 2956 1928 \n",
       "L 2956 0 \n",
       "L 2394 0 \n",
       "L 2394 2156 \n",
       "Q 2394 2531 2256 2718 \n",
       "Q 2119 2906 1806 2906 \n",
       "Q 1569 2906 1367 2781 \n",
       "Q 1166 2656 1075 2415 \n",
       "Q 984 2175 984 1722 \n",
       "L 984 0 \n",
       "L 422 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"ArialMT-6f\" d=\"M 213 1659 \n",
       "Q 213 2581 725 3025 \n",
       "Q 1153 3394 1769 3394 \n",
       "Q 2453 3394 2887 2945 \n",
       "Q 3322 2497 3322 1706 \n",
       "Q 3322 1066 3130 698 \n",
       "Q 2938 331 2570 128 \n",
       "Q 2203 -75 1769 -75 \n",
       "Q 1072 -75 642 372 \n",
       "Q 213 819 213 1659 \n",
       "z\n",
       "M 791 1659 \n",
       "Q 791 1022 1069 705 \n",
       "Q 1347 388 1769 388 \n",
       "Q 2188 388 2466 706 \n",
       "Q 2744 1025 2744 1678 \n",
       "Q 2744 2294 2464 2611 \n",
       "Q 2184 2928 1769 2928 \n",
       "Q 1347 2928 1069 2612 \n",
       "Q 791 2297 791 1659 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"ArialMT-75\" d=\"M 2597 0 \n",
       "L 2597 488 \n",
       "Q 2209 -75 1544 -75 \n",
       "Q 1250 -75 995 37 \n",
       "Q 741 150 617 320 \n",
       "Q 494 491 444 738 \n",
       "Q 409 903 409 1263 \n",
       "L 409 3319 \n",
       "L 972 3319 \n",
       "L 972 1478 \n",
       "Q 972 1038 1006 884 \n",
       "Q 1059 663 1231 536 \n",
       "Q 1403 409 1656 409 \n",
       "Q 1909 409 2131 539 \n",
       "Q 2353 669 2445 892 \n",
       "Q 2538 1116 2538 1541 \n",
       "L 2538 3319 \n",
       "L 3100 3319 \n",
       "L 3100 0 \n",
       "L 2597 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"ArialMT-73\" d=\"M 197 991 \n",
       "L 753 1078 \n",
       "Q 800 744 1014 566 \n",
       "Q 1228 388 1613 388 \n",
       "Q 2000 388 2187 545 \n",
       "Q 2375 703 2375 916 \n",
       "Q 2375 1106 2209 1216 \n",
       "Q 2094 1291 1634 1406 \n",
       "Q 1016 1563 777 1677 \n",
       "Q 538 1791 414 1992 \n",
       "Q 291 2194 291 2438 \n",
       "Q 291 2659 392 2848 \n",
       "Q 494 3038 669 3163 \n",
       "Q 800 3259 1026 3326 \n",
       "Q 1253 3394 1513 3394 \n",
       "Q 1903 3394 2198 3281 \n",
       "Q 2494 3169 2634 2976 \n",
       "Q 2775 2784 2828 2463 \n",
       "L 2278 2388 \n",
       "Q 2241 2644 2061 2787 \n",
       "Q 1881 2931 1553 2931 \n",
       "Q 1166 2931 1000 2803 \n",
       "Q 834 2675 834 2503 \n",
       "Q 834 2394 903 2306 \n",
       "Q 972 2216 1119 2156 \n",
       "Q 1203 2125 1616 2013 \n",
       "Q 2213 1853 2448 1751 \n",
       "Q 2684 1650 2818 1456 \n",
       "Q 2953 1263 2953 975 \n",
       "Q 2953 694 2789 445 \n",
       "Q 2625 197 2315 61 \n",
       "Q 2006 -75 1616 -75 \n",
       "Q 969 -75 630 194 \n",
       "Q 291 463 197 991 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"ArialMT-65\" d=\"M 2694 1069 \n",
       "L 3275 997 \n",
       "Q 3138 488 2766 206 \n",
       "Q 2394 -75 1816 -75 \n",
       "Q 1088 -75 661 373 \n",
       "Q 234 822 234 1631 \n",
       "Q 234 2469 665 2931 \n",
       "Q 1097 3394 1784 3394 \n",
       "Q 2450 3394 2872 2941 \n",
       "Q 3294 2488 3294 1666 \n",
       "Q 3294 1616 3291 1516 \n",
       "L 816 1516 \n",
       "Q 847 969 1125 678 \n",
       "Q 1403 388 1819 388 \n",
       "Q 2128 388 2347 550 \n",
       "Q 2566 713 2694 1069 \n",
       "z\n",
       "M 847 1978 \n",
       "L 2700 1978 \n",
       "Q 2663 2397 2488 2606 \n",
       "Q 2219 2931 1791 2931 \n",
       "Q 1403 2931 1139 2672 \n",
       "Q 875 2413 847 1978 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"ArialMT-2e\" d=\"M 581 0 \n",
       "L 581 641 \n",
       "L 1222 641 \n",
       "L 1222 0 \n",
       "L 581 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"ArialMT-6e\" d=\"M 422 0 \n",
       "L 422 3319 \n",
       "L 928 3319 \n",
       "L 928 2847 \n",
       "Q 1294 3394 1984 3394 \n",
       "Q 2284 3394 2536 3286 \n",
       "Q 2788 3178 2913 3003 \n",
       "Q 3038 2828 3088 2588 \n",
       "Q 3119 2431 3119 2041 \n",
       "L 3119 0 \n",
       "L 2556 0 \n",
       "L 2556 2019 \n",
       "Q 2556 2363 2490 2533 \n",
       "Q 2425 2703 2258 2804 \n",
       "Q 2091 2906 1866 2906 \n",
       "Q 1506 2906 1245 2678 \n",
       "Q 984 2450 984 1813 \n",
       "L 984 0 \n",
       "L 422 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"ArialMT-74\" d=\"M 1650 503 \n",
       "L 1731 6 \n",
       "Q 1494 -44 1306 -44 \n",
       "Q 1000 -44 831 53 \n",
       "Q 663 150 594 308 \n",
       "Q 525 466 525 972 \n",
       "L 525 2881 \n",
       "L 113 2881 \n",
       "L 113 3319 \n",
       "L 525 3319 \n",
       "L 525 4141 \n",
       "L 1084 4478 \n",
       "L 1084 3319 \n",
       "L 1650 3319 \n",
       "L 1650 2881 \n",
       "L 1084 2881 \n",
       "L 1084 941 \n",
       "Q 1084 700 1114 631 \n",
       "Q 1144 563 1211 522 \n",
       "Q 1278 481 1403 481 \n",
       "Q 1497 481 1650 503 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"ArialMT-2d\" d=\"M 203 1375 \n",
       "L 203 1941 \n",
       "L 1931 1941 \n",
       "L 1931 1375 \n",
       "L 203 1375 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"ArialMT-61\" d=\"M 2588 409 \n",
       "Q 2275 144 1986 34 \n",
       "Q 1697 -75 1366 -75 \n",
       "Q 819 -75 525 192 \n",
       "Q 231 459 231 875 \n",
       "Q 231 1119 342 1320 \n",
       "Q 453 1522 633 1644 \n",
       "Q 813 1766 1038 1828 \n",
       "Q 1203 1872 1538 1913 \n",
       "Q 2219 1994 2541 2106 \n",
       "Q 2544 2222 2544 2253 \n",
       "Q 2544 2597 2384 2738 \n",
       "Q 2169 2928 1744 2928 \n",
       "Q 1347 2928 1158 2789 \n",
       "Q 969 2650 878 2297 \n",
       "L 328 2372 \n",
       "Q 403 2725 575 2942 \n",
       "Q 747 3159 1072 3276 \n",
       "Q 1397 3394 1825 3394 \n",
       "Q 2250 3394 2515 3294 \n",
       "Q 2781 3194 2906 3042 \n",
       "Q 3031 2891 3081 2659 \n",
       "Q 3109 2516 3109 2141 \n",
       "L 3109 1391 \n",
       "Q 3109 606 3145 398 \n",
       "Q 3181 191 3288 0 \n",
       "L 2700 0 \n",
       "Q 2613 175 2588 409 \n",
       "z\n",
       "M 2541 1666 \n",
       "Q 2234 1541 1622 1453 \n",
       "Q 1275 1403 1131 1340 \n",
       "Q 988 1278 909 1158 \n",
       "Q 831 1038 831 891 \n",
       "Q 831 666 1001 516 \n",
       "Q 1172 366 1500 366 \n",
       "Q 1825 366 2078 508 \n",
       "Q 2331 650 2450 897 \n",
       "Q 2541 1088 2541 1459 \n",
       "L 2541 1666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"ArialMT-6c\" d=\"M 409 0 \n",
       "L 409 4581 \n",
       "L 972 4581 \n",
       "L 972 0 \n",
       "L 409 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"ArialMT-69\" d=\"M 425 3934 \n",
       "L 425 4581 \n",
       "L 988 4581 \n",
       "L 988 3934 \n",
       "L 425 3934 \n",
       "z\n",
       "M 425 0 \n",
       "L 425 3319 \n",
       "L 988 3319 \n",
       "L 988 0 \n",
       "L 425 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"ArialMT-67\" d=\"M 319 -275 \n",
       "L 866 -356 \n",
       "Q 900 -609 1056 -725 \n",
       "Q 1266 -881 1628 -881 \n",
       "Q 2019 -881 2231 -725 \n",
       "Q 2444 -569 2519 -288 \n",
       "Q 2563 -116 2559 434 \n",
       "Q 2191 0 1641 0 \n",
       "Q 956 0 581 494 \n",
       "Q 206 988 206 1678 \n",
       "Q 206 2153 378 2554 \n",
       "Q 550 2956 876 3175 \n",
       "Q 1203 3394 1644 3394 \n",
       "Q 2231 3394 2613 2919 \n",
       "L 2613 3319 \n",
       "L 3131 3319 \n",
       "L 3131 450 \n",
       "Q 3131 -325 2973 -648 \n",
       "Q 2816 -972 2473 -1159 \n",
       "Q 2131 -1347 1631 -1347 \n",
       "Q 1038 -1347 672 -1080 \n",
       "Q 306 -813 319 -275 \n",
       "z\n",
       "M 784 1719 \n",
       "Q 784 1066 1043 766 \n",
       "Q 1303 466 1694 466 \n",
       "Q 2081 466 2343 764 \n",
       "Q 2606 1063 2606 1700 \n",
       "Q 2606 2309 2336 2618 \n",
       "Q 2066 2928 1684 2928 \n",
       "Q 1309 2928 1046 2623 \n",
       "Q 784 2319 784 1719 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"ArialMT-64\" d=\"M 2575 0 \n",
       "L 2575 419 \n",
       "Q 2259 -75 1647 -75 \n",
       "Q 1250 -75 917 144 \n",
       "Q 584 363 401 755 \n",
       "Q 219 1147 219 1656 \n",
       "Q 219 2153 384 2558 \n",
       "Q 550 2963 881 3178 \n",
       "Q 1213 3394 1622 3394 \n",
       "Q 1922 3394 2156 3267 \n",
       "Q 2391 3141 2538 2938 \n",
       "L 2538 4581 \n",
       "L 3097 4581 \n",
       "L 3097 0 \n",
       "L 2575 0 \n",
       "z\n",
       "M 797 1656 \n",
       "Q 797 1019 1065 703 \n",
       "Q 1334 388 1700 388 \n",
       "Q 2069 388 2326 689 \n",
       "Q 2584 991 2584 1609 \n",
       "Q 2584 2291 2321 2609 \n",
       "Q 2059 2928 1675 2928 \n",
       "Q 1300 2928 1048 2622 \n",
       "Q 797 2316 797 1656 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"ArialMT-5f\" d=\"M -97 -1272 \n",
       "L -97 -866 \n",
       "L 3631 -866 \n",
       "L 3631 -1272 \n",
       "L -97 -1272 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"ArialMT-70\" d=\"M 422 -1272 \n",
       "L 422 3319 \n",
       "L 934 3319 \n",
       "L 934 2888 \n",
       "Q 1116 3141 1344 3267 \n",
       "Q 1572 3394 1897 3394 \n",
       "Q 2322 3394 2647 3175 \n",
       "Q 2972 2956 3137 2557 \n",
       "Q 3303 2159 3303 1684 \n",
       "Q 3303 1175 3120 767 \n",
       "Q 2938 359 2589 142 \n",
       "Q 2241 -75 1856 -75 \n",
       "Q 1575 -75 1351 44 \n",
       "Q 1128 163 984 344 \n",
       "L 984 -1272 \n",
       "L 422 -1272 \n",
       "z\n",
       "M 931 1641 \n",
       "Q 931 1000 1190 694 \n",
       "Q 1450 388 1819 388 \n",
       "Q 2194 388 2461 705 \n",
       "Q 2728 1022 2728 1688 \n",
       "Q 2728 2322 2467 2637 \n",
       "Q 2206 2953 1844 2953 \n",
       "Q 1484 2953 1207 2617 \n",
       "Q 931 2281 931 1641 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"ArialMT-63\" d=\"M 2588 1216 \n",
       "L 3141 1144 \n",
       "Q 3050 572 2676 248 \n",
       "Q 2303 -75 1759 -75 \n",
       "Q 1078 -75 664 370 \n",
       "Q 250 816 250 1647 \n",
       "Q 250 2184 428 2587 \n",
       "Q 606 2991 970 3192 \n",
       "Q 1334 3394 1763 3394 \n",
       "Q 2303 3394 2647 3120 \n",
       "Q 2991 2847 3088 2344 \n",
       "L 2541 2259 \n",
       "Q 2463 2594 2264 2762 \n",
       "Q 2066 2931 1784 2931 \n",
       "Q 1359 2931 1093 2626 \n",
       "Q 828 2322 828 1663 \n",
       "Q 828 994 1084 691 \n",
       "Q 1341 388 1753 388 \n",
       "Q 2084 388 2306 591 \n",
       "Q 2528 794 2588 1216 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"ArialMT-66\" d=\"M 556 0 \n",
       "L 556 2881 \n",
       "L 59 2881 \n",
       "L 59 3319 \n",
       "L 556 3319 \n",
       "L 556 3672 \n",
       "Q 556 4006 616 4169 \n",
       "Q 697 4388 901 4523 \n",
       "Q 1106 4659 1475 4659 \n",
       "Q 1713 4659 2000 4603 \n",
       "L 1916 4113 \n",
       "Q 1741 4144 1584 4144 \n",
       "Q 1328 4144 1222 4034 \n",
       "Q 1116 3925 1116 3625 \n",
       "L 1116 3319 \n",
       "L 1763 3319 \n",
       "L 1763 2881 \n",
       "L 1116 2881 \n",
       "L 1116 0 \n",
       "L 556 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#ArialMT-6d\"/>\n",
       "       <use xlink:href=\"#ArialMT-6f\" x=\"83.300781\"/>\n",
       "       <use xlink:href=\"#ArialMT-75\" x=\"138.916016\"/>\n",
       "       <use xlink:href=\"#ArialMT-73\" x=\"194.53125\"/>\n",
       "       <use xlink:href=\"#ArialMT-65\" x=\"244.53125\"/>\n",
       "       <use xlink:href=\"#ArialMT-2e\" x=\"300.146484\"/>\n",
       "       <use xlink:href=\"#ArialMT-6e\" x=\"327.929688\"/>\n",
       "       <use xlink:href=\"#ArialMT-6f\" x=\"383.544922\"/>\n",
       "       <use xlink:href=\"#ArialMT-74\" x=\"439.160156\"/>\n",
       "       <use xlink:href=\"#ArialMT-2d\" x=\"466.943359\"/>\n",
       "       <use xlink:href=\"#ArialMT-61\" x=\"500.244141\"/>\n",
       "       <use xlink:href=\"#ArialMT-6c\" x=\"555.859375\"/>\n",
       "       <use xlink:href=\"#ArialMT-69\" x=\"578.076172\"/>\n",
       "       <use xlink:href=\"#ArialMT-67\" x=\"600.292969\"/>\n",
       "       <use xlink:href=\"#ArialMT-6e\" x=\"655.908203\"/>\n",
       "       <use xlink:href=\"#ArialMT-65\" x=\"711.523438\"/>\n",
       "       <use xlink:href=\"#ArialMT-64\" x=\"767.138672\"/>\n",
       "       <use xlink:href=\"#ArialMT-5f\" x=\"822.753906\"/>\n",
       "       <use xlink:href=\"#ArialMT-73\" x=\"878.369141\"/>\n",
       "       <use xlink:href=\"#ArialMT-70\" x=\"928.369141\"/>\n",
       "       <use xlink:href=\"#ArialMT-65\" x=\"983.984375\"/>\n",
       "       <use xlink:href=\"#ArialMT-63\" x=\"1039.599609\"/>\n",
       "       <use xlink:href=\"#ArialMT-69\" x=\"1089.599609\"/>\n",
       "       <use xlink:href=\"#ArialMT-65\" x=\"1111.816406\"/>\n",
       "       <use xlink:href=\"#ArialMT-73\" x=\"1167.431641\"/>\n",
       "       <use xlink:href=\"#ArialMT-2d\" x=\"1217.431641\"/>\n",
       "       <use xlink:href=\"#ArialMT-73\" x=\"1250.732422\"/>\n",
       "       <use xlink:href=\"#ArialMT-70\" x=\"1300.732422\"/>\n",
       "       <use xlink:href=\"#ArialMT-65\" x=\"1356.347656\"/>\n",
       "       <use xlink:href=\"#ArialMT-63\" x=\"1411.962891\"/>\n",
       "       <use xlink:href=\"#ArialMT-69\" x=\"1461.962891\"/>\n",
       "       <use xlink:href=\"#ArialMT-66\" x=\"1484.179688\"/>\n",
       "       <use xlink:href=\"#ArialMT-69\" x=\"1511.962891\"/>\n",
       "       <use xlink:href=\"#ArialMT-63\" x=\"1534.179688\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_2\">\n",
       "     <g id=\"line2d_2\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m33278d9a19\" x=\"151.106563\" y=\"320.78\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_2\">\n",
       "      <!-- mouse.aligned_species-specific -->\n",
       "      <g transform=\"translate(146.897188 610.13625) rotate(-90) scale(0.2 -0.2)\">\n",
       "       <use xlink:href=\"#ArialMT-6d\"/>\n",
       "       <use xlink:href=\"#ArialMT-6f\" x=\"83.300781\"/>\n",
       "       <use xlink:href=\"#ArialMT-75\" x=\"138.916016\"/>\n",
       "       <use xlink:href=\"#ArialMT-73\" x=\"194.53125\"/>\n",
       "       <use xlink:href=\"#ArialMT-65\" x=\"244.53125\"/>\n",
       "       <use xlink:href=\"#ArialMT-2e\" x=\"300.146484\"/>\n",
       "       <use xlink:href=\"#ArialMT-61\" x=\"327.929688\"/>\n",
       "       <use xlink:href=\"#ArialMT-6c\" x=\"383.544922\"/>\n",
       "       <use xlink:href=\"#ArialMT-69\" x=\"405.761719\"/>\n",
       "       <use xlink:href=\"#ArialMT-67\" x=\"427.978516\"/>\n",
       "       <use xlink:href=\"#ArialMT-6e\" x=\"483.59375\"/>\n",
       "       <use xlink:href=\"#ArialMT-65\" x=\"539.208984\"/>\n",
       "       <use xlink:href=\"#ArialMT-64\" x=\"594.824219\"/>\n",
       "       <use xlink:href=\"#ArialMT-5f\" x=\"650.439453\"/>\n",
       "       <use xlink:href=\"#ArialMT-73\" x=\"706.054688\"/>\n",
       "       <use xlink:href=\"#ArialMT-70\" x=\"756.054688\"/>\n",
       "       <use xlink:href=\"#ArialMT-65\" x=\"811.669922\"/>\n",
       "       <use xlink:href=\"#ArialMT-63\" x=\"867.285156\"/>\n",
       "       <use xlink:href=\"#ArialMT-69\" x=\"917.285156\"/>\n",
       "       <use xlink:href=\"#ArialMT-65\" x=\"939.501953\"/>\n",
       "       <use xlink:href=\"#ArialMT-73\" x=\"995.117188\"/>\n",
       "       <use xlink:href=\"#ArialMT-2d\" x=\"1045.117188\"/>\n",
       "       <use xlink:href=\"#ArialMT-73\" x=\"1078.417969\"/>\n",
       "       <use xlink:href=\"#ArialMT-70\" x=\"1128.417969\"/>\n",
       "       <use xlink:href=\"#ArialMT-65\" x=\"1184.033203\"/>\n",
       "       <use xlink:href=\"#ArialMT-63\" x=\"1239.648438\"/>\n",
       "       <use xlink:href=\"#ArialMT-69\" x=\"1289.648438\"/>\n",
       "       <use xlink:href=\"#ArialMT-66\" x=\"1311.865234\"/>\n",
       "       <use xlink:href=\"#ArialMT-69\" x=\"1339.648438\"/>\n",
       "       <use xlink:href=\"#ArialMT-63\" x=\"1361.865234\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_3\">\n",
       "     <g id=\"line2d_3\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m33278d9a19\" x=\"206.346563\" y=\"320.78\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_3\">\n",
       "      <!-- mouse.not-aligned_bound-ortholog -->\n",
       "      <g transform=\"translate(202.137187 637.976875) rotate(-90) scale(0.2 -0.2)\">\n",
       "       <defs>\n",
       "        <path id=\"ArialMT-62\" d=\"M 941 0 \n",
       "L 419 0 \n",
       "L 419 4581 \n",
       "L 981 4581 \n",
       "L 981 2947 \n",
       "Q 1338 3394 1891 3394 \n",
       "Q 2197 3394 2470 3270 \n",
       "Q 2744 3147 2920 2923 \n",
       "Q 3097 2700 3197 2384 \n",
       "Q 3297 2069 3297 1709 \n",
       "Q 3297 856 2875 390 \n",
       "Q 2453 -75 1863 -75 \n",
       "Q 1275 -75 941 416 \n",
       "L 941 0 \n",
       "z\n",
       "M 934 1684 \n",
       "Q 934 1088 1097 822 \n",
       "Q 1363 388 1816 388 \n",
       "Q 2184 388 2453 708 \n",
       "Q 2722 1028 2722 1663 \n",
       "Q 2722 2313 2464 2622 \n",
       "Q 2206 2931 1841 2931 \n",
       "Q 1472 2931 1203 2611 \n",
       "Q 934 2291 934 1684 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"ArialMT-72\" d=\"M 416 0 \n",
       "L 416 3319 \n",
       "L 922 3319 \n",
       "L 922 2816 \n",
       "Q 1116 3169 1280 3281 \n",
       "Q 1444 3394 1641 3394 \n",
       "Q 1925 3394 2219 3213 \n",
       "L 2025 2691 \n",
       "Q 1819 2813 1613 2813 \n",
       "Q 1428 2813 1281 2702 \n",
       "Q 1134 2591 1072 2394 \n",
       "Q 978 2094 978 1738 \n",
       "L 978 0 \n",
       "L 416 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"ArialMT-68\" d=\"M 422 0 \n",
       "L 422 4581 \n",
       "L 984 4581 \n",
       "L 984 2938 \n",
       "Q 1378 3394 1978 3394 \n",
       "Q 2347 3394 2619 3248 \n",
       "Q 2891 3103 3008 2847 \n",
       "Q 3125 2591 3125 2103 \n",
       "L 3125 0 \n",
       "L 2563 0 \n",
       "L 2563 2103 \n",
       "Q 2563 2525 2380 2717 \n",
       "Q 2197 2909 1863 2909 \n",
       "Q 1613 2909 1392 2779 \n",
       "Q 1172 2650 1078 2428 \n",
       "Q 984 2206 984 1816 \n",
       "L 984 0 \n",
       "L 422 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#ArialMT-6d\"/>\n",
       "       <use xlink:href=\"#ArialMT-6f\" x=\"83.300781\"/>\n",
       "       <use xlink:href=\"#ArialMT-75\" x=\"138.916016\"/>\n",
       "       <use xlink:href=\"#ArialMT-73\" x=\"194.53125\"/>\n",
       "       <use xlink:href=\"#ArialMT-65\" x=\"244.53125\"/>\n",
       "       <use xlink:href=\"#ArialMT-2e\" x=\"300.146484\"/>\n",
       "       <use xlink:href=\"#ArialMT-6e\" x=\"327.929688\"/>\n",
       "       <use xlink:href=\"#ArialMT-6f\" x=\"383.544922\"/>\n",
       "       <use xlink:href=\"#ArialMT-74\" x=\"439.160156\"/>\n",
       "       <use xlink:href=\"#ArialMT-2d\" x=\"466.943359\"/>\n",
       "       <use xlink:href=\"#ArialMT-61\" x=\"500.244141\"/>\n",
       "       <use xlink:href=\"#ArialMT-6c\" x=\"555.859375\"/>\n",
       "       <use xlink:href=\"#ArialMT-69\" x=\"578.076172\"/>\n",
       "       <use xlink:href=\"#ArialMT-67\" x=\"600.292969\"/>\n",
       "       <use xlink:href=\"#ArialMT-6e\" x=\"655.908203\"/>\n",
       "       <use xlink:href=\"#ArialMT-65\" x=\"711.523438\"/>\n",
       "       <use xlink:href=\"#ArialMT-64\" x=\"767.138672\"/>\n",
       "       <use xlink:href=\"#ArialMT-5f\" x=\"822.753906\"/>\n",
       "       <use xlink:href=\"#ArialMT-62\" x=\"878.369141\"/>\n",
       "       <use xlink:href=\"#ArialMT-6f\" x=\"933.984375\"/>\n",
       "       <use xlink:href=\"#ArialMT-75\" x=\"989.599609\"/>\n",
       "       <use xlink:href=\"#ArialMT-6e\" x=\"1045.214844\"/>\n",
       "       <use xlink:href=\"#ArialMT-64\" x=\"1100.830078\"/>\n",
       "       <use xlink:href=\"#ArialMT-2d\" x=\"1156.445312\"/>\n",
       "       <use xlink:href=\"#ArialMT-6f\" x=\"1189.746094\"/>\n",
       "       <use xlink:href=\"#ArialMT-72\" x=\"1245.361328\"/>\n",
       "       <use xlink:href=\"#ArialMT-74\" x=\"1278.662109\"/>\n",
       "       <use xlink:href=\"#ArialMT-68\" x=\"1306.445312\"/>\n",
       "       <use xlink:href=\"#ArialMT-6f\" x=\"1362.060547\"/>\n",
       "       <use xlink:href=\"#ArialMT-6c\" x=\"1417.675781\"/>\n",
       "       <use xlink:href=\"#ArialMT-6f\" x=\"1439.892578\"/>\n",
       "       <use xlink:href=\"#ArialMT-67\" x=\"1495.507812\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_4\">\n",
       "     <g id=\"line2d_4\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m33278d9a19\" x=\"261.586563\" y=\"320.78\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_4\">\n",
       "      <!-- mouse.aligned_bound-ortholog_distant -->\n",
       "      <g transform=\"translate(257.377187 673.56125) rotate(-90) scale(0.2 -0.2)\">\n",
       "       <use xlink:href=\"#ArialMT-6d\"/>\n",
       "       <use xlink:href=\"#ArialMT-6f\" x=\"83.300781\"/>\n",
       "       <use xlink:href=\"#ArialMT-75\" x=\"138.916016\"/>\n",
       "       <use xlink:href=\"#ArialMT-73\" x=\"194.53125\"/>\n",
       "       <use xlink:href=\"#ArialMT-65\" x=\"244.53125\"/>\n",
       "       <use xlink:href=\"#ArialMT-2e\" x=\"300.146484\"/>\n",
       "       <use xlink:href=\"#ArialMT-61\" x=\"327.929688\"/>\n",
       "       <use xlink:href=\"#ArialMT-6c\" x=\"383.544922\"/>\n",
       "       <use xlink:href=\"#ArialMT-69\" x=\"405.761719\"/>\n",
       "       <use xlink:href=\"#ArialMT-67\" x=\"427.978516\"/>\n",
       "       <use xlink:href=\"#ArialMT-6e\" x=\"483.59375\"/>\n",
       "       <use xlink:href=\"#ArialMT-65\" x=\"539.208984\"/>\n",
       "       <use xlink:href=\"#ArialMT-64\" x=\"594.824219\"/>\n",
       "       <use xlink:href=\"#ArialMT-5f\" x=\"650.439453\"/>\n",
       "       <use xlink:href=\"#ArialMT-62\" x=\"706.054688\"/>\n",
       "       <use xlink:href=\"#ArialMT-6f\" x=\"761.669922\"/>\n",
       "       <use xlink:href=\"#ArialMT-75\" x=\"817.285156\"/>\n",
       "       <use xlink:href=\"#ArialMT-6e\" x=\"872.900391\"/>\n",
       "       <use xlink:href=\"#ArialMT-64\" x=\"928.515625\"/>\n",
       "       <use xlink:href=\"#ArialMT-2d\" x=\"984.130859\"/>\n",
       "       <use xlink:href=\"#ArialMT-6f\" x=\"1017.431641\"/>\n",
       "       <use xlink:href=\"#ArialMT-72\" x=\"1073.046875\"/>\n",
       "       <use xlink:href=\"#ArialMT-74\" x=\"1106.347656\"/>\n",
       "       <use xlink:href=\"#ArialMT-68\" x=\"1134.130859\"/>\n",
       "       <use xlink:href=\"#ArialMT-6f\" x=\"1189.746094\"/>\n",
       "       <use xlink:href=\"#ArialMT-6c\" x=\"1245.361328\"/>\n",
       "       <use xlink:href=\"#ArialMT-6f\" x=\"1267.578125\"/>\n",
       "       <use xlink:href=\"#ArialMT-67\" x=\"1323.193359\"/>\n",
       "       <use xlink:href=\"#ArialMT-5f\" x=\"1378.808594\"/>\n",
       "       <use xlink:href=\"#ArialMT-64\" x=\"1434.423828\"/>\n",
       "       <use xlink:href=\"#ArialMT-69\" x=\"1490.039062\"/>\n",
       "       <use xlink:href=\"#ArialMT-73\" x=\"1512.255859\"/>\n",
       "       <use xlink:href=\"#ArialMT-74\" x=\"1562.255859\"/>\n",
       "       <use xlink:href=\"#ArialMT-61\" x=\"1590.039062\"/>\n",
       "       <use xlink:href=\"#ArialMT-6e\" x=\"1645.654297\"/>\n",
       "       <use xlink:href=\"#ArialMT-74\" x=\"1701.269531\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_5\">\n",
       "     <g id=\"line2d_5\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m33278d9a19\" x=\"316.826563\" y=\"320.78\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_5\">\n",
       "      <!-- mouse.aligned_bound-ortholog_matched -->\n",
       "      <g transform=\"translate(312.617188 691.3425) rotate(-90) scale(0.2 -0.2)\">\n",
       "       <use xlink:href=\"#ArialMT-6d\"/>\n",
       "       <use xlink:href=\"#ArialMT-6f\" x=\"83.300781\"/>\n",
       "       <use xlink:href=\"#ArialMT-75\" x=\"138.916016\"/>\n",
       "       <use xlink:href=\"#ArialMT-73\" x=\"194.53125\"/>\n",
       "       <use xlink:href=\"#ArialMT-65\" x=\"244.53125\"/>\n",
       "       <use xlink:href=\"#ArialMT-2e\" x=\"300.146484\"/>\n",
       "       <use xlink:href=\"#ArialMT-61\" x=\"327.929688\"/>\n",
       "       <use xlink:href=\"#ArialMT-6c\" x=\"383.544922\"/>\n",
       "       <use xlink:href=\"#ArialMT-69\" x=\"405.761719\"/>\n",
       "       <use xlink:href=\"#ArialMT-67\" x=\"427.978516\"/>\n",
       "       <use xlink:href=\"#ArialMT-6e\" x=\"483.59375\"/>\n",
       "       <use xlink:href=\"#ArialMT-65\" x=\"539.208984\"/>\n",
       "       <use xlink:href=\"#ArialMT-64\" x=\"594.824219\"/>\n",
       "       <use xlink:href=\"#ArialMT-5f\" x=\"650.439453\"/>\n",
       "       <use xlink:href=\"#ArialMT-62\" x=\"706.054688\"/>\n",
       "       <use xlink:href=\"#ArialMT-6f\" x=\"761.669922\"/>\n",
       "       <use xlink:href=\"#ArialMT-75\" x=\"817.285156\"/>\n",
       "       <use xlink:href=\"#ArialMT-6e\" x=\"872.900391\"/>\n",
       "       <use xlink:href=\"#ArialMT-64\" x=\"928.515625\"/>\n",
       "       <use xlink:href=\"#ArialMT-2d\" x=\"984.130859\"/>\n",
       "       <use xlink:href=\"#ArialMT-6f\" x=\"1017.431641\"/>\n",
       "       <use xlink:href=\"#ArialMT-72\" x=\"1073.046875\"/>\n",
       "       <use xlink:href=\"#ArialMT-74\" x=\"1106.347656\"/>\n",
       "       <use xlink:href=\"#ArialMT-68\" x=\"1134.130859\"/>\n",
       "       <use xlink:href=\"#ArialMT-6f\" x=\"1189.746094\"/>\n",
       "       <use xlink:href=\"#ArialMT-6c\" x=\"1245.361328\"/>\n",
       "       <use xlink:href=\"#ArialMT-6f\" x=\"1267.578125\"/>\n",
       "       <use xlink:href=\"#ArialMT-67\" x=\"1323.193359\"/>\n",
       "       <use xlink:href=\"#ArialMT-5f\" x=\"1378.808594\"/>\n",
       "       <use xlink:href=\"#ArialMT-6d\" x=\"1434.423828\"/>\n",
       "       <use xlink:href=\"#ArialMT-61\" x=\"1517.724609\"/>\n",
       "       <use xlink:href=\"#ArialMT-74\" x=\"1573.339844\"/>\n",
       "       <use xlink:href=\"#ArialMT-63\" x=\"1601.123047\"/>\n",
       "       <use xlink:href=\"#ArialMT-68\" x=\"1651.123047\"/>\n",
       "       <use xlink:href=\"#ArialMT-65\" x=\"1706.738281\"/>\n",
       "       <use xlink:href=\"#ArialMT-64\" x=\"1762.353516\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_6\">\n",
       "     <g id=\"line2d_6\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m33278d9a19\" x=\"372.066563\" y=\"320.78\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_6\">\n",
       "      <!-- human.aligned_bound-ortholog_matched -->\n",
       "      <g transform=\"translate(367.857188 692.464375) rotate(-90) scale(0.2 -0.2)\">\n",
       "       <use xlink:href=\"#ArialMT-68\"/>\n",
       "       <use xlink:href=\"#ArialMT-75\" x=\"55.615234\"/>\n",
       "       <use xlink:href=\"#ArialMT-6d\" x=\"111.230469\"/>\n",
       "       <use xlink:href=\"#ArialMT-61\" x=\"194.53125\"/>\n",
       "       <use xlink:href=\"#ArialMT-6e\" x=\"250.146484\"/>\n",
       "       <use xlink:href=\"#ArialMT-2e\" x=\"305.761719\"/>\n",
       "       <use xlink:href=\"#ArialMT-61\" x=\"333.544922\"/>\n",
       "       <use xlink:href=\"#ArialMT-6c\" x=\"389.160156\"/>\n",
       "       <use xlink:href=\"#ArialMT-69\" x=\"411.376953\"/>\n",
       "       <use xlink:href=\"#ArialMT-67\" x=\"433.59375\"/>\n",
       "       <use xlink:href=\"#ArialMT-6e\" x=\"489.208984\"/>\n",
       "       <use xlink:href=\"#ArialMT-65\" x=\"544.824219\"/>\n",
       "       <use xlink:href=\"#ArialMT-64\" x=\"600.439453\"/>\n",
       "       <use xlink:href=\"#ArialMT-5f\" x=\"656.054688\"/>\n",
       "       <use xlink:href=\"#ArialMT-62\" x=\"711.669922\"/>\n",
       "       <use xlink:href=\"#ArialMT-6f\" x=\"767.285156\"/>\n",
       "       <use xlink:href=\"#ArialMT-75\" x=\"822.900391\"/>\n",
       "       <use xlink:href=\"#ArialMT-6e\" x=\"878.515625\"/>\n",
       "       <use xlink:href=\"#ArialMT-64\" x=\"934.130859\"/>\n",
       "       <use xlink:href=\"#ArialMT-2d\" x=\"989.746094\"/>\n",
       "       <use xlink:href=\"#ArialMT-6f\" x=\"1023.046875\"/>\n",
       "       <use xlink:href=\"#ArialMT-72\" x=\"1078.662109\"/>\n",
       "       <use xlink:href=\"#ArialMT-74\" x=\"1111.962891\"/>\n",
       "       <use xlink:href=\"#ArialMT-68\" x=\"1139.746094\"/>\n",
       "       <use xlink:href=\"#ArialMT-6f\" x=\"1195.361328\"/>\n",
       "       <use xlink:href=\"#ArialMT-6c\" x=\"1250.976562\"/>\n",
       "       <use xlink:href=\"#ArialMT-6f\" x=\"1273.193359\"/>\n",
       "       <use xlink:href=\"#ArialMT-67\" x=\"1328.808594\"/>\n",
       "       <use xlink:href=\"#ArialMT-5f\" x=\"1384.423828\"/>\n",
       "       <use xlink:href=\"#ArialMT-6d\" x=\"1440.039062\"/>\n",
       "       <use xlink:href=\"#ArialMT-61\" x=\"1523.339844\"/>\n",
       "       <use xlink:href=\"#ArialMT-74\" x=\"1578.955078\"/>\n",
       "       <use xlink:href=\"#ArialMT-63\" x=\"1606.738281\"/>\n",
       "       <use xlink:href=\"#ArialMT-68\" x=\"1656.738281\"/>\n",
       "       <use xlink:href=\"#ArialMT-65\" x=\"1712.353516\"/>\n",
       "       <use xlink:href=\"#ArialMT-64\" x=\"1767.96875\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_7\">\n",
       "     <g id=\"line2d_7\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m33278d9a19\" x=\"427.306563\" y=\"320.78\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_7\">\n",
       "      <!-- human.aligned_bound-ortholog_distant -->\n",
       "      <g transform=\"translate(423.097188 674.683125) rotate(-90) scale(0.2 -0.2)\">\n",
       "       <use xlink:href=\"#ArialMT-68\"/>\n",
       "       <use xlink:href=\"#ArialMT-75\" x=\"55.615234\"/>\n",
       "       <use xlink:href=\"#ArialMT-6d\" x=\"111.230469\"/>\n",
       "       <use xlink:href=\"#ArialMT-61\" x=\"194.53125\"/>\n",
       "       <use xlink:href=\"#ArialMT-6e\" x=\"250.146484\"/>\n",
       "       <use xlink:href=\"#ArialMT-2e\" x=\"305.761719\"/>\n",
       "       <use xlink:href=\"#ArialMT-61\" x=\"333.544922\"/>\n",
       "       <use xlink:href=\"#ArialMT-6c\" x=\"389.160156\"/>\n",
       "       <use xlink:href=\"#ArialMT-69\" x=\"411.376953\"/>\n",
       "       <use xlink:href=\"#ArialMT-67\" x=\"433.59375\"/>\n",
       "       <use xlink:href=\"#ArialMT-6e\" x=\"489.208984\"/>\n",
       "       <use xlink:href=\"#ArialMT-65\" x=\"544.824219\"/>\n",
       "       <use xlink:href=\"#ArialMT-64\" x=\"600.439453\"/>\n",
       "       <use xlink:href=\"#ArialMT-5f\" x=\"656.054688\"/>\n",
       "       <use xlink:href=\"#ArialMT-62\" x=\"711.669922\"/>\n",
       "       <use xlink:href=\"#ArialMT-6f\" x=\"767.285156\"/>\n",
       "       <use xlink:href=\"#ArialMT-75\" x=\"822.900391\"/>\n",
       "       <use xlink:href=\"#ArialMT-6e\" x=\"878.515625\"/>\n",
       "       <use xlink:href=\"#ArialMT-64\" x=\"934.130859\"/>\n",
       "       <use xlink:href=\"#ArialMT-2d\" x=\"989.746094\"/>\n",
       "       <use xlink:href=\"#ArialMT-6f\" x=\"1023.046875\"/>\n",
       "       <use xlink:href=\"#ArialMT-72\" x=\"1078.662109\"/>\n",
       "       <use xlink:href=\"#ArialMT-74\" x=\"1111.962891\"/>\n",
       "       <use xlink:href=\"#ArialMT-68\" x=\"1139.746094\"/>\n",
       "       <use xlink:href=\"#ArialMT-6f\" x=\"1195.361328\"/>\n",
       "       <use xlink:href=\"#ArialMT-6c\" x=\"1250.976562\"/>\n",
       "       <use xlink:href=\"#ArialMT-6f\" x=\"1273.193359\"/>\n",
       "       <use xlink:href=\"#ArialMT-67\" x=\"1328.808594\"/>\n",
       "       <use xlink:href=\"#ArialMT-5f\" x=\"1384.423828\"/>\n",
       "       <use xlink:href=\"#ArialMT-64\" x=\"1440.039062\"/>\n",
       "       <use xlink:href=\"#ArialMT-69\" x=\"1495.654297\"/>\n",
       "       <use xlink:href=\"#ArialMT-73\" x=\"1517.871094\"/>\n",
       "       <use xlink:href=\"#ArialMT-74\" x=\"1567.871094\"/>\n",
       "       <use xlink:href=\"#ArialMT-61\" x=\"1595.654297\"/>\n",
       "       <use xlink:href=\"#ArialMT-6e\" x=\"1651.269531\"/>\n",
       "       <use xlink:href=\"#ArialMT-74\" x=\"1706.884766\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_8\">\n",
       "     <g id=\"line2d_8\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m33278d9a19\" x=\"482.546563\" y=\"320.78\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_8\">\n",
       "      <!-- human.aligned_species-specific -->\n",
       "      <g transform=\"translate(478.337188 611.258125) rotate(-90) scale(0.2 -0.2)\">\n",
       "       <use xlink:href=\"#ArialMT-68\"/>\n",
       "       <use xlink:href=\"#ArialMT-75\" x=\"55.615234\"/>\n",
       "       <use xlink:href=\"#ArialMT-6d\" x=\"111.230469\"/>\n",
       "       <use xlink:href=\"#ArialMT-61\" x=\"194.53125\"/>\n",
       "       <use xlink:href=\"#ArialMT-6e\" x=\"250.146484\"/>\n",
       "       <use xlink:href=\"#ArialMT-2e\" x=\"305.761719\"/>\n",
       "       <use xlink:href=\"#ArialMT-61\" x=\"333.544922\"/>\n",
       "       <use xlink:href=\"#ArialMT-6c\" x=\"389.160156\"/>\n",
       "       <use xlink:href=\"#ArialMT-69\" x=\"411.376953\"/>\n",
       "       <use xlink:href=\"#ArialMT-67\" x=\"433.59375\"/>\n",
       "       <use xlink:href=\"#ArialMT-6e\" x=\"489.208984\"/>\n",
       "       <use xlink:href=\"#ArialMT-65\" x=\"544.824219\"/>\n",
       "       <use xlink:href=\"#ArialMT-64\" x=\"600.439453\"/>\n",
       "       <use xlink:href=\"#ArialMT-5f\" x=\"656.054688\"/>\n",
       "       <use xlink:href=\"#ArialMT-73\" x=\"711.669922\"/>\n",
       "       <use xlink:href=\"#ArialMT-70\" x=\"761.669922\"/>\n",
       "       <use xlink:href=\"#ArialMT-65\" x=\"817.285156\"/>\n",
       "       <use xlink:href=\"#ArialMT-63\" x=\"872.900391\"/>\n",
       "       <use xlink:href=\"#ArialMT-69\" x=\"922.900391\"/>\n",
       "       <use xlink:href=\"#ArialMT-65\" x=\"945.117188\"/>\n",
       "       <use xlink:href=\"#ArialMT-73\" x=\"1000.732422\"/>\n",
       "       <use xlink:href=\"#ArialMT-2d\" x=\"1050.732422\"/>\n",
       "       <use xlink:href=\"#ArialMT-73\" x=\"1084.033203\"/>\n",
       "       <use xlink:href=\"#ArialMT-70\" x=\"1134.033203\"/>\n",
       "       <use xlink:href=\"#ArialMT-65\" x=\"1189.648438\"/>\n",
       "       <use xlink:href=\"#ArialMT-63\" x=\"1245.263672\"/>\n",
       "       <use xlink:href=\"#ArialMT-69\" x=\"1295.263672\"/>\n",
       "       <use xlink:href=\"#ArialMT-66\" x=\"1317.480469\"/>\n",
       "       <use xlink:href=\"#ArialMT-69\" x=\"1345.263672\"/>\n",
       "       <use xlink:href=\"#ArialMT-63\" x=\"1367.480469\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_9\">\n",
       "     <g id=\"line2d_9\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m33278d9a19\" x=\"537.786563\" y=\"320.78\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_9\">\n",
       "      <!-- human.not-aligned_bound-ortholog -->\n",
       "      <g transform=\"translate(533.577188 639.09875) rotate(-90) scale(0.2 -0.2)\">\n",
       "       <use xlink:href=\"#ArialMT-68\"/>\n",
       "       <use xlink:href=\"#ArialMT-75\" x=\"55.615234\"/>\n",
       "       <use xlink:href=\"#ArialMT-6d\" x=\"111.230469\"/>\n",
       "       <use xlink:href=\"#ArialMT-61\" x=\"194.53125\"/>\n",
       "       <use xlink:href=\"#ArialMT-6e\" x=\"250.146484\"/>\n",
       "       <use xlink:href=\"#ArialMT-2e\" x=\"305.761719\"/>\n",
       "       <use xlink:href=\"#ArialMT-6e\" x=\"333.544922\"/>\n",
       "       <use xlink:href=\"#ArialMT-6f\" x=\"389.160156\"/>\n",
       "       <use xlink:href=\"#ArialMT-74\" x=\"444.775391\"/>\n",
       "       <use xlink:href=\"#ArialMT-2d\" x=\"472.558594\"/>\n",
       "       <use xlink:href=\"#ArialMT-61\" x=\"505.859375\"/>\n",
       "       <use xlink:href=\"#ArialMT-6c\" x=\"561.474609\"/>\n",
       "       <use xlink:href=\"#ArialMT-69\" x=\"583.691406\"/>\n",
       "       <use xlink:href=\"#ArialMT-67\" x=\"605.908203\"/>\n",
       "       <use xlink:href=\"#ArialMT-6e\" x=\"661.523438\"/>\n",
       "       <use xlink:href=\"#ArialMT-65\" x=\"717.138672\"/>\n",
       "       <use xlink:href=\"#ArialMT-64\" x=\"772.753906\"/>\n",
       "       <use xlink:href=\"#ArialMT-5f\" x=\"828.369141\"/>\n",
       "       <use xlink:href=\"#ArialMT-62\" x=\"883.984375\"/>\n",
       "       <use xlink:href=\"#ArialMT-6f\" x=\"939.599609\"/>\n",
       "       <use xlink:href=\"#ArialMT-75\" x=\"995.214844\"/>\n",
       "       <use xlink:href=\"#ArialMT-6e\" x=\"1050.830078\"/>\n",
       "       <use xlink:href=\"#ArialMT-64\" x=\"1106.445312\"/>\n",
       "       <use xlink:href=\"#ArialMT-2d\" x=\"1162.060547\"/>\n",
       "       <use xlink:href=\"#ArialMT-6f\" x=\"1195.361328\"/>\n",
       "       <use xlink:href=\"#ArialMT-72\" x=\"1250.976562\"/>\n",
       "       <use xlink:href=\"#ArialMT-74\" x=\"1284.277344\"/>\n",
       "       <use xlink:href=\"#ArialMT-68\" x=\"1312.060547\"/>\n",
       "       <use xlink:href=\"#ArialMT-6f\" x=\"1367.675781\"/>\n",
       "       <use xlink:href=\"#ArialMT-6c\" x=\"1423.291016\"/>\n",
       "       <use xlink:href=\"#ArialMT-6f\" x=\"1445.507812\"/>\n",
       "       <use xlink:href=\"#ArialMT-67\" x=\"1501.123047\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_10\">\n",
       "     <g id=\"line2d_10\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m33278d9a19\" x=\"593.026563\" y=\"320.78\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_10\">\n",
       "      <!-- human.not-aligned_species-specific -->\n",
       "      <g transform=\"translate(588.817188 645.7175) rotate(-90) scale(0.2 -0.2)\">\n",
       "       <use xlink:href=\"#ArialMT-68\"/>\n",
       "       <use xlink:href=\"#ArialMT-75\" x=\"55.615234\"/>\n",
       "       <use xlink:href=\"#ArialMT-6d\" x=\"111.230469\"/>\n",
       "       <use xlink:href=\"#ArialMT-61\" x=\"194.53125\"/>\n",
       "       <use xlink:href=\"#ArialMT-6e\" x=\"250.146484\"/>\n",
       "       <use xlink:href=\"#ArialMT-2e\" x=\"305.761719\"/>\n",
       "       <use xlink:href=\"#ArialMT-6e\" x=\"333.544922\"/>\n",
       "       <use xlink:href=\"#ArialMT-6f\" x=\"389.160156\"/>\n",
       "       <use xlink:href=\"#ArialMT-74\" x=\"444.775391\"/>\n",
       "       <use xlink:href=\"#ArialMT-2d\" x=\"472.558594\"/>\n",
       "       <use xlink:href=\"#ArialMT-61\" x=\"505.859375\"/>\n",
       "       <use xlink:href=\"#ArialMT-6c\" x=\"561.474609\"/>\n",
       "       <use xlink:href=\"#ArialMT-69\" x=\"583.691406\"/>\n",
       "       <use xlink:href=\"#ArialMT-67\" x=\"605.908203\"/>\n",
       "       <use xlink:href=\"#ArialMT-6e\" x=\"661.523438\"/>\n",
       "       <use xlink:href=\"#ArialMT-65\" x=\"717.138672\"/>\n",
       "       <use xlink:href=\"#ArialMT-64\" x=\"772.753906\"/>\n",
       "       <use xlink:href=\"#ArialMT-5f\" x=\"828.369141\"/>\n",
       "       <use xlink:href=\"#ArialMT-73\" x=\"883.984375\"/>\n",
       "       <use xlink:href=\"#ArialMT-70\" x=\"933.984375\"/>\n",
       "       <use xlink:href=\"#ArialMT-65\" x=\"989.599609\"/>\n",
       "       <use xlink:href=\"#ArialMT-63\" x=\"1045.214844\"/>\n",
       "       <use xlink:href=\"#ArialMT-69\" x=\"1095.214844\"/>\n",
       "       <use xlink:href=\"#ArialMT-65\" x=\"1117.431641\"/>\n",
       "       <use xlink:href=\"#ArialMT-73\" x=\"1173.046875\"/>\n",
       "       <use xlink:href=\"#ArialMT-2d\" x=\"1223.046875\"/>\n",
       "       <use xlink:href=\"#ArialMT-73\" x=\"1256.347656\"/>\n",
       "       <use xlink:href=\"#ArialMT-70\" x=\"1306.347656\"/>\n",
       "       <use xlink:href=\"#ArialMT-65\" x=\"1361.962891\"/>\n",
       "       <use xlink:href=\"#ArialMT-63\" x=\"1417.578125\"/>\n",
       "       <use xlink:href=\"#ArialMT-69\" x=\"1467.578125\"/>\n",
       "       <use xlink:href=\"#ArialMT-66\" x=\"1489.794922\"/>\n",
       "       <use xlink:href=\"#ArialMT-69\" x=\"1517.578125\"/>\n",
       "       <use xlink:href=\"#ArialMT-63\" x=\"1539.794922\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_2\">\n",
       "    <g id=\"ytick_1\">\n",
       "     <g id=\"line2d_11\">\n",
       "      <defs>\n",
       "       <path id=\"m2e775e365d\" d=\"M 0 0 \n",
       "L -3.5 0 \n",
       "\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </defs>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2e775e365d\" x=\"68.246563\" y=\"320.78\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_11\">\n",
       "      <!-- 0 -->\n",
       "      <g transform=\"translate(52.905156 326.148359) scale(0.15 -0.15)\">\n",
       "       <defs>\n",
       "        <path id=\"ArialMT-30\" d=\"M 266 2259 \n",
       "Q 266 3072 433 3567 \n",
       "Q 600 4063 929 4331 \n",
       "Q 1259 4600 1759 4600 \n",
       "Q 2128 4600 2406 4451 \n",
       "Q 2684 4303 2865 4023 \n",
       "Q 3047 3744 3150 3342 \n",
       "Q 3253 2941 3253 2259 \n",
       "Q 3253 1453 3087 958 \n",
       "Q 2922 463 2592 192 \n",
       "Q 2263 -78 1759 -78 \n",
       "Q 1097 -78 719 397 \n",
       "Q 266 969 266 2259 \n",
       "z\n",
       "M 844 2259 \n",
       "Q 844 1131 1108 757 \n",
       "Q 1372 384 1759 384 \n",
       "Q 2147 384 2411 759 \n",
       "Q 2675 1134 2675 2259 \n",
       "Q 2675 3391 2411 3762 \n",
       "Q 2147 4134 1753 4134 \n",
       "Q 1366 4134 1134 3806 \n",
       "Q 844 3388 844 2259 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#ArialMT-30\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_2\">\n",
       "     <g id=\"line2d_12\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2e775e365d\" x=\"68.246563\" y=\"253.205021\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_12\">\n",
       "      <!-- 2000 -->\n",
       "      <g transform=\"translate(27.880938 258.573381) scale(0.15 -0.15)\">\n",
       "       <defs>\n",
       "        <path id=\"ArialMT-32\" d=\"M 3222 541 \n",
       "L 3222 0 \n",
       "L 194 0 \n",
       "Q 188 203 259 391 \n",
       "Q 375 700 629 1000 \n",
       "Q 884 1300 1366 1694 \n",
       "Q 2113 2306 2375 2664 \n",
       "Q 2638 3022 2638 3341 \n",
       "Q 2638 3675 2398 3904 \n",
       "Q 2159 4134 1775 4134 \n",
       "Q 1369 4134 1125 3890 \n",
       "Q 881 3647 878 3216 \n",
       "L 300 3275 \n",
       "Q 359 3922 746 4261 \n",
       "Q 1134 4600 1788 4600 \n",
       "Q 2447 4600 2831 4234 \n",
       "Q 3216 3869 3216 3328 \n",
       "Q 3216 3053 3103 2787 \n",
       "Q 2991 2522 2730 2228 \n",
       "Q 2469 1934 1863 1422 \n",
       "Q 1356 997 1212 845 \n",
       "Q 1069 694 975 541 \n",
       "L 3222 541 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#ArialMT-32\"/>\n",
       "       <use xlink:href=\"#ArialMT-30\" x=\"55.615234\"/>\n",
       "       <use xlink:href=\"#ArialMT-30\" x=\"111.230469\"/>\n",
       "       <use xlink:href=\"#ArialMT-30\" x=\"166.845703\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_3\">\n",
       "     <g id=\"line2d_13\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2e775e365d\" x=\"68.246563\" y=\"185.630042\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_13\">\n",
       "      <!-- 4000 -->\n",
       "      <g transform=\"translate(27.880938 190.998402) scale(0.15 -0.15)\">\n",
       "       <defs>\n",
       "        <path id=\"ArialMT-34\" d=\"M 2069 0 \n",
       "L 2069 1097 \n",
       "L 81 1097 \n",
       "L 81 1613 \n",
       "L 2172 4581 \n",
       "L 2631 4581 \n",
       "L 2631 1613 \n",
       "L 3250 1613 \n",
       "L 3250 1097 \n",
       "L 2631 1097 \n",
       "L 2631 0 \n",
       "L 2069 0 \n",
       "z\n",
       "M 2069 1613 \n",
       "L 2069 3678 \n",
       "L 634 1613 \n",
       "L 2069 1613 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#ArialMT-34\"/>\n",
       "       <use xlink:href=\"#ArialMT-30\" x=\"55.615234\"/>\n",
       "       <use xlink:href=\"#ArialMT-30\" x=\"111.230469\"/>\n",
       "       <use xlink:href=\"#ArialMT-30\" x=\"166.845703\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_4\">\n",
       "     <g id=\"line2d_14\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2e775e365d\" x=\"68.246563\" y=\"118.055063\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_14\">\n",
       "      <!-- 6000 -->\n",
       "      <g transform=\"translate(27.880938 123.423423) scale(0.15 -0.15)\">\n",
       "       <defs>\n",
       "        <path id=\"ArialMT-36\" d=\"M 3184 3459 \n",
       "L 2625 3416 \n",
       "Q 2550 3747 2413 3897 \n",
       "Q 2184 4138 1850 4138 \n",
       "Q 1581 4138 1378 3988 \n",
       "Q 1113 3794 959 3422 \n",
       "Q 806 3050 800 2363 \n",
       "Q 1003 2672 1297 2822 \n",
       "Q 1591 2972 1913 2972 \n",
       "Q 2475 2972 2870 2558 \n",
       "Q 3266 2144 3266 1488 \n",
       "Q 3266 1056 3080 686 \n",
       "Q 2894 316 2569 119 \n",
       "Q 2244 -78 1831 -78 \n",
       "Q 1128 -78 684 439 \n",
       "Q 241 956 241 2144 \n",
       "Q 241 3472 731 4075 \n",
       "Q 1159 4600 1884 4600 \n",
       "Q 2425 4600 2770 4297 \n",
       "Q 3116 3994 3184 3459 \n",
       "z\n",
       "M 888 1484 \n",
       "Q 888 1194 1011 928 \n",
       "Q 1134 663 1356 523 \n",
       "Q 1578 384 1822 384 \n",
       "Q 2178 384 2434 671 \n",
       "Q 2691 959 2691 1453 \n",
       "Q 2691 1928 2437 2201 \n",
       "Q 2184 2475 1800 2475 \n",
       "Q 1419 2475 1153 2201 \n",
       "Q 888 1928 888 1484 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#ArialMT-36\"/>\n",
       "       <use xlink:href=\"#ArialMT-30\" x=\"55.615234\"/>\n",
       "       <use xlink:href=\"#ArialMT-30\" x=\"111.230469\"/>\n",
       "       <use xlink:href=\"#ArialMT-30\" x=\"166.845703\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_5\">\n",
       "     <g id=\"line2d_15\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m2e775e365d\" x=\"68.246563\" y=\"50.480085\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_15\">\n",
       "      <!-- 8000 -->\n",
       "      <g transform=\"translate(27.880938 55.848444) scale(0.15 -0.15)\">\n",
       "       <defs>\n",
       "        <path id=\"ArialMT-38\" d=\"M 1131 2484 \n",
       "Q 781 2613 612 2850 \n",
       "Q 444 3088 444 3419 \n",
       "Q 444 3919 803 4259 \n",
       "Q 1163 4600 1759 4600 \n",
       "Q 2359 4600 2725 4251 \n",
       "Q 3091 3903 3091 3403 \n",
       "Q 3091 3084 2923 2848 \n",
       "Q 2756 2613 2416 2484 \n",
       "Q 2838 2347 3058 2040 \n",
       "Q 3278 1734 3278 1309 \n",
       "Q 3278 722 2862 322 \n",
       "Q 2447 -78 1769 -78 \n",
       "Q 1091 -78 675 323 \n",
       "Q 259 725 259 1325 \n",
       "Q 259 1772 486 2073 \n",
       "Q 713 2375 1131 2484 \n",
       "z\n",
       "M 1019 3438 \n",
       "Q 1019 3113 1228 2906 \n",
       "Q 1438 2700 1772 2700 \n",
       "Q 2097 2700 2305 2904 \n",
       "Q 2513 3109 2513 3406 \n",
       "Q 2513 3716 2298 3927 \n",
       "Q 2084 4138 1766 4138 \n",
       "Q 1444 4138 1231 3931 \n",
       "Q 1019 3725 1019 3438 \n",
       "z\n",
       "M 838 1322 \n",
       "Q 838 1081 952 856 \n",
       "Q 1066 631 1291 507 \n",
       "Q 1516 384 1775 384 \n",
       "Q 2178 384 2440 643 \n",
       "Q 2703 903 2703 1303 \n",
       "Q 2703 1709 2433 1975 \n",
       "Q 2163 2241 1756 2241 \n",
       "Q 1359 2241 1098 1978 \n",
       "Q 838 1716 838 1322 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#ArialMT-38\"/>\n",
       "       <use xlink:href=\"#ArialMT-30\" x=\"55.615234\"/>\n",
       "       <use xlink:href=\"#ArialMT-30\" x=\"111.230469\"/>\n",
       "       <use xlink:href=\"#ArialMT-30\" x=\"166.845703\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"text_16\">\n",
       "     <!-- Number of peaks -->\n",
       "     <g transform=\"translate(20.303438 232.517969) rotate(-90) scale(0.18 -0.18)\">\n",
       "      <defs>\n",
       "       <path id=\"ArialMT-4e\" d=\"M 488 0 \n",
       "L 488 4581 \n",
       "L 1109 4581 \n",
       "L 3516 984 \n",
       "L 3516 4581 \n",
       "L 4097 4581 \n",
       "L 4097 0 \n",
       "L 3475 0 \n",
       "L 1069 3600 \n",
       "L 1069 0 \n",
       "L 488 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"ArialMT-20\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"ArialMT-6b\" d=\"M 425 0 \n",
       "L 425 4581 \n",
       "L 988 4581 \n",
       "L 988 1969 \n",
       "L 2319 3319 \n",
       "L 3047 3319 \n",
       "L 1778 2088 \n",
       "L 3175 0 \n",
       "L 2481 0 \n",
       "L 1384 1697 \n",
       "L 988 1316 \n",
       "L 988 0 \n",
       "L 425 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#ArialMT-4e\"/>\n",
       "      <use xlink:href=\"#ArialMT-75\" x=\"72.216797\"/>\n",
       "      <use xlink:href=\"#ArialMT-6d\" x=\"127.832031\"/>\n",
       "      <use xlink:href=\"#ArialMT-62\" x=\"211.132812\"/>\n",
       "      <use xlink:href=\"#ArialMT-65\" x=\"266.748047\"/>\n",
       "      <use xlink:href=\"#ArialMT-72\" x=\"322.363281\"/>\n",
       "      <use xlink:href=\"#ArialMT-20\" x=\"355.664062\"/>\n",
       "      <use xlink:href=\"#ArialMT-6f\" x=\"383.447266\"/>\n",
       "      <use xlink:href=\"#ArialMT-66\" x=\"439.0625\"/>\n",
       "      <use xlink:href=\"#ArialMT-20\" x=\"466.845703\"/>\n",
       "      <use xlink:href=\"#ArialMT-70\" x=\"494.628906\"/>\n",
       "      <use xlink:href=\"#ArialMT-65\" x=\"550.244141\"/>\n",
       "      <use xlink:href=\"#ArialMT-61\" x=\"605.859375\"/>\n",
       "      <use xlink:href=\"#ArialMT-6b\" x=\"661.474609\"/>\n",
       "      <use xlink:href=\"#ArialMT-73\" x=\"711.474609\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"patch_3\">\n",
       "    <path d=\"M 73.770563 320.78 \n",
       "L 117.962563 320.78 \n",
       "L 117.962563 307.129854 \n",
       "L 73.770563 307.129854 \n",
       "z\n",
       "\" clip-path=\"url(#pec2aaa0362)\" style=\"fill: #bbbbbb\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_4\">\n",
       "    <path d=\"M 129.010563 320.78 \n",
       "L 173.202563 320.78 \n",
       "L 173.202563 227.357592 \n",
       "L 129.010563 227.357592 \n",
       "z\n",
       "\" clip-path=\"url(#pec2aaa0362)\" style=\"fill: #bbbbbb\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_5\">\n",
       "    <path d=\"M 184.250563 320.78 \n",
       "L 228.442563 320.78 \n",
       "L 228.442563 311.285715 \n",
       "L 184.250563 311.285715 \n",
       "z\n",
       "\" clip-path=\"url(#pec2aaa0362)\" style=\"fill: #bbbbbb\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_6\">\n",
       "    <path d=\"M 239.490563 320.78 \n",
       "L 283.682563 320.78 \n",
       "L 283.682563 119.541713 \n",
       "L 239.490563 119.541713 \n",
       "z\n",
       "\" clip-path=\"url(#pec2aaa0362)\" style=\"fill: #bbbbbb\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_7\">\n",
       "    <path d=\"M 294.730563 320.78 \n",
       "L 338.922563 320.78 \n",
       "L 338.922563 214.045321 \n",
       "L 294.730563 214.045321 \n",
       "z\n",
       "\" clip-path=\"url(#pec2aaa0362)\" style=\"fill: #bbbbbb\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_8\">\n",
       "    <path d=\"M 349.970563 320.78 \n",
       "L 394.162563 320.78 \n",
       "L 394.162563 211.240959 \n",
       "L 349.970563 211.240959 \n",
       "z\n",
       "\" clip-path=\"url(#pec2aaa0362)\" style=\"fill: #bbbbbb\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_9\">\n",
       "    <path d=\"M 405.210563 320.78 \n",
       "L 449.402563 320.78 \n",
       "L 449.402563 22.132381 \n",
       "L 405.210563 22.132381 \n",
       "z\n",
       "\" clip-path=\"url(#pec2aaa0362)\" style=\"fill: #bbbbbb\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_10\">\n",
       "    <path d=\"M 460.450563 320.78 \n",
       "L 504.642563 320.78 \n",
       "L 504.642563 29.362904 \n",
       "L 460.450563 29.362904 \n",
       "z\n",
       "\" clip-path=\"url(#pec2aaa0362)\" style=\"fill: #bbbbbb\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_11\">\n",
       "    <path d=\"M 515.690563 320.78 \n",
       "L 559.882563 320.78 \n",
       "L 559.882563 301.081894 \n",
       "L 515.690563 301.081894 \n",
       "z\n",
       "\" clip-path=\"url(#pec2aaa0362)\" style=\"fill: #bbbbbb\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_12\">\n",
       "    <path d=\"M 570.930563 320.78 \n",
       "L 615.122563 320.78 \n",
       "L 615.122563 285.336924 \n",
       "L 570.930563 285.336924 \n",
       "z\n",
       "\" clip-path=\"url(#pec2aaa0362)\" style=\"fill: #bbbbbb\"/>\n",
       "   </g>\n",
       "   <g id=\"line2d_16\">\n",
       "    <path clip-path=\"url(#pec2aaa0362)\" style=\"fill: none; stroke: #424242; stroke-width: 2.7; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"line2d_17\">\n",
       "    <path clip-path=\"url(#pec2aaa0362)\" style=\"fill: none; stroke: #424242; stroke-width: 2.7; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"line2d_18\">\n",
       "    <path clip-path=\"url(#pec2aaa0362)\" style=\"fill: none; stroke: #424242; stroke-width: 2.7; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"line2d_19\">\n",
       "    <path clip-path=\"url(#pec2aaa0362)\" style=\"fill: none; stroke: #424242; stroke-width: 2.7; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"line2d_20\">\n",
       "    <path clip-path=\"url(#pec2aaa0362)\" style=\"fill: none; stroke: #424242; stroke-width: 2.7; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"line2d_21\">\n",
       "    <path clip-path=\"url(#pec2aaa0362)\" style=\"fill: none; stroke: #424242; stroke-width: 2.7; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"line2d_22\">\n",
       "    <path clip-path=\"url(#pec2aaa0362)\" style=\"fill: none; stroke: #424242; stroke-width: 2.7; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"line2d_23\">\n",
       "    <path clip-path=\"url(#pec2aaa0362)\" style=\"fill: none; stroke: #424242; stroke-width: 2.7; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"line2d_24\">\n",
       "    <path clip-path=\"url(#pec2aaa0362)\" style=\"fill: none; stroke: #424242; stroke-width: 2.7; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"line2d_25\">\n",
       "    <path clip-path=\"url(#pec2aaa0362)\" style=\"fill: none; stroke: #424242; stroke-width: 2.7; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_13\">\n",
       "    <path d=\"M 68.246563 320.78 \n",
       "L 68.246563 7.2 \n",
       "\" style=\"fill: none; stroke: #000000; stroke-width: 0.8; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_14\">\n",
       "    <path d=\"M 620.646562 320.78 \n",
       "L 620.646562 7.2 \n",
       "\" style=\"fill: none; stroke: #000000; stroke-width: 0.8; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_15\">\n",
       "    <path d=\"M 68.246563 320.78 \n",
       "L 620.646562 320.78 \n",
       "\" style=\"fill: none; stroke: #000000; stroke-width: 0.8; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_16\">\n",
       "    <path d=\"M 68.246563 7.2 \n",
       "L 620.646562 7.2 \n",
       "\" style=\"fill: none; stroke: #000000; stroke-width: 0.8; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"text_17\">\n",
       "    <!-- 404 -->\n",
       "    <g transform=\"translate(84.188594 304.347354) scale(0.14 -0.14)\">\n",
       "     <use xlink:href=\"#ArialMT-34\"/>\n",
       "     <use xlink:href=\"#ArialMT-30\" x=\"55.615234\"/>\n",
       "     <use xlink:href=\"#ArialMT-34\" x=\"111.230469\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_18\">\n",
       "    <!-- 2,765 -->\n",
       "    <g transform=\"translate(133.59125 224.575092) scale(0.14 -0.14)\">\n",
       "     <defs>\n",
       "      <path id=\"ArialMT-2c\" d=\"M 569 0 \n",
       "L 569 641 \n",
       "L 1209 641 \n",
       "L 1209 0 \n",
       "Q 1209 -353 1084 -570 \n",
       "Q 959 -788 688 -906 \n",
       "L 531 -666 \n",
       "Q 709 -588 793 -436 \n",
       "Q 878 -284 888 0 \n",
       "L 569 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-37\" d=\"M 303 3981 \n",
       "L 303 4522 \n",
       "L 3269 4522 \n",
       "L 3269 4084 \n",
       "Q 2831 3619 2401 2847 \n",
       "Q 1972 2075 1738 1259 \n",
       "Q 1569 684 1522 0 \n",
       "L 944 0 \n",
       "Q 953 541 1156 1306 \n",
       "Q 1359 2072 1739 2783 \n",
       "Q 2119 3494 2547 3981 \n",
       "L 303 3981 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-35\" d=\"M 266 1200 \n",
       "L 856 1250 \n",
       "Q 922 819 1161 601 \n",
       "Q 1400 384 1738 384 \n",
       "Q 2144 384 2425 690 \n",
       "Q 2706 997 2706 1503 \n",
       "Q 2706 1984 2436 2262 \n",
       "Q 2166 2541 1728 2541 \n",
       "Q 1456 2541 1237 2417 \n",
       "Q 1019 2294 894 2097 \n",
       "L 366 2166 \n",
       "L 809 4519 \n",
       "L 3088 4519 \n",
       "L 3088 3981 \n",
       "L 1259 3981 \n",
       "L 1013 2750 \n",
       "Q 1425 3038 1878 3038 \n",
       "Q 2478 3038 2890 2622 \n",
       "Q 3303 2206 3303 1553 \n",
       "Q 3303 931 2941 478 \n",
       "Q 2500 -78 1738 -78 \n",
       "Q 1113 -78 717 272 \n",
       "Q 322 622 266 1200 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "     </defs>\n",
       "     <use xlink:href=\"#ArialMT-32\"/>\n",
       "     <use xlink:href=\"#ArialMT-2c\" x=\"55.615234\"/>\n",
       "     <use xlink:href=\"#ArialMT-37\" x=\"83.398438\"/>\n",
       "     <use xlink:href=\"#ArialMT-36\" x=\"139.013672\"/>\n",
       "     <use xlink:href=\"#ArialMT-35\" x=\"194.628906\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_19\">\n",
       "    <!-- 281 -->\n",
       "    <g transform=\"translate(194.668594 308.503215) scale(0.14 -0.14)\">\n",
       "     <defs>\n",
       "      <path id=\"ArialMT-31\" d=\"M 2384 0 \n",
       "L 1822 0 \n",
       "L 1822 3584 \n",
       "Q 1619 3391 1289 3197 \n",
       "Q 959 3003 697 2906 \n",
       "L 697 3450 \n",
       "Q 1169 3672 1522 3987 \n",
       "Q 1875 4303 2022 4600 \n",
       "L 2384 4600 \n",
       "L 2384 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "     </defs>\n",
       "     <use xlink:href=\"#ArialMT-32\"/>\n",
       "     <use xlink:href=\"#ArialMT-38\" x=\"55.615234\"/>\n",
       "     <use xlink:href=\"#ArialMT-31\" x=\"111.230469\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_20\">\n",
       "    <!-- 5,956 -->\n",
       "    <g transform=\"translate(244.07125 116.759213) scale(0.14 -0.14)\">\n",
       "     <defs>\n",
       "      <path id=\"ArialMT-39\" d=\"M 350 1059 \n",
       "L 891 1109 \n",
       "Q 959 728 1153 556 \n",
       "Q 1347 384 1650 384 \n",
       "Q 1909 384 2104 503 \n",
       "Q 2300 622 2425 820 \n",
       "Q 2550 1019 2634 1356 \n",
       "Q 2719 1694 2719 2044 \n",
       "Q 2719 2081 2716 2156 \n",
       "Q 2547 1888 2255 1720 \n",
       "Q 1963 1553 1622 1553 \n",
       "Q 1053 1553 659 1965 \n",
       "Q 266 2378 266 3053 \n",
       "Q 266 3750 677 4175 \n",
       "Q 1088 4600 1706 4600 \n",
       "Q 2153 4600 2523 4359 \n",
       "Q 2894 4119 3086 3673 \n",
       "Q 3278 3228 3278 2384 \n",
       "Q 3278 1506 3087 986 \n",
       "Q 2897 466 2520 194 \n",
       "Q 2144 -78 1638 -78 \n",
       "Q 1100 -78 759 220 \n",
       "Q 419 519 350 1059 \n",
       "z\n",
       "M 2653 3081 \n",
       "Q 2653 3566 2395 3850 \n",
       "Q 2138 4134 1775 4134 \n",
       "Q 1400 4134 1122 3828 \n",
       "Q 844 3522 844 3034 \n",
       "Q 844 2597 1108 2323 \n",
       "Q 1372 2050 1759 2050 \n",
       "Q 2150 2050 2401 2323 \n",
       "Q 2653 2597 2653 3081 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "     </defs>\n",
       "     <use xlink:href=\"#ArialMT-35\"/>\n",
       "     <use xlink:href=\"#ArialMT-2c\" x=\"55.615234\"/>\n",
       "     <use xlink:href=\"#ArialMT-39\" x=\"83.398438\"/>\n",
       "     <use xlink:href=\"#ArialMT-35\" x=\"139.013672\"/>\n",
       "     <use xlink:href=\"#ArialMT-36\" x=\"194.628906\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_21\">\n",
       "    <!-- 3,159 -->\n",
       "    <g transform=\"translate(299.31125 211.262821) scale(0.14 -0.14)\">\n",
       "     <defs>\n",
       "      <path id=\"ArialMT-33\" d=\"M 269 1209 \n",
       "L 831 1284 \n",
       "Q 928 806 1161 595 \n",
       "Q 1394 384 1728 384 \n",
       "Q 2125 384 2398 659 \n",
       "Q 2672 934 2672 1341 \n",
       "Q 2672 1728 2419 1979 \n",
       "Q 2166 2231 1775 2231 \n",
       "Q 1616 2231 1378 2169 \n",
       "L 1441 2663 \n",
       "Q 1497 2656 1531 2656 \n",
       "Q 1891 2656 2178 2843 \n",
       "Q 2466 3031 2466 3422 \n",
       "Q 2466 3731 2256 3934 \n",
       "Q 2047 4138 1716 4138 \n",
       "Q 1388 4138 1169 3931 \n",
       "Q 950 3725 888 3313 \n",
       "L 325 3413 \n",
       "Q 428 3978 793 4289 \n",
       "Q 1159 4600 1703 4600 \n",
       "Q 2078 4600 2393 4439 \n",
       "Q 2709 4278 2876 4000 \n",
       "Q 3044 3722 3044 3409 \n",
       "Q 3044 3113 2884 2869 \n",
       "Q 2725 2625 2413 2481 \n",
       "Q 2819 2388 3044 2092 \n",
       "Q 3269 1797 3269 1353 \n",
       "Q 3269 753 2831 336 \n",
       "Q 2394 -81 1725 -81 \n",
       "Q 1122 -81 723 278 \n",
       "Q 325 638 269 1209 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "     </defs>\n",
       "     <use xlink:href=\"#ArialMT-33\"/>\n",
       "     <use xlink:href=\"#ArialMT-2c\" x=\"55.615234\"/>\n",
       "     <use xlink:href=\"#ArialMT-31\" x=\"83.398438\"/>\n",
       "     <use xlink:href=\"#ArialMT-35\" x=\"139.013672\"/>\n",
       "     <use xlink:href=\"#ArialMT-39\" x=\"194.628906\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_22\">\n",
       "    <!-- 3,242 -->\n",
       "    <g transform=\"translate(354.55125 208.458459) scale(0.14 -0.14)\">\n",
       "     <use xlink:href=\"#ArialMT-33\"/>\n",
       "     <use xlink:href=\"#ArialMT-2c\" x=\"55.615234\"/>\n",
       "     <use xlink:href=\"#ArialMT-32\" x=\"83.398438\"/>\n",
       "     <use xlink:href=\"#ArialMT-34\" x=\"139.013672\"/>\n",
       "     <use xlink:href=\"#ArialMT-32\" x=\"194.628906\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_23\">\n",
       "    <!-- 8,839 -->\n",
       "    <g transform=\"translate(409.79125 19.349881) scale(0.14 -0.14)\">\n",
       "     <use xlink:href=\"#ArialMT-38\"/>\n",
       "     <use xlink:href=\"#ArialMT-2c\" x=\"55.615234\"/>\n",
       "     <use xlink:href=\"#ArialMT-38\" x=\"83.398438\"/>\n",
       "     <use xlink:href=\"#ArialMT-33\" x=\"139.013672\"/>\n",
       "     <use xlink:href=\"#ArialMT-39\" x=\"194.628906\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_24\">\n",
       "    <!-- 8,625 -->\n",
       "    <g transform=\"translate(465.03125 26.580404) scale(0.14 -0.14)\">\n",
       "     <use xlink:href=\"#ArialMT-38\"/>\n",
       "     <use xlink:href=\"#ArialMT-2c\" x=\"55.615234\"/>\n",
       "     <use xlink:href=\"#ArialMT-36\" x=\"83.398438\"/>\n",
       "     <use xlink:href=\"#ArialMT-32\" x=\"139.013672\"/>\n",
       "     <use xlink:href=\"#ArialMT-35\" x=\"194.628906\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_25\">\n",
       "    <!-- 583 -->\n",
       "    <g transform=\"translate(526.108594 298.299394) scale(0.14 -0.14)\">\n",
       "     <use xlink:href=\"#ArialMT-35\"/>\n",
       "     <use xlink:href=\"#ArialMT-38\" x=\"55.615234\"/>\n",
       "     <use xlink:href=\"#ArialMT-33\" x=\"111.230469\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_26\">\n",
       "    <!-- 1,049 -->\n",
       "    <g transform=\"translate(575.51125 282.554424) scale(0.14 -0.14)\">\n",
       "     <use xlink:href=\"#ArialMT-31\"/>\n",
       "     <use xlink:href=\"#ArialMT-2c\" x=\"55.615234\"/>\n",
       "     <use xlink:href=\"#ArialMT-30\" x=\"83.398438\"/>\n",
       "     <use xlink:href=\"#ArialMT-34\" x=\"139.013672\"/>\n",
       "     <use xlink:href=\"#ArialMT-39\" x=\"194.628906\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "  </g>\n",
       " </g>\n",
       " <defs>\n",
       "  <clipPath id=\"pec2aaa0362\">\n",
       "   <rect x=\"68.246563\" y=\"7.2\" width=\"552.4\" height=\"313.58\"/>\n",
       "  </clipPath>\n",
       " </defs>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<Figure size 900x1000 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig = plt.figure(figsize=(9, 10))\n",
    "ax1 = fig.add_subplot(1, 1, 1)\n",
    "\n",
    "sns.barplot(\n",
    "    data=counts_peaks_orthology_label_3utr.to_frame().reset_index(),\n",
    "    x=\"index\",\n",
    "    y=\"count\",\n",
    "    order=PEAKS_COUNTS_ORDERED_LABELS,\n",
    "    color=\"#BBBBBB\",\n",
    "    ax=ax1,\n",
    ")\n",
    "\n",
    "_ = ax1.set_xticklabels(ax1.get_xticklabels(), rotation=90, ha=\"right\")\n",
    "_ = ax1.set_yticklabels(ax1.get_yticklabels(), fontsize=15)\n",
    "\n",
    "ax1.set_xlabel(\"\")\n",
    "ax1.set_ylabel(\"Count\")\n",
    "\n",
    "# Add the numbers above each bar in the plot\n",
    "for p in ax1.patches:\n",
    "    ax1.annotate(\n",
    "        f\"{int(p.get_height()):,}\",\n",
    "        (p.get_x() + p.get_width() / 2.0, p.get_height()),\n",
    "        ha=\"center\",\n",
    "        va=\"bottom\",\n",
    "        color=\"black\",\n",
    "        size=14,\n",
    "    )\n",
    "\n",
    "ax1.set_ylabel(\"Number of peaks\", fontsize=18)\n",
    "ax1.set_xlabel(\"\", fontsize=18)\n",
    "\n",
    "\n",
    "plt.tight_layout()\n",
    "\n",
    "# if params_export_flag:\n",
    "#    filepath = outputdir / \"cross_species_mappability.barplot.svg\"\n",
    "#    plt.savefig(filepath, bbox_inches=\"tight\", dpi=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a24c897",
   "metadata": {},
   "source": [
    "## EXPORT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6c89c15",
   "metadata": {},
   "outputs": [],
   "source": [
    "h_bound_genes = set(h_full[\"gene_id\"].unique())\n",
    "m_bound_genes = set(m_full[\"gene_id\"].unique())\n",
    "\n",
    "h_bound_genes_3utr = set(h_id_to_meta_3utr.loc[h_umap_3utr[\"identifier\"].values, \"gene_id\"].unique()) | set(\n",
    "    h_id_to_meta_3utr.loc[h_lift_3utr[\"identifier.human_source\"].values, \"gene_id\"].unique()\n",
    ")\n",
    "m_bound_genes_3utr = set(m_id_to_meta_3utr.loc[m_umap_3utr[\"identifier\"].values, \"gene_id\"].unique()) | set(\n",
    "    m_id_to_meta_3utr.loc[m_lift_3utr[\"identifier.mouse_source\"].values, \"gene_id\"].unique()\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b7e54a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TABLES TO BUILD :\n",
    "\n",
    "# - Orthologies mapping table\n",
    "# - Gene lists : [gene_id, status]\n",
    "# - Peak lists : [identifier, status]\n",
    "\n",
    "\n",
    "if params_export_flag:\n",
    "    # ORTHOLOGIES - MAPPING TABLE\n",
    "\n",
    "    export_orth_table = (\n",
    "        DATA[\"orthologs\"]\n",
    "        .loc[\n",
    "            lambda df: ((df[\"mouse_ensid\"].isin(m_bound_genes_3utr)) | (df[\"human_ensid\"].isin(h_bound_genes_3utr))),\n",
    "            :,\n",
    "        ]\n",
    "        .drop_duplicates()\n",
    "        .sort_values(by=[\"mouse_ensid\"])\n",
    "        .reset_index(drop=True)\n",
    "    )\n",
    "\n",
    "    export_orth_table.to_csv(\n",
    "        outputdir / \"mappability.orthologies_table.tsv.gz\",\n",
    "        header=True,\n",
    "        index=False,\n",
    "        sep=\"\\t\",\n",
    "        compression=\"gzip\",\n",
    "    )\n",
    "\n",
    "    # ORTHOLOGIES - ANY BIOTYPE\n",
    "    # =========================\n",
    "\n",
    "    export_h_genes = pd.concat(\n",
    "        [\n",
    "            pd.Series(list(h_genes_bound_orth_complete), name=\"ensembl_id\").to_frame().assign(label=\"bound_ortholog\"),\n",
    "            pd.Series(list(h_genes_no_bound_orth_complete), name=\"ensembl_id\")\n",
    "            .to_frame()\n",
    "            .assign(label=\"species_specific\"),\n",
    "        ],\n",
    "        axis=0,\n",
    "    )\n",
    "\n",
    "    export_m_genes = pd.concat(\n",
    "        [\n",
    "            pd.Series(list(m_genes_bound_orth_complete), name=\"ensembl_id\").to_frame().assign(label=\"bound_ortholog\"),\n",
    "            pd.Series(list(m_genes_no_bound_orth_complete), name=\"ensembl_id\")\n",
    "            .to_frame()\n",
    "            .assign(label=\"species_specific\"),\n",
    "        ],\n",
    "        axis=0,\n",
    "    )\n",
    "\n",
    "    export_h_genes.to_csv(\n",
    "        outputdir / \"mappability.any_biotype.human_genes_labeled.tsv.gz\",\n",
    "        header=True,\n",
    "        index=False,\n",
    "        sep=\"\\t\",\n",
    "        compression=\"gzip\",\n",
    "    )\n",
    "\n",
    "    export_m_genes.to_csv(\n",
    "        outputdir / \"mappability.any_biotype.mouse_genes_labeled.tsv.gz\",\n",
    "        header=True,\n",
    "        index=False,\n",
    "        sep=\"\\t\",\n",
    "        compression=\"gzip\",\n",
    "    )\n",
    "\n",
    "    # ORTHOLOGIES - 3'UTR\n",
    "    # ===================\n",
    "\n",
    "    export_h_genes_3utr = pd.concat(\n",
    "        [\n",
    "            pd.Series(list(h_genes_bound_orth_complete_3utr), name=\"ensembl_id\")\n",
    "            .to_frame()\n",
    "            .assign(label=\"bound_ortholog\"),\n",
    "            pd.Series(list(h_genes_no_bound_orth_complete_3utr), name=\"ensembl_id\")\n",
    "            .to_frame()\n",
    "            .assign(label=\"species_specific\"),\n",
    "        ],\n",
    "        axis=0,\n",
    "    )\n",
    "\n",
    "    export_m_genes_3utr = pd.concat(\n",
    "        [\n",
    "            pd.Series(list(m_genes_bound_orth_complete_3utr), name=\"ensembl_id\")\n",
    "            .to_frame()\n",
    "            .assign(label=\"bound_ortholog\"),\n",
    "            pd.Series(list(m_genes_no_bound_orth_complete_3utr), name=\"ensembl_id\")\n",
    "            .to_frame()\n",
    "            .assign(label=\"species_specific\"),\n",
    "        ],\n",
    "        axis=0,\n",
    "    )\n",
    "\n",
    "    export_h_genes_3utr.to_csv(\n",
    "        outputdir / \"mappability.3UTR.human_genes_labeled.tsv.gz\",\n",
    "        header=True,\n",
    "        index=False,\n",
    "        sep=\"\\t\",\n",
    "        compression=\"gzip\",\n",
    "    )\n",
    "\n",
    "    export_m_genes_3utr.to_csv(\n",
    "        outputdir / \"mappability.3UTR.mouse_genes_labeled.tsv.gz\",\n",
    "        header=True,\n",
    "        index=False,\n",
    "        sep=\"\\t\",\n",
    "        compression=\"gzip\",\n",
    "    )\n",
    "\n",
    "    # PEAKS - ORTHOLOGY LABELS\n",
    "    # ========================\n",
    "\n",
    "    # Export the table of orthologies of peak labels and their labels.\n",
    "\n",
    "    export_h_peaks_orthology_label_3utr = h_peaks_orthology_label_3utr\n",
    "    export_m_peaks_orthology_label_3utr = m_peaks_orthology_label_3utr\n",
    "\n",
    "    export_h_peaks_orthology_label_3utr.to_csv(\n",
    "        outputdir / \"mappability.3UTR.human_peaks_labeled.tsv.gz\",\n",
    "        header=True,\n",
    "        index=False,\n",
    "        sep=\"\\t\",\n",
    "        compression=\"gzip\",\n",
    "    )\n",
    "\n",
    "    export_m_peaks_orthology_label_3utr.to_csv(\n",
    "        outputdir / \"mappability.3UTR.mouse_peaks_labeled.tsv.gz\",\n",
    "        header=True,\n",
    "        index=False,\n",
    "        sep=\"\\t\",\n",
    "        compression=\"gzip\",\n",
    "    )\n",
    "\n",
    "    ###\n",
    "\n",
    "#    with pd.ExcelWriter(\n",
    "#        outputdir / \"mappability.orthologous_binding_genes_and_peaks.xlsx\", engine=\"xlsxwriter\"\n",
    "#    ) as writer:\n",
    "#        export_h_genes.to_excel(writer, sheet_name=\"1_human_anybiotype_genes\")\n",
    "#        export_m_genes.to_excel(writer, sheet_name=\"2_mouse_anybiotype_genes\")\n",
    "#        export_h_genes.to_excel(writer, sheet_name=\"3_human_3UTR_genes\")\n",
    "#        export_m_genes.to_excel(writer, sheet_name=\"4_mouse_3UTR_genes\")\n",
    "#        export_h_peaks_orthology_label_3utr.to_excel(writer, sheet_name=\"5_human_3UTR_peaks\")\n",
    "#        export_m_peaks_orthology_label_3utr.to_excel(writer, sheet_name=\"6_mouse_3UTR_peaks\")\n",
    "#        export_orth_table.to_excel(writer, sheet_name=\"7_orthologies\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hzmdev",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
