{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "63af1ed2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set to True to export results\n",
    "params_export_flag = True\n",
    "\n",
    "# Set to true to include timestamp in output directory name\n",
    "# TODO: actually a bad idea as data dependencies between notebooks rely on stable paths.\n",
    "params_export_timestamp = False\n",
    "\n",
    "# String for a filepath to a config file mapping file identifiers to filepaths.\n",
    "input_config_filepaths = None\n",
    "\n",
    "# Parent directory for output files\n",
    "output_directory = \"../results/\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3295020",
   "metadata": {},
   "source": [
    "# Figure - Comparative genomics - mouse / human gene sets orthologies from annotations of binding sites"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db1bf23b",
   "metadata": {},
   "source": [
    "## Overview"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87aabe7a",
   "metadata": {},
   "source": [
    "Venn diagram to show the species-specific and orthologously bound genes in human and mouse."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56c8f87b",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "445f4f6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import datetime\n",
    "import importlib\n",
    "import os\n",
    "import sys\n",
    "from pathlib import Path\n",
    "import pandera\n",
    "\n",
    "from numbers import Number\n",
    "from typing import Sequence, List, Set, Union, Tuple, Optional\n",
    "\n",
    "import itertools as itt\n",
    "import functools\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import yaml\n",
    "\n",
    "import pybedtools as pbt\n",
    "import pyBigWig as pbw\n",
    "\n",
    "import scipy.stats\n",
    "import statsmodels.stats.multitest\n",
    "from dotmap import DotMap\n",
    "\n",
    "import matplotlib as mpl\n",
    "\n",
    "from dotenv import load_dotenv, find_dotenv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "52f7dea4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib as mpl\n",
    "import matplotlib.font_manager\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib_venn as venn\n",
    "import seaborn as sns\n",
    "\n",
    "from matplotlib import rc\n",
    "import matplotlib_venn as venn\n",
    "import matplotlib_inline\n",
    "\n",
    "try:\n",
    "    matplotlib.font_manager.findfont(\"Arial\")\n",
    "    mpl.rcParams[\"font.sans-serif\"] = [\"Arial\"]\n",
    "except Exception as e:\n",
    "    print(f\"Arial font not found: {e}\")\n",
    "    mpl.rcParams[\"font.sans-serif\"] = [\"DejaVu Sans\"]\n",
    "\n",
    "rc(\"text\", usetex=False)\n",
    "\n",
    "\n",
    "FONT_SMALL_SIZE = 16\n",
    "FONT_MEDIUM_SIZE = 20\n",
    "FONT_BIGGER_SIZE = 22\n",
    "\n",
    "mpl.rc(\"font\", size=FONT_SMALL_SIZE)  # controls default text sizes\n",
    "mpl.rc(\"axes\", titlesize=FONT_SMALL_SIZE)  # fontsize of the axes title\n",
    "mpl.rc(\"axes\", labelsize=FONT_MEDIUM_SIZE)  # fontsize of the x and y labels\n",
    "mpl.rc(\"xtick\", labelsize=FONT_SMALL_SIZE)  # fontsize of the tick labels\n",
    "mpl.rc(\"ytick\", labelsize=FONT_SMALL_SIZE)  # fontsize of the tick labels\n",
    "mpl.rc(\"legend\", fontsize=FONT_SMALL_SIZE)  # legend fontsize\n",
    "mpl.rc(\"figure\", titlesize=FONT_BIGGER_SIZE)  # fontsize of the figure title\n",
    "\n",
    "matplotlib_inline.backend_inline.set_matplotlib_formats(\"svg\")\n",
    "\n",
    "DEFAULT_FIGSIZE = (10, 6)\n",
    "plt.rcParams[\"figure.figsize\"] = DEFAULT_FIGSIZE\n",
    "plt.rcParams[\"axes.spines.top\"] = True\n",
    "plt.rcParams[\"axes.spines.right\"] = True\n",
    "plt.rcParams[\"font.size\"] = FONT_SMALL_SIZE\n",
    "plt.rcParams[\"figure.titlesize\"] = FONT_BIGGER_SIZE\n",
    "plt.rcParams[\"xtick.labelsize\"] = FONT_MEDIUM_SIZE\n",
    "plt.rcParams[\"ytick.labelsize\"] = FONT_MEDIUM_SIZE\n",
    "plt.rcParams[\"axes.axisbelow\"] = True"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "765e4949",
   "metadata": {},
   "source": [
    "## Initialization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "69f1e928",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../results/gene_sets_overlaps\n"
     ]
    }
   ],
   "source": [
    "if params_export_flag:\n",
    "    if params_export_timestamp:\n",
    "        timestamp_format = \"%Y-%m-%d\"\n",
    "\n",
    "        outputdir = Path(output_directory) / f\"gene_sets_overlaps.{datetime.datetime.now().strftime(timestamp_format)}\"\n",
    "\n",
    "    else:\n",
    "        outputdir = Path(output_directory) / \"gene_sets_overlaps\"\n",
    "\n",
    "    if not outputdir.exists():\n",
    "        outputdir.mkdir(parents=True)\n",
    "\n",
    "    print(outputdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "00c529f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "if input_config_filepaths is not None:\n",
    "    if Path(input_config_filepaths).exists():\n",
    "        with open(input_config_filepaths, \"r\") as f:\n",
    "            filepaths = yaml.safe_load(f)\n",
    "    else:\n",
    "        filepaths = {}\n",
    "\n",
    "# NOTE: hard-coded filepaths override config filepaths if defined\n",
    "local_filepaths = {\n",
    "    \"supp_table_binding_sites\": None,\n",
    "    \"mouse_3utr_bs_mappability\": None,\n",
    "    \"human_3utr_bs_mappability\": None,\n",
    "}\n",
    "\n",
    "filepaths = {\n",
    "    **{k: v for k, v in local_filepaths.items() if v is not None},\n",
    "    **filepaths,\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "7334fe6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "FONT_SMALL_SIZE = 16\n",
    "FONT_MEDIUM_SIZE = 20\n",
    "FONT_BIGGER_SIZE = 22\n",
    "\n",
    "mpl.rc(\"font\", size=FONT_SMALL_SIZE)  # controls default text sizes\n",
    "mpl.rc(\"axes\", titlesize=FONT_SMALL_SIZE)  # fontsize of the axes title\n",
    "mpl.rc(\"axes\", labelsize=FONT_MEDIUM_SIZE)  # fontsize of the x and y labels\n",
    "mpl.rc(\"xtick\", labelsize=FONT_SMALL_SIZE)  # fontsize of the tick labels\n",
    "mpl.rc(\"ytick\", labelsize=FONT_SMALL_SIZE)  # fontsize of the tick labels\n",
    "mpl.rc(\"legend\", fontsize=FONT_SMALL_SIZE)  # legend fontsize\n",
    "mpl.rc(\"figure\", titlesize=FONT_BIGGER_SIZE)  # fontsize of the figure title\n",
    "\n",
    "matplotlib_inline.backend_inline.set_matplotlib_formats(\"svg\")\n",
    "\n",
    "plt.rcParams[\"axes.spines.top\"] = False\n",
    "plt.rcParams[\"axes.spines.right\"] = False\n",
    "plt.rcParams[\"font.size\"] = FONT_SMALL_SIZE\n",
    "plt.rcParams[\"figure.titlesize\"] = FONT_BIGGER_SIZE\n",
    "plt.rcParams[\"xtick.labelsize\"] = FONT_MEDIUM_SIZE\n",
    "plt.rcParams[\"ytick.labelsize\"] = FONT_MEDIUM_SIZE\n",
    "plt.rcParams[\"axes.axisbelow\"] = True"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad2ac92f",
   "metadata": {},
   "source": [
    "## Definitions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6795d7b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def has_array_one_value_greater_than_others(arr: Sequence[Number]) -> bool:\n",
    "    if len(arr) == 0:\n",
    "        raise ValueError(\"Empty array\")\n",
    "\n",
    "    if len(arr) == 1:\n",
    "        return True\n",
    "\n",
    "    else:\n",
    "        arr_s = sorted(arr, reverse=True)\n",
    "        if arr_s[0] > arr_s[1]:\n",
    "            return True\n",
    "        else:\n",
    "            return False\n",
    "\n",
    "\n",
    "def build_identifier_bed6(row_bed6: pd.Series) -> str:\n",
    "    \"\"\"Build a sequence identifier as {chrom}:{start}-{end}:{strand} from BED6 row.\"\"\"\n",
    "    return row_bed6[\"chrom\"] + \":\" + str(row_bed6[\"start\"]) + \"-\" + str(row_bed6[\"end\"]) + \":\" + row_bed6[\"strand\"]\n",
    "\n",
    "\n",
    "def series_extract_set_annotation(column: pd.Series) -> list[set[str]]:\n",
    "    expected_format_delimiters = \"NULL;NULL.NULL;;NULL\"\n",
    "    list_extracted_annotations = (\n",
    "        column.replace(np.nan, expected_format_delimiters)\n",
    "        .str.split(\";;\")\n",
    "        .map(lambda v: list(map(lambda elem: elem.split(\";\")[0], v)))\n",
    "        .map(set)\n",
    "    ).tolist()\n",
    "    return list_extracted_annotations\n",
    "\n",
    "\n",
    "def series_extract_first_annotation(column: pd.Series) -> list[str]:\n",
    "    expected_format_delimiters = \"NULL;NULL.NULL;;NULL\"\n",
    "    list_extracted_annotations = (\n",
    "        column.replace(np.nan, expected_format_delimiters).str.split(\";;\").str[0].str.split(\";\").str[0]\n",
    "    ).tolist()\n",
    "    return list_extracted_annotations\n",
    "\n",
    "\n",
    "def series_extract_first_gene_id(column: pd.Series) -> list[str]:\n",
    "    expected_format_delimiters = \";NULL.NULL;;NULL\"\n",
    "    list_extracted_ids = (\n",
    "        column.replace(np.nan, expected_format_delimiters)\n",
    "        .str.split(\";;\")\n",
    "        .str[0]\n",
    "        .str.split(\";\")\n",
    "        .str[-1]\n",
    "        .str.split(\".\")\n",
    "        .str[0]\n",
    "    ).tolist()\n",
    "    return list_extracted_ids\n",
    "\n",
    "\n",
    "def series_extract_set_geneid(column: pd.Series) -> list[set[str]]:\n",
    "    expected_format_delimiters = \"NULL;NULL.NULL;;NULL\"\n",
    "    list_extracted_annotations = (\n",
    "        column.replace(np.nan, expected_format_delimiters)\n",
    "        .str.split(\";;\")\n",
    "        .map(lambda v: list(map(lambda elem: elem.split(\";\")[-1].split(\".\")[0], v)))\n",
    "        .map(set)\n",
    "    ).tolist()\n",
    "    return list_extracted_annotations\n",
    "\n",
    "\n",
    "def complete_bed6_table_with_additional_columns(df: pd.DataFrame) -> pd.DataFrame:\n",
    "    df[\"identifier\"] = df.apply(lambda row: build_identifier_bed6(row), axis=1)\n",
    "    df[\"gene_id\"] = series_extract_first_gene_id(df[\"name\"])\n",
    "    df[\"main_annot\"] = series_extract_first_annotation(df[\"name\"])\n",
    "    df[\"annot_set\"] = series_extract_set_annotation(df[\"name\"])\n",
    "    df[\"gene_id_set\"] = series_extract_set_geneid(df[\"name\"])\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "def load_bed6_table_and_complete_with_additional_columns(path: os.PathLike, columns: list[str]) -> pd.DataFrame:\n",
    "    df = pd.read_csv(\n",
    "        path,\n",
    "        header=None,\n",
    "        index_col=None,\n",
    "        sep=\"\\t\",\n",
    "        names=columns,\n",
    "        comment=\"#\",\n",
    "    )\n",
    "\n",
    "    df = complete_bed6_table_with_additional_columns(df)\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10cac488",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_mouse_orthologies_first_order(\n",
    "    mouse_identifier: str,\n",
    "    human_identifier: str,\n",
    "    m_id_to_meta: pd.DataFrame,\n",
    "    h_id_to_meta: pd.DataFrame,\n",
    "    m_orth_sets: dict[str, set[str]],\n",
    ") -> Optional[bool]:\n",
    "    \"\"\"Check whether the \"best\" gene ID of the source mouse peak is orthologous to the \"best\" gene ID from its lifted coordinates.\"\"\"\n",
    "    m_gene_id = m_id_to_meta.loc[mouse_identifier][\"gene_id\"]\n",
    "    h_gene_id = h_id_to_meta.loc[human_identifier][\"gene_id\"]\n",
    "\n",
    "    if m_gene_id in m_orth_sets:\n",
    "        if h_gene_id in m_orth_sets[m_gene_id]:\n",
    "            return True\n",
    "        else:\n",
    "            return False\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "\n",
    "def check_mouse_orthologies_second_order(\n",
    "    mouse_identifier: str,\n",
    "    human_identifier: str,\n",
    "    m_id_to_meta: pd.DataFrame,\n",
    "    h_id_to_meta: pd.DataFrame,\n",
    "    m_orth_sets: dict[str, set[str]],\n",
    ") -> Optional[bool]:\n",
    "    \"\"\"Check whether the \"best\" gene ID of the source mouse peak is orthologous to any of the gene IDs from its lifted coordinates.\"\"\"\n",
    "    m_gene_id = m_id_to_meta.loc[mouse_identifier][\"gene_id\"]\n",
    "    h_gene_id_set = h_id_to_meta.loc[human_identifier][\"gene_id_set\"]\n",
    "\n",
    "    if m_gene_id in m_orth_sets:\n",
    "        for h_gene_id in h_gene_id_set:\n",
    "            if h_gene_id in m_orth_sets[m_gene_id]:\n",
    "                return True\n",
    "        else:\n",
    "            return False\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "\n",
    "def check_mouse_orthologies_third_order(\n",
    "    mouse_identifier: str,\n",
    "    human_identifier: str,\n",
    "    m_id_to_meta: pd.DataFrame,\n",
    "    h_id_to_meta: pd.DataFrame,\n",
    "    m_orth_sets: dict[str, set[str]],\n",
    ") -> list[tuple[str, Union[list[str], bool, None]]]:\n",
    "    \"\"\"Check whether any  gene ID of the source mouse peak is orthologous to any of the gene IDs from its lifted coordinates.\n",
    "\n",
    "    NOTE: Here the return structure is a bit more elaborated,\n",
    "    as it return all the instances of mouse gene IDs along\n",
    "    with either:\n",
    "\n",
    "    - a string for one of the human gene IDs (found as ortholog)\n",
    "    - or False (if none of the human gene IDs are orthologous)\n",
    "    - or np.nan (if the mouse gene ID is not found in the orthologs table).\n",
    "\n",
    "    \"\"\"\n",
    "    m_gene_id_set = m_id_to_meta.loc[mouse_identifier][\"gene_id_set\"] - set(\n",
    "        \"\",\n",
    "    )\n",
    "    h_gene_id_set = h_id_to_meta.loc[human_identifier][\"gene_id_set\"] - set(\n",
    "        \"\",\n",
    "    )\n",
    "\n",
    "    return_list = []\n",
    "    for m_gene_id in m_gene_id_set:\n",
    "        if m_gene_id in m_orth_sets:\n",
    "            orths = m_orth_sets[m_gene_id]\n",
    "\n",
    "            if orths & h_gene_id_set:\n",
    "                return_list.append((m_gene_id, list(orths & h_gene_id_set)))\n",
    "            else:\n",
    "                return_list.append((m_gene_id, False))\n",
    "\n",
    "        else:\n",
    "            return_list.append((m_gene_id, None))\n",
    "\n",
    "    return return_list\n",
    "\n",
    "\n",
    "####\n",
    "\n",
    "\n",
    "def check_human_orthologies_first_order(\n",
    "    human_identifier: str,\n",
    "    mouse_identifier: str,\n",
    "    h_id_to_meta: pd.DataFrame,\n",
    "    m_id_to_meta: pd.DataFrame,\n",
    "    h_orth_sets: dict[str, set[str]],\n",
    ") -> Optional[bool]:\n",
    "    \"\"\"Check whether the \"best\" gene ID of the source human peak is orthologous to the \"best\" gene ID from its lifted coordinates.\"\"\"\n",
    "    h_gene_id = h_id_to_meta.loc[human_identifier][\"gene_id\"]\n",
    "    m_gene_id = m_id_to_meta.loc[mouse_identifier][\"gene_id\"]\n",
    "\n",
    "    if h_gene_id in h_orth_sets:\n",
    "        if m_gene_id in h_orth_sets[h_gene_id]:\n",
    "            return True\n",
    "        else:\n",
    "            return False\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "\n",
    "def check_human_orthologies_second_order(\n",
    "    human_identifier: str,\n",
    "    mouse_identifier: str,\n",
    "    h_id_to_meta: pd.DataFrame,\n",
    "    m_id_to_meta: pd.DataFrame,\n",
    "    h_orth_sets: dict[str, set[str]],\n",
    ") -> Optional[bool]:\n",
    "    \"\"\"Check whether the \"best\" gene ID of the source human peak is orthologous to any of the gene IDs from its lifted coordinates.\"\"\"\n",
    "    h_gene_id = h_id_to_meta.loc[human_identifier][\"gene_id\"]\n",
    "    m_gene_id_set = m_id_to_meta.loc[mouse_identifier][\"gene_id_set\"]\n",
    "\n",
    "    if h_gene_id in h_orth_sets:\n",
    "        for m_gene_id in m_gene_id_set:\n",
    "            if m_gene_id in h_orth_sets[h_gene_id]:\n",
    "                return True\n",
    "        else:\n",
    "            return False\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "\n",
    "def check_human_orthologies_third_order(\n",
    "    human_identifier: str,\n",
    "    mouse_identifier: str,\n",
    "    h_id_to_meta: pd.DataFrame,\n",
    "    m_id_to_meta: pd.DataFrame,\n",
    "    h_orth_sets: dict[str, set[str]],\n",
    ") -> list[tuple[str, Union[list[str], bool, None]]]:\n",
    "    \"\"\"Check whether any  gene ID of the source mouse peak is orthologous to any of the gene IDs from its lifted coordinates.\n",
    "\n",
    "    NOTE: Here the return structure is a bit more elaborated,\n",
    "    as it return all the instances of mouse gene IDs along\n",
    "    with either:\n",
    "\n",
    "    - a string for one of the human gene IDs (found as ortholog)\n",
    "    - or False (if none of the human gene IDs are orthologous)\n",
    "    - or np.nan (if the mouse gene ID is not found in the orthologs table).\n",
    "    \"\"\"\n",
    "    m_gene_id_set = m_id_to_meta.loc[mouse_identifier][\"gene_id_set\"] - set(\n",
    "        \"\",\n",
    "    )\n",
    "    h_gene_id_set = h_id_to_meta.loc[human_identifier][\"gene_id_set\"] - set(\n",
    "        \"\",\n",
    "    )\n",
    "\n",
    "    return_list = []\n",
    "    for h_gene_id in h_gene_id_set:\n",
    "        if h_gene_id in h_orth_sets:\n",
    "            orths = h_orth_sets[h_gene_id]\n",
    "\n",
    "            if orths & m_gene_id_set:\n",
    "                return_list.append((h_gene_id, list(orths & m_gene_id_set)))\n",
    "            else:\n",
    "                return_list.append((h_gene_id, False))\n",
    "\n",
    "        else:\n",
    "            return_list.append((h_gene_id, None))\n",
    "\n",
    "    return return_list\n",
    "\n",
    "\n",
    "####\n",
    "\n",
    "\n",
    "def flag_from_third_order_orthologies(\n",
    "    orthologies: list[tuple[str, Union[list[str], bool, None]]],\n",
    ") -> bool:\n",
    "    \"\"\"Flag the orthologies from the third order orthologies.\"\"\"\n",
    "    as_bool_or_none_associated_values = []\n",
    "\n",
    "    for gene_id, ortho in orthologies:\n",
    "        if isinstance(ortho, list):\n",
    "            as_bool_or_none_associated_values.append(True)\n",
    "\n",
    "        elif isinstance(ortho, bool) and not ortho:\n",
    "            as_bool_or_none_associated_values.append(False)\n",
    "\n",
    "        else:\n",
    "            as_bool_or_none_associated_values.append(None)\n",
    "\n",
    "    if any(as_bool_or_none_associated_values):\n",
    "        return True\n",
    "\n",
    "    else:\n",
    "        if all([v is None for v in as_bool_or_none_associated_values]):\n",
    "            return None\n",
    "        else:\n",
    "            return False"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed659eec",
   "metadata": {},
   "source": [
    "## Initialization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f740ad7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed6_cols = [\"chrom\", \"start\", \"end\", \"name\", \"score\", \"strand\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "5ed93593",
   "metadata": {},
   "outputs": [],
   "source": [
    "if params_export_flag:\n",
    "    if params_export_timestamp:\n",
    "        timestamp_format = \"%Y-%m-%d\"\n",
    "\n",
    "        outputdir = Path(output_directory) / f\"gene_sets_overlaps.{datetime.datetime.now().strftime(timestamp_format)}\"\n",
    "    else:\n",
    "        outputdir = Path(output_directory) / \"gene_sets_overlaps\"\n",
    "\n",
    "    if not outputdir.exists():\n",
    "        outputdir.mkdir(parents=True)\n",
    "\n",
    "    print(outputdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "810631f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "if input_config_filepaths is not None:\n",
    "    if Path(input_config_filepaths).exists():\n",
    "        with open(input_config_filepaths, \"r\") as f:\n",
    "            filepaths = yaml.safe_load(f)\n",
    "    else:\n",
    "        filepaths = {}\n",
    "\n",
    "# NOTE: hard-coded filepaths override config filepaths if defined\n",
    "local_filepaths = {\n",
    "    \"mouse_chromsizes\": \"../../../roquin_lmoyon/general_storage/ucsc/chromsizes/mm10.chrom.sizes\",\n",
    "    \"human_chromsizes\": \"../../../roquin_lmoyon/general_storage/ucsc/chromsizes/hg38.chrom.sizes\",\n",
    "    \"orthologs\": \"../../../roquin_lmoyon/roquin_storage/orthologs/merged.human_mouse.tsv.gz\",\n",
    "    #\n",
    "    \"mouse_peaks_full\": \"../../../roquin_lmoyon/roquin_storage/20240715-liftover_results/roquin/mouse/test/bs_mouse.bed\",\n",
    "    \"mouse_peaks_mappable\": \"../../../roquin_lmoyon/roquin_storage/20240715-liftover_results/roquin/mouse/test/bs_mouse_mappable.bed\",\n",
    "    \"mouse_peaks_unmappable\": \"../../../roquin_lmoyon/roquin_storage/20240715-liftover_results/roquin/mouse/test/unmapped_bs.txt\",\n",
    "    \"mouse_peaks_lifted\": \"../../../roquin_lmoyon/roquin_storage/20240715-liftover_results/roquin/mouse/test/lifted_mm_to_hg.bed\",\n",
    "    #\n",
    "    \"human_peaks_full\": \"../../../roquin_lmoyon/roquin_storage/20240715-liftover_results/roquin/human/test/bs_human.bed\",\n",
    "    \"human_peaks_mappable\": \"../../../roquin_lmoyon/roquin_storage/20240715-liftover_results/roquin/human/test/bs_human_mappable.bed\",\n",
    "    \"human_peaks_unmappable\": \"../../../roquin_lmoyon/roquin_storage/20240715-liftover_results/roquin/human/test/unmapped_bs.txt\",\n",
    "    \"human_peaks_lifted\": \"../../../roquin_lmoyon/roquin_storage/20240715-liftover_results/roquin/human/test/lifted_hg_to_mm.bed\",\n",
    "    #\n",
    "    #\n",
    "}\n",
    "\n",
    "filepaths = {\n",
    "    **{k: v for k, v in local_filepaths.items() if v is not None},\n",
    "    **local_filepaths,\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5299976a",
   "metadata": {},
   "source": [
    "## Load"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "ff69ae4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA = DotMap()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df037bdf",
   "metadata": {},
   "source": [
    "### Misc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "7ebedbe2",
   "metadata": {},
   "outputs": [],
   "source": [
    "fp = filepaths[\"mouse_chromsizes\"]\n",
    "DATA[\"mouse\"][\"chromsizes_fp\"] = fp\n",
    "DATA[\"mouse\"][\"chromsizes\"] = pd.read_csv(fp, sep=\"\\t\", header=None).set_index(0)[1].to_dict()\n",
    "\n",
    "fp = filepaths[\"human_chromsizes\"]\n",
    "DATA[\"human\"][\"chromsizes_fp\"] = fp\n",
    "DATA[\"human\"][\"chromsizes\"] = pd.read_csv(fp, sep=\"\\t\", header=None).set_index(0)[1].to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "57c8466b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>human_ensid</th>\n",
       "      <th>human_name</th>\n",
       "      <th>mouse_ensid</th>\n",
       "      <th>mouse_name</th>\n",
       "      <th>relationtype_id</th>\n",
       "      <th>supported_by_resource</th>\n",
       "      <th>support_level</th>\n",
       "      <th>support_details</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>ENSMUSG00000067377</td>\n",
       "      <td>Tspan6</td>\n",
       "      <td>1:1</td>\n",
       "      <td>biomart;;omabrowser;;mgi</td>\n",
       "      <td>5</td>\n",
       "      <td>biomart:human_to_mouse;mouse_to_human;;omabrow...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>ENSMUSG00000031250</td>\n",
       "      <td>Tnmd</td>\n",
       "      <td>1:1</td>\n",
       "      <td>biomart;;omabrowser;;mgi</td>\n",
       "      <td>5</td>\n",
       "      <td>biomart:human_to_mouse;mouse_to_human;;omabrow...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>ENSMUSG00000078919</td>\n",
       "      <td>Dpm1</td>\n",
       "      <td>n:1</td>\n",
       "      <td>biomart;;omabrowser;;mgi</td>\n",
       "      <td>5</td>\n",
       "      <td>biomart:human_to_mouse;mouse_to_human;;omabrow...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       human_ensid human_name         mouse_ensid mouse_name relationtype_id  \\\n",
       "0  ENSG00000000003     TSPAN6  ENSMUSG00000067377     Tspan6             1:1   \n",
       "1  ENSG00000000005       TNMD  ENSMUSG00000031250       Tnmd             1:1   \n",
       "2  ENSG00000000419       DPM1  ENSMUSG00000078919       Dpm1             n:1   \n",
       "\n",
       "      supported_by_resource  support_level  \\\n",
       "0  biomart;;omabrowser;;mgi              5   \n",
       "1  biomart;;omabrowser;;mgi              5   \n",
       "2  biomart;;omabrowser;;mgi              5   \n",
       "\n",
       "                                     support_details  \n",
       "0  biomart:human_to_mouse;mouse_to_human;;omabrow...  \n",
       "1  biomart:human_to_mouse;mouse_to_human;;omabrow...  \n",
       "2  biomart:human_to_mouse;mouse_to_human;;omabrow...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(32490, 8)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Table of orthologous genes\n",
    "fp = filepaths[\"orthologs\"]\n",
    "df = pd.read_csv(\n",
    "    fp,\n",
    "    header=0,\n",
    "    index_col=None,\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "\n",
    "\n",
    "display(df.head(3))\n",
    "display(df.shape)\n",
    "\n",
    "DATA[\"orthologs\"] = df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcda7e93",
   "metadata": {},
   "source": [
    "### Mouse peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "25e7c679",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MOUSE PEAKS\n",
      "Number of peaks in the 'full set': 19,759\n",
      "Number of peaks in the 'mappable' set: 17,426\n",
      "Number of peaks in the 'unmapped' set: 2,333\n",
      "Number of peaks in the 'lifted' set: 17,426\n"
     ]
    }
   ],
   "source": [
    "print(\"MOUSE PEAKS\")\n",
    "\n",
    "####\n",
    "\n",
    "df = load_bed6_table_and_complete_with_additional_columns(path=filepaths[\"mouse_peaks_full\"], columns=bed6_cols)\n",
    "\n",
    "print(f\"Number of peaks in the 'full set': {df.shape[0]:,}\")\n",
    "DATA[\"mouse\"][\"peaks\"][\"full\"] = df\n",
    "\n",
    "\n",
    "####\n",
    "\n",
    "df = load_bed6_table_and_complete_with_additional_columns(\n",
    "    path=filepaths[\"mouse_peaks_mappable\"],  # NOTE: content is the same as mouse_bs_mapped.bed\n",
    "    columns=bed6_cols,\n",
    ")\n",
    "print(f\"Number of peaks in the 'mappable' set: {df.shape[0]:,}\")\n",
    "DATA[\"mouse\"][\"peaks\"][\"mappable\"] = df\n",
    "\n",
    "####\n",
    "\n",
    "df = load_bed6_table_and_complete_with_additional_columns(path=filepaths[\"mouse_peaks_unmappable\"], columns=bed6_cols)\n",
    "\n",
    "print(f\"Number of peaks in the 'unmapped' set: {df.shape[0]:,}\")\n",
    "DATA[\"mouse\"][\"peaks\"][\"unmapped\"] = df\n",
    "\n",
    "####\n",
    "\n",
    "\n",
    "df = load_bed6_table_and_complete_with_additional_columns(\n",
    "    path=filepaths[\"mouse_peaks_lifted\"],\n",
    "    columns=bed6_cols + [\"source_identifier\"],\n",
    ")\n",
    "\n",
    "DATA[\"mouse\"][\"peaks\"][\"lifted\"] = df\n",
    "print(f\"Number of peaks in the 'lifted' set: {df.shape[0]:,}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df241802",
   "metadata": {},
   "source": [
    "### Human peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "f325c2c1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HUMAN PEAKS\n",
      "Number of peaks in the 'full' set: 33,492\n",
      "Number of peaks in the 'mappable' set: 29,383\n",
      "Number of peaks in the 'unmapped' set: 4,109\n",
      "Number of peaks in the 'lifted' set: 29,383\n"
     ]
    }
   ],
   "source": [
    "print(\"HUMAN PEAKS\")\n",
    "\n",
    "####\n",
    "\n",
    "df = load_bed6_table_and_complete_with_additional_columns(path=filepaths[\"human_peaks_full\"], columns=bed6_cols)\n",
    "\n",
    "print(f\"Number of peaks in the 'full' set: {df.shape[0]:,}\")\n",
    "DATA[\"human\"][\"peaks\"][\"full\"] = df\n",
    "\n",
    "####\n",
    "\n",
    "df = load_bed6_table_and_complete_with_additional_columns(\n",
    "    path=filepaths[\"human_peaks_mappable\"],  # Same as human_bs_mapped.bed\n",
    "    columns=bed6_cols,\n",
    ")\n",
    "\n",
    "print(f\"Number of peaks in the 'mappable' set: {df.shape[0]:,}\")\n",
    "DATA[\"human\"][\"peaks\"][\"mappable\"] = df\n",
    "\n",
    "####\n",
    "\n",
    "df = load_bed6_table_and_complete_with_additional_columns(path=filepaths[\"human_peaks_unmappable\"], columns=bed6_cols)\n",
    "\n",
    "print(f\"Number of peaks in the 'unmapped' set: {df.shape[0]:,}\")\n",
    "DATA[\"human\"][\"peaks\"][\"unmapped\"] = df\n",
    "\n",
    "####\n",
    "\n",
    "df = load_bed6_table_and_complete_with_additional_columns(\n",
    "    path=filepaths[\"human_peaks_lifted\"],\n",
    "    columns=bed6_cols + [\"source_identifier\"],\n",
    ")\n",
    "\n",
    "print(f\"Number of peaks in the 'lifted' set: {df.shape[0]:,}\")\n",
    "DATA[\"human\"][\"peaks\"][\"lifted\"] = df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6aa0d375",
   "metadata": {},
   "source": [
    "### Verifications"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4dce7d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Verifications :\n",
    "# - regions should be unique\n",
    "# - mappable + unmappable = full\n",
    "\n",
    "m_full = DATA.mouse.peaks.full.loc[:, [\"identifier\", \"name\", \"gene_id\"]].copy()\n",
    "m_map = DATA.mouse.peaks.mappable.loc[:, [\"identifier\", \"name\", \"gene_id\"]].copy()\n",
    "m_umap = DATA.mouse.peaks.unmapped.loc[:, [\"identifier\", \"name\", \"gene_id\"]].copy()\n",
    "\n",
    "assert m_full[\"identifier\"].duplicated().sum() == 0\n",
    "assert m_map[\"identifier\"].duplicated().sum() == 0\n",
    "assert m_umap[\"identifier\"].duplicated().sum() == 0\n",
    "assert set(m_map[\"identifier\"].unique()) | set(m_umap[\"identifier\"].unique()) == set(m_full[\"identifier\"].unique())\n",
    "\n",
    "\n",
    "h_full = DATA.human.peaks.full.loc[:, [\"identifier\", \"name\", \"gene_id\"]].copy()\n",
    "h_map = DATA.human.peaks.mappable.loc[:, [\"identifier\", \"name\", \"gene_id\"]].copy()\n",
    "h_umap = DATA.human.peaks.unmapped.loc[:, [\"identifier\", \"name\", \"gene_id\"]].copy()\n",
    "\n",
    "assert h_full[\"identifier\"].duplicated().sum() == 0\n",
    "assert h_map[\"identifier\"].duplicated().sum() == 0\n",
    "assert h_umap[\"identifier\"].duplicated().sum() == 0\n",
    "assert set(h_map[\"identifier\"].unique()) | set(h_umap[\"identifier\"].unique()) == set(h_full[\"identifier\"].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80407533",
   "metadata": {},
   "source": [
    "## Orthologies - any biotype"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3e7b360",
   "metadata": {},
   "source": [
    "- We have assembled a table of mouse-human orthologous genes from multiple resources.\n",
    "- We have lifted-over peaks from mouse to human genome and vice versa.\n",
    "- We can establish the types of orthologies from lifted peaks establishing the orthology between the source and lifted annotation (orthologous gene or not, same biotype or not, etc.)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55da9471",
   "metadata": {},
   "source": [
    "### Gather tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "55556da0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'m_id_to_meta.shape=(48126, 5)'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>main_annot</th>\n",
       "      <th>annot_set</th>\n",
       "      <th>gene_id_set</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:6274997-6275006:+</th>\n",
       "      <td>three_prime_UTR;Rb1cc1;protein_coding;ENSMUSG0...</td>\n",
       "      <td>ENSMUSG00000025907</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{three_prime_UTR}</td>\n",
       "      <td>{ENSMUSG00000025907}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:7170406-7170415:+</th>\n",
       "      <td>three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...</td>\n",
       "      <td>ENSMUSG00000051285</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{three_prime_UTR}</td>\n",
       "      <td>{ENSMUSG00000051285}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:7170431-7170440:+</th>\n",
       "      <td>three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...</td>\n",
       "      <td>ENSMUSG00000051285</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{three_prime_UTR}</td>\n",
       "      <td>{ENSMUSG00000051285}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                     name  \\\n",
       "identifier                                                                  \n",
       "chr1:6274997-6275006:+  three_prime_UTR;Rb1cc1;protein_coding;ENSMUSG0...   \n",
       "chr1:7170406-7170415:+  three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...   \n",
       "chr1:7170431-7170440:+  three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...   \n",
       "\n",
       "                                   gene_id       main_annot  \\\n",
       "identifier                                                    \n",
       "chr1:6274997-6275006:+  ENSMUSG00000025907  three_prime_UTR   \n",
       "chr1:7170406-7170415:+  ENSMUSG00000051285  three_prime_UTR   \n",
       "chr1:7170431-7170440:+  ENSMUSG00000051285  three_prime_UTR   \n",
       "\n",
       "                                annot_set           gene_id_set  \n",
       "identifier                                                       \n",
       "chr1:6274997-6275006:+  {three_prime_UTR}  {ENSMUSG00000025907}  \n",
       "chr1:7170406-7170415:+  {three_prime_UTR}  {ENSMUSG00000051285}  \n",
       "chr1:7170431-7170440:+  {three_prime_UTR}  {ENSMUSG00000051285}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'m_id_to_bed6.shape=(49141, 6)'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:6274997-6275006:+</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6274997</td>\n",
       "      <td>6275006</td>\n",
       "      <td>three_prime_UTR;Rb1cc1;protein_coding;ENSMUSG0...</td>\n",
       "      <td>9.344450</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:7170406-7170415:+</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7170406</td>\n",
       "      <td>7170415</td>\n",
       "      <td>three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...</td>\n",
       "      <td>5.094060</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:7170431-7170440:+</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7170431</td>\n",
       "      <td>7170440</td>\n",
       "      <td>three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...</td>\n",
       "      <td>26.054101</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       chrom    start      end  \\\n",
       "identifier                                       \n",
       "chr1:6274997-6275006:+  chr1  6274997  6275006   \n",
       "chr1:7170406-7170415:+  chr1  7170406  7170415   \n",
       "chr1:7170431-7170440:+  chr1  7170431  7170440   \n",
       "\n",
       "                                                                     name  \\\n",
       "identifier                                                                  \n",
       "chr1:6274997-6275006:+  three_prime_UTR;Rb1cc1;protein_coding;ENSMUSG0...   \n",
       "chr1:7170406-7170415:+  three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...   \n",
       "chr1:7170431-7170440:+  three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...   \n",
       "\n",
       "                            score strand  \n",
       "identifier                                \n",
       "chr1:6274997-6275006:+   9.344450      +  \n",
       "chr1:7170406-7170415:+   5.094060      +  \n",
       "chr1:7170431-7170440:+  26.054101      +  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'m_lift.shape=(17426, 2)'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier.mouse_source</th>\n",
       "      <th>identifier.human_target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:6274997-6275006:+</td>\n",
       "      <td>chr8:52623049-52623059:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:7170406-7170415:+</td>\n",
       "      <td>chr8:51819766-51819775:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1:7170431-7170440:+</td>\n",
       "      <td>chr8:51819741-51819750:-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  identifier.mouse_source   identifier.human_target\n",
       "0  chr1:6274997-6275006:+  chr8:52623049-52623059:-\n",
       "1  chr1:7170406-7170415:+  chr8:51819766-51819775:-\n",
       "2  chr1:7170431-7170440:+  chr8:51819741-51819750:-"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "m_id_to_meta = (\n",
    "    pd.concat(\n",
    "        [\n",
    "            DATA.mouse.peaks.full.loc[\n",
    "                :, [\"identifier\", \"name\", \"gene_id\", \"main_annot\", \"annot_set\", \"gene_id_set\"]\n",
    "            ].copy(),\n",
    "            DATA.human.peaks.lifted.loc[\n",
    "                :, [\"identifier\", \"name\", \"gene_id\", \"main_annot\", \"annot_set\", \"gene_id_set\"]\n",
    "            ].copy(),\n",
    "        ],\n",
    "        axis=0,\n",
    "    )\n",
    "    .drop_duplicates(subset=\"identifier\")\n",
    "    .set_index(\"identifier\")\n",
    ")\n",
    "\n",
    "display(f\"{m_id_to_meta.shape=}\")\n",
    "display(m_id_to_meta.head(3))\n",
    "\n",
    "m_id_to_bed6 = (\n",
    "    pd.concat(\n",
    "        [\n",
    "            DATA.mouse.peaks.full.loc[:, bed6_cols + [\"identifier\"]],\n",
    "            DATA.human.peaks.lifted.loc[:, bed6_cols + [\"identifier\"]],\n",
    "        ],\n",
    "        axis=0,\n",
    "    )\n",
    "    .drop_duplicates()\n",
    "    .set_index(\"identifier\")\n",
    ")\n",
    "\n",
    "display(f\"{m_id_to_bed6.shape=}\")\n",
    "display(m_id_to_bed6.head(3))\n",
    "\n",
    "m_lift = (\n",
    "    DATA.mouse.peaks.lifted.loc[:, [\"source_identifier\", \"identifier\"]]\n",
    "    .copy()\n",
    "    .rename(columns={\"source_identifier\": \"identifier.mouse_source\", \"identifier\": \"identifier.human_target\"})\n",
    ")\n",
    "display(f\"{m_lift.shape=}\")\n",
    "display(m_lift.head(3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "5e8905a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'h_id_to_meta.shape=(49904, 5)'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>main_annot</th>\n",
       "      <th>annot_set</th>\n",
       "      <th>gene_id_set</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:1211398-1211407:-</th>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>ENSG00000186827</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{three_prime_UTR}</td>\n",
       "      <td>{ENSG00000186827}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:1211415-1211424:-</th>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>ENSG00000186827</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{three_prime_UTR}</td>\n",
       "      <td>{ENSG00000186827}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:1211431-1211440:-</th>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>ENSG00000186827</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{three_prime_UTR}</td>\n",
       "      <td>{ENSG00000186827}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                     name  \\\n",
       "identifier                                                                  \n",
       "chr1:1211398-1211407:-  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "chr1:1211415-1211424:-  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "chr1:1211431-1211440:-  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "\n",
       "                                gene_id       main_annot          annot_set  \\\n",
       "identifier                                                                    \n",
       "chr1:1211398-1211407:-  ENSG00000186827  three_prime_UTR  {three_prime_UTR}   \n",
       "chr1:1211415-1211424:-  ENSG00000186827  three_prime_UTR  {three_prime_UTR}   \n",
       "chr1:1211431-1211440:-  ENSG00000186827  three_prime_UTR  {three_prime_UTR}   \n",
       "\n",
       "                              gene_id_set  \n",
       "identifier                                 \n",
       "chr1:1211398-1211407:-  {ENSG00000186827}  \n",
       "chr1:1211415-1211424:-  {ENSG00000186827}  \n",
       "chr1:1211431-1211440:-  {ENSG00000186827}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'h_id_to_bed6.shape=(50917, 6)'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:1211398-1211407:-</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1211398</td>\n",
       "      <td>1211407</td>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>281.428788</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:1211415-1211424:-</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1211415</td>\n",
       "      <td>1211424</td>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>231.774900</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:1211431-1211440:-</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1211431</td>\n",
       "      <td>1211440</td>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>206.707600</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       chrom    start      end  \\\n",
       "identifier                                       \n",
       "chr1:1211398-1211407:-  chr1  1211398  1211407   \n",
       "chr1:1211415-1211424:-  chr1  1211415  1211424   \n",
       "chr1:1211431-1211440:-  chr1  1211431  1211440   \n",
       "\n",
       "                                                                     name  \\\n",
       "identifier                                                                  \n",
       "chr1:1211398-1211407:-  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "chr1:1211415-1211424:-  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "chr1:1211431-1211440:-  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "\n",
       "                             score strand  \n",
       "identifier                                 \n",
       "chr1:1211398-1211407:-  281.428788      -  \n",
       "chr1:1211415-1211424:-  231.774900      -  \n",
       "chr1:1211431-1211440:-  206.707600      -  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'h_lift.shape=(29383, 2)'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier.human_source</th>\n",
       "      <th>identifier.mouse_target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:1211398-1211407:-</td>\n",
       "      <td>chr4:156016532-156016539:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:1211415-1211424:-</td>\n",
       "      <td>chr4:156016517-156016524:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1:1211431-1211440:-</td>\n",
       "      <td>chr4:156016501-156016510:+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  identifier.human_source     identifier.mouse_target\n",
       "0  chr1:1211398-1211407:-  chr4:156016532-156016539:+\n",
       "1  chr1:1211415-1211424:-  chr4:156016517-156016524:+\n",
       "2  chr1:1211431-1211440:-  chr4:156016501-156016510:+"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "h_id_to_meta = (\n",
    "    pd.concat(\n",
    "        [\n",
    "            DATA.human.peaks.full.loc[\n",
    "                :, [\"identifier\", \"name\", \"gene_id\", \"main_annot\", \"annot_set\", \"gene_id_set\"]\n",
    "            ].copy(),\n",
    "            DATA.mouse.peaks.lifted.loc[\n",
    "                :, [\"identifier\", \"name\", \"gene_id\", \"main_annot\", \"annot_set\", \"gene_id_set\"]\n",
    "            ].copy(),\n",
    "        ],\n",
    "        axis=0,\n",
    "    )\n",
    "    .drop_duplicates(subset=\"identifier\")\n",
    "    .set_index(\"identifier\")\n",
    ")\n",
    "\n",
    "\n",
    "display(f\"{h_id_to_meta.shape=}\")\n",
    "display(h_id_to_meta.head(3))\n",
    "\n",
    "h_id_to_bed6 = (\n",
    "    pd.concat(\n",
    "        [\n",
    "            DATA.human.peaks.full.loc[:, bed6_cols + [\"identifier\"]],\n",
    "            DATA.mouse.peaks.lifted.loc[:, bed6_cols + [\"identifier\"]],\n",
    "        ],\n",
    "        axis=0,\n",
    "    )\n",
    "    .drop_duplicates()\n",
    "    .set_index(\"identifier\")\n",
    ")\n",
    "\n",
    "display(f\"{h_id_to_bed6.shape=}\")\n",
    "display(h_id_to_bed6.head(3))\n",
    "\n",
    "h_lift = (\n",
    "    DATA.human.peaks.lifted.loc[:, [\"source_identifier\", \"identifier\"]]\n",
    "    .copy()\n",
    "    .rename(columns={\"source_identifier\": \"identifier.human_source\", \"identifier\": \"identifier.mouse_target\"})\n",
    ")\n",
    "\n",
    "display(f\"{h_lift.shape=}\")\n",
    "display(h_lift.head(3))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2dfa3d93",
   "metadata": {},
   "source": [
    "### Demo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf46b161",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier.mouse_source</th>\n",
       "      <th>identifier.human_target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14339</th>\n",
       "      <td>chr1:160973494-160973503:+</td>\n",
       "      <td>chr1:173932787-173932789:-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          identifier.mouse_source     identifier.human_target\n",
       "14339  chr1:160973494-160973503:+  chr1:173932787-173932789:-"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier.human_source</th>\n",
       "      <th>identifier.mouse_target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1840</th>\n",
       "      <td>chr1:173932765-173932774:-</td>\n",
       "      <td>chr1:160973516-160973527:+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         identifier.human_source     identifier.mouse_target\n",
       "1840  chr1:173932765-173932774:-  chr1:160973516-160973527:+"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>main_annot</th>\n",
       "      <th>annot_set</th>\n",
       "      <th>gene_id_set</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:160973494-160973503:+</th>\n",
       "      <td>three_prime_UTR;Rc3h1;protein_coding;ENSMUSG00...</td>\n",
       "      <td>ENSMUSG00000040423</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{three_prime_UTR}</td>\n",
       "      <td>{ENSMUSG00000040423}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:160973516-160973527:+</th>\n",
       "      <td>three_prime_UTR;Rc3h1;protein_coding;ENSMUSG00...</td>\n",
       "      <td>ENSMUSG00000040423</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{three_prime_UTR}</td>\n",
       "      <td>{ENSMUSG00000040423}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                         name  \\\n",
       "identifier                                                                      \n",
       "chr1:160973494-160973503:+  three_prime_UTR;Rc3h1;protein_coding;ENSMUSG00...   \n",
       "chr1:160973516-160973527:+  three_prime_UTR;Rc3h1;protein_coding;ENSMUSG00...   \n",
       "\n",
       "                                       gene_id       main_annot  \\\n",
       "identifier                                                        \n",
       "chr1:160973494-160973503:+  ENSMUSG00000040423  three_prime_UTR   \n",
       "chr1:160973516-160973527:+  ENSMUSG00000040423  three_prime_UTR   \n",
       "\n",
       "                                    annot_set           gene_id_set  \n",
       "identifier                                                           \n",
       "chr1:160973494-160973503:+  {three_prime_UTR}  {ENSMUSG00000040423}  \n",
       "chr1:160973516-160973527:+  {three_prime_UTR}  {ENSMUSG00000040423}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Here : example of Rc3h1 (Roquin) with two close-by peaks in each species.\n",
    "\n",
    "display(m_lift.loc[lambda df: df[\"identifier.mouse_source\"] == \"chr1:160973494-160973503:+\", :])\n",
    "display(h_lift.loc[lambda df: df[\"identifier.human_source\"] == \"chr1:173932765-173932774:-\"])\n",
    "\n",
    "display(\n",
    "    m_id_to_meta.loc[\n",
    "        [\n",
    "            \"chr1:160973494-160973503:+\",\n",
    "            \"chr1:160973516-160973527:+\",\n",
    "        ],\n",
    "        :,\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fe7bc41",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1    49730\n",
       "2      174\n",
       "Name: gene_id_set, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>main_annot</th>\n",
       "      <th>annot_set</th>\n",
       "      <th>gene_id_set</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr5:131462001-131462010:-</th>\n",
       "      <td>CDS;ENSG00000273217;protein_coding;ENSG0000027...</td>\n",
       "      <td>ENSG00000273217</td>\n",
       "      <td>CDS</td>\n",
       "      <td>{CDS}</td>\n",
       "      <td>{ENSG00000158987, ENSG00000273217}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr5:131427093-131427102:-</th>\n",
       "      <td>three_prime_UTR;ENSG00000273217;protein_coding...</td>\n",
       "      <td>ENSG00000273217</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{three_prime_UTR}</td>\n",
       "      <td>{ENSG00000158987, ENSG00000273217}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr5:131427067-131427076:-</th>\n",
       "      <td>three_prime_UTR;ENSG00000273217;protein_coding...</td>\n",
       "      <td>ENSG00000273217</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{three_prime_UTR}</td>\n",
       "      <td>{ENSG00000158987, ENSG00000273217}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr5:131426963-131426972:-</th>\n",
       "      <td>three_prime_UTR;ENSG00000273217;protein_coding...</td>\n",
       "      <td>ENSG00000273217</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{three_prime_UTR}</td>\n",
       "      <td>{ENSG00000158987, ENSG00000273217}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr5:131426926-131426935:-</th>\n",
       "      <td>three_prime_UTR;ENSG00000273217;protein_coding...</td>\n",
       "      <td>ENSG00000273217</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{three_prime_UTR}</td>\n",
       "      <td>{ENSG00000158987, ENSG00000273217}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                         name  \\\n",
       "identifier                                                                      \n",
       "chr5:131462001-131462010:-  CDS;ENSG00000273217;protein_coding;ENSG0000027...   \n",
       "chr5:131427093-131427102:-  three_prime_UTR;ENSG00000273217;protein_coding...   \n",
       "chr5:131427067-131427076:-  three_prime_UTR;ENSG00000273217;protein_coding...   \n",
       "chr5:131426963-131426972:-  three_prime_UTR;ENSG00000273217;protein_coding...   \n",
       "chr5:131426926-131426935:-  three_prime_UTR;ENSG00000273217;protein_coding...   \n",
       "\n",
       "                                    gene_id       main_annot  \\\n",
       "identifier                                                     \n",
       "chr5:131462001-131462010:-  ENSG00000273217              CDS   \n",
       "chr5:131427093-131427102:-  ENSG00000273217  three_prime_UTR   \n",
       "chr5:131427067-131427076:-  ENSG00000273217  three_prime_UTR   \n",
       "chr5:131426963-131426972:-  ENSG00000273217  three_prime_UTR   \n",
       "chr5:131426926-131426935:-  ENSG00000273217  three_prime_UTR   \n",
       "\n",
       "                                    annot_set  \\\n",
       "identifier                                      \n",
       "chr5:131462001-131462010:-              {CDS}   \n",
       "chr5:131427093-131427102:-  {three_prime_UTR}   \n",
       "chr5:131427067-131427076:-  {three_prime_UTR}   \n",
       "chr5:131426963-131426972:-  {three_prime_UTR}   \n",
       "chr5:131426926-131426935:-  {three_prime_UTR}   \n",
       "\n",
       "                                                   gene_id_set  \n",
       "identifier                                                      \n",
       "chr5:131462001-131462010:-  {ENSG00000158987, ENSG00000273217}  \n",
       "chr5:131427093-131427102:-  {ENSG00000158987, ENSG00000273217}  \n",
       "chr5:131427067-131427076:-  {ENSG00000158987, ENSG00000273217}  \n",
       "chr5:131426963-131426972:-  {ENSG00000158987, ENSG00000273217}  \n",
       "chr5:131426926-131426935:-  {ENSG00000158987, ENSG00000273217}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "1    47988\n",
       "2      138\n",
       "Name: gene_id_set, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>main_annot</th>\n",
       "      <th>annot_set</th>\n",
       "      <th>gene_id_set</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr4:151056815-151056821:-</th>\n",
       "      <td>intron;Camta1;protein_coding;ENSMUSG0000001459...</td>\n",
       "      <td>ENSMUSG00000014592</td>\n",
       "      <td>intron</td>\n",
       "      <td>{intron}</td>\n",
       "      <td>{ENSMUSG00000028955, ENSMUSG00000014592}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr3:90605025-90605034:+</th>\n",
       "      <td>CDS;Gm42674;protein_coding;ENSMUSG00000105518....</td>\n",
       "      <td>ENSMUSG00000105518</td>\n",
       "      <td>CDS</td>\n",
       "      <td>{CDS}</td>\n",
       "      <td>{ENSMUSG00000001020, ENSMUSG00000105518}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr3:90056248-90056257:+</th>\n",
       "      <td>CDS;4933434E20Rik;protein_coding;ENSMUSG000000...</td>\n",
       "      <td>ENSMUSG00000027942</td>\n",
       "      <td>CDS</td>\n",
       "      <td>{CDS}</td>\n",
       "      <td>{ENSMUSG00000027942, ENSMUSG00000118504}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr3:90056222-90056231:+</th>\n",
       "      <td>CDS;4933434E20Rik;protein_coding;ENSMUSG000000...</td>\n",
       "      <td>ENSMUSG00000027942</td>\n",
       "      <td>CDS</td>\n",
       "      <td>{CDS}</td>\n",
       "      <td>{ENSMUSG00000027942, ENSMUSG00000118504}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr3:90053594-90053603:+</th>\n",
       "      <td>CDS;4933434E20Rik;protein_coding;ENSMUSG000000...</td>\n",
       "      <td>ENSMUSG00000027942</td>\n",
       "      <td>CDS</td>\n",
       "      <td>{CDS}</td>\n",
       "      <td>{ENSMUSG00000027942, ENSMUSG00000118504}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                         name  \\\n",
       "identifier                                                                      \n",
       "chr4:151056815-151056821:-  intron;Camta1;protein_coding;ENSMUSG0000001459...   \n",
       "chr3:90605025-90605034:+    CDS;Gm42674;protein_coding;ENSMUSG00000105518....   \n",
       "chr3:90056248-90056257:+    CDS;4933434E20Rik;protein_coding;ENSMUSG000000...   \n",
       "chr3:90056222-90056231:+    CDS;4933434E20Rik;protein_coding;ENSMUSG000000...   \n",
       "chr3:90053594-90053603:+    CDS;4933434E20Rik;protein_coding;ENSMUSG000000...   \n",
       "\n",
       "                                       gene_id main_annot annot_set  \\\n",
       "identifier                                                            \n",
       "chr4:151056815-151056821:-  ENSMUSG00000014592     intron  {intron}   \n",
       "chr3:90605025-90605034:+    ENSMUSG00000105518        CDS     {CDS}   \n",
       "chr3:90056248-90056257:+    ENSMUSG00000027942        CDS     {CDS}   \n",
       "chr3:90056222-90056231:+    ENSMUSG00000027942        CDS     {CDS}   \n",
       "chr3:90053594-90053603:+    ENSMUSG00000027942        CDS     {CDS}   \n",
       "\n",
       "                                                         gene_id_set  \n",
       "identifier                                                            \n",
       "chr4:151056815-151056821:-  {ENSMUSG00000028955, ENSMUSG00000014592}  \n",
       "chr3:90605025-90605034:+    {ENSMUSG00000001020, ENSMUSG00000105518}  \n",
       "chr3:90056248-90056257:+    {ENSMUSG00000027942, ENSMUSG00000118504}  \n",
       "chr3:90056222-90056231:+    {ENSMUSG00000027942, ENSMUSG00000118504}  \n",
       "chr3:90053594-90053603:+    {ENSMUSG00000027942, ENSMUSG00000118504}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Some peaks have multiple possible gene IDs.\n",
    "\n",
    "display(h_id_to_meta[\"gene_id_set\"].apply(len).value_counts())\n",
    "display(h_id_to_meta.loc[lambda df: df[\"gene_id_set\"].apply(len) > 1, :].head(5))\n",
    "\n",
    "\n",
    "display(m_id_to_meta[\"gene_id_set\"].apply(len).value_counts())\n",
    "display(m_id_to_meta.loc[lambda df: df[\"gene_id_set\"].apply(len) > 1, :].head(5))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d4f4bb4",
   "metadata": {},
   "source": [
    "### Establish orthologies"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f4d1f09",
   "metadata": {},
   "source": [
    "We will use the sets of human and mouse gene IDs associated to the lifted peaks\n",
    "and check whether there is any orthology between the two sets for each gene.\n",
    "\n",
    "NOTE: we can define \"first orthology\" and \"second orthology\" as checking that the \"best\" gene ID are orthologous, and then checking that whether the gene sets contain orthologies.\n",
    "\n",
    "\n",
    "Further, we add back the unmappable peaks to check whether their bound genes are within \n",
    "the orthologous gene sets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e2208a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "orth = DATA.orthologs.copy()\n",
    "\n",
    "m_orth_sets = orth.groupby(\"mouse_ensid\").apply(lambda g: set(g[\"human_ensid\"].values)).to_dict()\n",
    "h_orth_sets = orth.groupby(\"human_ensid\").apply(lambda g: set(g[\"mouse_ensid\"].values)).to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe8715b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "m_lift[\"orth_first_order\"] = m_lift.apply(\n",
    "    lambda row: check_mouse_orthologies_first_order(row[\"identifier.mouse_source\"], row[\"identifier.human_target\"]),\n",
    "    axis=1,\n",
    ")\n",
    "m_lift[\"orth_second_order\"] = m_lift.apply(\n",
    "    lambda row: check_mouse_orthologies_second_order(row[\"identifier.mouse_source\"], row[\"identifier.human_target\"]),\n",
    "    axis=1,\n",
    ")\n",
    "m_lift[\"orth_third_order.details\"] = m_lift.apply(\n",
    "    lambda row: check_mouse_orthologies_third_order(row[\"identifier.mouse_source\"], row[\"identifier.human_target\"]),\n",
    "    axis=1,\n",
    ")\n",
    "m_lift[\"orth_third_order\"] = m_lift[\"orth_third_order.details\"].apply(flag_from_third_order_orthologies)\n",
    "\n",
    "\n",
    "# Now assign the orthology status of the gene.\n",
    "m_lift[\"gene_orthology_status\"] = \"not_orthologous\"\n",
    "m_lift.loc[lambda df: (df[\"orth_first_order\"].replace(np.nan, False)), \"gene_orthology_status\"] = \"first\"\n",
    "m_lift.loc[\n",
    "    lambda df: (~df[\"orth_first_order\"].replace(np.nan, False)) & (df[\"orth_second_order\"].replace(np.nan, False)),\n",
    "    \"gene_orthology_status\",\n",
    "] = \"second\"\n",
    "m_lift.loc[\n",
    "    lambda df: (\n",
    "        (~df[\"orth_first_order\"].replace(np.nan, False))\n",
    "        & (~df[\"orth_second_order\"].replace(np.nan, False))\n",
    "        & (df[\"orth_third_order\"].replace(np.nan, False))\n",
    "    ),\n",
    "    \"gene_orthology_status\",\n",
    "] = \"third\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43cde57d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "first              17146\n",
       "not_orthologous      240\n",
       "second                40\n",
       "Name: gene_orthology_status, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(m_lift[\"gene_orthology_status\"].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "068ade93",
   "metadata": {},
   "outputs": [],
   "source": [
    "h_lift[\"orth_first_order\"] = h_lift.apply(\n",
    "    lambda row: check_human_orthologies_first_order(\n",
    "        mouse_identifier=row[\"identifier.mouse_target\"], human_identifier=row[\"identifier.human_source\"]\n",
    "    ),\n",
    "    axis=1,\n",
    ")\n",
    "h_lift[\"orth_second_order\"] = h_lift.apply(\n",
    "    lambda row: check_human_orthologies_second_order(\n",
    "        mouse_identifier=row[\"identifier.mouse_target\"], human_identifier=row[\"identifier.human_source\"]\n",
    "    ),\n",
    "    axis=1,\n",
    ")\n",
    "h_lift[\"orth_third_order.details\"] = h_lift.apply(\n",
    "    lambda row: check_human_orthologies_third_order(\n",
    "        mouse_identifier=row[\"identifier.mouse_target\"], human_identifier=row[\"identifier.human_source\"]\n",
    "    ),\n",
    "    axis=1,\n",
    ")\n",
    "h_lift[\"orth_third_order\"] = h_lift[\"orth_third_order.details\"].apply(flag_from_third_order_orthologies)\n",
    "\n",
    "\n",
    "# Now assign the orthology status of the gene.\n",
    "h_lift[\"gene_orthology_status\"] = \"not_orthologous\"\n",
    "h_lift.loc[lambda df: (df[\"orth_first_order\"].replace(np.nan, False)), \"gene_orthology_status\"] = \"first\"\n",
    "h_lift.loc[\n",
    "    lambda df: (~df[\"orth_first_order\"].replace(np.nan, False)) & (df[\"orth_second_order\"].replace(np.nan, False)),\n",
    "    \"gene_orthology_status\",\n",
    "] = \"second\"\n",
    "h_lift.loc[\n",
    "    lambda df: (\n",
    "        (~df[\"orth_first_order\"].replace(np.nan, False))\n",
    "        & (~df[\"orth_second_order\"].replace(np.nan, False))\n",
    "        & (df[\"orth_third_order\"].replace(np.nan, False))\n",
    "    ),\n",
    "    \"gene_orthology_status\",\n",
    "] = \"third\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "655d3cad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "first              28499\n",
       "not_orthologous      868\n",
       "second                16\n",
       "Name: gene_orthology_status, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(h_lift[\"gene_orthology_status\"].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "0366cfe2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    {ENSG00000023287}\n",
       "1    {ENSG00000168300}\n",
       "2    {ENSG00000168300}\n",
       "Name: orthologous_gene_set, dtype: object"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "17426"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Here: extract a set of annotated genes from the target-species lifted coordinates.\n",
    "# i.e. for a given mouse peak we produce a set of human genes annotated at the location\n",
    "# of the lifted coordinates of the mouse peak.\n",
    "\n",
    "m_lift[\"orthologous_gene_set\"] = m_lift[\"orth_third_order.details\"].apply(\n",
    "    lambda v: set(list(itt.chain(*[ortho for gene_id, ortho in v if isinstance(ortho, list)])))\n",
    ")\n",
    "\n",
    "h_lift[\"orthologous_gene_set\"] = h_lift[\"orth_third_order.details\"].apply(\n",
    "    lambda v: set(list(itt.chain(*[ortho for gene_id, ortho in v if isinstance(ortho, list)])))\n",
    ")\n",
    "\n",
    "display(m_lift[\"orthologous_gene_set\"].head(3))\n",
    "display(m_lift.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "ec8666aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier.mouse_source</th>\n",
       "      <th>identifier.human_target</th>\n",
       "      <th>orth_first_order</th>\n",
       "      <th>orth_second_order</th>\n",
       "      <th>orth_third_order.details</th>\n",
       "      <th>orth_third_order</th>\n",
       "      <th>gene_orthology_status</th>\n",
       "      <th>orthologous_gene_set</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:6274997-6275006:+</td>\n",
       "      <td>chr8:52623049-52623059:-</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSMUSG00000025907, [ENSG00000023287])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSG00000023287}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:7170406-7170415:+</td>\n",
       "      <td>chr8:51819766-51819775:-</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSMUSG00000051285, [ENSG00000168300])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSG00000168300}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1:7170431-7170440:+</td>\n",
       "      <td>chr8:51819741-51819750:-</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSMUSG00000051285, [ENSG00000168300])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSG00000168300}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  identifier.mouse_source   identifier.human_target orth_first_order  \\\n",
       "0  chr1:6274997-6275006:+  chr8:52623049-52623059:-             True   \n",
       "1  chr1:7170406-7170415:+  chr8:51819766-51819775:-             True   \n",
       "2  chr1:7170431-7170440:+  chr8:51819741-51819750:-             True   \n",
       "\n",
       "  orth_second_order                   orth_third_order.details  \\\n",
       "0              True  [(ENSMUSG00000025907, [ENSG00000023287])]   \n",
       "1              True  [(ENSMUSG00000051285, [ENSG00000168300])]   \n",
       "2              True  [(ENSMUSG00000051285, [ENSG00000168300])]   \n",
       "\n",
       "  orth_third_order gene_orthology_status orthologous_gene_set  \n",
       "0             True                 first    {ENSG00000023287}  \n",
       "1             True                 first    {ENSG00000168300}  \n",
       "2             True                 first    {ENSG00000168300}  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m_lift.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "c8ccd8a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Here (example with mouse peaks): take the identifier of a given (mouse) peak's lifted coordinates (human)\n",
    "# and access the set of genes annotated at the lifted coordinates (human genome).\n",
    "# Compare this set with the set of genes \"bound\" in human.\n",
    "\n",
    "# NOTE: we only consider the \"best\" annotated gene for the source-species \"bound\" genes.\n",
    "h_bound_genes = set(h_full[\"gene_id\"].unique())\n",
    "\n",
    "m_lift[\"annotated_lifted_region_bound_gene_set\"] = m_lift.apply(\n",
    "    lambda row: (h_id_to_meta[\"gene_id_set\"].get(row[\"identifier.human_target\"], set({})) & h_bound_genes), axis=1\n",
    ")\n",
    "\n",
    "###\n",
    "\n",
    "# Repeat with human peaks, comparing against mouse bound genes.\n",
    "\n",
    "m_bound_genes = set(m_full[\"gene_id\"].unique())\n",
    "\n",
    "h_lift[\"annotated_lifted_region_bound_gene_set\"] = h_lift.apply(\n",
    "    lambda row: (m_id_to_meta[\"gene_id_set\"].get(row[\"identifier.mouse_target\"], set({})) & m_bound_genes), axis=1\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "950831dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier.mouse_source</th>\n",
       "      <th>identifier.human_target</th>\n",
       "      <th>orth_first_order</th>\n",
       "      <th>orth_second_order</th>\n",
       "      <th>orth_third_order.details</th>\n",
       "      <th>orth_third_order</th>\n",
       "      <th>gene_orthology_status</th>\n",
       "      <th>orthologous_gene_set</th>\n",
       "      <th>annotated_lifted_region_bound_gene_set</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:6274997-6275006:+</td>\n",
       "      <td>chr8:52623049-52623059:-</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSMUSG00000025907, [ENSG00000023287])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSG00000023287}</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:7170406-7170415:+</td>\n",
       "      <td>chr8:51819766-51819775:-</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSMUSG00000051285, [ENSG00000168300])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSG00000168300}</td>\n",
       "      <td>{ENSG00000168300}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1:7170431-7170440:+</td>\n",
       "      <td>chr8:51819741-51819750:-</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSMUSG00000051285, [ENSG00000168300])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSG00000168300}</td>\n",
       "      <td>{ENSG00000168300}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  identifier.mouse_source   identifier.human_target orth_first_order  \\\n",
       "0  chr1:6274997-6275006:+  chr8:52623049-52623059:-             True   \n",
       "1  chr1:7170406-7170415:+  chr8:51819766-51819775:-             True   \n",
       "2  chr1:7170431-7170440:+  chr8:51819741-51819750:-             True   \n",
       "\n",
       "  orth_second_order                   orth_third_order.details  \\\n",
       "0              True  [(ENSMUSG00000025907, [ENSG00000023287])]   \n",
       "1              True  [(ENSMUSG00000051285, [ENSG00000168300])]   \n",
       "2              True  [(ENSMUSG00000051285, [ENSG00000168300])]   \n",
       "\n",
       "  orth_third_order gene_orthology_status orthologous_gene_set  \\\n",
       "0             True                 first    {ENSG00000023287}   \n",
       "1             True                 first    {ENSG00000168300}   \n",
       "2             True                 first    {ENSG00000168300}   \n",
       "\n",
       "  annotated_lifted_region_bound_gene_set  \n",
       "0                                     {}  \n",
       "1                      {ENSG00000168300}  \n",
       "2                      {ENSG00000168300}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Example rows for illustration\n",
    "display(m_lift.head(3))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "afa3bac9",
   "metadata": {},
   "source": [
    "So for instance \"chr1:6274997-6275006:+\" is a mouse peak found in \"ENSMUSG00000025907\", \n",
    "and the lifted-over region \"chr8:52623049-52623059:-\" is found in \"ENSG00000023287\" which is the orthologous gene.\n",
    "**BUT THIS ORTHOLOGOUS GENE IS NOT BOUND IN HUMAN**."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d07db453",
   "metadata": {},
   "source": [
    "### Gene counts - Mouse"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39fdeb03",
   "metadata": {},
   "source": [
    "Now simply count the number of genes in each species, separating the genes with orthologs bound from the genes without orthologs or with an ortholog gene but not bound.\n",
    "\n",
    "NOTE: do not forget to consider unmappable peaks."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d630ae7f",
   "metadata": {},
   "source": [
    "#### First attempt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71556ae5",
   "metadata": {},
   "outputs": [],
   "source": [
    "m_has_gene_bound_orth = (\n",
    "    (m_lift[\"gene_orthology_status\"] != \"not_orthologous\")\n",
    "    & (m_lift[\"annotated_lifted_region_bound_gene_set\"].apply(len) > 0)\n",
    ").values\n",
    "\n",
    "###\n",
    "\n",
    "m_peaks_w_bound_orth = m_lift.loc[m_has_gene_bound_orth, \"identifier.mouse_source\"].values\n",
    "m_peaks_wo_bound_orth = m_lift.loc[~m_has_gene_bound_orth, \"identifier.mouse_source\"].values\n",
    "\n",
    "###\n",
    "\n",
    "m_genes_bound_orth = set(m_id_to_meta.loc[m_peaks_w_bound_orth, \"gene_id\"].unique())\n",
    "m_genes_lifted_but_wo_annotated_bound_orth = set(m_id_to_meta.loc[m_peaks_wo_bound_orth, \"gene_id\"].unique())\n",
    "\n",
    "###\n",
    "\n",
    "# Here: aggregate the list of mouse ENS IDs identified as orthologs for the human peaks' annotated genes.\n",
    "# i.e. : we know these genes mouse ENS-ID have a human ortholog that has a peak.\n",
    "m_genes_from_human_peaks_w_orth = functools.reduce(set.union, h_lift[\"annotated_lifted_region_bound_gene_set\"].values)\n",
    "\n",
    "###\n",
    "\n",
    "m_genes_from_umap = set(m_id_to_meta.loc[m_umap[\"identifier\"].values, \"gene_id\"].unique())\n",
    "\n",
    "###\n",
    "\n",
    "# Finally : build the list of genes without bound orthologous gene (i.e. mouse-specific bound genes)\n",
    "# as the set of genes from unmappable peaks or peaks mapped but for which the lifted region is not an orthologous-bound gene.\n",
    "\n",
    "m_genes_no_bound_orth = (\n",
    "    (m_genes_from_umap | m_genes_lifted_but_wo_annotated_bound_orth)\n",
    "    - m_genes_from_human_peaks_w_orth\n",
    "    - m_genes_bound_orth\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2508d1d1",
   "metadata": {},
   "source": [
    "#### Missing genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3c1243d",
   "metadata": {},
   "source": [
    "We established a set of genes \"ortholog is bound\" and \"ortholog is not bound\" for the mouse peaks.\n",
    "\n",
    "But there are a few additional genes that appear to not have any status assigned from the previous section."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1c2f4f6",
   "metadata": {},
   "source": [
    "##### Identify"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "696ed8fb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True     57\n",
       "False    21\n",
       "Name: is_in_umap, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    }
   ],
   "source": [
    "# Identify genes missed in the previous step\n",
    "# ------------------------------------------\n",
    "\n",
    "m_missing = set(m_full[\"gene_id\"].unique()) - m_genes_bound_orth - m_genes_no_bound_orth\n",
    "print(len(m_missing))\n",
    "\n",
    "\n",
    "m_peaks_in_missing_genes = m_full.loc[lambda df: df[\"gene_id\"].isin(m_missing), :].assign(\n",
    "    is_in_umap=lambda df: df[\"identifier\"].isin(m_umap[\"identifier\"].values)\n",
    ")\n",
    "\n",
    "\n",
    "display(m_peaks_in_missing_genes[\"is_in_umap\"].value_counts())\n",
    "\n",
    "# EXPECTED:\n",
    "# True     57\n",
    "# False    21\n",
    "\n",
    "# So most peaks are not mapped,\n",
    "\n",
    "# ALL THESE GENES ARE FOUND TO BE BOUND IN HUMAN PEAKS LIFTED TO MOUSE.\n",
    "# i.e. I discard them wrongly in the last set operation, as they are bound in mouse but from unmapped peaks.\n",
    "\n",
    "print(all([g in m_genes_from_human_peaks_w_orth for g in m_missing]))\n",
    "\n",
    "# A subset of genes apparently are mappable, so I assume there it is a problem of gene annotation.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fdfca5a2",
   "metadata": {},
   "source": [
    "##### Example 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eac3755a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "g in m_bound_genes=True\n",
      "g in m_genes_bound_orth=False\n",
      "g in m_genes_no_bound_orth=False\n",
      "g in m_genes_lifted_but_wo_annotated_bound_orth=False\n",
      "g in m_genes_from_umap=True\n",
      "g in m_genes_from_human_peaks_w_orth=True\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>name</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1:13337373-13337382:-</td>\n",
       "      <td>intron;Ncoa2;protein_coding;ENSMUSG00000005886.14</td>\n",
       "      <td>ENSMUSG00000005886</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 identifier  \\\n",
       "5  chr1:13337373-13337382:-   \n",
       "\n",
       "                                                name             gene_id  \n",
       "5  intron;Ncoa2;protein_coding;ENSMUSG00000005886.14  ENSMUSG00000005886  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "20"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier.human_source</th>\n",
       "      <th>identifier.mouse_target</th>\n",
       "      <th>orth_first_order</th>\n",
       "      <th>orth_second_order</th>\n",
       "      <th>orth_third_order.details</th>\n",
       "      <th>orth_third_order</th>\n",
       "      <th>gene_orthology_status</th>\n",
       "      <th>orthologous_gene_set</th>\n",
       "      <th>annotated_lifted_region_bound_gene_set</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23076</th>\n",
       "      <td>chr8:70109825-70109834:-</td>\n",
       "      <td>chr1:13139168-13139176:-</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSG00000140396, [ENSMUSG00000005886])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSMUSG00000005886}</td>\n",
       "      <td>{ENSMUSG00000005886}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23077</th>\n",
       "      <td>chr8:70109861-70109870:-</td>\n",
       "      <td>chr1:13139203-13139212:-</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSG00000140396, [ENSMUSG00000005886])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSMUSG00000005886}</td>\n",
       "      <td>{ENSMUSG00000005886}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23078</th>\n",
       "      <td>chr8:70109880-70109889:-</td>\n",
       "      <td>chr1:13139222-13139231:-</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSG00000140396, [ENSMUSG00000005886])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSMUSG00000005886}</td>\n",
       "      <td>{ENSMUSG00000005886}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        identifier.human_source   identifier.mouse_target orth_first_order  \\\n",
       "23076  chr8:70109825-70109834:-  chr1:13139168-13139176:-             True   \n",
       "23077  chr8:70109861-70109870:-  chr1:13139203-13139212:-             True   \n",
       "23078  chr8:70109880-70109889:-  chr1:13139222-13139231:-             True   \n",
       "\n",
       "      orth_second_order                   orth_third_order.details  \\\n",
       "23076              True  [(ENSG00000140396, [ENSMUSG00000005886])]   \n",
       "23077              True  [(ENSG00000140396, [ENSMUSG00000005886])]   \n",
       "23078              True  [(ENSG00000140396, [ENSMUSG00000005886])]   \n",
       "\n",
       "      orth_third_order gene_orthology_status  orthologous_gene_set  \\\n",
       "23076             True                 first  {ENSMUSG00000005886}   \n",
       "23077             True                 first  {ENSMUSG00000005886}   \n",
       "23078             True                 first  {ENSMUSG00000005886}   \n",
       "\n",
       "      annotated_lifted_region_bound_gene_set  \n",
       "23076                   {ENSMUSG00000005886}  \n",
       "23077                   {ENSMUSG00000005886}  \n",
       "23078                   {ENSMUSG00000005886}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Example unmapped peak and associated gene: chr1:13337373-13337382:- : ENSMUSG00000005886 (Ncoa2)\n",
    "\n",
    "g = \"ENSMUSG00000005886\"\n",
    "\n",
    "# Verify status in the full list and in the pseudo-final lists.\n",
    "print(f\"{g in m_bound_genes=}\")\n",
    "print(f\"{g in m_genes_bound_orth=}\")\n",
    "print(f\"{g in m_genes_no_bound_orth=}\")\n",
    "# Here : True, False False ;\n",
    "# i.e. this gene is in the full set of bound genes, but not in the \"pseudo-final\" lists.\n",
    "\n",
    "# Track where this gene can be found from the subset lists.\n",
    "print(f\"{g in m_genes_lifted_but_wo_annotated_bound_orth=}\")\n",
    "print(f\"{g in m_genes_from_umap=}\")\n",
    "print(f\"{g in m_genes_from_human_peaks_w_orth=}\")\n",
    "# Here : False, True, True ; i.e. the gene comes from an unmapped peak from Mouse.\n",
    "# But in human there is a peak that can be mapped to mouse, and this peak is annotated in both species with the orthologs.\n",
    "\n",
    "# Here we identify which MOUSE PEAKS are annotated with this gene:\n",
    "\n",
    "display(m_full.loc[lambda df: df[\"gene_id\"] == g, :].shape[0])\n",
    "display(m_full.loc[lambda df: df[\"gene_id\"] == g, :].head(3))\n",
    "\n",
    "# There should be only one peak retrieve: chr1:13337373-13337382:-\n",
    "# Which is an unmappable peak, as shown here:\n",
    "print(\"chr1:13337373-13337382:-\" in m_umap[\"identifier\"].values)\n",
    "\n",
    "# On the other hand there are many human peaks that 1. are annotated with this gene,\n",
    "# 2. can be lifted, and 3. have their lifted-over region in the orthologous gene in mouse.\n",
    "\n",
    "display(h_lift.loc[lambda df: df[\"annotated_lifted_region_bound_gene_set\"].apply(lambda s: g in s), :].shape[0])\n",
    "display(h_lift.loc[lambda df: df[\"annotated_lifted_region_bound_gene_set\"].apply(lambda s: g in s), :].head(3))\n",
    "# This should be 20 peaks."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0725a68c",
   "metadata": {},
   "source": [
    "##### Example 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94e23be3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "g in m_bound_genes=True\n",
      "g in m_genes_bound_orth=False\n",
      "g in m_genes_no_bound_orth=False\n",
      "g in m_genes_lifted_but_wo_annotated_bound_orth=True\n",
      "g in m_genes_from_umap=False\n",
      "g in m_genes_from_human_peaks_w_orth=True\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>name</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2966</th>\n",
       "      <td>chr11:62894762-62894771:+</td>\n",
       "      <td>three_prime_UTR;Tvp23b;protein_coding;ENSMUSG0...</td>\n",
       "      <td>ENSMUSG00000014177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2967</th>\n",
       "      <td>chr11:62894772-62894781:+</td>\n",
       "      <td>three_prime_UTR;Tvp23b;protein_coding;ENSMUSG0...</td>\n",
       "      <td>ENSMUSG00000014177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2968</th>\n",
       "      <td>chr11:62894850-62894859:+</td>\n",
       "      <td>three_prime_UTR;Tvp23b;protein_coding;ENSMUSG0...</td>\n",
       "      <td>ENSMUSG00000014177</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     identifier  \\\n",
       "2966  chr11:62894762-62894771:+   \n",
       "2967  chr11:62894772-62894781:+   \n",
       "2968  chr11:62894850-62894859:+   \n",
       "\n",
       "                                                   name             gene_id  \n",
       "2966  three_prime_UTR;Tvp23b;protein_coding;ENSMUSG0...  ENSMUSG00000014177  \n",
       "2967  three_prime_UTR;Tvp23b;protein_coding;ENSMUSG0...  ENSMUSG00000014177  \n",
       "2968  three_prime_UTR;Tvp23b;protein_coding;ENSMUSG0...  ENSMUSG00000014177  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier.mouse_source</th>\n",
       "      <th>identifier.human_target</th>\n",
       "      <th>orth_first_order</th>\n",
       "      <th>orth_second_order</th>\n",
       "      <th>orth_third_order.details</th>\n",
       "      <th>orth_third_order</th>\n",
       "      <th>gene_orthology_status</th>\n",
       "      <th>orthologous_gene_set</th>\n",
       "      <th>annotated_lifted_region_bound_gene_set</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2650</th>\n",
       "      <td>chr11:62894762-62894771:+</td>\n",
       "      <td>chr17:15537415-15537423:-</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSMUSG00000014177, [ENSG00000175106])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSG00000175106}</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651</th>\n",
       "      <td>chr11:62894772-62894781:+</td>\n",
       "      <td>chr17:15537405-15537414:-</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSMUSG00000014177, [ENSG00000175106])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSG00000175106}</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2652</th>\n",
       "      <td>chr11:62894850-62894859:+</td>\n",
       "      <td>chr17:15537333-15537342:-</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSMUSG00000014177, [ENSG00000175106])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSG00000175106}</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        identifier.mouse_source    identifier.human_target orth_first_order  \\\n",
       "2650  chr11:62894762-62894771:+  chr17:15537415-15537423:-             True   \n",
       "2651  chr11:62894772-62894781:+  chr17:15537405-15537414:-             True   \n",
       "2652  chr11:62894850-62894859:+  chr17:15537333-15537342:-             True   \n",
       "\n",
       "     orth_second_order                   orth_third_order.details  \\\n",
       "2650              True  [(ENSMUSG00000014177, [ENSG00000175106])]   \n",
       "2651              True  [(ENSMUSG00000014177, [ENSG00000175106])]   \n",
       "2652              True  [(ENSMUSG00000014177, [ENSG00000175106])]   \n",
       "\n",
       "     orth_third_order gene_orthology_status orthologous_gene_set  \\\n",
       "2650             True                 first    {ENSG00000175106}   \n",
       "2651             True                 first    {ENSG00000175106}   \n",
       "2652             True                 first    {ENSG00000175106}   \n",
       "\n",
       "     annotated_lifted_region_bound_gene_set  \n",
       "2650                                     {}  \n",
       "2651                                     {}  \n",
       "2652                                     {}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Example: chr11:62894762-62894771:+ ENSMUSG00000014177 (Tvp23b)\n",
    "\n",
    "g = \"ENSMUSG00000014177\"\n",
    "\n",
    "# Verify status in the full list and in the pseudo-final lists.\n",
    "print(f\"{g in m_bound_genes=}\")\n",
    "print(f\"{g in m_genes_bound_orth=}\")\n",
    "print(f\"{g in m_genes_no_bound_orth=}\")\n",
    "# Here : True, False False ;\n",
    "# i.e. this gene is in the full set of bound genes, but not in the \"pseudo-final\" lists.\n",
    "\n",
    "# Track where this gene can be found from the subset lists.\n",
    "print(f\"{g in m_genes_lifted_but_wo_annotated_bound_orth=}\")\n",
    "print(f\"{g in m_genes_from_umap=}\")\n",
    "print(f\"{g in m_genes_from_human_peaks_w_orth=}\")\n",
    "# Here : True, False, True ; i.e. the gene comes from an mapped peak from Mouse, but where the annotated gene\n",
    "# in human did not match or was absent.\n",
    "# But in human there is a peak that can be mapped to mouse, and this peak is annotated in both species with the orthologs.\n",
    "\n",
    "\n",
    "# Here we identify which MOUSE PEAKS are annotated with this gene:\n",
    "\n",
    "display(m_full.loc[lambda df: df[\"gene_id\"] == g, :].shape[0])\n",
    "display(m_full.loc[lambda df: df[\"gene_id\"] == g, :].head(3))\n",
    "\n",
    "# There should be 3 peaks retrieved.\n",
    "# Here we check that all these peaks are indeed among the set of mappable peaks and could be lifted.\n",
    "print(\n",
    "    all(\n",
    "        [\n",
    "            identifier in m_lift[\"identifier.mouse_source\"].values\n",
    "            for identifier in m_full.loc[lambda df: df[\"gene_id\"] == g, \"identifier\"].values\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "# This should be True, i.e. all the peaks can be lifted to human.\n",
    "\n",
    "# Finally: we display the set of lifted peaks, and so as to confirm that the lifted regions do not bear any orthologous gene.\n",
    "display(\n",
    "    m_lift.loc[\n",
    "        lambda df: df[\"identifier.mouse_source\"].isin(m_full.loc[lambda df: df[\"gene_id\"] == g, \"identifier\"].values), :\n",
    "    ]\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3672eac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>name</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2966</th>\n",
       "      <td>chr11:62894762-62894771:+</td>\n",
       "      <td>three_prime_UTR;Tvp23b;protein_coding;ENSMUSG0...</td>\n",
       "      <td>ENSMUSG00000014177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2967</th>\n",
       "      <td>chr11:62894772-62894781:+</td>\n",
       "      <td>three_prime_UTR;Tvp23b;protein_coding;ENSMUSG0...</td>\n",
       "      <td>ENSMUSG00000014177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2968</th>\n",
       "      <td>chr11:62894850-62894859:+</td>\n",
       "      <td>three_prime_UTR;Tvp23b;protein_coding;ENSMUSG0...</td>\n",
       "      <td>ENSMUSG00000014177</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     identifier  \\\n",
       "2966  chr11:62894762-62894771:+   \n",
       "2967  chr11:62894772-62894781:+   \n",
       "2968  chr11:62894850-62894859:+   \n",
       "\n",
       "                                                   name             gene_id  \n",
       "2966  three_prime_UTR;Tvp23b;protein_coding;ENSMUSG0...  ENSMUSG00000014177  \n",
       "2967  three_prime_UTR;Tvp23b;protein_coding;ENSMUSG0...  ENSMUSG00000014177  \n",
       "2968  three_prime_UTR;Tvp23b;protein_coding;ENSMUSG0...  ENSMUSG00000014177  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier.mouse_source</th>\n",
       "      <th>identifier.human_target</th>\n",
       "      <th>orth_first_order</th>\n",
       "      <th>orth_second_order</th>\n",
       "      <th>orth_third_order.details</th>\n",
       "      <th>orth_third_order</th>\n",
       "      <th>gene_orthology_status</th>\n",
       "      <th>orthologous_gene_set</th>\n",
       "      <th>annotated_lifted_region_bound_gene_set</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2650</th>\n",
       "      <td>chr11:62894762-62894771:+</td>\n",
       "      <td>chr17:15537415-15537423:-</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSMUSG00000014177, [ENSG00000175106])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSG00000175106}</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        identifier.mouse_source    identifier.human_target orth_first_order  \\\n",
       "2650  chr11:62894762-62894771:+  chr17:15537415-15537423:-             True   \n",
       "\n",
       "     orth_second_order                   orth_third_order.details  \\\n",
       "2650              True  [(ENSMUSG00000014177, [ENSG00000175106])]   \n",
       "\n",
       "     orth_third_order gene_orthology_status orthologous_gene_set  \\\n",
       "2650             True                 first    {ENSG00000175106}   \n",
       "\n",
       "     annotated_lifted_region_bound_gene_set  \n",
       "2650                                     {}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>main_annot</th>\n",
       "      <th>annot_set</th>\n",
       "      <th>gene_id_set</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr11:62894762-62894771:+</th>\n",
       "      <td>three_prime_UTR;Tvp23b;protein_coding;ENSMUSG0...</td>\n",
       "      <td>ENSMUSG00000014177</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{three_prime_UTR}</td>\n",
       "      <td>{ENSMUSG00000014177}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                        name  \\\n",
       "identifier                                                                     \n",
       "chr11:62894762-62894771:+  three_prime_UTR;Tvp23b;protein_coding;ENSMUSG0...   \n",
       "\n",
       "                                      gene_id       main_annot  \\\n",
       "identifier                                                       \n",
       "chr11:62894762-62894771:+  ENSMUSG00000014177  three_prime_UTR   \n",
       "\n",
       "                                   annot_set           gene_id_set  \n",
       "identifier                                                          \n",
       "chr11:62894762-62894771:+  {three_prime_UTR}  {ENSMUSG00000014177}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>main_annot</th>\n",
       "      <th>annot_set</th>\n",
       "      <th>gene_id_set</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr17:15537415-15537423:-</th>\n",
       "      <td>three_prime_UTR;TVP23C;protein_coding;ENSG0000...</td>\n",
       "      <td>ENSG00000175106</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{three_prime_UTR}</td>\n",
       "      <td>{ENSG00000175106}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                        name  \\\n",
       "identifier                                                                     \n",
       "chr17:15537415-15537423:-  three_prime_UTR;TVP23C;protein_coding;ENSG0000...   \n",
       "\n",
       "                                   gene_id       main_annot  \\\n",
       "identifier                                                    \n",
       "chr17:15537415-15537423:-  ENSG00000175106  three_prime_UTR   \n",
       "\n",
       "                                   annot_set        gene_id_set  \n",
       "identifier                                                       \n",
       "chr17:15537415-15537423:-  {three_prime_UTR}  {ENSG00000175106}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Going through the example again from the point of view of mouse peak:\n",
    "# IMPORTANT : using the table as is, we faced an issue with an example of (likely) species-specific duplicated genes,\n",
    "# that has all the n:1 orthologous relationships in one way but not the other.\n",
    "\n",
    "# Basically : from human to mouse, ENSMUSG00000014177 (bound) maps to 'ENSG00000175106' (not bound) while\n",
    "# the human gene bound is ENSG00000171928 that maps to ENSMUSG00000014177.\n",
    "# From gene names, it seems the \"best\" orthologous gene is actually ENSG00000171928, matching the name of ENSMUSG00000014177\n",
    "# (Tvp23b and TVP23B) while ENSG00000175106 is named differently (TVP23C).\n",
    "\n",
    "display(m_full.loc[lambda df: df[\"gene_id\"] == \"ENSMUSG00000014177\", :])\n",
    "display(m_lift.loc[lambda df: df[\"identifier.mouse_source\"] == \"chr11:62894762-62894771:+\", :])\n",
    "display(m_id_to_meta.loc[[\"chr11:62894762-62894771:+\"], :])\n",
    "display(h_id_to_meta.loc[[\"chr17:15537415-15537423:-\"], :])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d27db38",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>main_annot</th>\n",
       "      <th>annot_set</th>\n",
       "      <th>gene_id_set</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr11:62894762-62894771:+</th>\n",
       "      <td>three_prime_UTR;Tvp23b;protein_coding;ENSMUSG0...</td>\n",
       "      <td>ENSMUSG00000014177</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{three_prime_UTR}</td>\n",
       "      <td>{ENSMUSG00000014177}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                        name  \\\n",
       "identifier                                                                     \n",
       "chr11:62894762-62894771:+  three_prime_UTR;Tvp23b;protein_coding;ENSMUSG0...   \n",
       "\n",
       "                                      gene_id       main_annot  \\\n",
       "identifier                                                       \n",
       "chr11:62894762-62894771:+  ENSMUSG00000014177  three_prime_UTR   \n",
       "\n",
       "                                   annot_set           gene_id_set  \n",
       "identifier                                                          \n",
       "chr11:62894762-62894771:+  {three_prime_UTR}  {ENSMUSG00000014177}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>main_annot</th>\n",
       "      <th>annot_set</th>\n",
       "      <th>gene_id_set</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr17:15537415-15537423:-</th>\n",
       "      <td>three_prime_UTR;TVP23C;protein_coding;ENSG0000...</td>\n",
       "      <td>ENSG00000175106</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{three_prime_UTR}</td>\n",
       "      <td>{ENSG00000175106}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                        name  \\\n",
       "identifier                                                                     \n",
       "chr17:15537415-15537423:-  three_prime_UTR;TVP23C;protein_coding;ENSG0000...   \n",
       "\n",
       "                                   gene_id       main_annot  \\\n",
       "identifier                                                    \n",
       "chr17:15537415-15537423:-  ENSG00000175106  three_prime_UTR   \n",
       "\n",
       "                                   annot_set        gene_id_set  \n",
       "identifier                                                       \n",
       "chr17:15537415-15537423:-  {three_prime_UTR}  {ENSG00000175106}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier.mouse_source</th>\n",
       "      <th>identifier.human_target</th>\n",
       "      <th>orth_first_order</th>\n",
       "      <th>orth_second_order</th>\n",
       "      <th>orth_third_order.details</th>\n",
       "      <th>orth_third_order</th>\n",
       "      <th>gene_orthology_status</th>\n",
       "      <th>orthologous_gene_set</th>\n",
       "      <th>annotated_lifted_region_bound_gene_set</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2650</th>\n",
       "      <td>chr11:62894762-62894771:+</td>\n",
       "      <td>chr17:15537415-15537423:-</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSMUSG00000014177, [ENSG00000175106])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSG00000175106}</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        identifier.mouse_source    identifier.human_target orth_first_order  \\\n",
       "2650  chr11:62894762-62894771:+  chr17:15537415-15537423:-             True   \n",
       "\n",
       "     orth_second_order                   orth_third_order.details  \\\n",
       "2650              True  [(ENSMUSG00000014177, [ENSG00000175106])]   \n",
       "\n",
       "     orth_third_order gene_orthology_status orthologous_gene_set  \\\n",
       "2650             True                 first    {ENSG00000175106}   \n",
       "\n",
       "     annotated_lifted_region_bound_gene_set  \n",
       "2650                                     {}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>human_ensid</th>\n",
       "      <th>human_name</th>\n",
       "      <th>mouse_ensid</th>\n",
       "      <th>mouse_name</th>\n",
       "      <th>relationtype_id</th>\n",
       "      <th>supported_by_resource</th>\n",
       "      <th>support_level</th>\n",
       "      <th>support_details</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15041</th>\n",
       "      <td>ENSG00000175106</td>\n",
       "      <td>TVP23C</td>\n",
       "      <td>ENSMUSG00000014177</td>\n",
       "      <td>Tvp23b</td>\n",
       "      <td>1:n</td>\n",
       "      <td>omabrowser;;mgi</td>\n",
       "      <td>3</td>\n",
       "      <td>omabrowser:human_to_mouse;mouse_to_human;;mgi:...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           human_ensid human_name         mouse_ensid mouse_name  \\\n",
       "15041  ENSG00000175106     TVP23C  ENSMUSG00000014177     Tvp23b   \n",
       "\n",
       "      relationtype_id supported_by_resource  support_level  \\\n",
       "15041             1:n       omabrowser;;mgi              3   \n",
       "\n",
       "                                         support_details  \n",
       "15041  omabrowser:human_to_mouse;mouse_to_human;;mgi:...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>human_ensid</th>\n",
       "      <th>human_name</th>\n",
       "      <th>mouse_ensid</th>\n",
       "      <th>mouse_name</th>\n",
       "      <th>relationtype_id</th>\n",
       "      <th>supported_by_resource</th>\n",
       "      <th>support_level</th>\n",
       "      <th>support_details</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14295</th>\n",
       "      <td>ENSG00000171928</td>\n",
       "      <td>TVP23B</td>\n",
       "      <td>ENSMUSG00000014177</td>\n",
       "      <td>Tvp23b</td>\n",
       "      <td>1:n</td>\n",
       "      <td>biomart;;omabrowser;;mgi</td>\n",
       "      <td>4</td>\n",
       "      <td>biomart:mouse_to_human;;omabrowser:human_to_mo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15041</th>\n",
       "      <td>ENSG00000175106</td>\n",
       "      <td>TVP23C</td>\n",
       "      <td>ENSMUSG00000014177</td>\n",
       "      <td>Tvp23b</td>\n",
       "      <td>1:n</td>\n",
       "      <td>omabrowser;;mgi</td>\n",
       "      <td>3</td>\n",
       "      <td>omabrowser:human_to_mouse;mouse_to_human;;mgi:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29274</th>\n",
       "      <td>ENSG00000259024</td>\n",
       "      <td>TVP23C-CDRT4</td>\n",
       "      <td>ENSMUSG00000014177</td>\n",
       "      <td>Tvp23b</td>\n",
       "      <td>1:n</td>\n",
       "      <td>biomart;;omabrowser;;mgi</td>\n",
       "      <td>4</td>\n",
       "      <td>biomart:mouse_to_human;;omabrowser:human_to_mo...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           human_ensid    human_name         mouse_ensid mouse_name  \\\n",
       "14295  ENSG00000171928        TVP23B  ENSMUSG00000014177     Tvp23b   \n",
       "15041  ENSG00000175106        TVP23C  ENSMUSG00000014177     Tvp23b   \n",
       "29274  ENSG00000259024  TVP23C-CDRT4  ENSMUSG00000014177     Tvp23b   \n",
       "\n",
       "      relationtype_id     supported_by_resource  support_level  \\\n",
       "14295             1:n  biomart;;omabrowser;;mgi              4   \n",
       "15041             1:n           omabrowser;;mgi              3   \n",
       "29274             1:n  biomart;;omabrowser;;mgi              4   \n",
       "\n",
       "                                         support_details  \n",
       "14295  biomart:mouse_to_human;;omabrowser:human_to_mo...  \n",
       "15041  omabrowser:human_to_mouse;mouse_to_human;;mgi:...  \n",
       "29274  biomart:mouse_to_human;;omabrowser:human_to_mo...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier.human_source</th>\n",
       "      <th>identifier.mouse_target</th>\n",
       "      <th>orth_first_order</th>\n",
       "      <th>orth_second_order</th>\n",
       "      <th>orth_third_order.details</th>\n",
       "      <th>orth_third_order</th>\n",
       "      <th>gene_orthology_status</th>\n",
       "      <th>orthologous_gene_set</th>\n",
       "      <th>annotated_lifted_region_bound_gene_set</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10169</th>\n",
       "      <td>chr17:18806058-18806067:+</td>\n",
       "      <td>chr11:62894538-62894546:+</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSG00000171928, [ENSMUSG00000014177])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSMUSG00000014177}</td>\n",
       "      <td>{ENSMUSG00000014177}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10170</th>\n",
       "      <td>chr17:18806269-18806278:+</td>\n",
       "      <td>chr11:62894764-62894772:+</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSG00000171928, [ENSMUSG00000014177])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSMUSG00000014177}</td>\n",
       "      <td>{ENSMUSG00000014177}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26026</th>\n",
       "      <td>chr17:18806240-18806249:+</td>\n",
       "      <td>chr11:62894727-62894736:+</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSG00000171928, [ENSMUSG00000014177])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSMUSG00000014177}</td>\n",
       "      <td>{ENSMUSG00000014177}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         identifier.human_source    identifier.mouse_target orth_first_order  \\\n",
       "10169  chr17:18806058-18806067:+  chr11:62894538-62894546:+             True   \n",
       "10170  chr17:18806269-18806278:+  chr11:62894764-62894772:+             True   \n",
       "26026  chr17:18806240-18806249:+  chr11:62894727-62894736:+             True   \n",
       "\n",
       "      orth_second_order                   orth_third_order.details  \\\n",
       "10169              True  [(ENSG00000171928, [ENSMUSG00000014177])]   \n",
       "10170              True  [(ENSG00000171928, [ENSMUSG00000014177])]   \n",
       "26026              True  [(ENSG00000171928, [ENSMUSG00000014177])]   \n",
       "\n",
       "      orth_third_order gene_orthology_status  orthologous_gene_set  \\\n",
       "10169             True                 first  {ENSMUSG00000014177}   \n",
       "10170             True                 first  {ENSMUSG00000014177}   \n",
       "26026             True                 first  {ENSMUSG00000014177}   \n",
       "\n",
       "      annotated_lifted_region_bound_gene_set  \n",
       "10169                   {ENSMUSG00000014177}  \n",
       "10170                   {ENSMUSG00000014177}  \n",
       "26026                   {ENSMUSG00000014177}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "name           three_prime_UTR;TVP23B;protein_coding;ENSG0000...\n",
       "gene_id                                          ENSG00000171928\n",
       "main_annot                                       three_prime_UTR\n",
       "annot_set                                      {three_prime_UTR}\n",
       "gene_id_set                                    {ENSG00000171928}\n",
       "Name: chr17:18806058-18806067:+, dtype: object"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(m_id_to_meta.loc[[\"chr11:62894762-62894771:+\"], :])\n",
    "\n",
    "display(h_id_to_meta.loc[[\"chr17:15537415-15537423:-\"], :])\n",
    "display(m_lift.loc[lambda df: df[\"identifier.mouse_source\"] == \"chr11:62894762-62894771:+\", :])\n",
    "\n",
    "display(orth.loc[lambda df: df[\"human_ensid\"] == \"ENSG00000175106\", :])\n",
    "display(orth.loc[lambda df: df[\"mouse_ensid\"] == \"ENSMUSG00000014177\", :])\n",
    "\n",
    "\n",
    "x = h_id_to_meta.loc[lambda df: df[\"gene_id\"].isin([\"ENSG00000171928\", \"ENSG00000175106\", \"ENSG00000259024\"]), :]\n",
    "display(h_lift.loc[lambda df: df[\"identifier.human_source\"].isin(x.index), :])\n",
    "# First example of a human peak region's annotation\n",
    "display(h_id_to_meta.loc[\"chr17:18806058-18806067:+\", :])\n",
    "\n",
    "\n",
    "# This example is a situation of a likely species-specific duplication, where the duplicated genes are all mapped to\n",
    "# the same other-species ortholog, while in the other-way, only one of the duplicated genes is identified as orthologous.\n",
    "# Basically : from human to mouse, ENSMUSG00000014177 (bound) maps to 'ENSG00000175106' (not bound) while\n",
    "# the human gene bound is ENSG00000171928 that maps to ENSMUSG00000014177.\n",
    "# From gene names, it seems the \"best\" orthologous gene is actually ENSG00000171928, matching the name of ENSMUSG00000014177\n",
    "# (Tvp23b and TVP23B) while ENSG00000175106 is named differently (TVP23C).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e808c03f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>human_ensid</th>\n",
       "      <th>human_name</th>\n",
       "      <th>mouse_ensid</th>\n",
       "      <th>mouse_name</th>\n",
       "      <th>relationtype_id</th>\n",
       "      <th>supported_by_resource</th>\n",
       "      <th>support_level</th>\n",
       "      <th>support_details</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14295</th>\n",
       "      <td>ENSG00000171928</td>\n",
       "      <td>TVP23B</td>\n",
       "      <td>ENSMUSG00000014177</td>\n",
       "      <td>Tvp23b</td>\n",
       "      <td>1:n</td>\n",
       "      <td>biomart;;omabrowser;;mgi</td>\n",
       "      <td>4</td>\n",
       "      <td>biomart:mouse_to_human;;omabrowser:human_to_mo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15041</th>\n",
       "      <td>ENSG00000175106</td>\n",
       "      <td>TVP23C</td>\n",
       "      <td>ENSMUSG00000014177</td>\n",
       "      <td>Tvp23b</td>\n",
       "      <td>1:n</td>\n",
       "      <td>omabrowser;;mgi</td>\n",
       "      <td>3</td>\n",
       "      <td>omabrowser:human_to_mouse;mouse_to_human;;mgi:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29274</th>\n",
       "      <td>ENSG00000259024</td>\n",
       "      <td>TVP23C-CDRT4</td>\n",
       "      <td>ENSMUSG00000014177</td>\n",
       "      <td>Tvp23b</td>\n",
       "      <td>1:n</td>\n",
       "      <td>biomart;;omabrowser;;mgi</td>\n",
       "      <td>4</td>\n",
       "      <td>biomart:mouse_to_human;;omabrowser:human_to_mo...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           human_ensid    human_name         mouse_ensid mouse_name  \\\n",
       "14295  ENSG00000171928        TVP23B  ENSMUSG00000014177     Tvp23b   \n",
       "15041  ENSG00000175106        TVP23C  ENSMUSG00000014177     Tvp23b   \n",
       "29274  ENSG00000259024  TVP23C-CDRT4  ENSMUSG00000014177     Tvp23b   \n",
       "\n",
       "      relationtype_id     supported_by_resource  support_level  \\\n",
       "14295             1:n  biomart;;omabrowser;;mgi              4   \n",
       "15041             1:n           omabrowser;;mgi              3   \n",
       "29274             1:n  biomart;;omabrowser;;mgi              4   \n",
       "\n",
       "                                         support_details  \n",
       "14295  biomart:mouse_to_human;;omabrowser:human_to_mo...  \n",
       "15041  omabrowser:human_to_mouse;mouse_to_human;;mgi:...  \n",
       "29274  biomart:mouse_to_human;;omabrowser:human_to_mo...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier.mouse_source</th>\n",
       "      <th>identifier.human_target</th>\n",
       "      <th>orth_first_order</th>\n",
       "      <th>orth_second_order</th>\n",
       "      <th>orth_third_order.details</th>\n",
       "      <th>orth_third_order</th>\n",
       "      <th>gene_orthology_status</th>\n",
       "      <th>orthologous_gene_set</th>\n",
       "      <th>annotated_lifted_region_bound_gene_set</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2650</th>\n",
       "      <td>chr11:62894762-62894771:+</td>\n",
       "      <td>chr17:15537415-15537423:-</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSMUSG00000014177, [ENSG00000175106])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSG00000175106}</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651</th>\n",
       "      <td>chr11:62894772-62894781:+</td>\n",
       "      <td>chr17:15537405-15537414:-</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSMUSG00000014177, [ENSG00000175106])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSG00000175106}</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2652</th>\n",
       "      <td>chr11:62894850-62894859:+</td>\n",
       "      <td>chr17:15537333-15537342:-</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSMUSG00000014177, [ENSG00000175106])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSG00000175106}</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        identifier.mouse_source    identifier.human_target orth_first_order  \\\n",
       "2650  chr11:62894762-62894771:+  chr17:15537415-15537423:-             True   \n",
       "2651  chr11:62894772-62894781:+  chr17:15537405-15537414:-             True   \n",
       "2652  chr11:62894850-62894859:+  chr17:15537333-15537342:-             True   \n",
       "\n",
       "     orth_second_order                   orth_third_order.details  \\\n",
       "2650              True  [(ENSMUSG00000014177, [ENSG00000175106])]   \n",
       "2651              True  [(ENSMUSG00000014177, [ENSG00000175106])]   \n",
       "2652              True  [(ENSMUSG00000014177, [ENSG00000175106])]   \n",
       "\n",
       "     orth_third_order gene_orthology_status orthologous_gene_set  \\\n",
       "2650             True                 first    {ENSG00000175106}   \n",
       "2651             True                 first    {ENSG00000175106}   \n",
       "2652             True                 first    {ENSG00000175106}   \n",
       "\n",
       "     annotated_lifted_region_bound_gene_set  \n",
       "2650                                     {}  \n",
       "2651                                     {}  \n",
       "2652                                     {}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier.human_source</th>\n",
       "      <th>identifier.mouse_target</th>\n",
       "      <th>orth_first_order</th>\n",
       "      <th>orth_second_order</th>\n",
       "      <th>orth_third_order.details</th>\n",
       "      <th>orth_third_order</th>\n",
       "      <th>gene_orthology_status</th>\n",
       "      <th>orthologous_gene_set</th>\n",
       "      <th>annotated_lifted_region_bound_gene_set</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10169</th>\n",
       "      <td>chr17:18806058-18806067:+</td>\n",
       "      <td>chr11:62894538-62894546:+</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSG00000171928, [ENSMUSG00000014177])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSMUSG00000014177}</td>\n",
       "      <td>{ENSMUSG00000014177}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10170</th>\n",
       "      <td>chr17:18806269-18806278:+</td>\n",
       "      <td>chr11:62894764-62894772:+</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSG00000171928, [ENSMUSG00000014177])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSMUSG00000014177}</td>\n",
       "      <td>{ENSMUSG00000014177}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26026</th>\n",
       "      <td>chr17:18806240-18806249:+</td>\n",
       "      <td>chr11:62894727-62894736:+</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSG00000171928, [ENSMUSG00000014177])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSMUSG00000014177}</td>\n",
       "      <td>{ENSMUSG00000014177}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         identifier.human_source    identifier.mouse_target orth_first_order  \\\n",
       "10169  chr17:18806058-18806067:+  chr11:62894538-62894546:+             True   \n",
       "10170  chr17:18806269-18806278:+  chr11:62894764-62894772:+             True   \n",
       "26026  chr17:18806240-18806249:+  chr11:62894727-62894736:+             True   \n",
       "\n",
       "      orth_second_order                   orth_third_order.details  \\\n",
       "10169              True  [(ENSG00000171928, [ENSMUSG00000014177])]   \n",
       "10170              True  [(ENSG00000171928, [ENSMUSG00000014177])]   \n",
       "26026              True  [(ENSG00000171928, [ENSMUSG00000014177])]   \n",
       "\n",
       "      orth_third_order gene_orthology_status  orthologous_gene_set  \\\n",
       "10169             True                 first  {ENSMUSG00000014177}   \n",
       "10170             True                 first  {ENSMUSG00000014177}   \n",
       "26026             True                 first  {ENSMUSG00000014177}   \n",
       "\n",
       "      annotated_lifted_region_bound_gene_set  \n",
       "10169                   {ENSMUSG00000014177}  \n",
       "10170                   {ENSMUSG00000014177}  \n",
       "26026                   {ENSMUSG00000014177}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# CONCLUSION :\n",
    "#\n",
    "# - At the level of the GENE, this could be considered as a shared binding.\n",
    "# - But when considering the PEAKS, we will get back to a \"clearer\" situation that the local region is not orthologously bound.\n",
    "\n",
    "display(orth.loc[lambda df: df[\"mouse_ensid\"] == \"ENSMUSG00000014177\", :])\n",
    "\n",
    "# Showing mouse peaks\n",
    "display(\n",
    "    m_lift.loc[\n",
    "        m_lift[\"identifier.mouse_source\"].isin(\n",
    "            m_id_to_meta.loc[lambda df: df[\"gene_id\"] == \"ENSMUSG00000014177\", :].index.values\n",
    "        ),\n",
    "        :,\n",
    "    ]\n",
    ")\n",
    "\n",
    "# Showing human peaks:\n",
    "display(\n",
    "    h_lift.loc[\n",
    "        h_lift[\"annotated_lifted_region_bound_gene_set\"]\n",
    "        .apply(\n",
    "            lambda v: v\n",
    "            & set(\n",
    "                [\n",
    "                    \"ENSMUSG00000014177\",\n",
    "                ]\n",
    "            )\n",
    "        )\n",
    "        .apply(len)\n",
    "        > 0,\n",
    "        :,\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2d9a9ec",
   "metadata": {},
   "source": [
    "#### Final numbers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea0ed75d",
   "metadata": {},
   "source": [
    "In summary : we either have cases of unmapped peaks' genes that actually have human peaks in the orthologous gene,\n",
    "or cases where the duplicated genes' orthologies map correctly in one direction but not the other.\n",
    "\n",
    "We thus need to correct the \"species-specific genes\" from unmapped peaks as \"bound in common\".\n",
    "For the issues in annotations : we trust the orthologies to correctly disentangle the 1:n duplicated orthologies into 1:1,\n",
    "and thus consider these genes as \"species-specific\" bound genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "4c29879c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "len(m_genes_bound_orth_complete)=1312\n",
      "len(m_genes_no_bound_orth_complete)=824\n",
      "(len(m_genes_bound_orth_complete) + len(m_genes_no_bound_orth_complete))=2136\n",
      "len(m_bound_genes)=2136\n",
      "(len(m_genes_bound_orth_complete) + len(m_genes_no_bound_orth_complete)) == len(m_bound_genes)=True\n"
     ]
    }
   ],
   "source": [
    "m_genes_bound_orth_complete = m_genes_bound_orth | (\n",
    "    set(m_peaks_in_missing_genes.loc[lambda df: df[\"is_in_umap\"], \"gene_id\"].unique()) & m_bound_genes\n",
    ")\n",
    "\n",
    "m_genes_no_bound_orth_complete = m_genes_no_bound_orth | (\n",
    "    set(m_peaks_in_missing_genes.loc[lambda df: (~df[\"is_in_umap\"]), \"gene_id\"].unique()) & m_bound_genes\n",
    ")\n",
    "# Final correction on cases of n:n mappings : \"bound\" takes over \"no-bound\"\n",
    "m_genes_no_bound_orth_complete = m_genes_no_bound_orth_complete - m_genes_bound_orth_complete\n",
    "\n",
    "# FINAL TABLES\n",
    "\n",
    "print(f\"{len(m_genes_bound_orth_complete)=}\")\n",
    "print(f\"{len(m_genes_no_bound_orth_complete)=}\")\n",
    "print(f\"{(len(m_genes_bound_orth_complete) + len(m_genes_no_bound_orth_complete))=}\")\n",
    "print(f\"{len(m_bound_genes)=}\")\n",
    "print(f\"{(len(m_genes_bound_orth_complete) + len(m_genes_no_bound_orth_complete)) == len(m_bound_genes)=}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a00d261b",
   "metadata": {},
   "source": [
    "### Gene counts - Human"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f09254de",
   "metadata": {},
   "source": [
    "#### First attempt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85a1e6a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "h_has_gene_bound_orth = (\n",
    "    (h_lift[\"gene_orthology_status\"] != \"not_orthologous\")\n",
    "    & (h_lift[\"annotated_lifted_region_bound_gene_set\"].apply(len) > 0)\n",
    ").values\n",
    "\n",
    "###\n",
    "\n",
    "h_peaks_w_bound_orth = h_lift.loc[h_has_gene_bound_orth, \"identifier.human_source\"].values\n",
    "h_peaks_wo_bound_orth = h_lift.loc[~h_has_gene_bound_orth, \"identifier.human_source\"].values\n",
    "\n",
    "###\n",
    "\n",
    "h_genes_bound_orth = set(h_id_to_meta.loc[h_peaks_w_bound_orth, \"gene_id\"].unique())\n",
    "\n",
    "h_genes_from_mouse_peaks_w_orth = functools.reduce(set.union, m_lift[\"annotated_lifted_region_bound_gene_set\"].values)\n",
    "\n",
    "h_genes_from_umap = set(h_id_to_meta.loc[h_umap[\"identifier\"].values, \"gene_id\"].unique())\n",
    "\n",
    "\n",
    "h_genes_no_bound_orth = (\n",
    "    (h_genes_from_umap | set(h_id_to_meta.loc[h_peaks_wo_bound_orth, \"gene_id\"].unique()))\n",
    "    - h_genes_from_mouse_peaks_w_orth\n",
    "    - h_genes_bound_orth\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "fd347460",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "len(h_genes_bound_orth)=1268\n",
      "len(h_genes_no_bound_orth)=1651\n"
     ]
    }
   ],
   "source": [
    "print(f\"{len(h_genes_bound_orth)=}\")\n",
    "print(f\"{len(h_genes_no_bound_orth)=}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8ee2ff2",
   "metadata": {},
   "source": [
    "#### Missing genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "3ec0badf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of missing genes: 50\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True     133\n",
       "False    123\n",
       "Name: is_in_umap, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    }
   ],
   "source": [
    "h_missing = set(h_full[\"gene_id\"].unique()) - h_genes_bound_orth - h_genes_no_bound_orth\n",
    "\n",
    "print(f\"Number of missing genes: {len(h_missing):}\")\n",
    "\n",
    "h_peaks_in_missing_genes = h_full.loc[lambda df: df[\"gene_id\"].isin(h_missing), :].assign(\n",
    "    is_in_umap=lambda df: df[\"identifier\"].isin(h_umap[\"identifier\"].values)\n",
    ")\n",
    "\n",
    "display(h_peaks_in_missing_genes[\"is_in_umap\"].value_counts())\n",
    "\n",
    "# EXPECTED:\n",
    "# True     133\n",
    "# False    123\n",
    "\n",
    "\n",
    "# ALL THESE GENES ARE FOUND TO BE BOUND IN MOUSE PEAKS LIFTED TO HUMAN.\n",
    "# i.e. I discard them wrongly in the last set operation, as they are bound in human but from unmapped peaks.\n",
    "\n",
    "print(all([g in h_genes_from_mouse_peaks_w_orth for g in h_missing]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1870d2f6",
   "metadata": {},
   "source": [
    "#### Final numbers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64957983",
   "metadata": {},
   "source": [
    "In summary : we have either cases of unmapped peaks' genes that actually have human peaks in the orthologous gene,\n",
    "or cases where the duplicated genes' orthologies map correctly in one direction but not the other.\n",
    "\n",
    "We thus need to correct the \"species-specific genes\" from unmapped peaks as \"bound in common\".\n",
    "For the issues in annotations : we trust the orthologies to correctly disentangle the 1:n duplicated orthologies into 1:1,\n",
    "and thus consider these genes as \"species-specific\" bound genes.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "efe7b2f8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "len(h_genes_bound_orth_complete)=1302\n",
      "len(h_genes_no_bound_orth_complete)=1667\n",
      "(len(h_genes_bound_orth_complete) + len(h_genes_no_bound_orth_complete))=2969\n",
      "len(h_bound_genes)=2969\n",
      "(len(h_genes_bound_orth_complete) + len(h_genes_no_bound_orth_complete)) == len(h_bound_genes)=True\n"
     ]
    }
   ],
   "source": [
    "h_genes_bound_orth_complete = h_genes_bound_orth | (\n",
    "    set(h_peaks_in_missing_genes.loc[lambda df: df[\"is_in_umap\"], \"gene_id\"].unique()) & h_bound_genes\n",
    ")\n",
    "\n",
    "h_genes_no_bound_orth_complete = h_genes_no_bound_orth | (\n",
    "    set(h_peaks_in_missing_genes.loc[lambda df: (~df[\"is_in_umap\"]), \"gene_id\"].unique()) & h_bound_genes\n",
    ")\n",
    "# Final correction on cases of n:n mappings : \"bound\" takes over \"no-bound\"\n",
    "h_genes_no_bound_orth_complete = h_genes_no_bound_orth_complete - h_genes_bound_orth_complete\n",
    "\n",
    "# FINAL TABLES\n",
    "\n",
    "print(f\"{len(h_genes_bound_orth_complete)=}\")\n",
    "print(f\"{len(h_genes_no_bound_orth_complete)=}\")\n",
    "print(f\"{(len(h_genes_bound_orth_complete) + len(h_genes_no_bound_orth_complete))=}\")\n",
    "print(f\"{len(h_bound_genes)=}\")\n",
    "print(f\"{(len(h_genes_bound_orth_complete) + len(h_genes_no_bound_orth_complete)) == len(h_bound_genes)=}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "388c652c",
   "metadata": {},
   "source": [
    "### Plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8917a64",
   "metadata": {},
   "source": [
    "We actually have 4 numbers rather than three: \n",
    "\n",
    "- mouse-specific genes \n",
    "- mouse genes with an ortholog bound in human \n",
    "- human-specific genes\n",
    "- human genes with an ortholog bound in mouse\n",
    "\n",
    "But the number of gene in common is going to differ between human and mouse owing to multi-orthologies.\n",
    "\n",
    "To simplify the situation, we use the mouse-genes numbers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "a8f93f19",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "len(m_genes_no_bound_orth_complete)=824\n",
      "len(m_genes_bound_orth_complete)=1312\n",
      "len(h_genes_bound_orth_complete)=1302\n",
      "len(h_genes_no_bound_orth_complete)=1667\n"
     ]
    }
   ],
   "source": [
    "print(f\"{len(m_genes_no_bound_orth_complete)=}\")\n",
    "print(f\"{len(m_genes_bound_orth_complete)=}\")\n",
    "print(f\"{len(h_genes_bound_orth_complete)=}\")\n",
    "print(f\"{len(h_genes_no_bound_orth_complete)=}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "d2cc7d0f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"utf-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       "  \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<svg xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"424.392237pt\" height=\"384.401817pt\" viewBox=\"0 0 424.392237 384.401817\" xmlns=\"http://www.w3.org/2000/svg\" version=\"1.1\">\n",
       " <metadata>\n",
       "  <rdf:RDF xmlns:dc=\"http://purl.org/dc/elements/1.1/\" xmlns:cc=\"http://creativecommons.org/ns#\" xmlns:rdf=\"http://www.w3.org/1999/02/22-rdf-syntax-ns#\">\n",
       "   <cc:Work>\n",
       "    <dc:type rdf:resource=\"http://purl.org/dc/dcmitype/StillImage\"/>\n",
       "    <dc:date>2025-07-14T05:47:55.237070</dc:date>\n",
       "    <dc:format>image/svg+xml</dc:format>\n",
       "    <dc:creator>\n",
       "     <cc:Agent>\n",
       "      <dc:title>Matplotlib v3.8.2, https://matplotlib.org/</dc:title>\n",
       "     </cc:Agent>\n",
       "    </dc:creator>\n",
       "   </cc:Work>\n",
       "  </rdf:RDF>\n",
       " </metadata>\n",
       " <defs>\n",
       "  <style type=\"text/css\">*{stroke-linejoin: round; stroke-linecap: butt}</style>\n",
       " </defs>\n",
       " <g id=\"figure_1\">\n",
       "  <g id=\"patch_1\">\n",
       "   <path d=\"M 0 384.401817 \n",
       "L 424.392237 384.401817 \n",
       "L 424.392237 0 \n",
       "L 0 0 \n",
       "z\n",
       "\" style=\"fill: #ffffff\"/>\n",
       "  </g>\n",
       "  <g id=\"axes_1\">\n",
       "   <g id=\"patch_2\">\n",
       "    <path d=\"M 174.021428 323.583958 \n",
       "C 155.037614 310.931347 139.469021 293.784188 128.702907 273.670398 \n",
       "C 117.936793 253.556608 112.302721 231.091894 112.302721 208.278 \n",
       "C 112.302721 185.464106 117.936793 162.999392 128.702907 142.885602 \n",
       "C 139.469021 122.771812 155.037614 105.624653 174.021428 92.972042 \n",
       "C 157.070133 89.776827 139.620944 90.358628 122.920134 94.675888 \n",
       "C 106.219324 98.993149 90.674196 106.940579 77.395027 117.950552 \n",
       "C 64.115859 128.960524 53.426506 142.764524 46.090434 158.376629 \n",
       "C 38.754361 173.988735 34.950483 191.028194 34.950483 208.278 \n",
       "C 34.950483 225.527806 38.754361 242.567265 46.090434 258.179371 \n",
       "C 53.426506 273.791476 64.115859 287.595476 77.395027 298.605448 \n",
       "C 90.674196 309.615421 106.219324 317.562851 122.920134 321.880112 \n",
       "C 139.620944 326.197372 157.070133 326.779173 174.021428 323.583958 \n",
       "\" clip-path=\"url(#p707d8eeb67)\" style=\"fill: #bbbbbb; opacity: 0.4\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_3\">\n",
       "    <path d=\"M 174.021428 92.972042 \n",
       "C 200.890725 98.036738 225.162641 112.337407 242.614417 133.386059 \n",
       "C 260.066193 154.434711 269.623454 180.935538 269.623454 208.278 \n",
       "C 269.623454 235.620462 260.066193 262.121289 242.614417 283.169941 \n",
       "C 225.162641 304.218593 200.890725 318.519262 174.021428 323.583958 \n",
       "C 194.88884 337.491978 219.145518 345.477884 244.193894 346.686492 \n",
       "C 269.24227 347.8951 294.15513 342.281672 316.264635 330.44733 \n",
       "C 338.374141 318.612989 356.861866 300.995805 369.748021 279.482347 \n",
       "C 382.634176 257.968888 389.441754 233.355517 389.441754 208.278 \n",
       "C 389.441754 183.200483 382.634176 158.587112 369.748021 137.073653 \n",
       "C 356.861866 115.560195 338.374141 97.943011 316.264635 86.10867 \n",
       "C 294.15513 74.274328 269.24227 68.6609 244.193894 69.869508 \n",
       "C 219.145518 71.078116 194.88884 79.064022 174.021428 92.972042 \n",
       "\" clip-path=\"url(#p707d8eeb67)\" style=\"fill: #bbbbbb; opacity: 0.4\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_4\">\n",
       "    <path d=\"M 174.021428 323.583958 \n",
       "C 200.890725 318.519262 225.162641 304.218593 242.614417 283.169941 \n",
       "C 260.066193 262.121289 269.623454 235.620462 269.623454 208.278 \n",
       "C 269.623454 180.935538 260.066193 154.434711 242.614417 133.386059 \n",
       "C 225.162641 112.337407 200.890725 98.036738 174.021428 92.972042 \n",
       "C 155.037614 105.624653 139.469021 122.771812 128.702907 142.885602 \n",
       "C 117.936793 162.999392 112.302721 185.464106 112.302721 208.278 \n",
       "C 112.302721 231.091894 117.936793 253.556608 128.702907 273.670398 \n",
       "C 139.469021 293.784188 155.037614 310.931347 174.021428 323.583958 \n",
       "\" clip-path=\"url(#p707d8eeb67)\" style=\"fill: #ffffff; opacity: 0.4\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_5\">\n",
       "    <path d=\"M 152.286968 325.614485 \n",
       "C 183.404968 325.614485 213.252644 313.251173 235.256393 291.247424 \n",
       "C 257.260141 269.243676 269.623454 239.396 269.623454 208.278 \n",
       "C 269.623454 177.16 257.260141 147.312324 235.256393 125.308576 \n",
       "C 213.252644 103.304827 183.404968 90.941515 152.286968 90.941515 \n",
       "C 121.168969 90.941515 91.321293 103.304827 69.317544 125.308576 \n",
       "C 47.313796 147.312324 34.950483 177.16 34.950483 208.278 \n",
       "C 34.950483 239.396 47.313796 269.243676 69.317544 291.247424 \n",
       "C 91.321293 313.251173 121.168969 325.614485 152.286968 325.614485 \n",
       "L 152.286968 325.614485 \n",
       "z\n",
       "\" clip-path=\"url(#p707d8eeb67)\" style=\"fill: none; stroke: #000000; stroke-width: 2; stroke-linejoin: miter\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_6\">\n",
       "    <path d=\"M 250.872237 346.847517 \n",
       "C 287.621303 346.847517 322.870169 332.246958 348.855682 306.261445 \n",
       "C 374.841196 280.275932 389.441754 245.027065 389.441754 208.278 \n",
       "C 389.441754 171.528935 374.841196 136.280068 348.855682 110.294555 \n",
       "C 322.870169 84.309042 287.621303 69.708483 250.872237 69.708483 \n",
       "C 214.123172 69.708483 178.874306 84.309042 152.888792 110.294555 \n",
       "C 126.903279 136.280068 112.302721 171.528935 112.302721 208.278 \n",
       "C 112.302721 245.027065 126.903279 280.275932 152.888792 306.261445 \n",
       "C 178.874306 332.246958 214.123172 346.847517 250.872237 346.847517 \n",
       "L 250.872237 346.847517 \n",
       "z\n",
       "\" clip-path=\"url(#p707d8eeb67)\" style=\"fill: none; stroke: #000000; stroke-width: 2; stroke-linejoin: miter\"/>\n",
       "   </g>\n",
       "   <g id=\"text_1\">\n",
       "    <!-- 824 -->\n",
       "    <g transform=\"translate(60.280352 212.41425) scale(0.16 -0.16)\">\n",
       "     <defs>\n",
       "      <path id=\"ArialMT-38\" d=\"M 1131 2484 \n",
       "Q 781 2613 612 2850 \n",
       "Q 444 3088 444 3419 \n",
       "Q 444 3919 803 4259 \n",
       "Q 1163 4600 1759 4600 \n",
       "Q 2359 4600 2725 4251 \n",
       "Q 3091 3903 3091 3403 \n",
       "Q 3091 3084 2923 2848 \n",
       "Q 2756 2613 2416 2484 \n",
       "Q 2838 2347 3058 2040 \n",
       "Q 3278 1734 3278 1309 \n",
       "Q 3278 722 2862 322 \n",
       "Q 2447 -78 1769 -78 \n",
       "Q 1091 -78 675 323 \n",
       "Q 259 725 259 1325 \n",
       "Q 259 1772 486 2073 \n",
       "Q 713 2375 1131 2484 \n",
       "z\n",
       "M 1019 3438 \n",
       "Q 1019 3113 1228 2906 \n",
       "Q 1438 2700 1772 2700 \n",
       "Q 2097 2700 2305 2904 \n",
       "Q 2513 3109 2513 3406 \n",
       "Q 2513 3716 2298 3927 \n",
       "Q 2084 4138 1766 4138 \n",
       "Q 1444 4138 1231 3931 \n",
       "Q 1019 3725 1019 3438 \n",
       "z\n",
       "M 838 1322 \n",
       "Q 838 1081 952 856 \n",
       "Q 1066 631 1291 507 \n",
       "Q 1516 384 1775 384 \n",
       "Q 2178 384 2440 643 \n",
       "Q 2703 903 2703 1303 \n",
       "Q 2703 1709 2433 1975 \n",
       "Q 2163 2241 1756 2241 \n",
       "Q 1359 2241 1098 1978 \n",
       "Q 838 1716 838 1322 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-32\" d=\"M 3222 541 \n",
       "L 3222 0 \n",
       "L 194 0 \n",
       "Q 188 203 259 391 \n",
       "Q 375 700 629 1000 \n",
       "Q 884 1300 1366 1694 \n",
       "Q 2113 2306 2375 2664 \n",
       "Q 2638 3022 2638 3341 \n",
       "Q 2638 3675 2398 3904 \n",
       "Q 2159 4134 1775 4134 \n",
       "Q 1369 4134 1125 3890 \n",
       "Q 881 3647 878 3216 \n",
       "L 300 3275 \n",
       "Q 359 3922 746 4261 \n",
       "Q 1134 4600 1788 4600 \n",
       "Q 2447 4600 2831 4234 \n",
       "Q 3216 3869 3216 3328 \n",
       "Q 3216 3053 3103 2787 \n",
       "Q 2991 2522 2730 2228 \n",
       "Q 2469 1934 1863 1422 \n",
       "Q 1356 997 1212 845 \n",
       "Q 1069 694 975 541 \n",
       "L 3222 541 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-34\" d=\"M 2069 0 \n",
       "L 2069 1097 \n",
       "L 81 1097 \n",
       "L 81 1613 \n",
       "L 2172 4581 \n",
       "L 2631 4581 \n",
       "L 2631 1613 \n",
       "L 3250 1613 \n",
       "L 3250 1097 \n",
       "L 2631 1097 \n",
       "L 2631 0 \n",
       "L 2069 0 \n",
       "z\n",
       "M 2069 1613 \n",
       "L 2069 3678 \n",
       "L 634 1613 \n",
       "L 2069 1613 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "     </defs>\n",
       "     <use xlink:href=\"#ArialMT-38\"/>\n",
       "     <use xlink:href=\"#ArialMT-32\" x=\"55.615234\"/>\n",
       "     <use xlink:href=\"#ArialMT-34\" x=\"111.230469\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_2\">\n",
       "    <!-- 1,667 -->\n",
       "    <g transform=\"translate(309.515104 212.41425) scale(0.16 -0.16)\">\n",
       "     <defs>\n",
       "      <path id=\"ArialMT-31\" d=\"M 2384 0 \n",
       "L 1822 0 \n",
       "L 1822 3584 \n",
       "Q 1619 3391 1289 3197 \n",
       "Q 959 3003 697 2906 \n",
       "L 697 3450 \n",
       "Q 1169 3672 1522 3987 \n",
       "Q 1875 4303 2022 4600 \n",
       "L 2384 4600 \n",
       "L 2384 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-2c\" d=\"M 569 0 \n",
       "L 569 641 \n",
       "L 1209 641 \n",
       "L 1209 0 \n",
       "Q 1209 -353 1084 -570 \n",
       "Q 959 -788 688 -906 \n",
       "L 531 -666 \n",
       "Q 709 -588 793 -436 \n",
       "Q 878 -284 888 0 \n",
       "L 569 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-36\" d=\"M 3184 3459 \n",
       "L 2625 3416 \n",
       "Q 2550 3747 2413 3897 \n",
       "Q 2184 4138 1850 4138 \n",
       "Q 1581 4138 1378 3988 \n",
       "Q 1113 3794 959 3422 \n",
       "Q 806 3050 800 2363 \n",
       "Q 1003 2672 1297 2822 \n",
       "Q 1591 2972 1913 2972 \n",
       "Q 2475 2972 2870 2558 \n",
       "Q 3266 2144 3266 1488 \n",
       "Q 3266 1056 3080 686 \n",
       "Q 2894 316 2569 119 \n",
       "Q 2244 -78 1831 -78 \n",
       "Q 1128 -78 684 439 \n",
       "Q 241 956 241 2144 \n",
       "Q 241 3472 731 4075 \n",
       "Q 1159 4600 1884 4600 \n",
       "Q 2425 4600 2770 4297 \n",
       "Q 3116 3994 3184 3459 \n",
       "z\n",
       "M 888 1484 \n",
       "Q 888 1194 1011 928 \n",
       "Q 1134 663 1356 523 \n",
       "Q 1578 384 1822 384 \n",
       "Q 2178 384 2434 671 \n",
       "Q 2691 959 2691 1453 \n",
       "Q 2691 1928 2437 2201 \n",
       "Q 2184 2475 1800 2475 \n",
       "Q 1419 2475 1153 2201 \n",
       "Q 888 1928 888 1484 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-37\" d=\"M 303 3981 \n",
       "L 303 4522 \n",
       "L 3269 4522 \n",
       "L 3269 4084 \n",
       "Q 2831 3619 2401 2847 \n",
       "Q 1972 2075 1738 1259 \n",
       "Q 1569 684 1522 0 \n",
       "L 944 0 \n",
       "Q 953 541 1156 1306 \n",
       "Q 1359 2072 1739 2783 \n",
       "Q 2119 3494 2547 3981 \n",
       "L 303 3981 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "     </defs>\n",
       "     <use xlink:href=\"#ArialMT-31\"/>\n",
       "     <use xlink:href=\"#ArialMT-2c\" x=\"55.615234\"/>\n",
       "     <use xlink:href=\"#ArialMT-36\" x=\"83.398438\"/>\n",
       "     <use xlink:href=\"#ArialMT-36\" x=\"139.013672\"/>\n",
       "     <use xlink:href=\"#ArialMT-37\" x=\"194.628906\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_3\">\n",
       "    <!-- 1,312 | 1,302 -->\n",
       "    <g transform=\"translate(144.404337 212.418) scale(0.16 -0.16)\">\n",
       "     <defs>\n",
       "      <path id=\"ArialMT-33\" d=\"M 269 1209 \n",
       "L 831 1284 \n",
       "Q 928 806 1161 595 \n",
       "Q 1394 384 1728 384 \n",
       "Q 2125 384 2398 659 \n",
       "Q 2672 934 2672 1341 \n",
       "Q 2672 1728 2419 1979 \n",
       "Q 2166 2231 1775 2231 \n",
       "Q 1616 2231 1378 2169 \n",
       "L 1441 2663 \n",
       "Q 1497 2656 1531 2656 \n",
       "Q 1891 2656 2178 2843 \n",
       "Q 2466 3031 2466 3422 \n",
       "Q 2466 3731 2256 3934 \n",
       "Q 2047 4138 1716 4138 \n",
       "Q 1388 4138 1169 3931 \n",
       "Q 950 3725 888 3313 \n",
       "L 325 3413 \n",
       "Q 428 3978 793 4289 \n",
       "Q 1159 4600 1703 4600 \n",
       "Q 2078 4600 2393 4439 \n",
       "Q 2709 4278 2876 4000 \n",
       "Q 3044 3722 3044 3409 \n",
       "Q 3044 3113 2884 2869 \n",
       "Q 2725 2625 2413 2481 \n",
       "Q 2819 2388 3044 2092 \n",
       "Q 3269 1797 3269 1353 \n",
       "Q 3269 753 2831 336 \n",
       "Q 2394 -81 1725 -81 \n",
       "Q 1122 -81 723 278 \n",
       "Q 325 638 269 1209 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-20\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-7c\" d=\"M 588 -1347 \n",
       "L 588 4659 \n",
       "L 1078 4659 \n",
       "L 1078 -1347 \n",
       "L 588 -1347 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-30\" d=\"M 266 2259 \n",
       "Q 266 3072 433 3567 \n",
       "Q 600 4063 929 4331 \n",
       "Q 1259 4600 1759 4600 \n",
       "Q 2128 4600 2406 4451 \n",
       "Q 2684 4303 2865 4023 \n",
       "Q 3047 3744 3150 3342 \n",
       "Q 3253 2941 3253 2259 \n",
       "Q 3253 1453 3087 958 \n",
       "Q 2922 463 2592 192 \n",
       "Q 2263 -78 1759 -78 \n",
       "Q 1097 -78 719 397 \n",
       "Q 266 969 266 2259 \n",
       "z\n",
       "M 844 2259 \n",
       "Q 844 1131 1108 757 \n",
       "Q 1372 384 1759 384 \n",
       "Q 2147 384 2411 759 \n",
       "Q 2675 1134 2675 2259 \n",
       "Q 2675 3391 2411 3762 \n",
       "Q 2147 4134 1753 4134 \n",
       "Q 1366 4134 1134 3806 \n",
       "Q 844 3388 844 2259 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "     </defs>\n",
       "     <use xlink:href=\"#ArialMT-31\"/>\n",
       "     <use xlink:href=\"#ArialMT-2c\" x=\"55.615234\"/>\n",
       "     <use xlink:href=\"#ArialMT-33\" x=\"83.398438\"/>\n",
       "     <use xlink:href=\"#ArialMT-31\" x=\"139.013672\"/>\n",
       "     <use xlink:href=\"#ArialMT-32\" x=\"194.628906\"/>\n",
       "     <use xlink:href=\"#ArialMT-20\" x=\"250.244141\"/>\n",
       "     <use xlink:href=\"#ArialMT-7c\" x=\"278.027344\"/>\n",
       "     <use xlink:href=\"#ArialMT-20\" x=\"304.003906\"/>\n",
       "     <use xlink:href=\"#ArialMT-31\" x=\"331.787109\"/>\n",
       "     <use xlink:href=\"#ArialMT-2c\" x=\"387.402344\"/>\n",
       "     <use xlink:href=\"#ArialMT-33\" x=\"415.185547\"/>\n",
       "     <use xlink:href=\"#ArialMT-30\" x=\"470.800781\"/>\n",
       "     <use xlink:href=\"#ArialMT-32\" x=\"526.416016\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_4\">\n",
       "    <!-- Mouse -->\n",
       "    <g transform=\"translate(94.662968 352.152785) scale(0.192 -0.192)\">\n",
       "     <defs>\n",
       "      <path id=\"ArialMT-4d\" d=\"M 475 0 \n",
       "L 475 4581 \n",
       "L 1388 4581 \n",
       "L 2472 1338 \n",
       "Q 2622 884 2691 659 \n",
       "Q 2769 909 2934 1394 \n",
       "L 4031 4581 \n",
       "L 4847 4581 \n",
       "L 4847 0 \n",
       "L 4263 0 \n",
       "L 4263 3834 \n",
       "L 2931 0 \n",
       "L 2384 0 \n",
       "L 1059 3900 \n",
       "L 1059 0 \n",
       "L 475 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-6f\" d=\"M 213 1659 \n",
       "Q 213 2581 725 3025 \n",
       "Q 1153 3394 1769 3394 \n",
       "Q 2453 3394 2887 2945 \n",
       "Q 3322 2497 3322 1706 \n",
       "Q 3322 1066 3130 698 \n",
       "Q 2938 331 2570 128 \n",
       "Q 2203 -75 1769 -75 \n",
       "Q 1072 -75 642 372 \n",
       "Q 213 819 213 1659 \n",
       "z\n",
       "M 791 1659 \n",
       "Q 791 1022 1069 705 \n",
       "Q 1347 388 1769 388 \n",
       "Q 2188 388 2466 706 \n",
       "Q 2744 1025 2744 1678 \n",
       "Q 2744 2294 2464 2611 \n",
       "Q 2184 2928 1769 2928 \n",
       "Q 1347 2928 1069 2612 \n",
       "Q 791 2297 791 1659 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-75\" d=\"M 2597 0 \n",
       "L 2597 488 \n",
       "Q 2209 -75 1544 -75 \n",
       "Q 1250 -75 995 37 \n",
       "Q 741 150 617 320 \n",
       "Q 494 491 444 738 \n",
       "Q 409 903 409 1263 \n",
       "L 409 3319 \n",
       "L 972 3319 \n",
       "L 972 1478 \n",
       "Q 972 1038 1006 884 \n",
       "Q 1059 663 1231 536 \n",
       "Q 1403 409 1656 409 \n",
       "Q 1909 409 2131 539 \n",
       "Q 2353 669 2445 892 \n",
       "Q 2538 1116 2538 1541 \n",
       "L 2538 3319 \n",
       "L 3100 3319 \n",
       "L 3100 0 \n",
       "L 2597 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-73\" d=\"M 197 991 \n",
       "L 753 1078 \n",
       "Q 800 744 1014 566 \n",
       "Q 1228 388 1613 388 \n",
       "Q 2000 388 2187 545 \n",
       "Q 2375 703 2375 916 \n",
       "Q 2375 1106 2209 1216 \n",
       "Q 2094 1291 1634 1406 \n",
       "Q 1016 1563 777 1677 \n",
       "Q 538 1791 414 1992 \n",
       "Q 291 2194 291 2438 \n",
       "Q 291 2659 392 2848 \n",
       "Q 494 3038 669 3163 \n",
       "Q 800 3259 1026 3326 \n",
       "Q 1253 3394 1513 3394 \n",
       "Q 1903 3394 2198 3281 \n",
       "Q 2494 3169 2634 2976 \n",
       "Q 2775 2784 2828 2463 \n",
       "L 2278 2388 \n",
       "Q 2241 2644 2061 2787 \n",
       "Q 1881 2931 1553 2931 \n",
       "Q 1166 2931 1000 2803 \n",
       "Q 834 2675 834 2503 \n",
       "Q 834 2394 903 2306 \n",
       "Q 972 2216 1119 2156 \n",
       "Q 1203 2125 1616 2013 \n",
       "Q 2213 1853 2448 1751 \n",
       "Q 2684 1650 2818 1456 \n",
       "Q 2953 1263 2953 975 \n",
       "Q 2953 694 2789 445 \n",
       "Q 2625 197 2315 61 \n",
       "Q 2006 -75 1616 -75 \n",
       "Q 969 -75 630 194 \n",
       "Q 291 463 197 991 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-65\" d=\"M 2694 1069 \n",
       "L 3275 997 \n",
       "Q 3138 488 2766 206 \n",
       "Q 2394 -75 1816 -75 \n",
       "Q 1088 -75 661 373 \n",
       "Q 234 822 234 1631 \n",
       "Q 234 2469 665 2931 \n",
       "Q 1097 3394 1784 3394 \n",
       "Q 2450 3394 2872 2941 \n",
       "Q 3294 2488 3294 1666 \n",
       "Q 3294 1616 3291 1516 \n",
       "L 816 1516 \n",
       "Q 847 969 1125 678 \n",
       "Q 1403 388 1819 388 \n",
       "Q 2128 388 2347 550 \n",
       "Q 2566 713 2694 1069 \n",
       "z\n",
       "M 847 1978 \n",
       "L 2700 1978 \n",
       "Q 2663 2397 2488 2606 \n",
       "Q 2219 2931 1791 2931 \n",
       "Q 1403 2931 1139 2672 \n",
       "Q 875 2413 847 1978 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "     </defs>\n",
       "     <use xlink:href=\"#ArialMT-4d\"/>\n",
       "     <use xlink:href=\"#ArialMT-6f\" x=\"83.300781\"/>\n",
       "     <use xlink:href=\"#ArialMT-75\" x=\"138.916016\"/>\n",
       "     <use xlink:href=\"#ArialMT-73\" x=\"194.53125\"/>\n",
       "     <use xlink:href=\"#ArialMT-65\" x=\"244.53125\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_5\">\n",
       "    <!-- Human -->\n",
       "    <g transform=\"translate(250.872237 373.385817) scale(0.192 -0.192)\">\n",
       "     <defs>\n",
       "      <path id=\"ArialMT-48\" d=\"M 513 0 \n",
       "L 513 4581 \n",
       "L 1119 4581 \n",
       "L 1119 2700 \n",
       "L 3500 2700 \n",
       "L 3500 4581 \n",
       "L 4106 4581 \n",
       "L 4106 0 \n",
       "L 3500 0 \n",
       "L 3500 2159 \n",
       "L 1119 2159 \n",
       "L 1119 0 \n",
       "L 513 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-6d\" d=\"M 422 0 \n",
       "L 422 3319 \n",
       "L 925 3319 \n",
       "L 925 2853 \n",
       "Q 1081 3097 1340 3245 \n",
       "Q 1600 3394 1931 3394 \n",
       "Q 2300 3394 2536 3241 \n",
       "Q 2772 3088 2869 2813 \n",
       "Q 3263 3394 3894 3394 \n",
       "Q 4388 3394 4653 3120 \n",
       "Q 4919 2847 4919 2278 \n",
       "L 4919 0 \n",
       "L 4359 0 \n",
       "L 4359 2091 \n",
       "Q 4359 2428 4304 2576 \n",
       "Q 4250 2725 4106 2815 \n",
       "Q 3963 2906 3769 2906 \n",
       "Q 3419 2906 3187 2673 \n",
       "Q 2956 2441 2956 1928 \n",
       "L 2956 0 \n",
       "L 2394 0 \n",
       "L 2394 2156 \n",
       "Q 2394 2531 2256 2718 \n",
       "Q 2119 2906 1806 2906 \n",
       "Q 1569 2906 1367 2781 \n",
       "Q 1166 2656 1075 2415 \n",
       "Q 984 2175 984 1722 \n",
       "L 984 0 \n",
       "L 422 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-61\" d=\"M 2588 409 \n",
       "Q 2275 144 1986 34 \n",
       "Q 1697 -75 1366 -75 \n",
       "Q 819 -75 525 192 \n",
       "Q 231 459 231 875 \n",
       "Q 231 1119 342 1320 \n",
       "Q 453 1522 633 1644 \n",
       "Q 813 1766 1038 1828 \n",
       "Q 1203 1872 1538 1913 \n",
       "Q 2219 1994 2541 2106 \n",
       "Q 2544 2222 2544 2253 \n",
       "Q 2544 2597 2384 2738 \n",
       "Q 2169 2928 1744 2928 \n",
       "Q 1347 2928 1158 2789 \n",
       "Q 969 2650 878 2297 \n",
       "L 328 2372 \n",
       "Q 403 2725 575 2942 \n",
       "Q 747 3159 1072 3276 \n",
       "Q 1397 3394 1825 3394 \n",
       "Q 2250 3394 2515 3294 \n",
       "Q 2781 3194 2906 3042 \n",
       "Q 3031 2891 3081 2659 \n",
       "Q 3109 2516 3109 2141 \n",
       "L 3109 1391 \n",
       "Q 3109 606 3145 398 \n",
       "Q 3181 191 3288 0 \n",
       "L 2700 0 \n",
       "Q 2613 175 2588 409 \n",
       "z\n",
       "M 2541 1666 \n",
       "Q 2234 1541 1622 1453 \n",
       "Q 1275 1403 1131 1340 \n",
       "Q 988 1278 909 1158 \n",
       "Q 831 1038 831 891 \n",
       "Q 831 666 1001 516 \n",
       "Q 1172 366 1500 366 \n",
       "Q 1825 366 2078 508 \n",
       "Q 2331 650 2450 897 \n",
       "Q 2541 1088 2541 1459 \n",
       "L 2541 1666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-6e\" d=\"M 422 0 \n",
       "L 422 3319 \n",
       "L 928 3319 \n",
       "L 928 2847 \n",
       "Q 1294 3394 1984 3394 \n",
       "Q 2284 3394 2536 3286 \n",
       "Q 2788 3178 2913 3003 \n",
       "Q 3038 2828 3088 2588 \n",
       "Q 3119 2431 3119 2041 \n",
       "L 3119 0 \n",
       "L 2556 0 \n",
       "L 2556 2019 \n",
       "Q 2556 2363 2490 2533 \n",
       "Q 2425 2703 2258 2804 \n",
       "Q 2091 2906 1866 2906 \n",
       "Q 1506 2906 1245 2678 \n",
       "Q 984 2450 984 1813 \n",
       "L 984 0 \n",
       "L 422 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "     </defs>\n",
       "     <use xlink:href=\"#ArialMT-48\"/>\n",
       "     <use xlink:href=\"#ArialMT-75\" x=\"72.216797\"/>\n",
       "     <use xlink:href=\"#ArialMT-6d\" x=\"127.832031\"/>\n",
       "     <use xlink:href=\"#ArialMT-61\" x=\"211.132812\"/>\n",
       "     <use xlink:href=\"#ArialMT-6e\" x=\"266.748047\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_6\">\n",
       "    <!-- Species-specific and -shared binding counts -->\n",
       "    <g transform=\"translate(55.671119 18.8475) scale(0.16 -0.16)\">\n",
       "     <defs>\n",
       "      <path id=\"ArialMT-53\" d=\"M 288 1472 \n",
       "L 859 1522 \n",
       "Q 900 1178 1048 958 \n",
       "Q 1197 738 1509 602 \n",
       "Q 1822 466 2213 466 \n",
       "Q 2559 466 2825 569 \n",
       "Q 3091 672 3220 851 \n",
       "Q 3350 1031 3350 1244 \n",
       "Q 3350 1459 3225 1620 \n",
       "Q 3100 1781 2813 1891 \n",
       "Q 2628 1963 1997 2114 \n",
       "Q 1366 2266 1113 2400 \n",
       "Q 784 2572 623 2826 \n",
       "Q 463 3081 463 3397 \n",
       "Q 463 3744 659 4045 \n",
       "Q 856 4347 1234 4503 \n",
       "Q 1613 4659 2075 4659 \n",
       "Q 2584 4659 2973 4495 \n",
       "Q 3363 4331 3572 4012 \n",
       "Q 3781 3694 3797 3291 \n",
       "L 3216 3247 \n",
       "Q 3169 3681 2898 3903 \n",
       "Q 2628 4125 2100 4125 \n",
       "Q 1550 4125 1298 3923 \n",
       "Q 1047 3722 1047 3438 \n",
       "Q 1047 3191 1225 3031 \n",
       "Q 1400 2872 2139 2705 \n",
       "Q 2878 2538 3153 2413 \n",
       "Q 3553 2228 3743 1945 \n",
       "Q 3934 1663 3934 1294 \n",
       "Q 3934 928 3725 604 \n",
       "Q 3516 281 3123 101 \n",
       "Q 2731 -78 2241 -78 \n",
       "Q 1619 -78 1198 103 \n",
       "Q 778 284 539 648 \n",
       "Q 300 1013 288 1472 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-70\" d=\"M 422 -1272 \n",
       "L 422 3319 \n",
       "L 934 3319 \n",
       "L 934 2888 \n",
       "Q 1116 3141 1344 3267 \n",
       "Q 1572 3394 1897 3394 \n",
       "Q 2322 3394 2647 3175 \n",
       "Q 2972 2956 3137 2557 \n",
       "Q 3303 2159 3303 1684 \n",
       "Q 3303 1175 3120 767 \n",
       "Q 2938 359 2589 142 \n",
       "Q 2241 -75 1856 -75 \n",
       "Q 1575 -75 1351 44 \n",
       "Q 1128 163 984 344 \n",
       "L 984 -1272 \n",
       "L 422 -1272 \n",
       "z\n",
       "M 931 1641 \n",
       "Q 931 1000 1190 694 \n",
       "Q 1450 388 1819 388 \n",
       "Q 2194 388 2461 705 \n",
       "Q 2728 1022 2728 1688 \n",
       "Q 2728 2322 2467 2637 \n",
       "Q 2206 2953 1844 2953 \n",
       "Q 1484 2953 1207 2617 \n",
       "Q 931 2281 931 1641 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-63\" d=\"M 2588 1216 \n",
       "L 3141 1144 \n",
       "Q 3050 572 2676 248 \n",
       "Q 2303 -75 1759 -75 \n",
       "Q 1078 -75 664 370 \n",
       "Q 250 816 250 1647 \n",
       "Q 250 2184 428 2587 \n",
       "Q 606 2991 970 3192 \n",
       "Q 1334 3394 1763 3394 \n",
       "Q 2303 3394 2647 3120 \n",
       "Q 2991 2847 3088 2344 \n",
       "L 2541 2259 \n",
       "Q 2463 2594 2264 2762 \n",
       "Q 2066 2931 1784 2931 \n",
       "Q 1359 2931 1093 2626 \n",
       "Q 828 2322 828 1663 \n",
       "Q 828 994 1084 691 \n",
       "Q 1341 388 1753 388 \n",
       "Q 2084 388 2306 591 \n",
       "Q 2528 794 2588 1216 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-69\" d=\"M 425 3934 \n",
       "L 425 4581 \n",
       "L 988 4581 \n",
       "L 988 3934 \n",
       "L 425 3934 \n",
       "z\n",
       "M 425 0 \n",
       "L 425 3319 \n",
       "L 988 3319 \n",
       "L 988 0 \n",
       "L 425 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-2d\" d=\"M 203 1375 \n",
       "L 203 1941 \n",
       "L 1931 1941 \n",
       "L 1931 1375 \n",
       "L 203 1375 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-66\" d=\"M 556 0 \n",
       "L 556 2881 \n",
       "L 59 2881 \n",
       "L 59 3319 \n",
       "L 556 3319 \n",
       "L 556 3672 \n",
       "Q 556 4006 616 4169 \n",
       "Q 697 4388 901 4523 \n",
       "Q 1106 4659 1475 4659 \n",
       "Q 1713 4659 2000 4603 \n",
       "L 1916 4113 \n",
       "Q 1741 4144 1584 4144 \n",
       "Q 1328 4144 1222 4034 \n",
       "Q 1116 3925 1116 3625 \n",
       "L 1116 3319 \n",
       "L 1763 3319 \n",
       "L 1763 2881 \n",
       "L 1116 2881 \n",
       "L 1116 0 \n",
       "L 556 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-64\" d=\"M 2575 0 \n",
       "L 2575 419 \n",
       "Q 2259 -75 1647 -75 \n",
       "Q 1250 -75 917 144 \n",
       "Q 584 363 401 755 \n",
       "Q 219 1147 219 1656 \n",
       "Q 219 2153 384 2558 \n",
       "Q 550 2963 881 3178 \n",
       "Q 1213 3394 1622 3394 \n",
       "Q 1922 3394 2156 3267 \n",
       "Q 2391 3141 2538 2938 \n",
       "L 2538 4581 \n",
       "L 3097 4581 \n",
       "L 3097 0 \n",
       "L 2575 0 \n",
       "z\n",
       "M 797 1656 \n",
       "Q 797 1019 1065 703 \n",
       "Q 1334 388 1700 388 \n",
       "Q 2069 388 2326 689 \n",
       "Q 2584 991 2584 1609 \n",
       "Q 2584 2291 2321 2609 \n",
       "Q 2059 2928 1675 2928 \n",
       "Q 1300 2928 1048 2622 \n",
       "Q 797 2316 797 1656 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-68\" d=\"M 422 0 \n",
       "L 422 4581 \n",
       "L 984 4581 \n",
       "L 984 2938 \n",
       "Q 1378 3394 1978 3394 \n",
       "Q 2347 3394 2619 3248 \n",
       "Q 2891 3103 3008 2847 \n",
       "Q 3125 2591 3125 2103 \n",
       "L 3125 0 \n",
       "L 2563 0 \n",
       "L 2563 2103 \n",
       "Q 2563 2525 2380 2717 \n",
       "Q 2197 2909 1863 2909 \n",
       "Q 1613 2909 1392 2779 \n",
       "Q 1172 2650 1078 2428 \n",
       "Q 984 2206 984 1816 \n",
       "L 984 0 \n",
       "L 422 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-72\" d=\"M 416 0 \n",
       "L 416 3319 \n",
       "L 922 3319 \n",
       "L 922 2816 \n",
       "Q 1116 3169 1280 3281 \n",
       "Q 1444 3394 1641 3394 \n",
       "Q 1925 3394 2219 3213 \n",
       "L 2025 2691 \n",
       "Q 1819 2813 1613 2813 \n",
       "Q 1428 2813 1281 2702 \n",
       "Q 1134 2591 1072 2394 \n",
       "Q 978 2094 978 1738 \n",
       "L 978 0 \n",
       "L 416 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-62\" d=\"M 941 0 \n",
       "L 419 0 \n",
       "L 419 4581 \n",
       "L 981 4581 \n",
       "L 981 2947 \n",
       "Q 1338 3394 1891 3394 \n",
       "Q 2197 3394 2470 3270 \n",
       "Q 2744 3147 2920 2923 \n",
       "Q 3097 2700 3197 2384 \n",
       "Q 3297 2069 3297 1709 \n",
       "Q 3297 856 2875 390 \n",
       "Q 2453 -75 1863 -75 \n",
       "Q 1275 -75 941 416 \n",
       "L 941 0 \n",
       "z\n",
       "M 934 1684 \n",
       "Q 934 1088 1097 822 \n",
       "Q 1363 388 1816 388 \n",
       "Q 2184 388 2453 708 \n",
       "Q 2722 1028 2722 1663 \n",
       "Q 2722 2313 2464 2622 \n",
       "Q 2206 2931 1841 2931 \n",
       "Q 1472 2931 1203 2611 \n",
       "Q 934 2291 934 1684 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-67\" d=\"M 319 -275 \n",
       "L 866 -356 \n",
       "Q 900 -609 1056 -725 \n",
       "Q 1266 -881 1628 -881 \n",
       "Q 2019 -881 2231 -725 \n",
       "Q 2444 -569 2519 -288 \n",
       "Q 2563 -116 2559 434 \n",
       "Q 2191 0 1641 0 \n",
       "Q 956 0 581 494 \n",
       "Q 206 988 206 1678 \n",
       "Q 206 2153 378 2554 \n",
       "Q 550 2956 876 3175 \n",
       "Q 1203 3394 1644 3394 \n",
       "Q 2231 3394 2613 2919 \n",
       "L 2613 3319 \n",
       "L 3131 3319 \n",
       "L 3131 450 \n",
       "Q 3131 -325 2973 -648 \n",
       "Q 2816 -972 2473 -1159 \n",
       "Q 2131 -1347 1631 -1347 \n",
       "Q 1038 -1347 672 -1080 \n",
       "Q 306 -813 319 -275 \n",
       "z\n",
       "M 784 1719 \n",
       "Q 784 1066 1043 766 \n",
       "Q 1303 466 1694 466 \n",
       "Q 2081 466 2343 764 \n",
       "Q 2606 1063 2606 1700 \n",
       "Q 2606 2309 2336 2618 \n",
       "Q 2066 2928 1684 2928 \n",
       "Q 1309 2928 1046 2623 \n",
       "Q 784 2319 784 1719 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-74\" d=\"M 1650 503 \n",
       "L 1731 6 \n",
       "Q 1494 -44 1306 -44 \n",
       "Q 1000 -44 831 53 \n",
       "Q 663 150 594 308 \n",
       "Q 525 466 525 972 \n",
       "L 525 2881 \n",
       "L 113 2881 \n",
       "L 113 3319 \n",
       "L 525 3319 \n",
       "L 525 4141 \n",
       "L 1084 4478 \n",
       "L 1084 3319 \n",
       "L 1650 3319 \n",
       "L 1650 2881 \n",
       "L 1084 2881 \n",
       "L 1084 941 \n",
       "Q 1084 700 1114 631 \n",
       "Q 1144 563 1211 522 \n",
       "Q 1278 481 1403 481 \n",
       "Q 1497 481 1650 503 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "     </defs>\n",
       "     <use xlink:href=\"#ArialMT-53\"/>\n",
       "     <use xlink:href=\"#ArialMT-70\" x=\"66.699219\"/>\n",
       "     <use xlink:href=\"#ArialMT-65\" x=\"122.314453\"/>\n",
       "     <use xlink:href=\"#ArialMT-63\" x=\"177.929688\"/>\n",
       "     <use xlink:href=\"#ArialMT-69\" x=\"227.929688\"/>\n",
       "     <use xlink:href=\"#ArialMT-65\" x=\"250.146484\"/>\n",
       "     <use xlink:href=\"#ArialMT-73\" x=\"305.761719\"/>\n",
       "     <use xlink:href=\"#ArialMT-2d\" x=\"355.761719\"/>\n",
       "     <use xlink:href=\"#ArialMT-73\" x=\"389.0625\"/>\n",
       "     <use xlink:href=\"#ArialMT-70\" x=\"439.0625\"/>\n",
       "     <use xlink:href=\"#ArialMT-65\" x=\"494.677734\"/>\n",
       "     <use xlink:href=\"#ArialMT-63\" x=\"550.292969\"/>\n",
       "     <use xlink:href=\"#ArialMT-69\" x=\"600.292969\"/>\n",
       "     <use xlink:href=\"#ArialMT-66\" x=\"622.509766\"/>\n",
       "     <use xlink:href=\"#ArialMT-69\" x=\"650.292969\"/>\n",
       "     <use xlink:href=\"#ArialMT-63\" x=\"672.509766\"/>\n",
       "     <use xlink:href=\"#ArialMT-20\" x=\"722.509766\"/>\n",
       "     <use xlink:href=\"#ArialMT-61\" x=\"750.292969\"/>\n",
       "     <use xlink:href=\"#ArialMT-6e\" x=\"805.908203\"/>\n",
       "     <use xlink:href=\"#ArialMT-64\" x=\"861.523438\"/>\n",
       "     <use xlink:href=\"#ArialMT-20\" x=\"917.138672\"/>\n",
       "     <use xlink:href=\"#ArialMT-2d\" x=\"944.921875\"/>\n",
       "     <use xlink:href=\"#ArialMT-73\" x=\"978.222656\"/>\n",
       "     <use xlink:href=\"#ArialMT-68\" x=\"1028.222656\"/>\n",
       "     <use xlink:href=\"#ArialMT-61\" x=\"1083.837891\"/>\n",
       "     <use xlink:href=\"#ArialMT-72\" x=\"1139.453125\"/>\n",
       "     <use xlink:href=\"#ArialMT-65\" x=\"1172.753906\"/>\n",
       "     <use xlink:href=\"#ArialMT-64\" x=\"1228.369141\"/>\n",
       "     <use xlink:href=\"#ArialMT-20\" x=\"1283.984375\"/>\n",
       "     <use xlink:href=\"#ArialMT-62\" x=\"1311.767578\"/>\n",
       "     <use xlink:href=\"#ArialMT-69\" x=\"1367.382812\"/>\n",
       "     <use xlink:href=\"#ArialMT-6e\" x=\"1389.599609\"/>\n",
       "     <use xlink:href=\"#ArialMT-64\" x=\"1445.214844\"/>\n",
       "     <use xlink:href=\"#ArialMT-69\" x=\"1500.830078\"/>\n",
       "     <use xlink:href=\"#ArialMT-6e\" x=\"1523.046875\"/>\n",
       "     <use xlink:href=\"#ArialMT-67\" x=\"1578.662109\"/>\n",
       "     <use xlink:href=\"#ArialMT-20\" x=\"1634.277344\"/>\n",
       "     <use xlink:href=\"#ArialMT-63\" x=\"1662.060547\"/>\n",
       "     <use xlink:href=\"#ArialMT-6f\" x=\"1712.060547\"/>\n",
       "     <use xlink:href=\"#ArialMT-75\" x=\"1767.675781\"/>\n",
       "     <use xlink:href=\"#ArialMT-6e\" x=\"1823.291016\"/>\n",
       "     <use xlink:href=\"#ArialMT-74\" x=\"1878.90625\"/>\n",
       "     <use xlink:href=\"#ArialMT-73\" x=\"1906.689453\"/>\n",
       "    </g>\n",
       "    <!-- at gene level, any biotype -->\n",
       "    <g transform=\"translate(121.472369 35.958) scale(0.16 -0.16)\">\n",
       "     <defs>\n",
       "      <path id=\"ArialMT-6c\" d=\"M 409 0 \n",
       "L 409 4581 \n",
       "L 972 4581 \n",
       "L 972 0 \n",
       "L 409 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-76\" d=\"M 1344 0 \n",
       "L 81 3319 \n",
       "L 675 3319 \n",
       "L 1388 1331 \n",
       "Q 1503 1009 1600 663 \n",
       "Q 1675 925 1809 1294 \n",
       "L 2547 3319 \n",
       "L 3125 3319 \n",
       "L 1869 0 \n",
       "L 1344 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-79\" d=\"M 397 -1278 \n",
       "L 334 -750 \n",
       "Q 519 -800 656 -800 \n",
       "Q 844 -800 956 -737 \n",
       "Q 1069 -675 1141 -563 \n",
       "Q 1194 -478 1313 -144 \n",
       "Q 1328 -97 1363 -6 \n",
       "L 103 3319 \n",
       "L 709 3319 \n",
       "L 1400 1397 \n",
       "Q 1534 1031 1641 628 \n",
       "Q 1738 1016 1872 1384 \n",
       "L 2581 3319 \n",
       "L 3144 3319 \n",
       "L 1881 -56 \n",
       "Q 1678 -603 1566 -809 \n",
       "Q 1416 -1088 1222 -1217 \n",
       "Q 1028 -1347 759 -1347 \n",
       "Q 597 -1347 397 -1278 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "     </defs>\n",
       "     <use xlink:href=\"#ArialMT-61\"/>\n",
       "     <use xlink:href=\"#ArialMT-74\" x=\"55.615234\"/>\n",
       "     <use xlink:href=\"#ArialMT-20\" x=\"83.398438\"/>\n",
       "     <use xlink:href=\"#ArialMT-67\" x=\"111.181641\"/>\n",
       "     <use xlink:href=\"#ArialMT-65\" x=\"166.796875\"/>\n",
       "     <use xlink:href=\"#ArialMT-6e\" x=\"222.412109\"/>\n",
       "     <use xlink:href=\"#ArialMT-65\" x=\"278.027344\"/>\n",
       "     <use xlink:href=\"#ArialMT-20\" x=\"333.642578\"/>\n",
       "     <use xlink:href=\"#ArialMT-6c\" x=\"361.425781\"/>\n",
       "     <use xlink:href=\"#ArialMT-65\" x=\"383.642578\"/>\n",
       "     <use xlink:href=\"#ArialMT-76\" x=\"439.257812\"/>\n",
       "     <use xlink:href=\"#ArialMT-65\" x=\"489.257812\"/>\n",
       "     <use xlink:href=\"#ArialMT-6c\" x=\"544.873047\"/>\n",
       "     <use xlink:href=\"#ArialMT-2c\" x=\"567.089844\"/>\n",
       "     <use xlink:href=\"#ArialMT-20\" x=\"594.873047\"/>\n",
       "     <use xlink:href=\"#ArialMT-61\" x=\"622.65625\"/>\n",
       "     <use xlink:href=\"#ArialMT-6e\" x=\"678.271484\"/>\n",
       "     <use xlink:href=\"#ArialMT-79\" x=\"733.886719\"/>\n",
       "     <use xlink:href=\"#ArialMT-20\" x=\"783.886719\"/>\n",
       "     <use xlink:href=\"#ArialMT-62\" x=\"811.669922\"/>\n",
       "     <use xlink:href=\"#ArialMT-69\" x=\"867.285156\"/>\n",
       "     <use xlink:href=\"#ArialMT-6f\" x=\"889.501953\"/>\n",
       "     <use xlink:href=\"#ArialMT-74\" x=\"945.117188\"/>\n",
       "     <use xlink:href=\"#ArialMT-79\" x=\"972.900391\"/>\n",
       "     <use xlink:href=\"#ArialMT-70\" x=\"1022.900391\"/>\n",
       "     <use xlink:href=\"#ArialMT-65\" x=\"1078.515625\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "  </g>\n",
       " </g>\n",
       " <defs>\n",
       "  <clipPath id=\"p707d8eeb67\">\n",
       "   <rect x=\"7.2\" y=\"41.958\" width=\"409.992237\" height=\"332.64\"/>\n",
       "  </clipPath>\n",
       " </defs>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<Figure size 1000x600 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "v = venn.venn2(\n",
    "    subsets=(\n",
    "        len(m_genes_no_bound_orth_complete),\n",
    "        len(h_genes_no_bound_orth_complete),\n",
    "        len(m_genes_bound_orth_complete),\n",
    "    ),\n",
    "    set_labels=(\"Mouse\", \"Human\"),\n",
    "    set_colors=(\n",
    "        \"#BBBBBB\",\n",
    "        \"#BBBBBB\",\n",
    "    ),\n",
    ")\n",
    "\n",
    "# Update the labels to include the info on mouse | human orthologous bound genes.\n",
    "v.subset_labels[0].set_text(f\"{len(m_genes_no_bound_orth_complete):,}\")\n",
    "v.subset_labels[1].set_text(f\"{len(h_genes_no_bound_orth_complete):,}\")\n",
    "v.subset_labels[2].set_text(f\"{len(m_genes_bound_orth_complete):,} | {len(h_genes_bound_orth_complete):,}\")\n",
    "\n",
    "circles = venn.venn2_circles(\n",
    "    subsets=(\n",
    "        len(m_genes_no_bound_orth_complete),\n",
    "        len(h_genes_no_bound_orth_complete),\n",
    "        len(m_genes_bound_orth_complete),\n",
    "    ),\n",
    "    linestyle=\"-\",\n",
    ")\n",
    "\n",
    "plt.title(\"Species-specific and -shared binding counts\\nat gene level, any biotype\")\n",
    "\n",
    "if params_export_flag:\n",
    "    plt.savefig(outputdir / \"venn_diagram_genes_w_peaks.any_biotype.svg\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "507fcfc3",
   "metadata": {},
   "source": [
    "## Orthologies - 3'UTR biotype only"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7721b2f4",
   "metadata": {},
   "source": [
    "### Gather tables"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e056bcd2",
   "metadata": {},
   "source": [
    "\n",
    "We can build tables from the previously established tables. \n",
    "\n",
    "We only have to restrict those to the 3'UTR peaks in each species.\n",
    "\n",
    "Then we re-establish the lists of genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4397e04e",
   "metadata": {},
   "outputs": [],
   "source": [
    "m_full_3utr = m_full.loc[lambda df: df[\"name\"].str.startswith(\"three_prime_UTR\"), :].copy()\n",
    "m_id_to_bed6_3utr = m_id_to_bed6.loc[lambda df: df[\"name\"].str.startswith(\"three_prime_UTR\"), :].copy()\n",
    "m_id_to_meta_3utr = m_id_to_meta.loc[lambda df: df[\"name\"].str.startswith(\"three_prime_UTR\"), :].copy()\n",
    "m_umap_3utr = m_umap.loc[lambda df: df[\"identifier\"].isin(m_id_to_bed6_3utr.index.values), :].copy()\n",
    "m_lift_3utr = m_lift.loc[lambda df: df[\"identifier.mouse_source\"].isin(m_id_to_bed6_3utr.index.values), :].copy()\n",
    "\n",
    "\n",
    "h_full_3utr = h_full.loc[lambda df: df[\"name\"].str.startswith(\"three_prime_UTR\"), :].copy()\n",
    "h_id_to_bed6_3utr = h_id_to_bed6.loc[lambda df: df[\"name\"].str.startswith(\"three_prime_UTR\"), :].copy()\n",
    "h_id_to_meta_3utr = h_id_to_meta.loc[lambda df: df[\"name\"].str.startswith(\"three_prime_UTR\"), :].copy()\n",
    "h_umap_3utr = h_umap.loc[lambda df: df[\"identifier\"].isin(h_id_to_bed6_3utr.index.values), :].copy()\n",
    "h_lift_3utr = h_lift.loc[lambda df: df[\"identifier.human_source\"].isin(h_id_to_bed6_3utr.index.values), :].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf417c07",
   "metadata": {},
   "outputs": [],
   "source": [
    "h_bound_genes_3utr = set(h_id_to_meta_3utr.loc[h_umap_3utr[\"identifier\"].values, \"gene_id\"].unique()) | set(\n",
    "    h_id_to_meta_3utr.loc[h_lift_3utr[\"identifier.human_source\"].values, \"gene_id\"].unique()\n",
    ")\n",
    "m_bound_genes_3utr = set(m_id_to_meta_3utr.loc[m_umap_3utr[\"identifier\"].values, \"gene_id\"].unique()) | set(\n",
    "    m_id_to_meta_3utr.loc[m_lift_3utr[\"identifier.mouse_source\"].values, \"gene_id\"].unique()\n",
    ")\n",
    "\n",
    "m_lift_3utr[\"annotated_lifted_region_bound_gene_set\"] = m_lift_3utr.apply(\n",
    "    lambda row: (h_id_to_meta_3utr[\"gene_id_set\"].get(row[\"identifier.human_target\"], set({})) & h_bound_genes_3utr),\n",
    "    axis=1,\n",
    ")\n",
    "\n",
    "h_lift_3utr[\"annotated_lifted_region_bound_gene_set\"] = h_lift_3utr.apply(\n",
    "    lambda row: (m_id_to_meta_3utr[\"gene_id_set\"].get(row[\"identifier.mouse_target\"], set({})) & m_bound_genes_3utr),\n",
    "    axis=1,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20cf362c",
   "metadata": {},
   "source": [
    "### Gene counts - Mouse"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0658bd7d",
   "metadata": {},
   "source": [
    "#### First attempt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e81b4cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "m_has_gene_bound_orth_3utr = (\n",
    "    (m_lift_3utr[\"gene_orthology_status\"] != \"not_orthologous\")\n",
    "    & (m_lift_3utr[\"annotated_lifted_region_bound_gene_set\"].apply(len) > 0)\n",
    ").values\n",
    "\n",
    "###\n",
    "\n",
    "m_peaks_w_bound_orth_3utr = m_lift_3utr.loc[m_has_gene_bound_orth_3utr, \"identifier.mouse_source\"].values\n",
    "m_peaks_wo_bound_orth_3utr = m_lift_3utr.loc[~m_has_gene_bound_orth_3utr, \"identifier.mouse_source\"].values\n",
    "\n",
    "###\n",
    "\n",
    "m_genes_bound_orth_3utr = set(m_id_to_meta_3utr.loc[m_peaks_w_bound_orth_3utr, \"gene_id\"].unique())\n",
    "m_genes_lifted_but_wo_annotated_bound_orth_3utr = set(\n",
    "    m_id_to_meta_3utr.loc[m_peaks_wo_bound_orth_3utr, \"gene_id\"].unique()\n",
    ")\n",
    "\n",
    "###\n",
    "\n",
    "# Here: aggregate the list of mouse ENS IDs identified as orthologs for the human peaks' annotated genes.\n",
    "# i.e. : we know these genes mouse ENS-ID have a human ortholog that has a peak.\n",
    "m_genes_from_human_peaks_w_orth_3utr = functools.reduce(\n",
    "    set.union, h_lift_3utr[\"annotated_lifted_region_bound_gene_set\"].values\n",
    ")\n",
    "\n",
    "###\n",
    "\n",
    "m_genes_from_umap_3utr = set(m_id_to_meta_3utr.loc[m_umap_3utr[\"identifier\"].values, \"gene_id\"].unique())\n",
    "\n",
    "###\n",
    "\n",
    "# Finally : build the list of genes without bound orthologous gene (i.e. mouse-specific bound genes)\n",
    "# as the set of genes from unmappable peaks or peaks mapped but for which the lifted region is not an orthologous-bound gene.\n",
    "\n",
    "m_genes_no_bound_orth_3utr = (\n",
    "    (m_genes_from_umap_3utr | m_genes_lifted_but_wo_annotated_bound_orth_3utr)\n",
    "    - m_genes_from_human_peaks_w_orth_3utr\n",
    "    - m_genes_bound_orth_3utr\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00c52621",
   "metadata": {},
   "source": [
    "#### Missing genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "298d809e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False    68\n",
       "True     17\n",
       "Name: is_in_umap, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    }
   ],
   "source": [
    "m_missing_3utr = set(m_full_3utr[\"gene_id\"].unique()) - m_genes_bound_orth_3utr - m_genes_no_bound_orth_3utr\n",
    "print(len(m_missing))\n",
    "\n",
    "\n",
    "m_peaks_in_missing_genes_3utr = m_full_3utr.loc[lambda df: df[\"gene_id\"].isin(m_missing_3utr), :].assign(\n",
    "    is_in_umap=lambda df: df[\"identifier\"].isin(m_umap_3utr[\"identifier\"].values)\n",
    ")\n",
    "\n",
    "\n",
    "display(m_peaks_in_missing_genes_3utr[\"is_in_umap\"].value_counts())\n",
    "\n",
    "# EXPECTED:\n",
    "# True     68\n",
    "# False    17\n",
    "\n",
    "# So most peaks are not mapped,\n",
    "\n",
    "# ALL THESE GENES ARE FOUND TO BE BOUND IN HUMAN PEAKS LIFTED TO MOUSE.\n",
    "# i.e. I discard them wrongly in the last set operation, as they are bound in mouse but from unmapped peaks.\n",
    "\n",
    "print(all([g in m_genes_from_human_peaks_w_orth_3utr for g in m_missing_3utr]))\n",
    "\n",
    "# A subset of genes apparently are mappable, so I assume there it is a problem of gene annotation.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37a3cc5b",
   "metadata": {},
   "source": [
    "#### Final numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "91650976",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "len(m_genes_bound_orth_complete_3utr)=1015\n",
      "len(m_genes_no_bound_orth_complete_3utr)=703\n",
      "(len(m_genes_bound_orth_complete_3utr) + len(m_genes_no_bound_orth_complete_3utr))=1718\n",
      "len(m_bound_genes_3utr)=1718\n",
      "(len(m_genes_bound_orth_complete_3utr) + len(m_genes_no_bound_orth_complete_3utr)) == len(m_bound_genes_3utr)=True\n"
     ]
    }
   ],
   "source": [
    "m_genes_bound_orth_complete_3utr = m_genes_bound_orth_3utr | (\n",
    "    set(m_peaks_in_missing_genes_3utr.loc[lambda df: df[\"is_in_umap\"], \"gene_id\"].unique()) & m_bound_genes_3utr\n",
    ")\n",
    "\n",
    "m_genes_no_bound_orth_complete_3utr = m_genes_no_bound_orth_3utr | (\n",
    "    set(m_peaks_in_missing_genes_3utr.loc[lambda df: (~df[\"is_in_umap\"]), \"gene_id\"].unique()) & m_bound_genes_3utr\n",
    ")\n",
    "# Final correction on cases of n:n mappings : \"bound\" takes over \"no-bound\"\n",
    "m_genes_no_bound_orth_complete_3utr = m_genes_no_bound_orth_complete_3utr - m_genes_bound_orth_complete_3utr\n",
    "\n",
    "# FINAL TABLES\n",
    "\n",
    "print(f\"{len(m_genes_bound_orth_complete_3utr)=}\")\n",
    "print(f\"{len(m_genes_no_bound_orth_complete_3utr)=}\")\n",
    "print(f\"{(len(m_genes_bound_orth_complete_3utr) + len(m_genes_no_bound_orth_complete_3utr))=}\")\n",
    "print(f\"{len(m_bound_genes_3utr)=}\")\n",
    "print(\n",
    "    f\"{(len(m_genes_bound_orth_complete_3utr) + len(m_genes_no_bound_orth_complete_3utr)) == len(m_bound_genes_3utr)=}\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b46320b",
   "metadata": {},
   "source": [
    "### Gene counts - Human"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e10a7cc",
   "metadata": {},
   "source": [
    "#### First attempt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df7b8092",
   "metadata": {},
   "outputs": [],
   "source": [
    "h_has_gene_bound_orth_3utr = (\n",
    "    (h_lift_3utr[\"gene_orthology_status\"] != \"not_orthologous\")\n",
    "    & (h_lift_3utr[\"annotated_lifted_region_bound_gene_set\"].apply(len) > 0)\n",
    ").values\n",
    "\n",
    "###\n",
    "\n",
    "h_peaks_w_bound_orth_3utr = h_lift_3utr.loc[h_has_gene_bound_orth_3utr, \"identifier.human_source\"].values\n",
    "h_peaks_wo_bound_orth_3utr = h_lift_3utr.loc[~h_has_gene_bound_orth_3utr, \"identifier.human_source\"].values\n",
    "\n",
    "###\n",
    "\n",
    "h_genes_bound_orth_3utr = set(h_id_to_meta_3utr.loc[h_peaks_w_bound_orth_3utr, \"gene_id\"].unique())\n",
    "h_genes_lifted_but_wo_annotated_bound_orth_3utr = set(\n",
    "    h_id_to_meta_3utr.loc[h_peaks_wo_bound_orth_3utr, \"gene_id\"].unique()\n",
    ")\n",
    "\n",
    "###\n",
    "\n",
    "# Here: aggregate the list of HUMAN ENS IDs identified as orthologs for the MOUSE peaks' annotated genes.\n",
    "# i.e. : we know these genes HUMAN ENS-ID have a MOUSE ortholog that has a peak.\n",
    "h_genes_from_mouse_peaks_w_orth_3utr = functools.reduce(\n",
    "    set.union, h_lift_3utr[\"annotated_lifted_region_bound_gene_set\"].values\n",
    ")\n",
    "\n",
    "###\n",
    "\n",
    "h_genes_from_umap_3utr = set(h_id_to_meta_3utr.loc[h_umap_3utr[\"identifier\"].values, \"gene_id\"].unique())\n",
    "\n",
    "###\n",
    "\n",
    "# Finally : build the list of genes without bound orthologous gene (i.e. HUMAN-specific bound genes)\n",
    "# as the set of genes from unmappable peaks or peaks mapped but for which the lifted region is not an orthologous-bound gene.\n",
    "\n",
    "h_genes_no_bound_orth_3utr = (\n",
    "    (h_genes_from_umap_3utr | h_genes_lifted_but_wo_annotated_bound_orth_3utr)\n",
    "    - h_genes_from_mouse_peaks_w_orth_3utr\n",
    "    - h_genes_bound_orth_3utr\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90b2f0fa",
   "metadata": {},
   "source": [
    "#### Missing genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95ccd57a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of missing genes: 0\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Series([], Name: is_in_umap, dtype: int64)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "h_missing_3utr = set(h_full_3utr[\"gene_id\"].unique()) - h_genes_bound_orth_3utr - h_genes_no_bound_orth_3utr\n",
    "print(f\"Number of missing genes: {len(h_missing_3utr):,}\")\n",
    "\n",
    "\n",
    "h_peaks_in_missing_genes_3utr = h_full_3utr.loc[lambda df: df[\"gene_id\"].isin(h_missing_3utr), :].assign(\n",
    "    is_in_umap=lambda df: df[\"identifier\"].isin(h_umap_3utr[\"identifier\"].values)\n",
    ")\n",
    "\n",
    "\n",
    "display(h_peaks_in_missing_genes_3utr[\"is_in_umap\"].value_counts())\n",
    "\n",
    "# EXPECTED:\n",
    "# None\n",
    "\n",
    "# all genes should be covered"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1f7e7a7",
   "metadata": {},
   "source": [
    "#### Final numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87bc1de1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "len(h_genes_bound_orth_complete_3utr)=1000\n",
      "len(h_genes_no_bound_orth_complete_3utr)=1480\n",
      "(len(h_genes_bound_orth_complete_3utr) + len(h_genes_no_bound_orth_complete_3utr))=2480\n",
      "2480\n",
      "(len(h_genes_bound_orth_complete_3utr) + len(h_genes_no_bound_orth_complete_3utr)) == len(h_bound_genes_3utr)=True\n"
     ]
    }
   ],
   "source": [
    "# Here in human : no missing genes, so we directly have access to the complete sets.\n",
    "\n",
    "h_genes_bound_orth_complete_3utr = h_genes_bound_orth_3utr\n",
    "\n",
    "\n",
    "h_genes_no_bound_orth_complete_3utr = h_genes_no_bound_orth_3utr\n",
    "\n",
    "\n",
    "# FINAL TABLES\n",
    "\n",
    "print(f\"{len(h_genes_bound_orth_complete_3utr)=}\")\n",
    "(print(f\"{len(h_genes_no_bound_orth_complete_3utr)=}\"),)\n",
    "print(f\"{(len(h_genes_bound_orth_complete_3utr) + len(h_genes_no_bound_orth_complete_3utr))=}\")\n",
    "print(f\"{len(h_bound_genes_3utr)}\")\n",
    "print(\n",
    "    f\"{(len(h_genes_bound_orth_complete_3utr) + len(h_genes_no_bound_orth_complete_3utr)) == len(h_bound_genes_3utr)=}\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "661a9a4c",
   "metadata": {},
   "source": [
    "### Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "c955b067",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"utf-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       "  \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<svg xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"427.612679pt\" height=\"384.290341pt\" viewBox=\"0 0 427.612679 384.290341\" xmlns=\"http://www.w3.org/2000/svg\" version=\"1.1\">\n",
       " <metadata>\n",
       "  <rdf:RDF xmlns:dc=\"http://purl.org/dc/elements/1.1/\" xmlns:cc=\"http://creativecommons.org/ns#\" xmlns:rdf=\"http://www.w3.org/1999/02/22-rdf-syntax-ns#\">\n",
       "   <cc:Work>\n",
       "    <dc:type rdf:resource=\"http://purl.org/dc/dcmitype/StillImage\"/>\n",
       "    <dc:date>2025-07-14T05:55:01.957139</dc:date>\n",
       "    <dc:format>image/svg+xml</dc:format>\n",
       "    <dc:creator>\n",
       "     <cc:Agent>\n",
       "      <dc:title>Matplotlib v3.8.2, https://matplotlib.org/</dc:title>\n",
       "     </cc:Agent>\n",
       "    </dc:creator>\n",
       "   </cc:Work>\n",
       "  </rdf:RDF>\n",
       " </metadata>\n",
       " <defs>\n",
       "  <style type=\"text/css\">*{stroke-linejoin: round; stroke-linecap: butt}</style>\n",
       " </defs>\n",
       " <g id=\"figure_1\">\n",
       "  <g id=\"patch_1\">\n",
       "   <path d=\"M 0 384.290341 \n",
       "L 427.612679 384.290341 \n",
       "L 427.612679 0 \n",
       "L 0 0 \n",
       "z\n",
       "\" style=\"fill: #ffffff\"/>\n",
       "  </g>\n",
       "  <g id=\"axes_1\">\n",
       "   <g id=\"patch_2\">\n",
       "    <path d=\"M 173.329767 320.87802 \n",
       "C 155.458621 308.053362 140.896577 291.154512 130.852896 271.584786 \n",
       "C 120.809216 252.01506 115.569831 230.331584 115.569831 208.335 \n",
       "C 115.569831 186.338416 120.809216 164.65494 130.852896 145.085214 \n",
       "C 140.896577 125.515488 155.458621 108.616638 173.329767 95.79198 \n",
       "C 156.608396 92.317405 139.320943 92.612741 122.728013 96.656452 \n",
       "C 106.135083 100.700163 90.650045 108.391512 77.40207 119.169643 \n",
       "C 64.154096 129.947774 53.473224 143.544179 46.138306 158.967404 \n",
       "C 38.803387 174.39063 34.997152 191.256449 34.997152 208.335 \n",
       "C 34.997152 225.413551 38.803387 242.27937 46.138306 257.702596 \n",
       "C 53.473224 273.125821 64.154096 286.722226 77.40207 297.500357 \n",
       "C 90.650045 308.278488 106.135083 315.969837 122.728013 320.013548 \n",
       "C 139.320943 324.057259 156.608396 324.352595 173.329767 320.87802 \n",
       "\" clip-path=\"url(#p95d04b2978)\" style=\"fill: #bbbbbb; opacity: 0.4\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_3\">\n",
       "    <path d=\"M 173.329767 95.79198 \n",
       "C 199.196822 101.16696 222.436345 115.299 239.109529 135.792898 \n",
       "C 255.782714 156.286797 264.891192 181.915406 264.891192 208.335 \n",
       "C 264.891192 234.754594 255.782714 260.383203 239.109529 280.877102 \n",
       "C 222.436345 301.371 199.196822 315.50304 173.329767 320.87802 \n",
       "C 194.027295 335.730942 218.426514 344.590279 243.831949 346.477306 \n",
       "C 269.237383 348.364334 294.677719 343.206906 317.342465 331.574783 \n",
       "C 340.00721 319.94266 359.029835 302.280568 372.308837 280.539705 \n",
       "C 385.587838 258.798843 392.615527 233.810419 392.615527 208.335 \n",
       "C 392.615527 182.859581 385.587838 157.871157 372.308837 136.130295 \n",
       "C 359.029835 114.389432 340.00721 96.72734 317.342465 85.095217 \n",
       "C 294.677719 73.463094 269.237383 68.305666 243.831949 70.192694 \n",
       "C 218.426514 72.079721 194.027295 80.939058 173.329767 95.79198 \n",
       "\" clip-path=\"url(#p95d04b2978)\" style=\"fill: #bbbbbb; opacity: 0.4\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_4\">\n",
       "    <path d=\"M 173.329767 320.87802 \n",
       "C 199.196822 315.50304 222.436345 301.371 239.109529 280.877102 \n",
       "C 255.782714 260.383203 264.891192 234.754594 264.891192 208.335 \n",
       "C 264.891192 181.915406 255.782714 156.286797 239.109529 135.792898 \n",
       "C 222.436345 115.299 199.196822 101.16696 173.329767 95.79198 \n",
       "C 155.458621 108.616638 140.896577 125.515488 130.852896 145.085214 \n",
       "C 120.809216 164.65494 115.569831 186.338416 115.569831 208.335 \n",
       "C 115.569831 230.331584 120.809216 252.01506 130.852896 271.584786 \n",
       "C 140.896577 291.154512 155.458621 308.053362 173.329767 320.87802 \n",
       "\" clip-path=\"url(#p95d04b2978)\" style=\"fill: #ffffff; opacity: 0.4\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_5\">\n",
       "    <path d=\"M 149.944172 323.28202 \n",
       "C 180.428478 323.28202 209.66833 311.170477 231.223989 289.614817 \n",
       "C 252.779649 268.059158 264.891192 238.819306 264.891192 208.335 \n",
       "C 264.891192 177.850694 252.779649 148.610842 231.223989 127.055183 \n",
       "C 209.66833 105.499523 180.428478 93.38798 149.944172 93.38798 \n",
       "C 119.459866 93.38798 90.220014 105.499523 68.664355 127.055183 \n",
       "C 47.108695 148.610842 34.997152 177.850694 34.997152 208.335 \n",
       "C 34.997152 238.819306 47.108695 268.059158 68.664355 289.614817 \n",
       "C 90.220014 311.170477 119.459866 323.28202 149.944172 323.28202 \n",
       "L 149.944172 323.28202 \n",
       "z\n",
       "\" clip-path=\"url(#p95d04b2978)\" style=\"fill: none; stroke: #000000; stroke-width: 2; stroke-linejoin: miter\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_6\">\n",
       "    <path d=\"M 254.092679 346.857848 \n",
       "C 290.829368 346.857848 326.066363 332.262207 352.043124 306.285445 \n",
       "C 378.019886 280.308683 392.615527 245.071689 392.615527 208.335 \n",
       "C 392.615527 171.598311 378.019886 136.361317 352.043124 110.384555 \n",
       "C 326.066363 84.407793 290.829368 69.812152 254.092679 69.812152 \n",
       "C 217.35599 69.812152 182.118996 84.407793 156.142234 110.384555 \n",
       "C 130.165472 136.361317 115.569831 171.598311 115.569831 208.335 \n",
       "C 115.569831 245.071689 130.165472 280.308683 156.142234 306.285445 \n",
       "C 182.118996 332.262207 217.35599 346.857848 254.092679 346.857848 \n",
       "L 254.092679 346.857848 \n",
       "z\n",
       "\" clip-path=\"url(#p95d04b2978)\" style=\"fill: none; stroke: #000000; stroke-width: 2; stroke-linejoin: miter\"/>\n",
       "   </g>\n",
       "   <g id=\"text_1\">\n",
       "    <!-- 703 -->\n",
       "    <g transform=\"translate(61.937242 212.47125) scale(0.16 -0.16)\">\n",
       "     <defs>\n",
       "      <path id=\"ArialMT-37\" d=\"M 303 3981 \n",
       "L 303 4522 \n",
       "L 3269 4522 \n",
       "L 3269 4084 \n",
       "Q 2831 3619 2401 2847 \n",
       "Q 1972 2075 1738 1259 \n",
       "Q 1569 684 1522 0 \n",
       "L 944 0 \n",
       "Q 953 541 1156 1306 \n",
       "Q 1359 2072 1739 2783 \n",
       "Q 2119 3494 2547 3981 \n",
       "L 303 3981 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-30\" d=\"M 266 2259 \n",
       "Q 266 3072 433 3567 \n",
       "Q 600 4063 929 4331 \n",
       "Q 1259 4600 1759 4600 \n",
       "Q 2128 4600 2406 4451 \n",
       "Q 2684 4303 2865 4023 \n",
       "Q 3047 3744 3150 3342 \n",
       "Q 3253 2941 3253 2259 \n",
       "Q 3253 1453 3087 958 \n",
       "Q 2922 463 2592 192 \n",
       "Q 2263 -78 1759 -78 \n",
       "Q 1097 -78 719 397 \n",
       "Q 266 969 266 2259 \n",
       "z\n",
       "M 844 2259 \n",
       "Q 844 1131 1108 757 \n",
       "Q 1372 384 1759 384 \n",
       "Q 2147 384 2411 759 \n",
       "Q 2675 1134 2675 2259 \n",
       "Q 2675 3391 2411 3762 \n",
       "Q 2147 4134 1753 4134 \n",
       "Q 1366 4134 1134 3806 \n",
       "Q 844 3388 844 2259 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-33\" d=\"M 269 1209 \n",
       "L 831 1284 \n",
       "Q 928 806 1161 595 \n",
       "Q 1394 384 1728 384 \n",
       "Q 2125 384 2398 659 \n",
       "Q 2672 934 2672 1341 \n",
       "Q 2672 1728 2419 1979 \n",
       "Q 2166 2231 1775 2231 \n",
       "Q 1616 2231 1378 2169 \n",
       "L 1441 2663 \n",
       "Q 1497 2656 1531 2656 \n",
       "Q 1891 2656 2178 2843 \n",
       "Q 2466 3031 2466 3422 \n",
       "Q 2466 3731 2256 3934 \n",
       "Q 2047 4138 1716 4138 \n",
       "Q 1388 4138 1169 3931 \n",
       "Q 950 3725 888 3313 \n",
       "L 325 3413 \n",
       "Q 428 3978 793 4289 \n",
       "Q 1159 4600 1703 4600 \n",
       "Q 2078 4600 2393 4439 \n",
       "Q 2709 4278 2876 4000 \n",
       "Q 3044 3722 3044 3409 \n",
       "Q 3044 3113 2884 2869 \n",
       "Q 2725 2625 2413 2481 \n",
       "Q 2819 2388 3044 2092 \n",
       "Q 3269 1797 3269 1353 \n",
       "Q 3269 753 2831 336 \n",
       "Q 2394 -81 1725 -81 \n",
       "Q 1122 -81 723 278 \n",
       "Q 325 638 269 1209 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "     </defs>\n",
       "     <use xlink:href=\"#ArialMT-37\"/>\n",
       "     <use xlink:href=\"#ArialMT-30\" x=\"55.615234\"/>\n",
       "     <use xlink:href=\"#ArialMT-33\" x=\"111.230469\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_2\">\n",
       "    <!-- 1,480 -->\n",
       "    <g transform=\"translate(308.73586 212.47125) scale(0.16 -0.16)\">\n",
       "     <defs>\n",
       "      <path id=\"ArialMT-31\" d=\"M 2384 0 \n",
       "L 1822 0 \n",
       "L 1822 3584 \n",
       "Q 1619 3391 1289 3197 \n",
       "Q 959 3003 697 2906 \n",
       "L 697 3450 \n",
       "Q 1169 3672 1522 3987 \n",
       "Q 1875 4303 2022 4600 \n",
       "L 2384 4600 \n",
       "L 2384 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-2c\" d=\"M 569 0 \n",
       "L 569 641 \n",
       "L 1209 641 \n",
       "L 1209 0 \n",
       "Q 1209 -353 1084 -570 \n",
       "Q 959 -788 688 -906 \n",
       "L 531 -666 \n",
       "Q 709 -588 793 -436 \n",
       "Q 878 -284 888 0 \n",
       "L 569 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-34\" d=\"M 2069 0 \n",
       "L 2069 1097 \n",
       "L 81 1097 \n",
       "L 81 1613 \n",
       "L 2172 4581 \n",
       "L 2631 4581 \n",
       "L 2631 1613 \n",
       "L 3250 1613 \n",
       "L 3250 1097 \n",
       "L 2631 1097 \n",
       "L 2631 0 \n",
       "L 2069 0 \n",
       "z\n",
       "M 2069 1613 \n",
       "L 2069 3678 \n",
       "L 634 1613 \n",
       "L 2069 1613 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-38\" d=\"M 1131 2484 \n",
       "Q 781 2613 612 2850 \n",
       "Q 444 3088 444 3419 \n",
       "Q 444 3919 803 4259 \n",
       "Q 1163 4600 1759 4600 \n",
       "Q 2359 4600 2725 4251 \n",
       "Q 3091 3903 3091 3403 \n",
       "Q 3091 3084 2923 2848 \n",
       "Q 2756 2613 2416 2484 \n",
       "Q 2838 2347 3058 2040 \n",
       "Q 3278 1734 3278 1309 \n",
       "Q 3278 722 2862 322 \n",
       "Q 2447 -78 1769 -78 \n",
       "Q 1091 -78 675 323 \n",
       "Q 259 725 259 1325 \n",
       "Q 259 1772 486 2073 \n",
       "Q 713 2375 1131 2484 \n",
       "z\n",
       "M 1019 3438 \n",
       "Q 1019 3113 1228 2906 \n",
       "Q 1438 2700 1772 2700 \n",
       "Q 2097 2700 2305 2904 \n",
       "Q 2513 3109 2513 3406 \n",
       "Q 2513 3716 2298 3927 \n",
       "Q 2084 4138 1766 4138 \n",
       "Q 1444 4138 1231 3931 \n",
       "Q 1019 3725 1019 3438 \n",
       "z\n",
       "M 838 1322 \n",
       "Q 838 1081 952 856 \n",
       "Q 1066 631 1291 507 \n",
       "Q 1516 384 1775 384 \n",
       "Q 2178 384 2440 643 \n",
       "Q 2703 903 2703 1303 \n",
       "Q 2703 1709 2433 1975 \n",
       "Q 2163 2241 1756 2241 \n",
       "Q 1359 2241 1098 1978 \n",
       "Q 838 1716 838 1322 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "     </defs>\n",
       "     <use xlink:href=\"#ArialMT-31\"/>\n",
       "     <use xlink:href=\"#ArialMT-2c\" x=\"55.615234\"/>\n",
       "     <use xlink:href=\"#ArialMT-34\" x=\"83.398438\"/>\n",
       "     <use xlink:href=\"#ArialMT-38\" x=\"139.013672\"/>\n",
       "     <use xlink:href=\"#ArialMT-30\" x=\"194.628906\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_3\">\n",
       "    <!-- 1,015 | 1,000 -->\n",
       "    <g transform=\"translate(143.671762 212.475) scale(0.16 -0.16)\">\n",
       "     <defs>\n",
       "      <path id=\"ArialMT-35\" d=\"M 266 1200 \n",
       "L 856 1250 \n",
       "Q 922 819 1161 601 \n",
       "Q 1400 384 1738 384 \n",
       "Q 2144 384 2425 690 \n",
       "Q 2706 997 2706 1503 \n",
       "Q 2706 1984 2436 2262 \n",
       "Q 2166 2541 1728 2541 \n",
       "Q 1456 2541 1237 2417 \n",
       "Q 1019 2294 894 2097 \n",
       "L 366 2166 \n",
       "L 809 4519 \n",
       "L 3088 4519 \n",
       "L 3088 3981 \n",
       "L 1259 3981 \n",
       "L 1013 2750 \n",
       "Q 1425 3038 1878 3038 \n",
       "Q 2478 3038 2890 2622 \n",
       "Q 3303 2206 3303 1553 \n",
       "Q 3303 931 2941 478 \n",
       "Q 2500 -78 1738 -78 \n",
       "Q 1113 -78 717 272 \n",
       "Q 322 622 266 1200 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-20\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-7c\" d=\"M 588 -1347 \n",
       "L 588 4659 \n",
       "L 1078 4659 \n",
       "L 1078 -1347 \n",
       "L 588 -1347 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "     </defs>\n",
       "     <use xlink:href=\"#ArialMT-31\"/>\n",
       "     <use xlink:href=\"#ArialMT-2c\" x=\"55.615234\"/>\n",
       "     <use xlink:href=\"#ArialMT-30\" x=\"83.398438\"/>\n",
       "     <use xlink:href=\"#ArialMT-31\" x=\"139.013672\"/>\n",
       "     <use xlink:href=\"#ArialMT-35\" x=\"194.628906\"/>\n",
       "     <use xlink:href=\"#ArialMT-20\" x=\"250.244141\"/>\n",
       "     <use xlink:href=\"#ArialMT-7c\" x=\"278.027344\"/>\n",
       "     <use xlink:href=\"#ArialMT-20\" x=\"304.003906\"/>\n",
       "     <use xlink:href=\"#ArialMT-31\" x=\"331.787109\"/>\n",
       "     <use xlink:href=\"#ArialMT-2c\" x=\"387.402344\"/>\n",
       "     <use xlink:href=\"#ArialMT-30\" x=\"415.185547\"/>\n",
       "     <use xlink:href=\"#ArialMT-30\" x=\"470.800781\"/>\n",
       "     <use xlink:href=\"#ArialMT-30\" x=\"526.416016\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_4\">\n",
       "    <!-- Mouse -->\n",
       "    <g transform=\"translate(92.320172 349.698513) scale(0.192 -0.192)\">\n",
       "     <defs>\n",
       "      <path id=\"ArialMT-4d\" d=\"M 475 0 \n",
       "L 475 4581 \n",
       "L 1388 4581 \n",
       "L 2472 1338 \n",
       "Q 2622 884 2691 659 \n",
       "Q 2769 909 2934 1394 \n",
       "L 4031 4581 \n",
       "L 4847 4581 \n",
       "L 4847 0 \n",
       "L 4263 0 \n",
       "L 4263 3834 \n",
       "L 2931 0 \n",
       "L 2384 0 \n",
       "L 1059 3900 \n",
       "L 1059 0 \n",
       "L 475 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-6f\" d=\"M 213 1659 \n",
       "Q 213 2581 725 3025 \n",
       "Q 1153 3394 1769 3394 \n",
       "Q 2453 3394 2887 2945 \n",
       "Q 3322 2497 3322 1706 \n",
       "Q 3322 1066 3130 698 \n",
       "Q 2938 331 2570 128 \n",
       "Q 2203 -75 1769 -75 \n",
       "Q 1072 -75 642 372 \n",
       "Q 213 819 213 1659 \n",
       "z\n",
       "M 791 1659 \n",
       "Q 791 1022 1069 705 \n",
       "Q 1347 388 1769 388 \n",
       "Q 2188 388 2466 706 \n",
       "Q 2744 1025 2744 1678 \n",
       "Q 2744 2294 2464 2611 \n",
       "Q 2184 2928 1769 2928 \n",
       "Q 1347 2928 1069 2612 \n",
       "Q 791 2297 791 1659 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-75\" d=\"M 2597 0 \n",
       "L 2597 488 \n",
       "Q 2209 -75 1544 -75 \n",
       "Q 1250 -75 995 37 \n",
       "Q 741 150 617 320 \n",
       "Q 494 491 444 738 \n",
       "Q 409 903 409 1263 \n",
       "L 409 3319 \n",
       "L 972 3319 \n",
       "L 972 1478 \n",
       "Q 972 1038 1006 884 \n",
       "Q 1059 663 1231 536 \n",
       "Q 1403 409 1656 409 \n",
       "Q 1909 409 2131 539 \n",
       "Q 2353 669 2445 892 \n",
       "Q 2538 1116 2538 1541 \n",
       "L 2538 3319 \n",
       "L 3100 3319 \n",
       "L 3100 0 \n",
       "L 2597 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-73\" d=\"M 197 991 \n",
       "L 753 1078 \n",
       "Q 800 744 1014 566 \n",
       "Q 1228 388 1613 388 \n",
       "Q 2000 388 2187 545 \n",
       "Q 2375 703 2375 916 \n",
       "Q 2375 1106 2209 1216 \n",
       "Q 2094 1291 1634 1406 \n",
       "Q 1016 1563 777 1677 \n",
       "Q 538 1791 414 1992 \n",
       "Q 291 2194 291 2438 \n",
       "Q 291 2659 392 2848 \n",
       "Q 494 3038 669 3163 \n",
       "Q 800 3259 1026 3326 \n",
       "Q 1253 3394 1513 3394 \n",
       "Q 1903 3394 2198 3281 \n",
       "Q 2494 3169 2634 2976 \n",
       "Q 2775 2784 2828 2463 \n",
       "L 2278 2388 \n",
       "Q 2241 2644 2061 2787 \n",
       "Q 1881 2931 1553 2931 \n",
       "Q 1166 2931 1000 2803 \n",
       "Q 834 2675 834 2503 \n",
       "Q 834 2394 903 2306 \n",
       "Q 972 2216 1119 2156 \n",
       "Q 1203 2125 1616 2013 \n",
       "Q 2213 1853 2448 1751 \n",
       "Q 2684 1650 2818 1456 \n",
       "Q 2953 1263 2953 975 \n",
       "Q 2953 694 2789 445 \n",
       "Q 2625 197 2315 61 \n",
       "Q 2006 -75 1616 -75 \n",
       "Q 969 -75 630 194 \n",
       "Q 291 463 197 991 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-65\" d=\"M 2694 1069 \n",
       "L 3275 997 \n",
       "Q 3138 488 2766 206 \n",
       "Q 2394 -75 1816 -75 \n",
       "Q 1088 -75 661 373 \n",
       "Q 234 822 234 1631 \n",
       "Q 234 2469 665 2931 \n",
       "Q 1097 3394 1784 3394 \n",
       "Q 2450 3394 2872 2941 \n",
       "Q 3294 2488 3294 1666 \n",
       "Q 3294 1616 3291 1516 \n",
       "L 816 1516 \n",
       "Q 847 969 1125 678 \n",
       "Q 1403 388 1819 388 \n",
       "Q 2128 388 2347 550 \n",
       "Q 2566 713 2694 1069 \n",
       "z\n",
       "M 847 1978 \n",
       "L 2700 1978 \n",
       "Q 2663 2397 2488 2606 \n",
       "Q 2219 2931 1791 2931 \n",
       "Q 1403 2931 1139 2672 \n",
       "Q 875 2413 847 1978 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "     </defs>\n",
       "     <use xlink:href=\"#ArialMT-4d\"/>\n",
       "     <use xlink:href=\"#ArialMT-6f\" x=\"83.300781\"/>\n",
       "     <use xlink:href=\"#ArialMT-75\" x=\"138.916016\"/>\n",
       "     <use xlink:href=\"#ArialMT-73\" x=\"194.53125\"/>\n",
       "     <use xlink:href=\"#ArialMT-65\" x=\"244.53125\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_5\">\n",
       "    <!-- Human -->\n",
       "    <g transform=\"translate(254.092679 373.274341) scale(0.192 -0.192)\">\n",
       "     <defs>\n",
       "      <path id=\"ArialMT-48\" d=\"M 513 0 \n",
       "L 513 4581 \n",
       "L 1119 4581 \n",
       "L 1119 2700 \n",
       "L 3500 2700 \n",
       "L 3500 4581 \n",
       "L 4106 4581 \n",
       "L 4106 0 \n",
       "L 3500 0 \n",
       "L 3500 2159 \n",
       "L 1119 2159 \n",
       "L 1119 0 \n",
       "L 513 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-6d\" d=\"M 422 0 \n",
       "L 422 3319 \n",
       "L 925 3319 \n",
       "L 925 2853 \n",
       "Q 1081 3097 1340 3245 \n",
       "Q 1600 3394 1931 3394 \n",
       "Q 2300 3394 2536 3241 \n",
       "Q 2772 3088 2869 2813 \n",
       "Q 3263 3394 3894 3394 \n",
       "Q 4388 3394 4653 3120 \n",
       "Q 4919 2847 4919 2278 \n",
       "L 4919 0 \n",
       "L 4359 0 \n",
       "L 4359 2091 \n",
       "Q 4359 2428 4304 2576 \n",
       "Q 4250 2725 4106 2815 \n",
       "Q 3963 2906 3769 2906 \n",
       "Q 3419 2906 3187 2673 \n",
       "Q 2956 2441 2956 1928 \n",
       "L 2956 0 \n",
       "L 2394 0 \n",
       "L 2394 2156 \n",
       "Q 2394 2531 2256 2718 \n",
       "Q 2119 2906 1806 2906 \n",
       "Q 1569 2906 1367 2781 \n",
       "Q 1166 2656 1075 2415 \n",
       "Q 984 2175 984 1722 \n",
       "L 984 0 \n",
       "L 422 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-61\" d=\"M 2588 409 \n",
       "Q 2275 144 1986 34 \n",
       "Q 1697 -75 1366 -75 \n",
       "Q 819 -75 525 192 \n",
       "Q 231 459 231 875 \n",
       "Q 231 1119 342 1320 \n",
       "Q 453 1522 633 1644 \n",
       "Q 813 1766 1038 1828 \n",
       "Q 1203 1872 1538 1913 \n",
       "Q 2219 1994 2541 2106 \n",
       "Q 2544 2222 2544 2253 \n",
       "Q 2544 2597 2384 2738 \n",
       "Q 2169 2928 1744 2928 \n",
       "Q 1347 2928 1158 2789 \n",
       "Q 969 2650 878 2297 \n",
       "L 328 2372 \n",
       "Q 403 2725 575 2942 \n",
       "Q 747 3159 1072 3276 \n",
       "Q 1397 3394 1825 3394 \n",
       "Q 2250 3394 2515 3294 \n",
       "Q 2781 3194 2906 3042 \n",
       "Q 3031 2891 3081 2659 \n",
       "Q 3109 2516 3109 2141 \n",
       "L 3109 1391 \n",
       "Q 3109 606 3145 398 \n",
       "Q 3181 191 3288 0 \n",
       "L 2700 0 \n",
       "Q 2613 175 2588 409 \n",
       "z\n",
       "M 2541 1666 \n",
       "Q 2234 1541 1622 1453 \n",
       "Q 1275 1403 1131 1340 \n",
       "Q 988 1278 909 1158 \n",
       "Q 831 1038 831 891 \n",
       "Q 831 666 1001 516 \n",
       "Q 1172 366 1500 366 \n",
       "Q 1825 366 2078 508 \n",
       "Q 2331 650 2450 897 \n",
       "Q 2541 1088 2541 1459 \n",
       "L 2541 1666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-6e\" d=\"M 422 0 \n",
       "L 422 3319 \n",
       "L 928 3319 \n",
       "L 928 2847 \n",
       "Q 1294 3394 1984 3394 \n",
       "Q 2284 3394 2536 3286 \n",
       "Q 2788 3178 2913 3003 \n",
       "Q 3038 2828 3088 2588 \n",
       "Q 3119 2431 3119 2041 \n",
       "L 3119 0 \n",
       "L 2556 0 \n",
       "L 2556 2019 \n",
       "Q 2556 2363 2490 2533 \n",
       "Q 2425 2703 2258 2804 \n",
       "Q 2091 2906 1866 2906 \n",
       "Q 1506 2906 1245 2678 \n",
       "Q 984 2450 984 1813 \n",
       "L 984 0 \n",
       "L 422 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "     </defs>\n",
       "     <use xlink:href=\"#ArialMT-48\"/>\n",
       "     <use xlink:href=\"#ArialMT-75\" x=\"72.216797\"/>\n",
       "     <use xlink:href=\"#ArialMT-6d\" x=\"127.832031\"/>\n",
       "     <use xlink:href=\"#ArialMT-61\" x=\"211.132812\"/>\n",
       "     <use xlink:href=\"#ArialMT-6e\" x=\"266.748047\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"text_6\">\n",
       "    <!-- Species-specific and -shared binding counts -->\n",
       "    <g transform=\"translate(57.28134 18.8475) scale(0.16 -0.16)\">\n",
       "     <defs>\n",
       "      <path id=\"ArialMT-53\" d=\"M 288 1472 \n",
       "L 859 1522 \n",
       "Q 900 1178 1048 958 \n",
       "Q 1197 738 1509 602 \n",
       "Q 1822 466 2213 466 \n",
       "Q 2559 466 2825 569 \n",
       "Q 3091 672 3220 851 \n",
       "Q 3350 1031 3350 1244 \n",
       "Q 3350 1459 3225 1620 \n",
       "Q 3100 1781 2813 1891 \n",
       "Q 2628 1963 1997 2114 \n",
       "Q 1366 2266 1113 2400 \n",
       "Q 784 2572 623 2826 \n",
       "Q 463 3081 463 3397 \n",
       "Q 463 3744 659 4045 \n",
       "Q 856 4347 1234 4503 \n",
       "Q 1613 4659 2075 4659 \n",
       "Q 2584 4659 2973 4495 \n",
       "Q 3363 4331 3572 4012 \n",
       "Q 3781 3694 3797 3291 \n",
       "L 3216 3247 \n",
       "Q 3169 3681 2898 3903 \n",
       "Q 2628 4125 2100 4125 \n",
       "Q 1550 4125 1298 3923 \n",
       "Q 1047 3722 1047 3438 \n",
       "Q 1047 3191 1225 3031 \n",
       "Q 1400 2872 2139 2705 \n",
       "Q 2878 2538 3153 2413 \n",
       "Q 3553 2228 3743 1945 \n",
       "Q 3934 1663 3934 1294 \n",
       "Q 3934 928 3725 604 \n",
       "Q 3516 281 3123 101 \n",
       "Q 2731 -78 2241 -78 \n",
       "Q 1619 -78 1198 103 \n",
       "Q 778 284 539 648 \n",
       "Q 300 1013 288 1472 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-70\" d=\"M 422 -1272 \n",
       "L 422 3319 \n",
       "L 934 3319 \n",
       "L 934 2888 \n",
       "Q 1116 3141 1344 3267 \n",
       "Q 1572 3394 1897 3394 \n",
       "Q 2322 3394 2647 3175 \n",
       "Q 2972 2956 3137 2557 \n",
       "Q 3303 2159 3303 1684 \n",
       "Q 3303 1175 3120 767 \n",
       "Q 2938 359 2589 142 \n",
       "Q 2241 -75 1856 -75 \n",
       "Q 1575 -75 1351 44 \n",
       "Q 1128 163 984 344 \n",
       "L 984 -1272 \n",
       "L 422 -1272 \n",
       "z\n",
       "M 931 1641 \n",
       "Q 931 1000 1190 694 \n",
       "Q 1450 388 1819 388 \n",
       "Q 2194 388 2461 705 \n",
       "Q 2728 1022 2728 1688 \n",
       "Q 2728 2322 2467 2637 \n",
       "Q 2206 2953 1844 2953 \n",
       "Q 1484 2953 1207 2617 \n",
       "Q 931 2281 931 1641 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-63\" d=\"M 2588 1216 \n",
       "L 3141 1144 \n",
       "Q 3050 572 2676 248 \n",
       "Q 2303 -75 1759 -75 \n",
       "Q 1078 -75 664 370 \n",
       "Q 250 816 250 1647 \n",
       "Q 250 2184 428 2587 \n",
       "Q 606 2991 970 3192 \n",
       "Q 1334 3394 1763 3394 \n",
       "Q 2303 3394 2647 3120 \n",
       "Q 2991 2847 3088 2344 \n",
       "L 2541 2259 \n",
       "Q 2463 2594 2264 2762 \n",
       "Q 2066 2931 1784 2931 \n",
       "Q 1359 2931 1093 2626 \n",
       "Q 828 2322 828 1663 \n",
       "Q 828 994 1084 691 \n",
       "Q 1341 388 1753 388 \n",
       "Q 2084 388 2306 591 \n",
       "Q 2528 794 2588 1216 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-69\" d=\"M 425 3934 \n",
       "L 425 4581 \n",
       "L 988 4581 \n",
       "L 988 3934 \n",
       "L 425 3934 \n",
       "z\n",
       "M 425 0 \n",
       "L 425 3319 \n",
       "L 988 3319 \n",
       "L 988 0 \n",
       "L 425 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-2d\" d=\"M 203 1375 \n",
       "L 203 1941 \n",
       "L 1931 1941 \n",
       "L 1931 1375 \n",
       "L 203 1375 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-66\" d=\"M 556 0 \n",
       "L 556 2881 \n",
       "L 59 2881 \n",
       "L 59 3319 \n",
       "L 556 3319 \n",
       "L 556 3672 \n",
       "Q 556 4006 616 4169 \n",
       "Q 697 4388 901 4523 \n",
       "Q 1106 4659 1475 4659 \n",
       "Q 1713 4659 2000 4603 \n",
       "L 1916 4113 \n",
       "Q 1741 4144 1584 4144 \n",
       "Q 1328 4144 1222 4034 \n",
       "Q 1116 3925 1116 3625 \n",
       "L 1116 3319 \n",
       "L 1763 3319 \n",
       "L 1763 2881 \n",
       "L 1116 2881 \n",
       "L 1116 0 \n",
       "L 556 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-64\" d=\"M 2575 0 \n",
       "L 2575 419 \n",
       "Q 2259 -75 1647 -75 \n",
       "Q 1250 -75 917 144 \n",
       "Q 584 363 401 755 \n",
       "Q 219 1147 219 1656 \n",
       "Q 219 2153 384 2558 \n",
       "Q 550 2963 881 3178 \n",
       "Q 1213 3394 1622 3394 \n",
       "Q 1922 3394 2156 3267 \n",
       "Q 2391 3141 2538 2938 \n",
       "L 2538 4581 \n",
       "L 3097 4581 \n",
       "L 3097 0 \n",
       "L 2575 0 \n",
       "z\n",
       "M 797 1656 \n",
       "Q 797 1019 1065 703 \n",
       "Q 1334 388 1700 388 \n",
       "Q 2069 388 2326 689 \n",
       "Q 2584 991 2584 1609 \n",
       "Q 2584 2291 2321 2609 \n",
       "Q 2059 2928 1675 2928 \n",
       "Q 1300 2928 1048 2622 \n",
       "Q 797 2316 797 1656 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-68\" d=\"M 422 0 \n",
       "L 422 4581 \n",
       "L 984 4581 \n",
       "L 984 2938 \n",
       "Q 1378 3394 1978 3394 \n",
       "Q 2347 3394 2619 3248 \n",
       "Q 2891 3103 3008 2847 \n",
       "Q 3125 2591 3125 2103 \n",
       "L 3125 0 \n",
       "L 2563 0 \n",
       "L 2563 2103 \n",
       "Q 2563 2525 2380 2717 \n",
       "Q 2197 2909 1863 2909 \n",
       "Q 1613 2909 1392 2779 \n",
       "Q 1172 2650 1078 2428 \n",
       "Q 984 2206 984 1816 \n",
       "L 984 0 \n",
       "L 422 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-72\" d=\"M 416 0 \n",
       "L 416 3319 \n",
       "L 922 3319 \n",
       "L 922 2816 \n",
       "Q 1116 3169 1280 3281 \n",
       "Q 1444 3394 1641 3394 \n",
       "Q 1925 3394 2219 3213 \n",
       "L 2025 2691 \n",
       "Q 1819 2813 1613 2813 \n",
       "Q 1428 2813 1281 2702 \n",
       "Q 1134 2591 1072 2394 \n",
       "Q 978 2094 978 1738 \n",
       "L 978 0 \n",
       "L 416 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-62\" d=\"M 941 0 \n",
       "L 419 0 \n",
       "L 419 4581 \n",
       "L 981 4581 \n",
       "L 981 2947 \n",
       "Q 1338 3394 1891 3394 \n",
       "Q 2197 3394 2470 3270 \n",
       "Q 2744 3147 2920 2923 \n",
       "Q 3097 2700 3197 2384 \n",
       "Q 3297 2069 3297 1709 \n",
       "Q 3297 856 2875 390 \n",
       "Q 2453 -75 1863 -75 \n",
       "Q 1275 -75 941 416 \n",
       "L 941 0 \n",
       "z\n",
       "M 934 1684 \n",
       "Q 934 1088 1097 822 \n",
       "Q 1363 388 1816 388 \n",
       "Q 2184 388 2453 708 \n",
       "Q 2722 1028 2722 1663 \n",
       "Q 2722 2313 2464 2622 \n",
       "Q 2206 2931 1841 2931 \n",
       "Q 1472 2931 1203 2611 \n",
       "Q 934 2291 934 1684 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-67\" d=\"M 319 -275 \n",
       "L 866 -356 \n",
       "Q 900 -609 1056 -725 \n",
       "Q 1266 -881 1628 -881 \n",
       "Q 2019 -881 2231 -725 \n",
       "Q 2444 -569 2519 -288 \n",
       "Q 2563 -116 2559 434 \n",
       "Q 2191 0 1641 0 \n",
       "Q 956 0 581 494 \n",
       "Q 206 988 206 1678 \n",
       "Q 206 2153 378 2554 \n",
       "Q 550 2956 876 3175 \n",
       "Q 1203 3394 1644 3394 \n",
       "Q 2231 3394 2613 2919 \n",
       "L 2613 3319 \n",
       "L 3131 3319 \n",
       "L 3131 450 \n",
       "Q 3131 -325 2973 -648 \n",
       "Q 2816 -972 2473 -1159 \n",
       "Q 2131 -1347 1631 -1347 \n",
       "Q 1038 -1347 672 -1080 \n",
       "Q 306 -813 319 -275 \n",
       "z\n",
       "M 784 1719 \n",
       "Q 784 1066 1043 766 \n",
       "Q 1303 466 1694 466 \n",
       "Q 2081 466 2343 764 \n",
       "Q 2606 1063 2606 1700 \n",
       "Q 2606 2309 2336 2618 \n",
       "Q 2066 2928 1684 2928 \n",
       "Q 1309 2928 1046 2623 \n",
       "Q 784 2319 784 1719 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-74\" d=\"M 1650 503 \n",
       "L 1731 6 \n",
       "Q 1494 -44 1306 -44 \n",
       "Q 1000 -44 831 53 \n",
       "Q 663 150 594 308 \n",
       "Q 525 466 525 972 \n",
       "L 525 2881 \n",
       "L 113 2881 \n",
       "L 113 3319 \n",
       "L 525 3319 \n",
       "L 525 4141 \n",
       "L 1084 4478 \n",
       "L 1084 3319 \n",
       "L 1650 3319 \n",
       "L 1650 2881 \n",
       "L 1084 2881 \n",
       "L 1084 941 \n",
       "Q 1084 700 1114 631 \n",
       "Q 1144 563 1211 522 \n",
       "Q 1278 481 1403 481 \n",
       "Q 1497 481 1650 503 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "     </defs>\n",
       "     <use xlink:href=\"#ArialMT-53\"/>\n",
       "     <use xlink:href=\"#ArialMT-70\" x=\"66.699219\"/>\n",
       "     <use xlink:href=\"#ArialMT-65\" x=\"122.314453\"/>\n",
       "     <use xlink:href=\"#ArialMT-63\" x=\"177.929688\"/>\n",
       "     <use xlink:href=\"#ArialMT-69\" x=\"227.929688\"/>\n",
       "     <use xlink:href=\"#ArialMT-65\" x=\"250.146484\"/>\n",
       "     <use xlink:href=\"#ArialMT-73\" x=\"305.761719\"/>\n",
       "     <use xlink:href=\"#ArialMT-2d\" x=\"355.761719\"/>\n",
       "     <use xlink:href=\"#ArialMT-73\" x=\"389.0625\"/>\n",
       "     <use xlink:href=\"#ArialMT-70\" x=\"439.0625\"/>\n",
       "     <use xlink:href=\"#ArialMT-65\" x=\"494.677734\"/>\n",
       "     <use xlink:href=\"#ArialMT-63\" x=\"550.292969\"/>\n",
       "     <use xlink:href=\"#ArialMT-69\" x=\"600.292969\"/>\n",
       "     <use xlink:href=\"#ArialMT-66\" x=\"622.509766\"/>\n",
       "     <use xlink:href=\"#ArialMT-69\" x=\"650.292969\"/>\n",
       "     <use xlink:href=\"#ArialMT-63\" x=\"672.509766\"/>\n",
       "     <use xlink:href=\"#ArialMT-20\" x=\"722.509766\"/>\n",
       "     <use xlink:href=\"#ArialMT-61\" x=\"750.292969\"/>\n",
       "     <use xlink:href=\"#ArialMT-6e\" x=\"805.908203\"/>\n",
       "     <use xlink:href=\"#ArialMT-64\" x=\"861.523438\"/>\n",
       "     <use xlink:href=\"#ArialMT-20\" x=\"917.138672\"/>\n",
       "     <use xlink:href=\"#ArialMT-2d\" x=\"944.921875\"/>\n",
       "     <use xlink:href=\"#ArialMT-73\" x=\"978.222656\"/>\n",
       "     <use xlink:href=\"#ArialMT-68\" x=\"1028.222656\"/>\n",
       "     <use xlink:href=\"#ArialMT-61\" x=\"1083.837891\"/>\n",
       "     <use xlink:href=\"#ArialMT-72\" x=\"1139.453125\"/>\n",
       "     <use xlink:href=\"#ArialMT-65\" x=\"1172.753906\"/>\n",
       "     <use xlink:href=\"#ArialMT-64\" x=\"1228.369141\"/>\n",
       "     <use xlink:href=\"#ArialMT-20\" x=\"1283.984375\"/>\n",
       "     <use xlink:href=\"#ArialMT-62\" x=\"1311.767578\"/>\n",
       "     <use xlink:href=\"#ArialMT-69\" x=\"1367.382812\"/>\n",
       "     <use xlink:href=\"#ArialMT-6e\" x=\"1389.599609\"/>\n",
       "     <use xlink:href=\"#ArialMT-64\" x=\"1445.214844\"/>\n",
       "     <use xlink:href=\"#ArialMT-69\" x=\"1500.830078\"/>\n",
       "     <use xlink:href=\"#ArialMT-6e\" x=\"1523.046875\"/>\n",
       "     <use xlink:href=\"#ArialMT-67\" x=\"1578.662109\"/>\n",
       "     <use xlink:href=\"#ArialMT-20\" x=\"1634.277344\"/>\n",
       "     <use xlink:href=\"#ArialMT-63\" x=\"1662.060547\"/>\n",
       "     <use xlink:href=\"#ArialMT-6f\" x=\"1712.060547\"/>\n",
       "     <use xlink:href=\"#ArialMT-75\" x=\"1767.675781\"/>\n",
       "     <use xlink:href=\"#ArialMT-6e\" x=\"1823.291016\"/>\n",
       "     <use xlink:href=\"#ArialMT-74\" x=\"1878.90625\"/>\n",
       "     <use xlink:href=\"#ArialMT-73\" x=\"1906.689453\"/>\n",
       "    </g>\n",
       "    <!-- at gene level, 3'UTR only -->\n",
       "    <g transform=\"translate(124.68259 36.015) scale(0.16 -0.16)\">\n",
       "     <defs>\n",
       "      <path id=\"ArialMT-6c\" d=\"M 409 0 \n",
       "L 409 4581 \n",
       "L 972 4581 \n",
       "L 972 0 \n",
       "L 409 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-76\" d=\"M 1344 0 \n",
       "L 81 3319 \n",
       "L 675 3319 \n",
       "L 1388 1331 \n",
       "Q 1503 1009 1600 663 \n",
       "Q 1675 925 1809 1294 \n",
       "L 2547 3319 \n",
       "L 3125 3319 \n",
       "L 1869 0 \n",
       "L 1344 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-27\" d=\"M 425 2959 \n",
       "L 281 3816 \n",
       "L 281 4581 \n",
       "L 922 4581 \n",
       "L 922 3816 \n",
       "L 772 2959 \n",
       "L 425 2959 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-55\" d=\"M 3500 4581 \n",
       "L 4106 4581 \n",
       "L 4106 1934 \n",
       "Q 4106 1244 3950 837 \n",
       "Q 3794 431 3386 176 \n",
       "Q 2978 -78 2316 -78 \n",
       "Q 1672 -78 1262 144 \n",
       "Q 853 366 678 786 \n",
       "Q 503 1206 503 1934 \n",
       "L 503 4581 \n",
       "L 1109 4581 \n",
       "L 1109 1938 \n",
       "Q 1109 1341 1220 1058 \n",
       "Q 1331 775 1601 622 \n",
       "Q 1872 469 2263 469 \n",
       "Q 2931 469 3215 772 \n",
       "Q 3500 1075 3500 1938 \n",
       "L 3500 4581 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-54\" d=\"M 1659 0 \n",
       "L 1659 4041 \n",
       "L 150 4041 \n",
       "L 150 4581 \n",
       "L 3781 4581 \n",
       "L 3781 4041 \n",
       "L 2266 4041 \n",
       "L 2266 0 \n",
       "L 1659 0 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-52\" d=\"M 503 0 \n",
       "L 503 4581 \n",
       "L 2534 4581 \n",
       "Q 3147 4581 3465 4457 \n",
       "Q 3784 4334 3975 4021 \n",
       "Q 4166 3709 4166 3331 \n",
       "Q 4166 2844 3850 2509 \n",
       "Q 3534 2175 2875 2084 \n",
       "Q 3116 1969 3241 1856 \n",
       "Q 3506 1613 3744 1247 \n",
       "L 4541 0 \n",
       "L 3778 0 \n",
       "L 3172 953 \n",
       "Q 2906 1366 2734 1584 \n",
       "Q 2563 1803 2427 1890 \n",
       "Q 2291 1978 2150 2013 \n",
       "Q 2047 2034 1813 2034 \n",
       "L 1109 2034 \n",
       "L 1109 0 \n",
       "L 503 0 \n",
       "z\n",
       "M 1109 2559 \n",
       "L 2413 2559 \n",
       "Q 2828 2559 3062 2645 \n",
       "Q 3297 2731 3419 2920 \n",
       "Q 3541 3109 3541 3331 \n",
       "Q 3541 3656 3305 3865 \n",
       "Q 3069 4075 2559 4075 \n",
       "L 1109 4075 \n",
       "L 1109 2559 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"ArialMT-79\" d=\"M 397 -1278 \n",
       "L 334 -750 \n",
       "Q 519 -800 656 -800 \n",
       "Q 844 -800 956 -737 \n",
       "Q 1069 -675 1141 -563 \n",
       "Q 1194 -478 1313 -144 \n",
       "Q 1328 -97 1363 -6 \n",
       "L 103 3319 \n",
       "L 709 3319 \n",
       "L 1400 1397 \n",
       "Q 1534 1031 1641 628 \n",
       "Q 1738 1016 1872 1384 \n",
       "L 2581 3319 \n",
       "L 3144 3319 \n",
       "L 1881 -56 \n",
       "Q 1678 -603 1566 -809 \n",
       "Q 1416 -1088 1222 -1217 \n",
       "Q 1028 -1347 759 -1347 \n",
       "Q 597 -1347 397 -1278 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "     </defs>\n",
       "     <use xlink:href=\"#ArialMT-61\"/>\n",
       "     <use xlink:href=\"#ArialMT-74\" x=\"55.615234\"/>\n",
       "     <use xlink:href=\"#ArialMT-20\" x=\"83.398438\"/>\n",
       "     <use xlink:href=\"#ArialMT-67\" x=\"111.181641\"/>\n",
       "     <use xlink:href=\"#ArialMT-65\" x=\"166.796875\"/>\n",
       "     <use xlink:href=\"#ArialMT-6e\" x=\"222.412109\"/>\n",
       "     <use xlink:href=\"#ArialMT-65\" x=\"278.027344\"/>\n",
       "     <use xlink:href=\"#ArialMT-20\" x=\"333.642578\"/>\n",
       "     <use xlink:href=\"#ArialMT-6c\" x=\"361.425781\"/>\n",
       "     <use xlink:href=\"#ArialMT-65\" x=\"383.642578\"/>\n",
       "     <use xlink:href=\"#ArialMT-76\" x=\"439.257812\"/>\n",
       "     <use xlink:href=\"#ArialMT-65\" x=\"489.257812\"/>\n",
       "     <use xlink:href=\"#ArialMT-6c\" x=\"544.873047\"/>\n",
       "     <use xlink:href=\"#ArialMT-2c\" x=\"567.089844\"/>\n",
       "     <use xlink:href=\"#ArialMT-20\" x=\"594.873047\"/>\n",
       "     <use xlink:href=\"#ArialMT-33\" x=\"622.65625\"/>\n",
       "     <use xlink:href=\"#ArialMT-27\" x=\"678.271484\"/>\n",
       "     <use xlink:href=\"#ArialMT-55\" x=\"697.363281\"/>\n",
       "     <use xlink:href=\"#ArialMT-54\" x=\"769.580078\"/>\n",
       "     <use xlink:href=\"#ArialMT-52\" x=\"830.664062\"/>\n",
       "     <use xlink:href=\"#ArialMT-20\" x=\"902.880859\"/>\n",
       "     <use xlink:href=\"#ArialMT-6f\" x=\"930.664062\"/>\n",
       "     <use xlink:href=\"#ArialMT-6e\" x=\"986.279297\"/>\n",
       "     <use xlink:href=\"#ArialMT-6c\" x=\"1041.894531\"/>\n",
       "     <use xlink:href=\"#ArialMT-79\" x=\"1064.111328\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "  </g>\n",
       " </g>\n",
       " <defs>\n",
       "  <clipPath id=\"p95d04b2978\">\n",
       "   <rect x=\"7.2\" y=\"42.015\" width=\"413.212679\" height=\"332.64\"/>\n",
       "  </clipPath>\n",
       " </defs>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<Figure size 1000x600 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "v = venn.venn2(\n",
    "    subsets=(\n",
    "        len(m_genes_no_bound_orth_complete_3utr),\n",
    "        len(h_genes_no_bound_orth_complete_3utr),\n",
    "        len(m_genes_bound_orth_complete_3utr),\n",
    "    ),\n",
    "    set_labels=(\"Mouse\", \"Human\"),\n",
    "    set_colors=(\n",
    "        \"#BBBBBB\",\n",
    "        \"#BBBBBB\",\n",
    "    ),\n",
    ")\n",
    "\n",
    "# Update the labels to include the info on mouse | human orthologous bound genes.\n",
    "v.subset_labels[0].set_text(f\"{len(m_genes_no_bound_orth_complete_3utr):,}\")\n",
    "v.subset_labels[1].set_text(f\"{len(h_genes_no_bound_orth_complete_3utr):,}\")\n",
    "v.subset_labels[2].set_text(f\"{len(m_genes_bound_orth_complete_3utr):,} | {len(h_genes_bound_orth_complete_3utr):,}\")\n",
    "\n",
    "circles = venn.venn2_circles(\n",
    "    subsets=(\n",
    "        len(m_genes_no_bound_orth_complete_3utr),\n",
    "        len(h_genes_no_bound_orth_complete_3utr),\n",
    "        len(m_genes_bound_orth_complete_3utr),\n",
    "    ),\n",
    "    linestyle=\"-\",\n",
    ")\n",
    "\n",
    "plt.title(\"Species-specific and -shared binding counts\\nat gene level, 3'UTR only\")\n",
    "\n",
    "if params_export_flag:\n",
    "    plt.savefig(outputdir / \"venn_diagram_genes_w_peaks.3UTR.svg\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b9198d7",
   "metadata": {},
   "source": [
    "## Export data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91966b13",
   "metadata": {},
   "source": [
    "### Mouse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "4ac473d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "m_lift_3utr.shape=(11880, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier.mouse_source</th>\n",
       "      <th>identifier.human_target</th>\n",
       "      <th>orth_first_order</th>\n",
       "      <th>orth_second_order</th>\n",
       "      <th>orth_third_order.details</th>\n",
       "      <th>orth_third_order</th>\n",
       "      <th>gene_orthology_status</th>\n",
       "      <th>orthologous_gene_set</th>\n",
       "      <th>annotated_lifted_region_bound_gene_set</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:6274997-6275006:+</td>\n",
       "      <td>chr8:52623049-52623059:-</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSMUSG00000025907, [ENSG00000023287])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSG00000023287}</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:7170406-7170415:+</td>\n",
       "      <td>chr8:51819766-51819775:-</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSMUSG00000051285, [ENSG00000168300])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSG00000168300}</td>\n",
       "      <td>{ENSG00000168300}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1:7170431-7170440:+</td>\n",
       "      <td>chr8:51819741-51819750:-</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSMUSG00000051285, [ENSG00000168300])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSG00000168300}</td>\n",
       "      <td>{ENSG00000168300}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  identifier.mouse_source   identifier.human_target orth_first_order  \\\n",
       "0  chr1:6274997-6275006:+  chr8:52623049-52623059:-             True   \n",
       "1  chr1:7170406-7170415:+  chr8:51819766-51819775:-             True   \n",
       "2  chr1:7170431-7170440:+  chr8:51819741-51819750:-             True   \n",
       "\n",
       "  orth_second_order                   orth_third_order.details  \\\n",
       "0              True  [(ENSMUSG00000025907, [ENSG00000023287])]   \n",
       "1              True  [(ENSMUSG00000051285, [ENSG00000168300])]   \n",
       "2              True  [(ENSMUSG00000051285, [ENSG00000168300])]   \n",
       "\n",
       "  orth_third_order gene_orthology_status orthologous_gene_set  \\\n",
       "0             True                 first    {ENSG00000023287}   \n",
       "1             True                 first    {ENSG00000168300}   \n",
       "2             True                 first    {ENSG00000168300}   \n",
       "\n",
       "  annotated_lifted_region_bound_gene_set  \n",
       "0                                     {}  \n",
       "1                      {ENSG00000168300}  \n",
       "2                      {ENSG00000168300}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wrote mouse 3'UTR with identified gene sets orthologies:  ../results/gene_sets_overlaps/m_lift_3utr.tsv.gz\n"
     ]
    }
   ],
   "source": [
    "print(f\"{m_lift_3utr.shape=}\")\n",
    "display(m_lift_3utr.head(3))\n",
    "\n",
    "\n",
    "if params_export_flag:\n",
    "    fp = outputdir / \"m_lift_3utr.tsv.gz\"\n",
    "    m_lift_3utr.to_csv(fp, sep=\"\\t\", index=False, compression=\"gzip\")\n",
    "    print(\"Wrote mouse 3'UTR with identified gene sets orthologies: \", fp)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "093a7e6c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "m_umap_3utr.shape=(685, 3)\n",
      "Wrote mouse 3'UTR unmapped peaks:  ../results/gene_sets_overlaps/m_umap_3utr.tsv.gz\n"
     ]
    }
   ],
   "source": [
    "print(f\"{m_umap_3utr.shape=}\")\n",
    "m_umap_3utr.head(3)\n",
    "\n",
    "if params_export_flag:\n",
    "    fp = outputdir / \"m_umap_3utr.tsv.gz\"\n",
    "    m_umap_3utr.to_csv(fp, sep=\"\\t\", index=False, compression=\"gzip\")\n",
    "    print(\"Wrote mouse 3'UTR unmapped peaks: \", fp)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "id": "19c86ea4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "len(m_genes_bound_orth_complete)=1312\n",
      "['ENSMUSG00000018143', 'ENSMUSG00000035021', 'ENSMUSG00000079509']\n",
      "len(m_genes_no_bound_orth_complete)=824\n",
      "['ENSMUSG00000037958', 'ENSMUSG00000027177', 'ENSMUSG00000062949']\n",
      "len(m_genes_bound_orth_complete_3utr)=1015\n",
      "['ENSMUSG00000018143', 'ENSMUSG00000031924', 'ENSMUSG00000035021']\n",
      "len(m_genes_no_bound_orth_complete_3utr)=703\n",
      "['ENSMUSG00000037958', 'ENSMUSG00000027177', 'ENSMUSG00000062949']\n"
     ]
    }
   ],
   "source": [
    "# GENE SETS\n",
    "\n",
    "print(f\"{len(m_genes_bound_orth_complete)=}\")\n",
    "print(list(m_genes_bound_orth_complete)[:3])\n",
    "\n",
    "if params_export_flag:\n",
    "    fp = outputdir / \"m_genes_bound_orth_complete.txt\"\n",
    "    with open(fp, \"w\") as f:\n",
    "        f.write(\"\\n\".join(m_genes_bound_orth_complete) + \"\\n\")\n",
    "\n",
    "\n",
    "print(f\"{len(m_genes_no_bound_orth_complete)=}\")\n",
    "print(list(m_genes_no_bound_orth_complete)[:3])\n",
    "\n",
    "if params_export_flag:\n",
    "    fp = outputdir / \"m_genes_no_bound_orth_complete.txt\"\n",
    "    with open(fp, \"w\") as f:\n",
    "        f.write(\"\\n\".join(m_genes_no_bound_orth_complete) + \"\\n\")\n",
    "\n",
    "\n",
    "###\n",
    "\n",
    "\n",
    "print(f\"{len(m_genes_bound_orth_complete_3utr)=}\")\n",
    "print(list(m_genes_bound_orth_complete_3utr)[:3])\n",
    "\n",
    "if params_export_flag:\n",
    "    fp = outputdir / \"m_genes_bound_orth_complete_3utr.txt\"\n",
    "    with open(fp, \"w\") as f:\n",
    "        f.write(\"\\n\".join(m_genes_bound_orth_complete_3utr) + \"\\n\")\n",
    "\n",
    "\n",
    "print(f\"{len(m_genes_no_bound_orth_complete_3utr)=}\")\n",
    "print(list(m_genes_no_bound_orth_complete_3utr)[:3])\n",
    "\n",
    "if params_export_flag:\n",
    "    fp = outputdir / \"m_genes_no_bound_orth_complete_3utr.txt\"\n",
    "    with open(fp, \"w\") as f:\n",
    "        f.write(\"\\n\".join(m_genes_no_bound_orth_complete_3utr) + \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "id": "87644126",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>main_annot</th>\n",
       "      <th>annot_set</th>\n",
       "      <th>gene_id_set</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:6274997-6275006:+</th>\n",
       "      <td>three_prime_UTR;Rb1cc1;protein_coding;ENSMUSG0...</td>\n",
       "      <td>ENSMUSG00000025907</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{three_prime_UTR}</td>\n",
       "      <td>{ENSMUSG00000025907}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:7170406-7170415:+</th>\n",
       "      <td>three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...</td>\n",
       "      <td>ENSMUSG00000051285</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{three_prime_UTR}</td>\n",
       "      <td>{ENSMUSG00000051285}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:7170431-7170440:+</th>\n",
       "      <td>three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...</td>\n",
       "      <td>ENSMUSG00000051285</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{three_prime_UTR}</td>\n",
       "      <td>{ENSMUSG00000051285}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                     name  \\\n",
       "identifier                                                                  \n",
       "chr1:6274997-6275006:+  three_prime_UTR;Rb1cc1;protein_coding;ENSMUSG0...   \n",
       "chr1:7170406-7170415:+  three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...   \n",
       "chr1:7170431-7170440:+  three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...   \n",
       "\n",
       "                                   gene_id       main_annot  \\\n",
       "identifier                                                    \n",
       "chr1:6274997-6275006:+  ENSMUSG00000025907  three_prime_UTR   \n",
       "chr1:7170406-7170415:+  ENSMUSG00000051285  three_prime_UTR   \n",
       "chr1:7170431-7170440:+  ENSMUSG00000051285  three_prime_UTR   \n",
       "\n",
       "                                annot_set           gene_id_set  \n",
       "identifier                                                       \n",
       "chr1:6274997-6275006:+  {three_prime_UTR}  {ENSMUSG00000025907}  \n",
       "chr1:7170406-7170415:+  {three_prime_UTR}  {ENSMUSG00000051285}  \n",
       "chr1:7170431-7170440:+  {three_prime_UTR}  {ENSMUSG00000051285}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "m_id_to_meta.shape=(48126, 5)\n",
      "Wrote mouse all metadata:  ../results/gene_sets_overlaps/m_id_to_meta.tsv.gz\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>main_annot</th>\n",
       "      <th>annot_set</th>\n",
       "      <th>gene_id_set</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:6274997-6275006:+</th>\n",
       "      <td>three_prime_UTR;Rb1cc1;protein_coding;ENSMUSG0...</td>\n",
       "      <td>ENSMUSG00000025907</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{three_prime_UTR}</td>\n",
       "      <td>{ENSMUSG00000025907}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:7170406-7170415:+</th>\n",
       "      <td>three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...</td>\n",
       "      <td>ENSMUSG00000051285</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{three_prime_UTR}</td>\n",
       "      <td>{ENSMUSG00000051285}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:7170431-7170440:+</th>\n",
       "      <td>three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...</td>\n",
       "      <td>ENSMUSG00000051285</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{three_prime_UTR}</td>\n",
       "      <td>{ENSMUSG00000051285}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                     name  \\\n",
       "identifier                                                                  \n",
       "chr1:6274997-6275006:+  three_prime_UTR;Rb1cc1;protein_coding;ENSMUSG0...   \n",
       "chr1:7170406-7170415:+  three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...   \n",
       "chr1:7170431-7170440:+  three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...   \n",
       "\n",
       "                                   gene_id       main_annot  \\\n",
       "identifier                                                    \n",
       "chr1:6274997-6275006:+  ENSMUSG00000025907  three_prime_UTR   \n",
       "chr1:7170406-7170415:+  ENSMUSG00000051285  three_prime_UTR   \n",
       "chr1:7170431-7170440:+  ENSMUSG00000051285  three_prime_UTR   \n",
       "\n",
       "                                annot_set           gene_id_set  \n",
       "identifier                                                       \n",
       "chr1:6274997-6275006:+  {three_prime_UTR}  {ENSMUSG00000025907}  \n",
       "chr1:7170406-7170415:+  {three_prime_UTR}  {ENSMUSG00000051285}  \n",
       "chr1:7170431-7170440:+  {three_prime_UTR}  {ENSMUSG00000051285}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "m_id_to_meta_3utr.shape=(30917, 5)\n",
      "Wrote mouse 3'UTR metadata:  ../results/gene_sets_overlaps/m_id_to_meta_3utr.tsv.gz\n"
     ]
    }
   ],
   "source": [
    "display(m_id_to_meta.head(3))\n",
    "print(f\"{m_id_to_meta.shape=}\")\n",
    "\n",
    "if params_export_flag:\n",
    "    fp = outputdir / \"m_id_to_meta.tsv.gz\"\n",
    "    m_id_to_meta.to_csv(fp, sep=\"\\t\", index=True, compression=\"gzip\")\n",
    "    print(\"Wrote mouse all metadata: \", fp)\n",
    "\n",
    "\n",
    "display(m_id_to_meta_3utr.head(3))\n",
    "print(f\"{m_id_to_meta_3utr.shape=}\")\n",
    "\n",
    "if params_export_flag:\n",
    "    fp = outputdir / \"m_id_to_meta_3utr.tsv.gz\"\n",
    "    m_id_to_meta_3utr.to_csv(fp, sep=\"\\t\", index=True, compression=\"gzip\")\n",
    "    print(\"Wrote mouse 3'UTR metadata: \", fp)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "id": "02af4fdd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "m_id_to_bed6_3utr.shape=(31695, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:6274997-6275006:+</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6274997</td>\n",
       "      <td>6275006</td>\n",
       "      <td>three_prime_UTR;Rb1cc1;protein_coding;ENSMUSG0...</td>\n",
       "      <td>9.344450</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:7170406-7170415:+</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7170406</td>\n",
       "      <td>7170415</td>\n",
       "      <td>three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...</td>\n",
       "      <td>5.094060</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:7170431-7170440:+</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7170431</td>\n",
       "      <td>7170440</td>\n",
       "      <td>three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...</td>\n",
       "      <td>26.054101</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       chrom    start      end  \\\n",
       "identifier                                       \n",
       "chr1:6274997-6275006:+  chr1  6274997  6275006   \n",
       "chr1:7170406-7170415:+  chr1  7170406  7170415   \n",
       "chr1:7170431-7170440:+  chr1  7170431  7170440   \n",
       "\n",
       "                                                                     name  \\\n",
       "identifier                                                                  \n",
       "chr1:6274997-6275006:+  three_prime_UTR;Rb1cc1;protein_coding;ENSMUSG0...   \n",
       "chr1:7170406-7170415:+  three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...   \n",
       "chr1:7170431-7170440:+  three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...   \n",
       "\n",
       "                            score strand  \n",
       "identifier                                \n",
       "chr1:6274997-6275006:+   9.344450      +  \n",
       "chr1:7170406-7170415:+   5.094060      +  \n",
       "chr1:7170431-7170440:+  26.054101      +  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wrote mouse 3'UTR bed6:  ../results/gene_sets_overlaps/m_id_to_bed6_3utr.tsv.gz\n"
     ]
    }
   ],
   "source": [
    "print(f\"{m_id_to_bed6_3utr.shape=}\")\n",
    "display(m_id_to_bed6_3utr.head(3))\n",
    "\n",
    "if params_export_flag:\n",
    "    fp = outputdir / \"m_id_to_bed6_3utr.tsv.gz\"\n",
    "    m_id_to_bed6_3utr.to_csv(fp, sep=\"\\t\", index=True, compression=\"gzip\")\n",
    "    print(\"Wrote mouse 3'UTR bed6: \", fp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "id": "d05e911e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "m_id_to_bed6.shape=(49141, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:6274997-6275006:+</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6274997</td>\n",
       "      <td>6275006</td>\n",
       "      <td>three_prime_UTR;Rb1cc1;protein_coding;ENSMUSG0...</td>\n",
       "      <td>9.344450</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:7170406-7170415:+</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7170406</td>\n",
       "      <td>7170415</td>\n",
       "      <td>three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...</td>\n",
       "      <td>5.094060</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:7170431-7170440:+</th>\n",
       "      <td>chr1</td>\n",
       "      <td>7170431</td>\n",
       "      <td>7170440</td>\n",
       "      <td>three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...</td>\n",
       "      <td>26.054101</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       chrom    start      end  \\\n",
       "identifier                                       \n",
       "chr1:6274997-6275006:+  chr1  6274997  6275006   \n",
       "chr1:7170406-7170415:+  chr1  7170406  7170415   \n",
       "chr1:7170431-7170440:+  chr1  7170431  7170440   \n",
       "\n",
       "                                                                     name  \\\n",
       "identifier                                                                  \n",
       "chr1:6274997-6275006:+  three_prime_UTR;Rb1cc1;protein_coding;ENSMUSG0...   \n",
       "chr1:7170406-7170415:+  three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...   \n",
       "chr1:7170431-7170440:+  three_prime_UTR;Pcmtd1;protein_coding;ENSMUSG0...   \n",
       "\n",
       "                            score strand  \n",
       "identifier                                \n",
       "chr1:6274997-6275006:+   9.344450      +  \n",
       "chr1:7170406-7170415:+   5.094060      +  \n",
       "chr1:7170431-7170440:+  26.054101      +  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wrote human 3'UTR bed6:  ../results/gene_sets_overlaps/m_id_to_bed6.tsv.gz\n"
     ]
    }
   ],
   "source": [
    "print(f\"{m_id_to_bed6.shape=}\")\n",
    "display(m_id_to_bed6.head(3))\n",
    "\n",
    "if params_export_flag:\n",
    "    fp = outputdir / \"m_id_to_bed6.tsv.gz\"\n",
    "    m_id_to_bed6.to_csv(fp, sep=\"\\t\", index=True, compression=\"gzip\")\n",
    "    print(\"Wrote human 3'UTR bed6: \", fp)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a8f9919",
   "metadata": {},
   "source": [
    "### Human"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "id": "d0a80f03",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "h_lift_3utr.shape=(20706, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier.human_source</th>\n",
       "      <th>identifier.mouse_target</th>\n",
       "      <th>orth_first_order</th>\n",
       "      <th>orth_second_order</th>\n",
       "      <th>orth_third_order.details</th>\n",
       "      <th>orth_third_order</th>\n",
       "      <th>gene_orthology_status</th>\n",
       "      <th>orthologous_gene_set</th>\n",
       "      <th>annotated_lifted_region_bound_gene_set</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:1211398-1211407:-</td>\n",
       "      <td>chr4:156016532-156016539:+</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSG00000186827, [ENSMUSG00000029075])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSMUSG00000029075}</td>\n",
       "      <td>{ENSMUSG00000029075}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:1211415-1211424:-</td>\n",
       "      <td>chr4:156016517-156016524:+</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSG00000186827, [ENSMUSG00000029075])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSMUSG00000029075}</td>\n",
       "      <td>{ENSMUSG00000029075}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1:1211431-1211440:-</td>\n",
       "      <td>chr4:156016501-156016510:+</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>[(ENSG00000186827, [ENSMUSG00000029075])]</td>\n",
       "      <td>True</td>\n",
       "      <td>first</td>\n",
       "      <td>{ENSMUSG00000029075}</td>\n",
       "      <td>{ENSMUSG00000029075}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  identifier.human_source     identifier.mouse_target orth_first_order  \\\n",
       "0  chr1:1211398-1211407:-  chr4:156016532-156016539:+             True   \n",
       "1  chr1:1211415-1211424:-  chr4:156016517-156016524:+             True   \n",
       "2  chr1:1211431-1211440:-  chr4:156016501-156016510:+             True   \n",
       "\n",
       "  orth_second_order                   orth_third_order.details  \\\n",
       "0              True  [(ENSG00000186827, [ENSMUSG00000029075])]   \n",
       "1              True  [(ENSG00000186827, [ENSMUSG00000029075])]   \n",
       "2              True  [(ENSG00000186827, [ENSMUSG00000029075])]   \n",
       "\n",
       "  orth_third_order gene_orthology_status  orthologous_gene_set  \\\n",
       "0             True                 first  {ENSMUSG00000029075}   \n",
       "1             True                 first  {ENSMUSG00000029075}   \n",
       "2             True                 first  {ENSMUSG00000029075}   \n",
       "\n",
       "  annotated_lifted_region_bound_gene_set  \n",
       "0                   {ENSMUSG00000029075}  \n",
       "1                   {ENSMUSG00000029075}  \n",
       "2                   {ENSMUSG00000029075}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wrote human 3'UTR with identified gene sets orthologies:  ../results/gene_sets_overlaps/h_lift_3utr.tsv.gz\n"
     ]
    }
   ],
   "source": [
    "print(f\"{h_lift_3utr.shape=}\")\n",
    "display(h_lift_3utr.head(3))\n",
    "\n",
    "\n",
    "if params_export_flag:\n",
    "    fp = outputdir / \"h_lift_3utr.tsv.gz\"\n",
    "    h_lift_3utr.to_csv(fp, sep=\"\\t\", index=False, compression=\"gzip\")\n",
    "    print(\"Wrote human 3'UTR with identified gene sets orthologies: \", fp)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "id": "a4d3c01f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "h_umap_3utr.shape=(1632, 3)\n",
      "Wrote mouse 3'UTR unmapped peaks:  ../results/gene_sets_overlaps/h_umap_3utr.tsv.gz\n"
     ]
    }
   ],
   "source": [
    "print(f\"{h_umap_3utr.shape=}\")\n",
    "h_umap_3utr.head(3)\n",
    "\n",
    "if params_export_flag:\n",
    "    fp = outputdir / \"h_umap_3utr.tsv.gz\"\n",
    "    h_umap_3utr.to_csv(fp, sep=\"\\t\", index=False, compression=\"gzip\")\n",
    "    print(\"Wrote mouse 3'UTR unmapped peaks: \", fp)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "id": "ce1de3d3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "len(h_genes_bound_orth_complete)=1302\n",
      "['ENSG00000188404', 'ENSG00000113384', 'ENSG00000130699']\n",
      "len(h_genes_no_bound_orth_complete)=1667\n",
      "['ENSG00000213741', 'ENSG00000146282', 'ENSG00000100567']\n",
      "len(h_genes_bound_orth_complete_3utr)=1000\n",
      "['ENSG00000137642', 'ENSG00000188404', 'ENSG00000113384']\n",
      "len(h_genes_no_bound_orth_complete_3utr)=1480\n",
      "['ENSG00000169813', 'ENSG00000213741', 'ENSG00000146282']\n"
     ]
    }
   ],
   "source": [
    "# GENE SETS\n",
    "\n",
    "print(f\"{len(h_genes_bound_orth_complete)=}\")\n",
    "print(list(h_genes_bound_orth_complete)[:3])\n",
    "\n",
    "if params_export_flag:\n",
    "    fp = outputdir / \"h_genes_bound_orth_complete.txt\"\n",
    "    with open(fp, \"w\") as f:\n",
    "        f.write(\"\\n\".join(h_genes_bound_orth_complete) + \"\\n\")\n",
    "\n",
    "\n",
    "print(f\"{len(h_genes_no_bound_orth_complete)=}\")\n",
    "print(list(h_genes_no_bound_orth_complete)[:3])\n",
    "\n",
    "if params_export_flag:\n",
    "    fp = outputdir / \"h_genes_no_bound_orth_complete.txt\"\n",
    "    with open(fp, \"w\") as f:\n",
    "        f.write(\"\\n\".join(h_genes_no_bound_orth_complete) + \"\\n\")\n",
    "\n",
    "\n",
    "###\n",
    "\n",
    "print(f\"{len(h_genes_bound_orth_complete_3utr)=}\")\n",
    "print(list(h_genes_bound_orth_complete_3utr)[:3])\n",
    "\n",
    "if params_export_flag:\n",
    "    fp = outputdir / \"h_genes_bound_orth_complete_3utr.txt\"\n",
    "    with open(fp, \"w\") as f:\n",
    "        f.write(\"\\n\".join(h_genes_bound_orth_complete_3utr) + \"\\n\")\n",
    "\n",
    "\n",
    "print(f\"{len(h_genes_no_bound_orth_complete_3utr)=}\")\n",
    "print(list(h_genes_no_bound_orth_complete_3utr)[:3])\n",
    "\n",
    "if params_export_flag:\n",
    "    fp = outputdir / \"h_genes_no_bound_orth_complete_3utr.txt\"\n",
    "    with open(fp, \"w\") as f:\n",
    "        f.write(\"\\n\".join(h_genes_no_bound_orth_complete_3utr) + \"\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "id": "330eb4e7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>main_annot</th>\n",
       "      <th>annot_set</th>\n",
       "      <th>gene_id_set</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:1211398-1211407:-</th>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>ENSG00000186827</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{three_prime_UTR}</td>\n",
       "      <td>{ENSG00000186827}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:1211415-1211424:-</th>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>ENSG00000186827</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{three_prime_UTR}</td>\n",
       "      <td>{ENSG00000186827}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:1211431-1211440:-</th>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>ENSG00000186827</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{three_prime_UTR}</td>\n",
       "      <td>{ENSG00000186827}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                     name  \\\n",
       "identifier                                                                  \n",
       "chr1:1211398-1211407:-  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "chr1:1211415-1211424:-  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "chr1:1211431-1211440:-  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "\n",
       "                                gene_id       main_annot          annot_set  \\\n",
       "identifier                                                                    \n",
       "chr1:1211398-1211407:-  ENSG00000186827  three_prime_UTR  {three_prime_UTR}   \n",
       "chr1:1211415-1211424:-  ENSG00000186827  three_prime_UTR  {three_prime_UTR}   \n",
       "chr1:1211431-1211440:-  ENSG00000186827  three_prime_UTR  {three_prime_UTR}   \n",
       "\n",
       "                              gene_id_set  \n",
       "identifier                                 \n",
       "chr1:1211398-1211407:-  {ENSG00000186827}  \n",
       "chr1:1211415-1211424:-  {ENSG00000186827}  \n",
       "chr1:1211431-1211440:-  {ENSG00000186827}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "h_id_to_meta.shape=(49904, 5)\n",
      "Wrote human all metadata:  ../results/gene_sets_overlaps/h_id_to_meta.tsv.gz\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>main_annot</th>\n",
       "      <th>annot_set</th>\n",
       "      <th>gene_id_set</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:1211398-1211407:-</th>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>ENSG00000186827</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{three_prime_UTR}</td>\n",
       "      <td>{ENSG00000186827}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:1211415-1211424:-</th>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>ENSG00000186827</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{three_prime_UTR}</td>\n",
       "      <td>{ENSG00000186827}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:1211431-1211440:-</th>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>ENSG00000186827</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>{three_prime_UTR}</td>\n",
       "      <td>{ENSG00000186827}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                     name  \\\n",
       "identifier                                                                  \n",
       "chr1:1211398-1211407:-  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "chr1:1211415-1211424:-  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "chr1:1211431-1211440:-  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "\n",
       "                                gene_id       main_annot          annot_set  \\\n",
       "identifier                                                                    \n",
       "chr1:1211398-1211407:-  ENSG00000186827  three_prime_UTR  {three_prime_UTR}   \n",
       "chr1:1211415-1211424:-  ENSG00000186827  three_prime_UTR  {three_prime_UTR}   \n",
       "chr1:1211431-1211440:-  ENSG00000186827  three_prime_UTR  {three_prime_UTR}   \n",
       "\n",
       "                              gene_id_set  \n",
       "identifier                                 \n",
       "chr1:1211398-1211407:-  {ENSG00000186827}  \n",
       "chr1:1211415-1211424:-  {ENSG00000186827}  \n",
       "chr1:1211431-1211440:-  {ENSG00000186827}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "h_id_to_meta_3utr.shape=(33847, 5)\n",
      "Wrote human 3'UTR metadata:  ../results/gene_sets_overlaps/h_id_to_meta_3utr.tsv.gz\n"
     ]
    }
   ],
   "source": [
    "display(h_id_to_meta.head(3))\n",
    "print(f\"{h_id_to_meta.shape=}\")\n",
    "\n",
    "if params_export_flag:\n",
    "    fp = outputdir / \"h_id_to_meta.tsv.gz\"\n",
    "    h_id_to_meta.to_csv(fp, sep=\"\\t\", index=True, compression=\"gzip\")\n",
    "    print(\"Wrote human all metadata: \", fp)\n",
    "\n",
    "\n",
    "display(h_id_to_meta_3utr.head(3))\n",
    "print(f\"{h_id_to_meta_3utr.shape=}\")\n",
    "\n",
    "if params_export_flag:\n",
    "    fp = outputdir / \"h_id_to_meta_3utr.tsv.gz\"\n",
    "    h_id_to_meta_3utr.to_csv(fp, sep=\"\\t\", index=True, compression=\"gzip\")\n",
    "    print(\"Wrote human 3'UTR metadata: \", fp)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c816f31",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "h_id_to_bed6_3utr.shape=(34652, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:1211398-1211407:-</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1211398</td>\n",
       "      <td>1211407</td>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>281.428788</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:1211415-1211424:-</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1211415</td>\n",
       "      <td>1211424</td>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>231.774900</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:1211431-1211440:-</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1211431</td>\n",
       "      <td>1211440</td>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>206.707600</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       chrom    start      end  \\\n",
       "identifier                                       \n",
       "chr1:1211398-1211407:-  chr1  1211398  1211407   \n",
       "chr1:1211415-1211424:-  chr1  1211415  1211424   \n",
       "chr1:1211431-1211440:-  chr1  1211431  1211440   \n",
       "\n",
       "                                                                     name  \\\n",
       "identifier                                                                  \n",
       "chr1:1211398-1211407:-  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "chr1:1211415-1211424:-  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "chr1:1211431-1211440:-  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "\n",
       "                             score strand  \n",
       "identifier                                 \n",
       "chr1:1211398-1211407:-  281.428788      -  \n",
       "chr1:1211415-1211424:-  231.774900      -  \n",
       "chr1:1211431-1211440:-  206.707600      -  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wrote human 3'UTR bed6:  ../results/gene_sets_overlaps/h_id_to_bed6_3utr.tsv.gz\n"
     ]
    }
   ],
   "source": [
    "print(f\"{h_id_to_bed6_3utr.shape=}\")\n",
    "display(h_id_to_bed6_3utr.head(3))\n",
    "\n",
    "if params_export_flag:\n",
    "    fp = outputdir / \"h_id_to_bed6_3utr.tsv.gz\"\n",
    "    h_id_to_bed6_3utr.to_csv(fp, sep=\"\\t\", index=True, compression=\"gzip\")\n",
    "    print(\"Wrote human 3'UTR bed6: \", fp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "id": "b928c270",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "h_id_to_bed6.shape=(50917, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:1211398-1211407:-</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1211398</td>\n",
       "      <td>1211407</td>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>281.428788</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:1211415-1211424:-</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1211415</td>\n",
       "      <td>1211424</td>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>231.774900</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:1211431-1211440:-</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1211431</td>\n",
       "      <td>1211440</td>\n",
       "      <td>three_prime_UTR;TNFRSF4;protein_coding;ENSG000...</td>\n",
       "      <td>206.707600</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       chrom    start      end  \\\n",
       "identifier                                       \n",
       "chr1:1211398-1211407:-  chr1  1211398  1211407   \n",
       "chr1:1211415-1211424:-  chr1  1211415  1211424   \n",
       "chr1:1211431-1211440:-  chr1  1211431  1211440   \n",
       "\n",
       "                                                                     name  \\\n",
       "identifier                                                                  \n",
       "chr1:1211398-1211407:-  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "chr1:1211415-1211424:-  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "chr1:1211431-1211440:-  three_prime_UTR;TNFRSF4;protein_coding;ENSG000...   \n",
       "\n",
       "                             score strand  \n",
       "identifier                                 \n",
       "chr1:1211398-1211407:-  281.428788      -  \n",
       "chr1:1211415-1211424:-  231.774900      -  \n",
       "chr1:1211431-1211440:-  206.707600      -  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wrote human 3'UTR bed6:  ../results/gene_sets_overlaps/h_id_to_bed6.tsv.gz\n"
     ]
    }
   ],
   "source": [
    "print(f\"{h_id_to_bed6.shape=}\")\n",
    "display(h_id_to_bed6.head(3))\n",
    "\n",
    "if params_export_flag:\n",
    "    fp = outputdir / \"h_id_to_bed6.tsv.gz\"\n",
    "    h_id_to_bed6.to_csv(fp, sep=\"\\t\", index=True, compression=\"gzip\")\n",
    "    print(\"Wrote human 3'UTR bed6: \", fp)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hzmdev",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
